---
title: "Broad Mutational Scanning of the Dihydrofolate Reductase Protein Family"
author: 'Authors: [Karl J. Romanowicz](https://kromanowicz.github.io/), Carmen Resnick, Samuel R. Hinton, Calin Plesa'
output:
  html_notebook:
    theme: spacelab
    toc: yes
    toc_depth: 5
    toc_float:
      collapsed: yes
      smooth_scroll: yes
  html_document:
    toc: yes
    toc_depth: '5'
    df_print: paged
  pdf_document:
    toc: yes
    toc_depth: '5'
---

**GitHub Repository:** [https://github.com/PlesaLab/DHFR_12288](https://github.com/PlesaLab/DHFR_12288)

# Experiment

This R Notebook provides complete reproducibility of the data analysis presented in ***"Broad mutational scanning of the dihydrofolate reductase protein family"*** by Romanowicz, Resnick, Hinton, and Plesa (*In Prep*).

<font color="green">This pipeline processes a library of 5,756 DHFR homologs and their associated mutants in two-fold redundancy (two codon versions) to derive fitness scores from multiple time points during a multiplexed in-vivo assay that characterizes the ability of these homologs and their mutants to complement functionality of an E. coli knockout strain (dFolA|dThyA). Sequence data were generated using the Illumina NovaSeq platform using paired-end sequencing read amplicons.</font>

![Methods overview to achieve a broad-mutational scan for DHFR homologs.](images/methods.png)

```{css}
.badCode {
background-color: lightpink;
font-weight: bold;
}

.goodCode {
background-color: lightgreen;
font-weight: bold;
}

.sharedCode {
background-color: lightblue;
font-weight: bold;
}

table {
  margin: auto;
  border-top: 1px solid #666;
  border-bottom: 1px solid #666;
}
table thead th { border-bottom: 1px solid #ddd; }
th, td { padding: 5px; }
thead, tfoot, tr:nth-child(even) { background: #eee; }
```

```{r setup, include=FALSE}
# Set global options for notebook
knitr::opts_chunk$set(echo = TRUE)
knitr::opts_chunk$set(warning = TRUE, message = TRUE)
knitr::opts_chunk$set(echo = TRUE, class.source = "bg-success")

# Getting the path of your current open file and set as wd
current_path = rstudioapi::getActiveDocumentContext()$path 
setwd(dirname(current_path))
print(getwd())
```

# Packages
The following R packages must be installed prior to loading into the R session. See the **Reproducibility** tab for a complete list of packages and their versions used in this workflow.
```{r message=FALSE, results='hide', warning=FALSE}
# Make a vector of required packages
required.packages <- c("Biostrings", "castor", "cowplot", "devtools", "dplyr", "ggExtra", "ggnewscale", "ggplot2", "ggridges", "ggtree", "ggtreeExtra", "gridExtra", "knitr", "patchwork", "pscl", "RColorBrewer", "reshape", "ROCR", "scales", "stringr", "stringi", "tidyr", "viridis")

# Load required packages
lapply(required.packages, library, character.only = TRUE)
```

```{r include=FALSE}
# set.seed is used to fix the random number generation to make the results repeatable
set.seed(123)
```

# Load Data
<font color="blue">**This section is based on the R file: "R_load_data.R".**</font> It describes how to load all of the pre-existing barcode mapping data necessary for downstream analysis.

**Pre-Existing Data Files to Import**

* **BCinfo:** Mapping Information (one for each library)
* **mutIDinfo:** Mutant Information (one for each library)
* **BCs:** Everything Sequenced (one for eacy library)
* **BCmut:** Only Mapped Barcodes (one for each library)

## Barcode Info
**Begin the analysis by loading in the barcode information for each codon-version library.** This data represents each **barcode** [`BC`] recovered from the complementation process, linked to each **homolog or mutant** [`mutID`], the **number of mutations** observed in the sequence relative to its designed homolog [`mutations`], a list of where the **mutations occurred** along the protein sequence[`cigar`], and the **number of reads** associated with each barcode [`reads`]:
```{r}
# Library 15 (Codon 1)
BCinfo15 = read.csv("for_Karl/15_HiFi_BC_mutID_all.csv", head=TRUE)  # read csv file

# Library 16 (Codon 2)
BCinfo16 = read.csv("for_Karl/16_HiFi_BC_mutID_all.csv", head=TRUE)  # read csv file
```

<font color="green">**Preview of BCinfo**</font>
```{r echo=FALSE}
# Library 15
head(BCinfo15)

# Library 16
head(BCinfo16)
```

```{r class.output="goodCode"}
# Count the number of unique barcodes for Library 15
BCinfo15.count <- length(unique(BCinfo15$BC))
format(BCinfo15.count, big.mark = ",")

# Count the number of unique barcodes for Library 16
BCinfo16.count <- length(unique(BCinfo16$BC))
format(BCinfo16.count, big.mark = ",")
```

## Combine BCinfo Libraries
Combine both libraries based on shared BCs:
```{r class.output="sharedCode"}
# Use 'merge' function to merge both libraries based on shared BCs
BCinfo_all <- merge(BCinfo15, BCinfo16, by = "BC", all = FALSE, suffixes = c("_Lib15", "_Lib16"))

# Count the number of unique barcodes shared between libraries
length(unique(BCinfo_all$BC))
```

Run the code again but keep all BCs regardless if they're shared between libraries
```{r class.output="sharedCode"}
# Use 'merge' function to merge both libraries based on shared BCs (all = TRUE)
BCinfo_all <- merge(BCinfo15, BCinfo16, all = TRUE)

# Count the number of unique barcodes retained between libraries
BCinfo_all.count <- length(unique(BCinfo_all$BC))
format(BCinfo_all.count, big.mark = ",")
```

## Mutant Info
**Load in the mutant information for each codon-version library.** This data represents each **mutant** sequence [`mutID`] from the dataset, the **designed homolog** it mutated from [`IDalign`], the **number of barcodes** associated with each mutant [`numBCs`], the **number of mutations** in the sequence relative to the designed homolog [`mutations`], the **full sequence** of the mutant [`seq`], the **percent identity of the mutant** to the designed homolog [`pct_ident`], a list of **all barcodes** associated with the mutant [`BCs`], and the **barcode code** [`BCcode`]:
```{r}
# Library 15 (Codon 1)
mutIDinfo15 = read.csv("for_Karl/15_HiFi_mutID_info_all.csv", head=TRUE)  # read csv file

# Library 16 (Codon 2)
mutIDinfo16 = read.csv("for_Karl/16_HiFi_mutID_info_all.csv", head=TRUE)  # read csv file
```

```{r class.output="goodCode"}
## Make sure each mutant has >=0 mutations by removing rows with negative values
mutIDinfo15 <- mutIDinfo15 %>%
  filter(mutations >= 0)

# Check mutate count for Lib15 (should be "0"):
min(mutIDinfo15$mutations)

## Make sure each mutant has >=0 mutations by removing rows with negative values
mutIDinfo16 <- mutIDinfo16 %>%
  filter(mutations >= 0)

# Check mutate count for Lib16 (should be "0"):
min(mutIDinfo16$mutations)
```

<font color="green">**Preview of mutIDinfo**</font>
```{r echo=FALSE}
# Library 15
head(mutIDinfo15)

# Library 16
head(mutIDinfo16)
```

```{r class.output="goodCode"}
# Count the number of unique barcodes for Library 15
mutIDinfo15.count <- length(unique(mutIDinfo15$BCs))
format(mutIDinfo15.count, big.mark = ",")

# Count the number of unique barcodes for Library 16
mutIDinfo16.count <- length(unique(mutIDinfo16$BCs))
format(mutIDinfo16.count, big.mark = ",")
```

## Combine Libraries
Combine both libraries based on shared BCs:
```{r class.output="sharedCode"}
#Use 'merge' function to merge both libraries based on shared BCs
mutIDinfo_all <- merge(mutIDinfo15, mutIDinfo16, by = "BCs", all = FALSE, suffixes = c("_Lib15", "_Lib16"))

# Count the number of unique barcodes shared between libraries
length(unique(mutIDinfo_all$BCs))
```

Combine both libraries again, but retain all unique barcodes:
```{r class.output="sharedCode"}
#Use 'merge' function to merge both libraries based on shared BCs
mutIDinfo_all <- merge(mutIDinfo15, mutIDinfo16, all = TRUE)

# Count the number of unique barcodes shared between libraries
mutIDinfo_all.count <- length(unique(mutIDinfo_all$BCs))
format(mutIDinfo_all.count, big.mark = ",")
```

## Perfects Data
Determine the number of true homologs ("perfects") that match the designed homologs without mutations:
```{r class.output="goodCode"}
# Library 15
perfects.count15 <- length(unique(mutIDinfo15$IDalign[which(mutIDinfo15$mutations==0)]))
  format(perfects.count15, big.mark = ",")

# Library 16
perfects.count16 <- length(unique(mutIDinfo16$IDalign[which(mutIDinfo16$mutations==0)]))
  format(perfects.count16, big.mark = ",")
```

```{r class.output="sharedCode"}
# Combined Libraries
perfects.count.all <- length(unique(mutIDinfo_all$IDalign[which(mutIDinfo_all$mutations==0)]))
  format(perfects.count.all, big.mark = ",")
```

## Count Data
**Import the count data (sequences) for each sampling condition in Library 15 (D01, D03, D05:D11, E07:F01, F9:G3, G11:H5) and Library 16 (D02, D04, D12:E6, F2:F8, G4:G10, H6:H12).** Each .csv file contains two columns: **(1)** [`BC`] represents the **unique barcode** recovered from the sampling condition, and **(2)** [`D0X`] represents the **sequence counts** matching each barcode within the sampling condition.

Import all Library 15 count files:
```{r}
### Library 15

## Library 15 - T0

# D01 - Lib15 - Overnight Growth on LB - T0
D01 = read.csv("counts/D01_collapse_d1.tsv", head=FALSE, sep="\t") %>%
  select(-V3)
colnames(D01) <- c("BC","D01")

# D03 - Lib15 - Growth on M9 - Full Supplement - T0
D03 = read.csv("counts/D03_collapse_d1.tsv", head=FALSE, sep="\t") %>%
  select(-V3)
colnames(D03) <- c("BC","D03")

## Library 15 - T1

# D05 - Lib15 - Growth on M9 - 0 ug/mL TMP - T1
D05 = read.csv("counts/D05_collapse_d1.tsv", head=FALSE, sep="\t") %>%
  select(-V3)
colnames(D05) <- c("BC","D05")

# D06 - Lib15 - Growth on M9 - 0.058 ug/ml TMP - T1
D06 = read.csv("counts/D06_collapse_d1.tsv", head=FALSE, sep="\t") %>%
  select(-V3)
colnames(D06) <- c("BC","D06")

# D07 - Lib15 - Growth on M9 - 0.5 ug/ml TMP - T1
D07 = read.csv("counts/D07_collapse_d1.tsv", head=FALSE, sep="\t") %>%
  select(-V3)
colnames(D07) <- c("BC","D07")

# D08 - Lib15 - Growth on M9 - 1.0 ug/ml TMP - T1
D08 = read.csv("counts/D08_collapse_d1.tsv", head=FALSE, sep="\t") %>%
  select(-V3)
colnames(D08) <- c("BC","D08")

# D09 - Lib15 - Growth on M9 - 10.0 ug/ml TMP - T1
D09 = read.csv("counts/D09_collapse_d1.tsv", head=FALSE, sep="\t") %>%
  select(-V3)
colnames(D09) <- c("BC","D09")

# D10 - Lib15 - Growth on M9 - 50.0 ug/ml TMP - T1
D10 = read.csv("counts/D10_collapse_d1.tsv", head=FALSE, sep="\t") %>%
  select(-V3)
colnames(D10) <- c("BC","D10")

# D11 - Lib15 - Growth on M9 - 200.0 ug/ml TMP - T1
D11 = read.csv("counts/D11_collapse_d1.tsv", head=FALSE, sep="\t") %>%
  select(-V3)
colnames(D11) <- c("BC","D11")

## Library 15 - T2

# E07 - Lib15 - Growth on M9 - 0 ug/mL TMP - T2
E07 = read.csv("counts/E07_collapse_d1.tsv", head=FALSE, sep="\t") %>%
  select(-V3)
colnames(E07) <- c("BC","E07")

# E08 - Lib15 - Growth on M9 - 0.058 ug/ml TMP - T2
E08 = read.csv("counts/E08_collapse_d1.tsv", head=FALSE, sep="\t") %>%
  select(-V3)
colnames(E08) <- c("BC","E08")

# E09 - Lib15 - Growth on M9 - 0.5 ug/ml TMP - T2
E09 = read.csv("counts/E09_collapse_d1.tsv", head=FALSE, sep="\t") %>%
  select(-V3)
colnames(E09) <- c("BC","E09")

# E10 - Lib15 - Growth on M9 - 1.0 ug/ml TMP - T2
E10 = read.csv("counts/E10_collapse_d1.tsv", head=FALSE, sep="\t") %>%
  select(-V3)
colnames(E10) <- c("BC","E10")

# E11 - Lib15 - Growth on M9 - 10.0 ug/ml TMP - T2
E11 = read.csv("counts/E11_collapse_d1.tsv", head=FALSE, sep="\t") %>%
  select(-V3)
colnames(E11) <- c("BC","E11")

# E12 - Lib15 - Growth on M9 - 50.0 ug/ml TMP - T2
E12 = read.csv("counts/E12_collapse_d1.tsv", head=FALSE, sep="\t") %>%
  select(-V3)
colnames(E12) <- c("BC","E12")

# F01 - Lib15 - Growth on M9 - 200.0 ug/ml TMP - T2
F01 = read.csv("counts/F01_collapse_d1.tsv", head=FALSE, sep="\t") %>%
  select(-V3)
colnames(F01) <- c("BC","F01")

## Library 15 - T3

# F09 - Lib15 - Growth on M9 - 0 ug/mL TMP - T3
F09 = read.csv("counts/F09_collapse_d1.tsv", head=FALSE, sep="\t") %>%
  select(-V3)
colnames(F09) <- c("BC","F09")

# F10 - Lib15 - Growth on M9 - 0.058 ug/ml TMP - T3
F10 = read.csv("counts/F10_collapse_d1.tsv", head=FALSE, sep="\t") %>%
  select(-V3)
colnames(F10) <- c("BC","F10")

# F11 - Lib15 - Growth on M9 - 0.5 ug/ml TMP - T3
F11 = read.csv("counts/F11_collapse_d1.tsv", head=FALSE, sep="\t") %>%
  select(-V3)
colnames(F11) <- c("BC","F11")

# F12 - Lib15 - Growth on M9 - 1.0 ug/ml TMP - T3
F12 = read.csv("counts/F12_collapse_d1.tsv", head=FALSE, sep="\t") %>%
  select(-V3)
colnames(F12) <- c("BC","F12")

# G01 - Lib15 - Growth on M9 - 10.0 ug/ml TMP - T3
G01 = read.csv("counts/G01_collapse_d1.tsv", head=FALSE, sep="\t") %>%
  select(-V3)
colnames(G01) <- c("BC","G01")

# G02 - Lib15 - Growth on M9 - 50.0 ug/ml TMP - T3
G02 = read.csv("counts/G02_collapse_d1.tsv", head=FALSE, sep="\t") %>%
  select(-V3)
colnames(G02) <- c("BC","G02")

# G03 - Lib15 - Growth on M9 - 200.0 ug/ml TMP - T3
G03 = read.csv("counts/G03_collapse_d1.tsv", head=FALSE, sep="\t") %>%
  select(-V3)
colnames(G03) <- c("BC","G03")

## Library 15 - T4

# G11 - Lib15 - Growth on M9 - 0 ug/mL TMP - T4
G11 = read.csv("counts/G11_collapse_d1.tsv", head=FALSE, sep="\t") %>%
  select(-V3)
colnames(G11) <- c("BC","G11")

# G12 - Lib15 - Growth on M9 - 0.058 ug/ml TMP - T4
G12 = read.csv("counts/G12_collapse_d1.tsv", head=FALSE, sep="\t") %>%
  select(-V3)
colnames(G12) <- c("BC","G12")

# H01 - Lib15 - Growth on M9 - 0.5 ug/ml TMP - T4
H01 = read.csv("counts/H01_collapse_d1.tsv", head=FALSE, sep="\t") %>%
  select(-V3)
colnames(H01) <- c("BC","H01")

# H02 - Lib15 - Growth on M9 - 1.0 ug/ml TMP - T4
H02 = read.csv("counts/H02_collapse_d1.tsv", head=FALSE, sep="\t") %>%
  select(-V3)
colnames(H02) <- c("BC","H02")

# H03 - Lib15 - Growth on M9 - 10.0 ug/ml TMP - T4
H03 = read.csv("counts/H03_collapse_d1.tsv", head=FALSE, sep="\t") %>%
  select(-V3)
colnames(H03) <- c("BC","H03")

# H04 - Lib15 - Growth on M9 - 50.0 ug/ml TMP - T4
H04 = read.csv("counts/H04_collapse_d1.tsv", head=FALSE, sep="\t") %>%
  select(-V3)
colnames(H04) <- c("BC","H04")

# H05 - Lib15 - Growth on M9 - 200.0 ug/ml TMP - T4
H05 = read.csv("counts/H05_collapse_d1.tsv", head=FALSE, sep="\t") %>%
  select(-V3)
colnames(H05) <- c("BC","H05")
```

Import all Library 16 count files:
```{r}
### Library 16

## Library 16 - T0

# D02 - Lib16 - Overnight Growth on LB - T0
D02 = read.csv("counts/D02_collapse_d1.tsv", head=FALSE, sep="\t") %>%
  select(-V3)
colnames(D02) <- c("BC","D02")

# D04 - Lib16 - Growth on M9 - Full Supplement - T0
D04 = read.csv("counts/D04_collapse_d1.tsv", head=FALSE, sep="\t") %>%
  select(-V3)
colnames(D04) <- c("BC","D04")

## Library 16 - T1

# D12 - Lib16 - Growth on M9 - 0 ug/mL TMP - T1
D12 = read.csv("counts/D12_collapse_d1.tsv", head=FALSE, sep="\t") %>%
  select(-V3)
colnames(D12) <- c("BC","D12")

# E01 - Lib16 - Growth on M9 - 0.058 ug/ml TMP - T1
E01 = read.csv("counts/E01_collapse_d1.tsv", head=FALSE, sep="\t") %>%
  select(-V3)
colnames(E01) <- c("BC","E01")

# E02 - Lib16 - Growth on M9 - 0.5 ug/ml TMP - T1
E02 = read.csv("counts/E02_collapse_d1.tsv", head=FALSE, sep="\t") %>%
  select(-V3)
colnames(E02) <- c("BC","E02")

# E03 - Lib16 - Growth on M9 - 1.0 ug/ml TMP - T1
E03 = read.csv("counts/E03_collapse_d1.tsv", head=FALSE, sep="\t") %>%
  select(-V3)
colnames(E03) <- c("BC","E03")

# E04 - Lib16 - Growth on M9 - 10.0 ug/ml TMP - T1
E04 = read.csv("counts/E04_collapse_d1.tsv", head=FALSE, sep="\t") %>%
  select(-V3)
colnames(E04) <- c("BC","E04")

# E05 - Lib16 - Growth on M9 - 50.0 ug/ml TMP - T1
E05 = read.csv("counts/E05_collapse_d1.tsv", head=FALSE, sep="\t") %>%
  select(-V3)
colnames(E05) <- c("BC","E05")

# E06 - Lib16 - Growth on M9 - 200.0 ug/ml TMP - T1
E06 = read.csv("counts/E06_collapse_d1.tsv", head=FALSE, sep="\t") %>%
  select(-V3)
colnames(E06) <- c("BC","E06")

## Library 16 - T2

# F02 - Lib16 - Growth on M9 - 0 ug/mL TMP - T2
F02 = read.csv("counts/F02_collapse_d1.tsv", head=FALSE, sep="\t") %>%
  select(-V3)
colnames(F02) <- c("BC","F02")

# F03 - Lib16 - Growth on M9 - 0.058 ug/ml TMP - T2
F03 = read.csv("counts/F03_collapse_d1.tsv", head=FALSE, sep="\t") %>%
  select(-V3)
colnames(F03) <- c("BC","F03")

# F04 - Lib16 - Growth on M9 - 0.5 ug/ml TMP - T2
F04 = read.csv("counts/F04_collapse_d1.tsv", head=FALSE, sep="\t") %>%
  select(-V3)
colnames(F04) <- c("BC","F04")

# F05 - Lib16 - Growth on M9 - 1.0 ug/ml TMP - T2
F05 = read.csv("counts/F05_collapse_d1.tsv", head=FALSE, sep="\t") %>%
  select(-V3)
colnames(F05) <- c("BC","F05")

# F06 - Lib16 - Growth on M9 - 10.0 ug/ml TMP - T2
F06 = read.csv("counts/F06_collapse_d1.tsv", head=FALSE, sep="\t") %>%
  select(-V3)
colnames(F06) <- c("BC","F06")

# F07 - Lib16 - Growth on M9 - 50.0 ug/ml TMP - T2
F07 = read.csv("counts/F07_collapse_d1.tsv", head=FALSE, sep="\t") %>%
  select(-V3)
colnames(F07) <- c("BC","F07")

# F08 - Lib16 - Growth on M9 - 200.0 ug/ml TMP - T2
F08 = read.csv("counts/F08_collapse_d1.tsv", head=FALSE, sep="\t") %>%
  select(-V3)
colnames(F08) <- c("BC","F08")

## Library 16 - T3

# G04 - Lib16 - Growth on M9 - 0 ug/mL TMP - T3
G04 = read.csv("counts/G04_collapse_d1.tsv", head=FALSE, sep="\t") %>%
  select(-V3)
colnames(G04) <- c("BC","G04")

# G05 - Lib16 - Growth on M9 - 0.058 ug/ml TMP - T3
G05 = read.csv("counts/G05_collapse_d1.tsv", head=FALSE, sep="\t") %>%
  select(-V3)
colnames(G05) <- c("BC","G05")

# G06 - Lib16 - Growth on M9 - 0.5 ug/ml TMP - T3
G06 = read.csv("counts/G06_collapse_d1.tsv", head=FALSE, sep="\t") %>%
  select(-V3)
colnames(G06) <- c("BC","G06")

# G07 - Lib16 - Growth on M9 - 1.0 ug/ml TMP - T3
G07 = read.csv("counts/G07_collapse_d1.tsv", head=FALSE, sep="\t") %>%
  select(-V3)
colnames(G07) <- c("BC","G07")

# G08 - Lib16 - Growth on M9 - 10.0 ug/ml TMP - T3
G08 = read.csv("counts/G08_collapse_d1.tsv", head=FALSE, sep="\t") %>%
  select(-V3)
colnames(G08) <- c("BC","G08")

# G09 - Lib16 - Growth on M9 - 50.0 ug/ml TMP - T3
G09 = read.csv("counts/G09_collapse_d1.tsv", head=FALSE, sep="\t") %>%
  select(-V3)
colnames(G09) <- c("BC","G09")

# G10 - Lib16 - Growth on M9 - 200.0 ug/ml TMP - T3
G10 = read.csv("counts/G10_collapse_d1.tsv", head=FALSE, sep="\t") %>%
  select(-V3)
colnames(G10) <- c("BC","G10")

## Library 16 - T4

# H06 - Lib16 - Growth on M9 - 0 ug/mL TMP - T4
H06 = read.csv("counts/H06_collapse_d1.tsv", head=FALSE, sep="\t") %>%
  select(-V3)
colnames(H06) <- c("BC","H06")

# H07 - Lib16 - Growth on M9 - 0.058 ug/ml TMP - T4
H07 = read.csv("counts/H07_collapse_d1.tsv", head=FALSE, sep="\t") %>%
  select(-V3)
colnames(H07) <- c("BC","H07")

# H08 - Lib16 - Growth on M9 - 0.5 ug/ml TMP - T4
H08 = read.csv("counts/H08_collapse_d1.tsv", head=FALSE, sep="\t") %>%
  select(-V3)
colnames(H08) <- c("BC","H08")

# H09 - Lib16 - Growth on M9 - 1.0 ug/ml TMP - T4
H09 = read.csv("counts/H09_collapse_d1.tsv", head=FALSE, sep="\t") %>%
  select(-V3)
colnames(H09) <- c("BC","H09")

# H10 - Lib16 - Growth on M9 - 10.0 ug/ml TMP - T4
H10 = read.csv("counts/H10_collapse_d1.tsv", head=FALSE, sep="\t") %>%
  select(-V3)
colnames(H10) <- c("BC","H10")

# H11 - Lib16 - Growth on M9 - 50.0 ug/ml TMP - T4
H11 = read.csv("counts/H11_collapse_d1.tsv", head=FALSE, sep="\t") %>%
  select(-V3)
colnames(H11) <- c("BC","H11")

# H12 - Lib16 - Growth on M9 - 200.0 ug/ml TMP - T4
H12 = read.csv("counts/H12_collapse_d1.tsv", head=FALSE, sep="\t") %>%
  select(-V3)
colnames(H12) <- c("BC","H12")
```

# Summarized Data

**Summarize the imported count data by sampling condition for each codon-version library.** Begin by combining the count data from each sampling condition into a single dataframe for each library. Next, sum the full count values for each sampling condition (total number of sequences recovered for each sampling condition). Then, sum the number of unique barcodes recovered from each sampling condition. Define the sampling condition associated with each sample code. Finally, write the data summary to a .csv file and save in the working directory.

**Summarize Library 15**
```{r}
## Library 15

# Merge data objects together into single dataframe
norm_vals_15 <- data.frame(sample=c("D01","D03","D05","D06","D07","D08","D09","D10","D11","E07","E08","E09","E10","E11","E12","F01","F09","F10","F11","F12","G01","G02","G03","G11","G12","H01","H02","H03","H04","H05"),
                        counts=c(sum(D01$D01),sum(D03$D03),sum(D05$D05),sum(D06$D06),sum(D07$D07),sum(D08$D08),
                                 sum(D09$D09),sum(D10$D10),sum(D11$D11),sum(E07$E07),sum(E08$E08),sum(E09$E09),
                                 sum(E10$E10),sum(E11$E11),sum(E12$E12),sum(F01$F01),sum(F09$F09),sum(F10$F10),
                                 sum(F11$F11),sum(F12$F12),sum(G01$G01),sum(G02$G02),sum(G03$G03),sum(G11$G11),
                                 sum(G12$G12),sum(H01$H01),sum(H02$H02),sum(H03$H03),sum(H04$H04),sum(H05$H05)),
                        numBCs=c(length(D01$D01),length(D03$D03),length(D05$D05),length(D06$D06),length(D07$D07),
                                 length(D08$D08),length(D09$D09),length(D10$D10),length(D11$D11),length(E07$E07),
                                 length(E08$E08),length(E09$E09),length(E10$E10),length(E11$E11),length(E12$E12),
                                 length(F01$F01),length(F09$F09),length(F10$F10),length(F11$F11),length(F12$F12),
                                 length(G01$G01),length(G02$G02),length(G03$G03),length(G11$G11),length(G12$G12),
                                 length(H01$H01),length(H02$H02),length(H03$H03),length(H04$H04),length(H05$H05)),
                        sampledesc=c("D01 - Lib15 - T0 - Overnight Growth on LB",
                                     "D03 - Lib15 - T0 - Growth on M9 - Full Supplement",
                                     "D05 - Lib15 - T1 - Growth on M9 - 0.0 ug/ml TMP",
                                     "D06 - Lib15 - T1 - Growth on M9 - 0.058 ug/ml TMP",
                                     "D07 - Lib15 - T1 - Growth on M9 - 0.5 ug/ml TMP",
                                     "D08 - Lib15 - T1 - Growth on M9 - 1.0 ug/ml TMP",
                                     "D09 - Lib15 - T1 - Growth on M9 - 10.0 ug/ml TMP",
                                     "D10 - Lib15 - T1 - Growth on M9 - 50.0 ug/ml TMP",
                                     "D11 - Lib15 - T1 - Growth on M9 - 200.0 ug/ml TMP",
                                     "E07 - Lib15 - T2 - Growth on M9 - 0.0 ug/ml TMP",
                                     "E08 - Lib15 - T2 - Growth on M9 - 0.058 ug/ml TMP",
                                     "E09 - Lib15 - T2 - Growth on M9 - 0.5 ug/ml TMP",
                                     "E10 - Lib15 - T2 - Growth on M9 - 1.0 ug/ml TMP",
                                     "E11 - Lib15 - T2 - Growth on M9 - 10.0 ug/ml TMP",
                                     "E12 - Lib15 - T2 - Growth on M9 - 50.0 ug/ml TMP",
                                     "F01 - Lib15 - T2 - Growth on M9 - 200.0 ug/ml TMP",
                                     "F09 - Lib15 - T3 - Growth on M9 - 0.0 ug/ml TMP",
                                     "F10 - Lib15 - T3 - Growth on M9 - 0.058 ug/ml TMP",
                                     "F11 - Lib15 - T3 - Growth on M9 - 0.5 ug/ml TMP",
                                     "F12 - Lib15 - T3 - Growth on M9 - 1.0 ug/ml TMP",
                                     "G01 - Lib15 - T3 - Growth on M9 - 10.0 ug/ml TMP",
                                     "G02 - Lib15 - T3 - Growth on M9 - 50.0 ug/ml TMP",
                                     "G03 - Lib15 - T3 - Growth on M9 - 200.0 ug/ml TMP",
                                     "G11 - Lib15 - T4 - Growth on M9 - 0.0 ug/ml TMP",
                                     "G12 - Lib15 - T4 - Growth on M9 - 0.058 ug/ml TMP",
                                     "H01 - Lib15 - T4 - Growth on M9 - 0.5 ug/ml TMP",
                                     "H02 - Lib15 - T4 - Growth on M9 - 1.0 ug/ml TMP",
                                     "H03 - Lib15 - T4 - Growth on M9 - 10.0 ug/ml TMP",
                                     "H04 - Lib15 - T4 - Growth on M9 - 50.0 ug/ml TMP",
                                     "H05 - Lib15 - T4 - Growth on M9 - 200.0 ug/ml TMP"))

# Write dataframe to .csv table
write.csv(norm_vals_15, 'novaseq_basic_stats_Lib15.csv', row.names = FALSE, quote=FALSE)
```

<font color="green">**Library 15 Summarized Count Data for Sequences and Barcodes (BC)**</font>
```{r echo=FALSE}
# Display dataframe with kable formatting
norm_vals_15_col1 <- norm_vals_15[,c(4,1,2,3)]

knitr::kable(norm_vals_15_col1,
  col.names = c('Sample Conditions', 'Sample ID', 'Seq Counts', 'BC Counts'),
  align = "lccl",
  format.args = list(big.mark = ","))
```

**Summarize Library 16**
```{r}
## Library 16

# Merge data objects together into single dataframe
norm_vals_16 <- data.frame(sample=c("D02","D04",
                                    "D12","E01","E02","E03","E04","E05","E06",
                                    "F02","F03","F04","F05","F06","F07","F08",
                                    "G04","G05","G06","G07","G08","G09","G10",
                                    "H06","H07","H08","H09","H10","H11","H12"),
                        counts=c(sum(D02$D02),sum(D04$D04),sum(D12$D12),
                                 sum(E01$E01),sum(E02$E02),sum(E03$E03),sum(E04$E04),sum(E05$E05),sum(E06$E06),
                                 sum(F02$F02),sum(F03$F03),sum(F04$F04),sum(F05$F05),sum(F06$F06),sum(F07$F07),
                                 sum(F08$F08),sum(G04$G04),sum(G05$G05),sum(G06$G06),sum(G07$G07),sum(G08$G08),
                                 sum(G09$G09),sum(G10$G10),sum(H06$H06),sum(H07$H07),sum(H08$H08),sum(H09$H09),
                                 sum(H10$H10),sum(H11$H11),sum(H12$H12)),
                        numBCs=c(length(D02$D02),length(D04$D04),length(D12$D12),length(E01$E01),length(E02$E02),
                                 length(E03$E03),length(E04$E04),length(E05$E05),length(E06$E06),length(F02$F02),
                                 length(F03$F03),length(F04$F04),length(F05$F05),length(F06$F06),length(F07$F07),
                                 length(F08$F08),length(G04$G04),length(G05$G05),length(G06$G06),length(G07$G07),
                                 length(G08$G08),length(G09$G09),length(G10$G10),length(H06$H06),length(H07$H07),
                                 length(H08$H08),length(H09$H09),length(H10$H10),length(H11$H11),length(H12$H12)),
                        sampledesc=c("D02 - Lib16 - T0 - Overnight Growth on LB",
                                     "D04 - Lib16 - T0 - Growth on M9 - Full Supplement",
                                     "D12 - Lib16 - T1 - Growth on M9 - 0.0 ug/ml TMP",
                                     "E01 - Lib16 - T1 - Growth on M9 - 0.058 ug/ml TMP",
                                     "E02 - Lib16 - T1 - Growth on M9 - 0.5 ug/ml TMP",
                                     "E03 - Lib16 - T1 - Growth on M9 - 1.0 ug/ml TMP",
                                     "E04 - Lib16 - T1 - Growth on M9 - 10.0 ug/ml TMP",
                                     "E05 - Lib16 - T1 - Growth on M9 - 50.0 ug/ml TMP",
                                     "E06 - Lib16 - T1 - Growth on M9 - 200.0 ug/ml TMP",
                                     "F02 - Lib16 - T2 - Growth on M9 - 0.0 ug/ml TMP",
                                     "F03 - Lib16 - T2 - Growth on M9 - 0.058 ug/ml TMP",
                                     "F04 - Lib16 - T2 - Growth on M9 - 0.5 ug/ml TMP",
                                     "F05 - Lib16 - T2 - Growth on M9 - 1.0 ug/ml TMP",
                                     "F06 - Lib16 - T2 - Growth on M9 - 10.0 ug/ml TMP",
                                     "F07 - Lib16 - T2 - Growth on M9 - 50.0 ug/ml TMP",
                                     "F08 - Lib16 - T2 - Growth on M9 - 200.0 ug/ml TMP",
                                     "G04 - Lib16 - T3 - Growth on M9 - 0.0 ug/ml TMP",
                                     "G05 - Lib16 - T3 - Growth on M9 - 0.058 ug/ml TMP",
                                     "G06 - Lib16 - T3 - Growth on M9 - 0.5 ug/ml TMP",
                                     "G07 - Lib16 - T3 - Growth on M9 - 1.0 ug/ml TMP",
                                     "G08 - Lib16 - T3 - Growth on M9 - 10.0 ug/ml TMP",
                                     "G09 - Lib16 - T3 - Growth on M9 - 50.0 ug/ml TMP",
                                     "G10 - Lib16 - T3 - Growth on M9 - 200.0 ug/ml TMP",
                                     "H06 - Lib16 - T4 - Growth on M9 - 0.0 ug/ml TMP",
                                     "H07 - Lib16 - T4 - Growth on M9 - 0.058 ug/ml TMP",
                                     "H08 - Lib16 - T4 - Growth on M9 - 0.5 ug/ml TMP",
                                     "H09 - Lib16 - T4 - Growth on M9 - 1.0 ug/ml TMP",
                                     "H10 - Lib16 - T4 - Growth on M9 - 10.0 ug/ml TMP",
                                     "H11 - Lib16 - T4 - Growth on M9 - 50.0 ug/ml TMP",
                                     "H12 - Lib16 - T4 - Growth on M9 - 200.0 ug/ml TMP"))

# Write dataframe to .csv table
write.csv(norm_vals_16, 'novaseq_basic_stats_Lib16.csv', row.names = FALSE, quote=FALSE)
```

<font color="green">**Table of Summarized Count Data for Sequences and Barcodes (BC)**</font>
```{r echo=FALSE}
# Display dataframe with kable formatting
norm_vals_16_col1 <- norm_vals_16[,c(4,1,2,3)]

knitr::kable(norm_vals_16_col1,
  col.names = c('Sample Conditions', 'Sample ID', 'Seq Counts', 'BC Counts'),
  align = "lccl",
  format.args = list(big.mark = ","))
```

## Merged Data
Merge the count data from each library for each sampling condition by barcode for downstream analysis. 
```{r}
# Library 15
BCs15 <- D01 %>%
  full_join(D03,by="BC") %>%
  full_join(D05,by="BC") %>%
  full_join(D06,by="BC") %>%
  full_join(D07,by="BC") %>%
  full_join(D08,by="BC") %>%
  full_join(D09,by="BC") %>%
  full_join(D10,by="BC") %>%
  full_join(D11,by="BC") %>%
  full_join(E07,by="BC") %>%
  full_join(E08,by="BC") %>%
  full_join(E09,by="BC") %>%
  full_join(E10,by="BC") %>%
  full_join(E11,by="BC") %>%
  full_join(E12,by="BC") %>%
  full_join(F01,by="BC") %>%
  full_join(F09,by="BC") %>%
  full_join(F10,by="BC") %>%
  full_join(F11,by="BC") %>%
  full_join(F12,by="BC") %>%
  full_join(G01,by="BC") %>%
  full_join(G02,by="BC") %>%
  full_join(G03,by="BC") %>%
  full_join(G11,by="BC") %>%
  full_join(G12,by="BC") %>%
  full_join(H01,by="BC") %>%
  full_join(H02,by="BC") %>%
  full_join(H03,by="BC") %>%
  full_join(H04,by="BC") %>%
  full_join(H05,by="BC")

# Library 16
BCs16 <- D02 %>%
  full_join(D04,by="BC") %>%
  full_join(D12,by="BC") %>%
  full_join(E01,by="BC") %>%
  full_join(E02,by="BC") %>%
  full_join(E03,by="BC") %>%
  full_join(E04,by="BC") %>%
  full_join(E05,by="BC") %>%
  full_join(E06,by="BC") %>%
  full_join(F02,by="BC") %>%
  full_join(F03,by="BC") %>%
  full_join(F04,by="BC") %>%
  full_join(F05,by="BC") %>%
  full_join(F06,by="BC") %>%
  full_join(F07,by="BC") %>%
  full_join(F08,by="BC") %>%
  full_join(G04,by="BC") %>%
  full_join(G05,by="BC") %>%
  full_join(G06,by="BC") %>%
  full_join(G07,by="BC") %>%
  full_join(G08,by="BC") %>%
  full_join(G09,by="BC") %>%
  full_join(G10,by="BC") %>%
  full_join(H06,by="BC") %>%
  full_join(H07,by="BC") %>%
  full_join(H08,by="BC") %>%
  full_join(H09,by="BC") %>%
  full_join(H10,by="BC") %>%
  full_join(H11,by="BC") %>%
  full_join(H12,by="BC")
```

<font color="green">**Preview of BCs count data with N/A for blank values**</font>
```{r echo=FALSE}
head(BCs15)
head(BCs16)
```

Convert all "NA" values to "0" in the dataframe:
```{r}
# Library 15
BCs15$D01[is.na(BCs15$D01)] <- 0
BCs15$D03[is.na(BCs15$D03)] <- 0
BCs15$D05[is.na(BCs15$D05)] <- 0
BCs15$D06[is.na(BCs15$D06)] <- 0
BCs15$D07[is.na(BCs15$D07)] <- 0
BCs15$D08[is.na(BCs15$D08)] <- 0
BCs15$D09[is.na(BCs15$D09)] <- 0
BCs15$D10[is.na(BCs15$D10)] <- 0
BCs15$D11[is.na(BCs15$D11)] <- 0
BCs15$E07[is.na(BCs15$E07)] <- 0
BCs15$E08[is.na(BCs15$E08)] <- 0
BCs15$E09[is.na(BCs15$E09)] <- 0
BCs15$E10[is.na(BCs15$E10)] <- 0
BCs15$E11[is.na(BCs15$E11)] <- 0
BCs15$E12[is.na(BCs15$E12)] <- 0
BCs15$F01[is.na(BCs15$F01)] <- 0
BCs15$F09[is.na(BCs15$F09)] <- 0
BCs15$F10[is.na(BCs15$F10)] <- 0
BCs15$F11[is.na(BCs15$F11)] <- 0
BCs15$F12[is.na(BCs15$F12)] <- 0
BCs15$G01[is.na(BCs15$G01)] <- 0
BCs15$G02[is.na(BCs15$G02)] <- 0
BCs15$G03[is.na(BCs15$G03)] <- 0
BCs15$G11[is.na(BCs15$G11)] <- 0
BCs15$G12[is.na(BCs15$G12)] <- 0
BCs15$H01[is.na(BCs15$H01)] <- 0
BCs15$H02[is.na(BCs15$H02)] <- 0
BCs15$H03[is.na(BCs15$H03)] <- 0
BCs15$H04[is.na(BCs15$H04)] <- 0
BCs15$H05[is.na(BCs15$H05)] <- 0

# Library 16
BCs16$D02[is.na(BCs16$D02)] <- 0
BCs16$D04[is.na(BCs16$D04)] <- 0
BCs16$D12[is.na(BCs16$D12)] <- 0
BCs16$E01[is.na(BCs16$E01)] <- 0
BCs16$E02[is.na(BCs16$E02)] <- 0
BCs16$E03[is.na(BCs16$E03)] <- 0
BCs16$E04[is.na(BCs16$E04)] <- 0
BCs16$E05[is.na(BCs16$E05)] <- 0
BCs16$E06[is.na(BCs16$E06)] <- 0
BCs16$F02[is.na(BCs16$F02)] <- 0
BCs16$F03[is.na(BCs16$F03)] <- 0
BCs16$F04[is.na(BCs16$F04)] <- 0
BCs16$F05[is.na(BCs16$F05)] <- 0
BCs16$F06[is.na(BCs16$F06)] <- 0
BCs16$F07[is.na(BCs16$F07)] <- 0
BCs16$F08[is.na(BCs16$F08)] <- 0
BCs16$G04[is.na(BCs16$G04)] <- 0
BCs16$G05[is.na(BCs16$G05)] <- 0
BCs16$G06[is.na(BCs16$G06)] <- 0
BCs16$G07[is.na(BCs16$G07)] <- 0
BCs16$G08[is.na(BCs16$G08)] <- 0
BCs16$G09[is.na(BCs16$G09)] <- 0
BCs16$G10[is.na(BCs16$G10)] <- 0
BCs16$H06[is.na(BCs16$H06)] <- 0
BCs16$H07[is.na(BCs16$H07)] <- 0
BCs16$H08[is.na(BCs16$H08)] <- 0
BCs16$H09[is.na(BCs16$H09)] <- 0
BCs16$H10[is.na(BCs16$H10)] <- 0
BCs16$H11[is.na(BCs16$H11)] <- 0
BCs16$H12[is.na(BCs16$H12)] <- 0
```

<font color="green">**Preview of BCs count data with zeros ("0") for blank values for both libraries**</font>
```{r echo=FALSE}
head(BCs15)
head(BCs16)
```

```{r class.output="goodCode"}
# Count the number of unique barcodes for Library 15
BCs15.count <- length(unique(BCs15$BC))
format(BCs15.count, big.mark = ",")

# Count the number of unique barcodes for Library 16
BCs16.count <- length(unique(BCs16$BC))
format(BCs16.count, big.mark = ",")
```

## Combine Libraries
Combine both libraries based on shared BCs:
```{r class.output="sharedCode"}
#Use 'merge' function to merge both libraries based on shared BCs
BCs_shared <- merge(BCs15, BCs16, by = "BC", all = FALSE)

# Count the number of unique barcodes shared between libraries
BCs_shared.count <- length(unique(BCs_shared$BC))
format(BCs_all.count, big.mark = ",")
```

```{r echo=FALSE}
head(BCs_shared)
```

**Summarize the Shared Libraries**
```{r}
## Shared Libraries

# Merge data objects together into single dataframe
norm_vals_shared <- data.frame(sample=c("D01","D02","D03","D04",
                                     "D05","D06","D07","D08","D09","D10","D11",
                                     "D12","E01","E02","E03","E04","E05","E06",
                                     "E07","E08","E09","E10","E11","E12","F01",
                                     "F02","F03","F04","F05","F06","F07","F08",
                                     "F09","F10","F11","F12","G01","G02","G03",
                                     "G04","G05","G06","G07","G08","G09","G10",
                                     "G11","G12","H01","H02","H03","H04","H05",
                                     "H06","H07","H08","H09","H10","H11","H12"),
                        counts=c(sum(BCs_shared$D01),sum(BCs_shared$D02),sum(BCs_shared$D03),sum(BCs_shared$D04),
                                 sum(BCs_shared$D05),sum(BCs_shared$D06),sum(BCs_shared$D07),sum(BCs_shared$D08),
                                 sum(BCs_shared$D09),sum(BCs_shared$D10),sum(BCs_shared$D11),sum(BCs_shared$D12),
                                 sum(BCs_shared$E01),sum(BCs_shared$E02),sum(BCs_shared$E03),sum(BCs_shared$E04),
                                 sum(BCs_shared$E05),sum(BCs_shared$E06),sum(BCs_shared$E07),sum(BCs_shared$E08),
                                 sum(BCs_shared$E09),sum(BCs_shared$E10),sum(BCs_shared$E11),sum(BCs_shared$E12),
                                 sum(BCs_shared$F01),sum(BCs_shared$F02),sum(BCs_shared$F03),sum(BCs_shared$F04),
                                 sum(BCs_shared$F05),sum(BCs_shared$F06),sum(BCs_shared$F07),sum(BCs_shared$F08),
                                 sum(BCs_shared$F09),sum(BCs_shared$F10),sum(BCs_shared$F11),sum(BCs_shared$F12),
                                 sum(BCs_shared$G01),sum(BCs_shared$G02),sum(BCs_shared$G03),sum(BCs_shared$G04),
                                 sum(BCs_shared$G05),sum(BCs_shared$G06),sum(BCs_shared$G07),sum(BCs_shared$G08),
                                 sum(BCs_shared$G09),sum(BCs_shared$G10),sum(BCs_shared$G11),sum(BCs_shared$G12),
                                 sum(BCs_shared$H01),sum(BCs_shared$H02),sum(BCs_shared$H03),sum(BCs_shared$H04),
                                 sum(BCs_shared$H05),sum(BCs_shared$H06),sum(BCs_shared$H07),sum(BCs_shared$H08),
                                 sum(BCs_shared$H09),sum(BCs_shared$H10),sum(BCs_shared$H11),sum(BCs_shared$H12)),
                        numBCs=c(length(BCs_shared$D01),length(BCs_shared$D02),length(BCs_shared$D03),
                                 length(BCs_shared$D04),length(BCs_shared$D05),length(BCs_shared$D06),
                                 length(BCs_shared$D07),length(BCs_shared$D08),length(BCs_shared$D09),
                                 length(BCs_shared$D10),length(BCs_shared$D11),length(BCs_shared$D12),
                                 length(BCs_shared$E01),length(BCs_shared$E02),length(BCs_shared$E03),
                                 length(BCs_shared$E04),length(BCs_shared$E05),length(BCs_shared$E06),
                                 length(BCs_shared$E07),length(BCs_shared$E08),length(BCs_shared$E09),
                                 length(BCs_shared$E10),length(BCs_shared$E11),length(BCs_shared$E12),
                                 length(BCs_shared$F01),length(BCs_shared$F02),length(BCs_shared$F03),
                                 length(BCs_shared$F04),length(BCs_shared$F05),length(BCs_shared$F06),
                                 length(BCs_shared$F07),length(BCs_shared$F08),length(BCs_shared$F09),
                                 length(BCs_shared$F10),length(BCs_shared$F11),length(BCs_shared$F12),
                                 length(BCs_shared$G01),length(BCs_shared$G02),length(BCs_shared$G03),
                                 length(BCs_shared$G04),length(BCs_shared$G05),length(BCs_shared$G06),
                                 length(BCs_shared$G07),length(BCs_shared$G08),length(BCs_shared$G09),
                                 length(BCs_shared$G10),length(BCs_shared$G11),length(BCs_shared$G12),
                                 length(BCs_shared$H01),length(BCs_shared$H02),length(BCs_shared$H03),
                                 length(BCs_shared$H04),length(BCs_shared$H05),length(BCs_shared$H06),
                                 length(BCs_shared$H07),length(BCs_shared$H08),length(BCs_shared$H09),
                                 length(BCs_shared$H10),length(BCs_shared$H11),length(BCs_shared$H12)),
                        sampledesc=c("D01 - Lib15 - T0 - Overnight Growth on LB",
                                     "D02 - Lib16 - T0 - Overnight Growth on LB",
                                     "D03 - Lib15 - T0 - Growth on M9 - Full Supplement",
                                     "D04 - Lib16 - T0 - Growth on M9 - Full Supplement",
                                     "D05 - Lib15 - T1 - Growth on M9 - 0.0 ug/ml TMP",
                                     "D06 - Lib15 - T1 - Growth on M9 - 0.058 ug/ml TMP",
                                     "D07 - Lib15 - T1 - Growth on M9 - 0.5 ug/ml TMP",
                                     "D08 - Lib15 - T1 - Growth on M9 - 1.0 ug/ml TMP",
                                     "D09 - Lib15 - T1 - Growth on M9 - 10.0 ug/ml TMP",
                                     "D10 - Lib15 - T1 - Growth on M9 - 50.0 ug/ml TMP",
                                     "D11 - Lib15 - T1 - Growth on M9 - 200.0 ug/ml TMP",
                                     "D12 - Lib16 - T1 - Growth on M9 - 0.0 ug/ml TMP",
                                     "E01 - Lib16 - T1 - Growth on M9 - 0.058 ug/ml TMP",
                                     "E02 - Lib16 - T1 - Growth on M9 - 0.5 ug/ml TMP",
                                     "E03 - Lib16 - T1 - Growth on M9 - 1.0 ug/ml TMP",
                                     "E04 - Lib16 - T1 - Growth on M9 - 10.0 ug/ml TMP",
                                     "E05 - Lib16 - T1 - Growth on M9 - 50.0 ug/ml TMP",
                                     "E06 - Lib16 - T1 - Growth on M9 - 200.0 ug/ml TMP",
                                     "E07 - Lib15 - T2 - Growth on M9 - 0.0 ug/ml TMP",
                                     "E08 - Lib15 - T2 - Growth on M9 - 0.058 ug/ml TMP",
                                     "E09 - Lib15 - T2 - Growth on M9 - 0.5 ug/ml TMP",
                                     "E10 - Lib15 - T2 - Growth on M9 - 1.0 ug/ml TMP",
                                     "E11 - Lib15 - T2 - Growth on M9 - 10.0 ug/ml TMP",
                                     "E12 - Lib15 - T2 - Growth on M9 - 50.0 ug/ml TMP",
                                     "F01 - Lib15 - T2 - Growth on M9 - 200.0 ug/ml TMP",
                                     "F02 - Lib16 - T2 - Growth on M9 - 0.0 ug/ml TMP",
                                     "F03 - Lib16 - T2 - Growth on M9 - 0.058 ug/ml TMP",
                                     "F04 - Lib16 - T2 - Growth on M9 - 0.5 ug/ml TMP",
                                     "F05 - Lib16 - T2 - Growth on M9 - 1.0 ug/ml TMP",
                                     "F06 - Lib16 - T2 - Growth on M9 - 10.0 ug/ml TMP",
                                     "F07 - Lib16 - T2 - Growth on M9 - 50.0 ug/ml TMP",
                                     "F08 - Lib16 - T2 - Growth on M9 - 200.0 ug/ml TMP",
                                     "F09 - Lib15 - T3 - Growth on M9 - 0.0 ug/ml TMP",
                                     "F10 - Lib15 - T3 - Growth on M9 - 0.058 ug/ml TMP",
                                     "F11 - Lib15 - T3 - Growth on M9 - 0.5 ug/ml TMP",
                                     "F12 - Lib15 - T3 - Growth on M9 - 1.0 ug/ml TMP",
                                     "G01 - Lib15 - T3 - Growth on M9 - 10.0 ug/ml TMP",
                                     "G02 - Lib15 - T3 - Growth on M9 - 50.0 ug/ml TMP",
                                     "G03 - Lib15 - T3 - Growth on M9 - 200.0 ug/ml TMP",
                                     "G04 - Lib16 - T3 - Growth on M9 - 0.0 ug/ml TMP",
                                     "G05 - Lib16 - T3 - Growth on M9 - 0.058 ug/ml TMP",
                                     "G06 - Lib16 - T3 - Growth on M9 - 0.5 ug/ml TMP",
                                     "G07 - Lib16 - T3 - Growth on M9 - 1.0 ug/ml TMP",
                                     "G08 - Lib16 - T3 - Growth on M9 - 10.0 ug/ml TMP",
                                     "G09 - Lib16 - T3 - Growth on M9 - 50.0 ug/ml TMP",
                                     "G10 - Lib16 - T3 - Growth on M9 - 200.0 ug/ml TMP",
                                     "G11 - Lib15 - T4 - Growth on M9 - 0.0 ug/ml TMP",
                                     "G12 - Lib15 - T4 - Growth on M9 - 0.058 ug/ml TMP",
                                     "H01 - Lib15 - T4 - Growth on M9 - 0.5 ug/ml TMP",
                                     "H02 - Lib15 - T4 - Growth on M9 - 1.0 ug/ml TMP",
                                     "H03 - Lib15 - T4 - Growth on M9 - 10.0 ug/ml TMP",
                                     "H04 - Lib15 - T4 - Growth on M9 - 50.0 ug/ml TMP",
                                     "H05 - Lib15 - T4 - Growth on M9 - 200.0 ug/ml TMP",
                                     "H06 - Lib16 - T4 - Growth on M9 - 0.0 ug/ml TMP",
                                     "H07 - Lib16 - T4 - Growth on M9 - 0.058 ug/ml TMP",
                                     "H08 - Lib16 - T4 - Growth on M9 - 0.5 ug/ml TMP",
                                     "H09 - Lib16 - T4 - Growth on M9 - 1.0 ug/ml TMP",
                                     "H10 - Lib16 - T4 - Growth on M9 - 10.0 ug/ml TMP",
                                     "H11 - Lib16 - T4 - Growth on M9 - 50.0 ug/ml TMP",
                                     "H12 - Lib16 - T4 - Growth on M9 - 200.0 ug/ml TMP"))

# Write dataframe to .csv table
write.csv(norm_vals_shared, 'novaseq_basic_stats_Lib15_Lib16_shared_BCs.csv', row.names = FALSE, quote=FALSE)
```

<font color="green">**Table of Summarized Count Data for Sequences shared by Barcodes (BC) between both Libraries (normlized)**</font>
```{r echo=FALSE}
# Display dataframe with kable formatting
norm_vals_shared_col1 <- norm_vals_shared[,c(4,1,2,3)]

knitr::kable(norm_vals_shared_col1,
  col.names = c('Sample Conditions', 'Sample ID', 'Seq Counts', 'Shared BCs'),
  align = "lccc",
  format.args = list(big.mark = ","))
```

```{r}
#Use 'merge' function to merge both libraries based on shared BCs
norm_vals_all <- bind_rows(norm_vals_15, norm_vals_16)
norm_vals_all <- arrange(norm_vals_all, sample)
```

```{r echo=FALSE}
# Display dataframe with kable formatting
norm_vals_all_col1 <- norm_vals_all[,c(4,1,2,3)]

knitr::kable(norm_vals_all_col1,
  col.names = c('Sample Conditions', 'Sample ID', 'Seq Counts', 'BC Count'),
  align = "lccc",
  format.args = list(big.mark = ","))
```











## Basic Checks

With the data from both libraries imported and merged, we can perform some basic checks to determine everything has been incorporated properly.

### Barcodes

How many of our barcodes (from count files) are in the mapping files for each library?
```{r class.output="goodCode"}
# Library 15
bc15_map <- inner_join(BCinfo15,BCs15,by="BC") %>%
  nrow()
format(bc15_map, big.mark = ",")
  
# Library 16
bc16_map <- inner_join(BCinfo16,BCs16,by="BC") %>%
  nrow()
format(bc16_map, big.mark = ",")
```

How many of our barcodes are in the mapping file for the shared libraries?
```{r class.output="sharedCode"}
# Shared Libraries
BCshared_map <- inner_join(BCinfo_all,BCs_shared,by="BC") %>%
  nrow()
format(BCshared_map, big.mark = ",")
```

How many of our barcodes are associated with perfects (perfect sequence match with a designed homolog) for each unique library?
```{r class.output="goodCode"}
# Library 15
bc15_perf <- inner_join(BCinfo15,BCs15,by="BC") %>%
  filter(mutations==0) %>%
  nrow()
format(bc15_perf, big.mark = ",")

# Library 16
bc16_perf <- inner_join(BCinfo16,BCs16,by="BC") %>%
  filter(mutations==0) %>%
  nrow()
format(bc16_perf, big.mark = ",")
```

How many of our barcodes are associated with perfects (perfect sequence match with a designed homolog) between the shared libraries?
```{r class.output="sharedCode"}
# Shared Libraries
BCshared_perf <- inner_join(BCinfo_all,BCs_shared,by="BC") %>%
  filter(mutations==0) %>%
  nrow()
format(BCshared_perf, big.mark = ",")
```

How many unique perfects did we recover from all sampling conditions for each library?
```{r class.output="goodCode"}
# Library 15
perf_uniq_15 <- inner_join(BCinfo15,BCs15,by="BC") %>%
  filter(mutations==0) %>%
  select(mutID) %>%
  distinct() %>%
  nrow()
format(perf_uniq_15, big.mark = ",")

# Library 16
perf_uniq_16 <- inner_join(BCinfo16,BCs16,by="BC") %>%
  filter(mutations==0) %>%
  select(mutID) %>%
  distinct() %>%
  nrow()
format(perf_uniq_16, big.mark = ",")
```

How many unique perfects did we recover from all sampling conditions between the shared libraries?
```{r class.output="sharedCode"}
# Shared Libraries
perf_uniq_shared <- inner_join(BCinfo_all,BCs_shared,by="BC") %>%
  filter(mutations==0) %>%
  select(mutID) %>%
  distinct() %>%
  nrow()
format(perf_uniq_shared, big.mark = ",")
```

How many unique mutants did we recover from all sampling conditions for unique libraries?
```{r class.output="goodCode"}
# Library 15
mutant_15 <- inner_join(BCinfo15,BCs15,by="BC") %>%
  filter(mutations!=0) %>%
  select(mutID) %>%
  distinct() %>%
  nrow()
format(mutant_15, big.mark = ",")

# Library 16
mutant_16 <- inner_join(BCinfo16,BCs16,by="BC") %>%
  filter(mutations!=0) %>%
  select(mutID) %>%
  distinct() %>%
  nrow()
format(mutant_16, big.mark = ",")
```

How many unique mutants did we recover from all sampling conditions between shared libraries?
```{r class.output="sharedCode"}
# Shared Libraries
mutant_shared <- inner_join(BCinfo_all,BCs_shared,by="BC") %>%
  filter(mutations!=0) %>%
  select(mutID) %>%
  distinct() %>%
  nrow()
format(mutant_shared, big.mark = ",")
```

#-------------------------------------------------------------------------------------------

### Normalization

Count data needs to be normalized across sampling conditions prior to making any downstream comparisons. Here, we scale the sequence counts against the "T0-Overnight Growth on LB" sampling condition. This represents "D01" for Lib15 and "D02" for Lib16.

#### Scaling Values by Condition

**Lib15 Scaling**
```{r}
# Calculating the scaling factor for each sampling condition based on D01 (overnight growth on LB) as the standard
norm_scaling_15 <- rep.int(norm_vals_15$counts[1],nrow(norm_vals_15))/norm_vals_15$counts

# Add the scaling values as a column to the "norm_vals_15" dataframe
norm_vals_15$scaling <- norm_scaling_15
```

<font color="green">**Table of Summarized Count Data including the Scaling Value for each Sample Condition**</font>
```{r echo=FALSE}
# Display dataframe with kable formatting
norm_vals_15_col2 <- norm_vals_15[,c(4,1,2,3,5)]

knitr::kable(norm_vals_15_col2,
  col.names = c('Sample Conditions', 'Sample ID', 'Seq Counts', 'BC Counts', 'Scaling'),
  align = "llccc",
  format.args = list(big.mark = ","),
  digits = 2)
```

**Lib16 Scaling**
```{r}
# Calculate scaling factor for each sampling condition based on D02 (overnight growth on LB) as the standard
norm_scaling_16 <- rep.int(norm_vals_16$counts[1],nrow(norm_vals_16))/norm_vals_16$counts

# Add the scaling values as a column to the "norm_vals_16" dataframe
norm_vals_16$scaling <- norm_scaling_16
```

<font color="green">**Table of Summarized Count Data including the Scaling Value for each Sample Condition**</font>
```{r echo=FALSE}
# Display dataframe with kable formatting
norm_vals_16_col2 <- norm_vals_16[,c(4,1,2,3,5)]

knitr::kable(norm_vals_16_col2,
  col.names = c('Sample Conditions', 'Sample ID', 'Seq Counts', 'BC Counts', 'Scaling'),
  align = "llccc",
  format.args = list(big.mark = ","),
  digits = 2)
```

#### Scaling Counts by Barcode
<font color="green">**The following table displays a preview of the original count values by unique barcode:**</font>

**Lib15 BC Scaling**
```{r echo=FALSE}
# Display dataframe with kable formatting
BCs15_head <- head(BCs15)

knitr::kable(BCs15_head,
  col.names = c('BC', 'D01', 'D03', 'D05', 'D06', 'D07', 'D08', 'D09', 'D10', 'D11', 
                'E07', 'E08', 'E09', 'E10', 'E11', 'E12', 'F01', 
                'F09', 'F10', 'F11', 'F12', 'G01', 'G02', 'G03', 
                'G11', 'G12', 'H01', 'H02', 'H03', 'H04', 'H05'),
  align = "lccccccccccc",
  format.args = list(big.mark = ","),
  digits = 0)
```

For each sample, make a normalized counts value based on sequencing depth using the scaling values derived for each sampling condition in the table above. 
```{r}
# Make new columns for the normalized counts in the BCs object:
oldnames15 <- colnames(BCs15)
newnamesfrac15 = c('D01n', 'D03n', 
                   'D05n', 'D06n', 'D07n', 'D08n', 'D09n', 'D10n', 'D11n', 
                   'E07n', 'E08n', 'E09n', 'E10n', 'E11n', 'E12n', 'F01n', 
                   'F09n', 'F10n', 'F11n', 'F12n', 'G01n', 'G02n', 'G03n', 
                   'G11n', 'G12n', 'H01n', 'H02n', 'H03n', 'H04n', 'H05n')

# Add the new columns to the BCs object based on the number of original columns for sampling conditions:
numcol_15 <- ncol(BCs15)
D01index = which( colnames(BCs15)=="D01" )

# For each dilution, re-scale the number of reads by relative sequencing depth of the sample using the scaling values in the norm_vals object:
for (i in 1:30){
  BCs15[,numcol_15+i] <- BCs15[,D01index+i-1]*as.numeric(norm_vals_15$scaling[i])
}

# Add the new column names to the BCs object:
colnames(BCs15) <- c(oldnames15,newnamesfrac15)

# Clean up objects from the global environment
#rm(oldnames15,newnamesfrac15,numcol15,D01index,i)
```

<font color="green">**The following table displays a preview of the normalized count values by unique barcode:**</font>
```{r echo=FALSE}
BCs_norm_15 <- BCs15[,c(1,32:61)]

# Display dataframe with kable formatting
BCs_norm_head_15 <- head(BCs_norm_15)

knitr::kable(BCs_norm_head_15,
  col.names = c('BC', 'D01', 'D03', 'D05', 'D06', 'D07', 'D08', 'D09', 'D10', 'D11', 
                'E07', 'E08', 'E09', 'E10', 'E11', 'E12', 'F01', 
                'F09', 'F10', 'F11', 'F12', 'G01', 'G02', 'G03', 
                'G11', 'G12', 'H01', 'H02', 'H03', 'H04', 'H05'),
  align = "lccccccccccc",
  format.args = list(big.mark = ","),
  digits = 0)
```

**Lib16 BC Scaling**
```{r echo=FALSE}
# Display dataframe with kable formatting
BCs16_head <- head(BCs16)

knitr::kable(BCs16_head,
  col.names = c('BC', 'D02', 'D04', 'D12', 'E01', 'E02', 'E03', 'E04', 'E05', 'E06', 
                'F02', 'F03', 'F04', 'F05', 'F06', 'F07', 'F08', 
                'G04', 'G05', 'G06', 'G07', 'G08', 'G09', 'G10', 
                'H06', 'H07', 'H08', 'H09', 'H10', 'H11', 'H12'),
  align = "lccccccccccc",
  format.args = list(big.mark = ","),
  digits = 0)
```

For each sample, make a normalized counts value based on sequencing depth using the scaling values derived for each sampling condition in the table above. 
```{r}
# Make new columns for the normalized counts in the BCs object:
oldnames16 <- colnames(BCs16)
newnamesfrac16 = c('D02n', 'D04n', 
                   'D12n', 'E01n', 'E02n', 'E03n', 'E04n', 'E05n', 'E06n', 
                   'F02n', 'F03n', 'F04n', 'F05n', 'F06n', 'F07n', 'F08n', 
                   'G04n', 'G05n', 'G06n', 'G07n', 'G08n', 'G09n', 'G10n', 
                   'H06n', 'H07n', 'H08n', 'H09n', 'H10n', 'H11n', 'H12n')

# Add the new columns to the BCs object based on the number of original columns for sampling conditions:
numcol_16 <- ncol(BCs16)
D02index = which( colnames(BCs16)=="D02" )

# For each dilution, re-scale the number of reads by relative sequencing depth of the sample using the scaling values in the norm_vals object:
for (i in 1:30){
  BCs16[,numcol_16+i] <- BCs16[,D02index+i-1]*as.numeric(norm_vals_16$scaling[i])
}

# Add the new column names to the BCs object:
colnames(BCs16) <- c(oldnames16,newnamesfrac16)

# Clean up objects from the global environment
#rm(oldnames16,newnamesfrac16,numcol16,D02index,i)
```

<font color="green">**The following table displays a preview of the normalized count values by unique barcode:**</font>
```{r echo=FALSE}
BCs_norm_16 <- BCs16[,c(1,32:61)]

# Display dataframe with kable formatting
BCs_norm_head_16 <- head(BCs_norm_16)

knitr::kable(BCs_norm_head_16,
  col.names = c('BC', 'D02', 'D04', 'D12', 'E01', 'E02', 'E03', 'E04', 'E05', 'E06', 
                'F02', 'F03', 'F04', 'F05', 'F06', 'F07', 'F08', 
                'G04', 'G05', 'G06', 'G07', 'G08', 'G09', 'G10', 
                'H06', 'H07', 'H08', 'H09', 'H10', 'H11', 'H12'),
  align = "lccccccccccc",
  format.args = list(big.mark = ","),
  digits = 0)
```

### Log2-Fold Change

Determine the log2-fold changes for each barcode relative to the M9-Supplementation condition (D03).
```{r}
pseudocount = 1

#Library 15
BCs15 <- BCs15 %>%
  mutate(D03D01fc=log2(D03n+pseudocount)-log2(D01n+pseudocount),
         D05D03fc=log2(D05n+pseudocount)-log2(D03n+pseudocount),
         D06D03fc=log2(D06n+pseudocount)-log2(D03n+pseudocount),
         D07D03fc=log2(D07n+pseudocount)-log2(D03n+pseudocount),
         D08D03fc=log2(D08n+pseudocount)-log2(D03n+pseudocount),
         D09D03fc=log2(D09n+pseudocount)-log2(D03n+pseudocount),
         D10D03fc=log2(D10n+pseudocount)-log2(D03n+pseudocount),
         D11D03fc=log2(D11n+pseudocount)-log2(D03n+pseudocount),
         E07D03fc=log2(E07n+pseudocount)-log2(D03n+pseudocount),
         E08D03fc=log2(E08n+pseudocount)-log2(D03n+pseudocount),
         E09D03fc=log2(E09n+pseudocount)-log2(D03n+pseudocount),
         E10D03fc=log2(E10n+pseudocount)-log2(D03n+pseudocount),
         E11D03fc=log2(E11n+pseudocount)-log2(D03n+pseudocount),
         E12D03fc=log2(E12n+pseudocount)-log2(D03n+pseudocount),
         F01D03fc=log2(F01n+pseudocount)-log2(D03n+pseudocount),
         F09D03fc=log2(F09n+pseudocount)-log2(D03n+pseudocount),
         F10D03fc=log2(F10n+pseudocount)-log2(D03n+pseudocount),
         F11D03fc=log2(F11n+pseudocount)-log2(D03n+pseudocount),
         F12D03fc=log2(F12n+pseudocount)-log2(D03n+pseudocount),
         G01D03fc=log2(G01n+pseudocount)-log2(D03n+pseudocount),
         G02D03fc=log2(G02n+pseudocount)-log2(D03n+pseudocount),
         G03D03fc=log2(G03n+pseudocount)-log2(D03n+pseudocount),
         G11D03fc=log2(G11n+pseudocount)-log2(D03n+pseudocount),
         G12D03fc=log2(G12n+pseudocount)-log2(D03n+pseudocount),
         H01D03fc=log2(H01n+pseudocount)-log2(D03n+pseudocount),
         H02D03fc=log2(H02n+pseudocount)-log2(D03n+pseudocount),
         H03D03fc=log2(H03n+pseudocount)-log2(D03n+pseudocount),
         H04D03fc=log2(H04n+pseudocount)-log2(D03n+pseudocount),
         H05D03fc=log2(H05n+pseudocount)-log2(D03n+pseudocount))

#Library 16
BCs16 <- BCs16 %>%
  mutate(D04D02fc=log2(D04n+pseudocount)-log2(D02n+pseudocount),
         D12D04fc=log2(D12n+pseudocount)-log2(D04n+pseudocount),
         E01D04fc=log2(E01n+pseudocount)-log2(D04n+pseudocount),
         E02D04fc=log2(E02n+pseudocount)-log2(D04n+pseudocount),
         E03D04fc=log2(E03n+pseudocount)-log2(D04n+pseudocount),
         E04D04fc=log2(E04n+pseudocount)-log2(D04n+pseudocount),
         E05D04fc=log2(E05n+pseudocount)-log2(D04n+pseudocount),
         E06D04fc=log2(E06n+pseudocount)-log2(D04n+pseudocount),
         F02D04fc=log2(F02n+pseudocount)-log2(D04n+pseudocount),
         F03D04fc=log2(F03n+pseudocount)-log2(D04n+pseudocount),
         F04D04fc=log2(F04n+pseudocount)-log2(D04n+pseudocount),
         F05D04fc=log2(F05n+pseudocount)-log2(D04n+pseudocount),
         F06D04fc=log2(F06n+pseudocount)-log2(D04n+pseudocount),
         F07D04fc=log2(F07n+pseudocount)-log2(D04n+pseudocount),
         F08D04fc=log2(F08n+pseudocount)-log2(D04n+pseudocount),
         G04D04fc=log2(G04n+pseudocount)-log2(D04n+pseudocount),
         G05D04fc=log2(G05n+pseudocount)-log2(D04n+pseudocount),
         G06D04fc=log2(G06n+pseudocount)-log2(D04n+pseudocount),
         G07D04fc=log2(G07n+pseudocount)-log2(D04n+pseudocount),
         G08D04fc=log2(G08n+pseudocount)-log2(D04n+pseudocount),
         G09D04fc=log2(G09n+pseudocount)-log2(D04n+pseudocount),
         G10D04fc=log2(G10n+pseudocount)-log2(D04n+pseudocount),
         H06D04fc=log2(H06n+pseudocount)-log2(D04n+pseudocount),
         H07D04fc=log2(H07n+pseudocount)-log2(D04n+pseudocount),
         H08D04fc=log2(H08n+pseudocount)-log2(D04n+pseudocount),
         H09D04fc=log2(H09n+pseudocount)-log2(D04n+pseudocount),
         H10D04fc=log2(H10n+pseudocount)-log2(D04n+pseudocount),
         H11D04fc=log2(H11n+pseudocount)-log2(D04n+pseudocount),
         H12D04fc=log2(H12n+pseudocount)-log2(D04n+pseudocount))
```

**Lib15 Log2-FC**
<font color="green">**The following table displays a preview of the log2-fold change values between sampling treatments:**</font>
```{r echo=FALSE}
BCs_norm_log_15 <- BCs15[,c(1,62:90)]

# Display dataframe with kable formatting
BCs_norm_log_head_15 <- head(BCs_norm_log_15)

knitr::kable(BCs_norm_log_head_15,
  col.names = c('BC', 'D03-D01fc', 
                'D05-D03fc', 'D06-D03fc', 'D07-D03fc', 'D08-D03fc', 'D09-D03fc', 'D10-D03fc', 'D11-D03fc', 
                'E07-D03fc', 'E08-D03fc', 'E09-D03fc', 'E10-D03fc', 'E11-D03fc', 'E12-D03fc', 'F01-D03fc', 
                'F09-D03fc', 'F10-D03fc', 'F11-D03fc', 'F12-D03fc', 'G01-D03fc', 'G02-D03fc', 'G03-D03fc', 
                'G11-D03fc', 'G12-D03fc', 'H01-D03fc', 'H02-D03fc', 'H03-D03fc', 'H04-D03fc', 'H05-D03fc'),
  align = "lcccccccccc",
  format.args = list(big.mark = ","),
  digits = 0)
```

**Lib16 Log2-FC**
<font color="green">**The following table displays a preview of the log2-fold change values between sampling treatments:**</font>
```{r echo=FALSE}
BCs_norm_log_16 <- BCs16[,c(1,62:90)]

# Display dataframe with kable formatting
BCs_norm_log_head_16 <- head(BCs_norm_log_16)

knitr::kable(BCs_norm_log_head_16,
  col.names = c('BC', 'D04-D02fc', 
                'D12-D04fc', 'E01-D04fc', 'E02-D04fc', 'E03-D04fc', 'E04-D04fc', 'E05-D04fc', 'E06-D04fc', 
                'F02-D04fc', 'F03-D04fc', 'F04-D04fc', 'F05-D04fc', 'F06-D04fc', 'F07-D04fc', 'F08-D04fc', 
                'G04-D04fc', 'G05-D04fc', 'G06-D04fc', 'G07-D04fc', 'G08-D04fc', 'G09-D04fc', 'G10-D04fc', 
                'H06-D04fc', 'H07-D04fc', 'H08-D04fc', 'H09-D04fc', 'H10-D04fc', 'H11-D04fc', 'H12-D04fc'),
  align = "lcccccccccc",
  format.args = list(big.mark = ","),
  digits = 0)
```


### Total Reads & BCs

Calculate the total number of reads and unique barcodes recovered from each sampling condition.
```{r}
# Library 15

# Add the count values, normalized count values, and log2-FC values to the BCinfo object
BCmut_15 <- inner_join(BCinfo15,BCs15,by="BC")

# Sum the total reads recovered from each sampling condition:
norm_vals_15$mappedcounts <- c(sum(BCmut_15$D01),sum(BCmut_15$D03),
                            sum(BCmut_15$D05),sum(BCmut_15$D06),
                            sum(BCmut_15$D07),sum(BCmut_15$D08),
                            sum(BCmut_15$D09),sum(BCmut_15$D10),
                            sum(BCmut_15$D11),sum(BCmut_15$E07),
                            sum(BCmut_15$E08),sum(BCmut_15$E09),
                            sum(BCmut_15$E10),sum(BCmut_15$E11),
                            sum(BCmut_15$E12),sum(BCmut_15$F01),
                            sum(BCmut_15$F09),sum(BCmut_15$F10),
                            sum(BCmut_15$F11),sum(BCmut_15$F12),
                            sum(BCmut_15$G01),sum(BCmut_15$G02),
                            sum(BCmut_15$G03),sum(BCmut_15$G11),
                            sum(BCmut_15$G12),sum(BCmut_15$H01),
                            sum(BCmut_15$H02),sum(BCmut_15$H03),
                            sum(BCmut_15$H04),sum(BCmut_15$H05))

# Determine the number of unique barcodes recovered from each sampling condition:
norm_vals_15$mappedBCs <- c(length(BCmut_15$D01[BCmut_15$D01 != 0]),
                         length(BCmut_15$D03[BCmut_15$D03 != 0]),
                         length(BCmut_15$D05[BCmut_15$D05 != 0]),
                         length(BCmut_15$D06[BCmut_15$D06 != 0]),
                         length(BCmut_15$D07[BCmut_15$D07 != 0]),
                         length(BCmut_15$D08[BCmut_15$D08 != 0]),
                         length(BCmut_15$D09[BCmut_15$D09 != 0]),
                         length(BCmut_15$D10[BCmut_15$D10 != 0]),
                         length(BCmut_15$D11[BCmut_15$D11 != 0]),
                         length(BCmut_15$E07[BCmut_15$E07 != 0]),
                         length(BCmut_15$E08[BCmut_15$E08 != 0]),
                         length(BCmut_15$E09[BCmut_15$E09 != 0]),
                         length(BCmut_15$E10[BCmut_15$E10 != 0]),
                         length(BCmut_15$E11[BCmut_15$E11 != 0]),
                         length(BCmut_15$E12[BCmut_15$E12 != 0]),
                         length(BCmut_15$F01[BCmut_15$F01 != 0]),
                         length(BCmut_15$F09[BCmut_15$F09 != 0]),
                         length(BCmut_15$F10[BCmut_15$F10 != 0]),
                         length(BCmut_15$F11[BCmut_15$F11 != 0]),
                         length(BCmut_15$F12[BCmut_15$F12 != 0]),
                         length(BCmut_15$G01[BCmut_15$G01 != 0]),
                         length(BCmut_15$G02[BCmut_15$G02 != 0]),
                         length(BCmut_15$G03[BCmut_15$G03 != 0]),
                         length(BCmut_15$G11[BCmut_15$G11 != 0]),
                         length(BCmut_15$G12[BCmut_15$G12 != 0]),
                         length(BCmut_15$H01[BCmut_15$H01 != 0]),
                         length(BCmut_15$H02[BCmut_15$H02 != 0]),
                         length(BCmut_15$H03[BCmut_15$H03 != 0]),
                         length(BCmut_15$H04[BCmut_15$H04 != 0]),
                         length(BCmut_15$H05[BCmut_15$H05 != 0]))

# Library 16

# Add the count values, normalized count values, and log2-FC values to the BCinfo object
BCmut_16 <- inner_join(BCinfo16,BCs16,by="BC")

# Sum the total reads recovered from each sampling condition:
norm_vals_16$mappedcounts <- c(sum(BCmut_16$D02),sum(BCmut_16$D04),
                            sum(BCmut_16$D12),sum(BCmut_16$E01),
                            sum(BCmut_16$E02),sum(BCmut_16$E03),
                            sum(BCmut_16$E04),sum(BCmut_16$E05),
                            sum(BCmut_16$E06),sum(BCmut_16$F02),
                            sum(BCmut_16$F03),sum(BCmut_16$F04),
                            sum(BCmut_16$F05),sum(BCmut_16$F06),
                            sum(BCmut_16$F07),sum(BCmut_16$F08),
                            sum(BCmut_16$G04),sum(BCmut_16$G05),
                            sum(BCmut_16$G06),sum(BCmut_16$G07),
                            sum(BCmut_16$G08),sum(BCmut_16$G09),
                            sum(BCmut_16$G10),sum(BCmut_16$H06),
                            sum(BCmut_16$H07),sum(BCmut_16$H08),
                            sum(BCmut_16$H09),sum(BCmut_16$H10),
                            sum(BCmut_16$H11),sum(BCmut_16$H12))

# Determine the number of unique barcodes recovered from each sampling condition:
norm_vals_16$mappedBCs <- c(length(BCmut_16$D02[BCmut_16$D02 != 0]),
                         length(BCmut_16$D04[BCmut_16$D04 != 0]),
                         length(BCmut_16$D12[BCmut_16$D12 != 0]),
                         length(BCmut_16$E01[BCmut_16$E01 != 0]),
                         length(BCmut_16$E02[BCmut_16$E02 != 0]),
                         length(BCmut_16$E03[BCmut_16$E03 != 0]),
                         length(BCmut_16$E04[BCmut_16$E04 != 0]),
                         length(BCmut_16$E05[BCmut_16$E05 != 0]),
                         length(BCmut_16$E06[BCmut_16$E06 != 0]),
                         length(BCmut_16$F02[BCmut_16$F02 != 0]),
                         length(BCmut_16$F03[BCmut_16$F03 != 0]),
                         length(BCmut_16$F04[BCmut_16$F04 != 0]),
                         length(BCmut_16$F05[BCmut_16$F05 != 0]),
                         length(BCmut_16$F06[BCmut_16$F06 != 0]),
                         length(BCmut_16$F07[BCmut_16$F07 != 0]),
                         length(BCmut_16$F08[BCmut_16$F08 != 0]),
                         length(BCmut_16$G04[BCmut_16$G04 != 0]),
                         length(BCmut_16$G05[BCmut_16$G05 != 0]),
                         length(BCmut_16$G06[BCmut_16$G06 != 0]),
                         length(BCmut_16$G07[BCmut_16$G07 != 0]),
                         length(BCmut_16$G08[BCmut_16$G08 != 0]),
                         length(BCmut_16$G09[BCmut_16$G09 != 0]),
                         length(BCmut_16$G10[BCmut_16$G10 != 0]),
                         length(BCmut_16$H06[BCmut_16$H06 != 0]),
                         length(BCmut_16$H07[BCmut_16$H07 != 0]),
                         length(BCmut_16$H08[BCmut_16$H08 != 0]),
                         length(BCmut_16$H09[BCmut_16$H09 != 0]),
                         length(BCmut_16$H10[BCmut_16$H10 != 0]),
                         length(BCmut_16$H11[BCmut_16$H11 != 0]),
                         length(BCmut_16$H12[BCmut_16$H12 != 0]))
```

#### Plot Mapped Reads

Plot the mapped reads <font color="red">(match with a barcode)</font> as a fraction of the total reads.
```{r}
# Library 15

# Calculate the fraction of mapped counts or barcodes relative to total counts or barcodes, respectively
norm_vals_15 <- norm_vals_15 %>%
  mutate(frac_mapped_counts=mappedcounts/counts,
         frac_mapped_BC=mappedBCs/numBCs)

# Library 16

# Calculate the fraction of mapped counts or barcodes relative to total counts or barcodes, respectively
norm_vals_16 <- norm_vals_16 %>%
  mutate(frac_mapped_counts=mappedcounts/counts,
         frac_mapped_BC=mappedBCs/numBCs)
```

**Lib15 Total Reads**
```{r}
# Plot the Fraction of Total Reads Mapped - Library 15
p1_15 <- ggplot(data=norm_vals_15, aes(x=sample, y=frac_mapped_counts)) +
  geom_bar(stat="identity",color="black", fill="aquamarine4", width = 0.8) + 
  xlab("Sample") + 
  ylab("Fraction of Total Reads Mapped") + 
  ggtitle('Library 15 Total Reads Mapped') +
  #theme_cowplot() +
  theme(axis.text.x = element_text(size = 12),
        axis.text.y = element_text(size = 12),
        axis.title.x = element_text(size = 14),
        axis.title.y = element_text(size = 14)) +
  theme(legend.position = "none") +
  theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1)) +
  scale_y_continuous(expand = c(0, 0), limits = c(0, 0.8)) +
  guides(linetype = "none") +
  panel_border(color = "black")

p1_15
```

**Lib16 Total Reads**
```{r}
# Plot the Fraction of Total Reads Mapped
p1_16 <- ggplot(data=norm_vals_16, aes(x=sample, y=frac_mapped_counts)) +
  geom_bar(stat="identity",color="black", fill="aquamarine4", width = 0.8) + 
  xlab("Sample") + 
  ylab("Fraction of Total Reads Mapped") + 
  ggtitle('Library 16 Total Reads Mapped') +
  #theme_cowplot() +
  theme(axis.text.x = element_text(size = 12),
        axis.text.y = element_text(size = 12),
        axis.title.x = element_text(size = 14),
        axis.title.y = element_text(size = 14)) +
  theme(legend.position = "none") +
  theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1)) +
  scale_y_continuous(expand = c(0, 0), limits = c(0, 0.8)) +
  guides(linetype = "none") +
  panel_border(color = "black")

p1_16
```

```{r echo=FALSE}
# Library 15
ggsave(file="plots/fraction_reads_mapped_lib15.png", plot=p1_15,
       dpi=600, width = 8, height = 6, units = "in")

# Library 16
ggsave(file="plots/fraction_reads_mapped_lib16.png", plot=p1_16,
       dpi=600, width = 8, height = 6, units = "in")
```

#### Plot BCs Mapped

**Lib15 Total BCs**
```{r}
# Plot the Fraction of Total Barcodes Mapped
p2_15 <- ggplot(data=norm_vals_15, aes(x=sample, y=frac_mapped_BC)) +
  geom_bar(stat="identity",color="black", fill="aquamarine4", width = 0.8) + 
  xlab("Sample") + 
  ylab("Fraction of Total Barcodes Mapped") +
  ggtitle('Library 15 Total BCs Mapped') +
  #theme_cowplot() +
  theme(axis.text.x = element_text(size = 12),
        axis.text.y = element_text(size = 12),
        axis.title.x = element_text(size = 14),
        axis.title.y = element_text(size = 14)) +
  theme(legend.position = "none") +
  theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1)) +
  scale_y_continuous(expand = c(0, 0), limits = c(0, 0.8)) +
  guides(linetype = "none") +
  panel_border(color = "black")

p2_15
```

**Lib16 Total BCs**
```{r}
# Plot the Fraction of Total Barcodes Mapped
p2_16 <- ggplot(data=norm_vals_16, aes(x=sample, y=frac_mapped_BC)) +
  geom_bar(stat="identity",color="black", fill="aquamarine4", width = 0.8) + 
  xlab("Sample") + 
  ylab("Fraction of Total Barcodes Mapped") +
  ggtitle('Library 16 Total BCs Mapped') +
  #theme_cowplot() +
  theme(axis.text.x = element_text(size = 12),
        axis.text.y = element_text(size = 12),
        axis.title.x = element_text(size = 14),
        axis.title.y = element_text(size = 14)) +
  theme(legend.position = "none") +
  theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1)) +
  scale_y_continuous(expand = c(0, 0), limits = c(0, 0.8)) +
  guides(linetype = "none") +
  panel_border(color = "black")

p2_16
```

```{r echo=FALSE}
# Library 15
ggsave(file="plots/fraction_bc_mapped_lib15.png", plot=p2_15,
       dpi=600, width = 8, height = 6, units = "in")

# Library 16
ggsave(file="plots/fraction_bc_mapped_lib16.png", plot=p2_16,
       dpi=600, width = 8, height = 6, units = "in")
```

```{r echo=FALSE}
patch1 <- (p1_15 | p1_16)/(p2_15 | p2_16)
patch1
```

```{r echo=FALSE}
ggsave(file="plots/lib15.16.total.reads.bcs.mapped.png", plot=patch1,
       dpi=600, width = 12, height = 8, units = "in")
```

### Controls
The following section summarizes the log2-fold change in **CONTROL** barcodes between A05 vs. A06 sampling conditions.

#### Control Barcodes
Load in the controls and their associated barcodes:
```{r}
#load controls
controls = read.csv("control_BCs.csv", head=FALSE)
colnames(controls) <- c("ctrlname","colnum","BClong","BC","BClen")
```

<font color="red">What is the following code doing?</font>
```{r}
controls <- controls %>%
  mutate(BCstub=str_sub(BClong,1,-4)) %>%
  dplyr::rename(BCold=BC) %>%
  dplyr::rename(BC=BCstub)
```

Summarize the number of control barcodes from the full `BCs` dataset:
```{r, class.output="goodCode"}
# Library 15

BCcontrols_15 <- inner_join(BCs15,controls,by="BC")

BCcontrols_15.count <- inner_join(BCs15,controls,by="BC") %>%
  nrow()

format(BCcontrols_15.count, big.mark = ",")

# Library 16

BCcontrols_16 <- inner_join(BCs16,controls,by="BC")

BCcontrols_16.count <- inner_join(BCs16,controls,by="BC") %>%
  nrow()

format(BCcontrols_16.count, big.mark = ",")
```

#### Plotting Controls

**Lib15 Controls**
Plot the control barcodes as the log2-fold change comparison between D05 vs. D03:
```{r}
# Plot the Fraction of Total Barcodes Mapped
p3_15 <- ggplot(data=BCcontrols_15, aes(x=ctrlname, y=D05D03fc)) +
  geom_boxplot(color="black", fill="aquamarine4", alpha=0.8) + 
  geom_point() +
  xlab("Control") + 
  ylab("log2 fold change") + 
  ggtitle("Log2-Fold Change (D05 vs. D03)") +
  #theme_cowplot() +
  theme(axis.text.x = element_text(size = 12),
        axis.text.y = element_text(size = 12),
        axis.title.x = element_text(size = 14),
        axis.title.y = element_text(size = 14)) +
  theme(legend.position = "none") +
  theme(plot.title = element_text(size=16, face="bold", hjust=0.5)) +
  panel_border(color = "black") +
  scale_y_continuous(expand = c(0, 0), limits = c(-3, 0))

p3_15
```

**Lib16 Controls**
Plot the control barcodes as the log2-fold change comparison between D12 vs. D04:
```{r}
# Plot the Fraction of Total Barcodes Mapped
p3_16 <- ggplot(data=BCcontrols_16, aes(x=ctrlname, y=D12D04fc)) +
  geom_boxplot(color="black", fill="aquamarine4", alpha=0.8) + 
  geom_point() +
  xlab("Control") + 
  ylab("log2 fold change") + 
  ggtitle("Log2-Fold Change (D12 vs. D04)") +
  #theme_cowplot() +
  theme(axis.text.x = element_text(size = 12),
        axis.text.y = element_text(size = 12),
        axis.title.x = element_text(size = 14),
        axis.title.y = element_text(size = 14)) +
  theme(legend.position = "none") +
  theme(plot.title = element_text(size=16, face="bold", hjust=0.5)) +
  panel_border(color = "black")+
  scale_y_continuous(expand = c(0, 0), limits = c(-3, 0))

p3_16
```

```{r echo=FALSE}
# Library 15
ggsave(file="plots/controls/controls_Lib15_D05D03fc.png", plot=p3_15,
       dpi=600, width = 6, height = 4, units = "in")

# Library 16
ggsave(file="plots/controls/controls_Lib16_D12D04fc.png", plot=p3_16,
       dpi=600, width = 6, height = 4, units = "in")
```

```{r echo=FALSE}
patch2 <- (p3_15 | p3_16)
patch2
```

```{r echo=FALSE}
ggsave(file="plots/controls/lib15.16.ctrls.png", plot=patch2,
       dpi=600, width = 8, height = 6, units = "in")
```

```{r echo=FALSE, results='hide'}
# Library 15
write.csv(BCcontrols_15,"BC_15_controls.csv", row.names = FALSE)

# Library 16
write.csv(BCcontrols_16,"BC_16_controls.csv", row.names = FALSE)
```

### Barcode Count

Determine the number of unique barcodes based on a minimum sequence count threshold across sampling conditions. Filter the barcodes by a minimum threshold of sequence reads per unique barcode. In this case, we use a mimimum cutoff of 9 sequence reads per barcode:

**Lib15 BC Count***
```{r class.output="goodCode"}
# Filter unique barcodes by minimum sequence count. This code uses "or" such that if any of the sampling conditions contain at least 9 sequences counts/barcode, then the barcode is retained for the full dataset.
BCprune15 <- BCs15 %>%
  filter(D01>5 | D03>9 | D05>9 | D06>9 | D07>9 | D08>9 | D09>9 | D10>9 | D11>9 | 
           E07>9 | E08>9 | E09>9 | E10>9 | E11>9 | E12>9 | F01>9 | 
           F09>9 | F10>9 | F11>9 | F12>9 | G01>9 | G02>9 | G03>9 | 
           G11>9 | G12>9 | H01>9 | H02>9 | H03>9 | H04>9 | H05>9)

# Trim the "BCmut" dataset based on the "BCprune" barcodes retained:
BCmut_15 <- inner_join(BCinfo15,BCprune15,by="BC")

BCmut_15.count <- inner_join(BCinfo15,BCprune15,by="BC") %>%
  nrow()

format(BCmut_15.count, big.mark = ",")
```

**Lib16 BC Count**
```{r class.output="goodCode"}
# Filter unique barcodes by minimum sequence count. This code uses "or" such that if any of the sampling conditions contain at least 9 sequences counts/barcode, then the barcode is retained for the full dataset.
BCprune16 <- BCs16 %>%
  filter(D02>9 | D04>9 | D12>9 | E01>9 | E02>9 | E03>9 | E04>9 | E05>9 | E06>9 | 
          F02>9 | F03>9 | F04>9 | F05>9 | F06>9 | F07>9 | F08>9 | 
           G04>9 | G05>9 | G06>9 | G07>9 | G08>9 | G09>9 | G10>9 | 
           H06>9 | H07>9 | H08>9 | H09>9 | H10>9 | H11>9 | H12>9)

# Trim the "BCmut" dataset based on the "BCprune" barcodes retained:
BCmut_16 <- inner_join(BCinfo16,BCprune16,by="BC")

BCmut_16.count <- inner_join(BCinfo16,BCprune16,by="BC") %>%
  nrow()

format(BCmut_16.count, big.mark = ",")
```

### Miscellaneous
<font color="red">The following section contains code for miscellaneous calculations.  These sections may or may not be required for downstream analysis.</font>

#### Mutant Count
Begin by summarizing the mutant counts for each unique barcode and set a minimum sequence count threshold for each unique barcode. <font color="red">What is the following code doing"?</font>

**Lib15 Mutant Count**
```{r}
# Lib15
#how many barcodes for each mutant
mutIDinfo15 <- BCmut_15 %>%
  group_by(mutID) %>%
  summarise(numprunedBCs = n()) %>%
  right_join(mutIDinfo15, by="mutID")
```

```{r}
# Lib15
#replace NAs with 0
mutIDinfo15$numprunedBCs[is.na(mutIDinfo15$numprunedBCs)] <- 0
```

```{r}
# Lib15
#mutant must have at least 1 BC after pruning:
mutIDinfo15_1BC <- mutIDinfo15 %>%
  filter(numprunedBCs>0)
```

**Lib16 Mutant Count**
```{r}
# Lib16
#how many barcodes for each mutant
mutIDinfo16 <- BCmut_16 %>%
  group_by(mutID) %>%
  summarise(numprunedBCs = n()) %>%
  right_join(mutIDinfo16, by="mutID")
```

```{r}
# Lib16
#replace NAs with 0
mutIDinfo16$numprunedBCs[is.na(mutIDinfo16$numprunedBCs)] <- 0
```

```{r}
# Lib16
#mutant must have at least 1 BC after pruning:
mutIDinfo16_1BC <- mutIDinfo16 %>%
  filter(numprunedBCs>0)
```

# Perfects Only
<font color="blue">**This section is based on the R file: "R_perfects_only.R".**</font> It describes how to select only for perfects (perfect match to designed homolog sequence).

## Select Perfects

Begin by selecting perfects unique mutant IDs from the "mutIDinfo" object where mutations = 0:
```{r}
# Lib15
perfects15 <- mutIDinfo15 %>%
  filter(numprunedBCs>0) %>%
  filter(mutations==0) %>%
  dplyr::rename(ID = IDalign)

# Lib16
perfects16 <- mutIDinfo16 %>%
  filter(numprunedBCs>0) %>%
  filter(mutations==0) %>%
  dplyr::rename(ID = IDalign)
```

Select perfects BCs from "BCmut" object where mutations = 0:
```{r}
# Lib15
BCperfect15 <- BCmut_15 %>%
  filter(mutations==0) %>%
  dplyr::rename(ID=mutID)

# Lib16
BCperfect16 <- BCmut_16 %>%
  filter(mutations==0) %>%
  dplyr::rename(ID=mutID)
```

Perform a quick check to verify no IDs are missing:
```{r class.output="goodCode"}
# Lib15
#check how many are missing ID
nrow(BCperfect15 %>% filter(ID == ""))

# Lib16
#check how many are missing ID
nrow(BCperfect16 %>% filter(ID == ""))
```

Create a function to check if each unique perfect sequence represents a **"true perfect"** based on the designed homolog sequence or a **"synonymous mutation"** that is functionally redundant.
```{r}
#check if this is a synonymous mutant
cigar_syn_mutant_check <- function(cigarin){
  flag_out <- "Perfect"
  if (grepl("X",cigarin, fixed=TRUE)) {
    flag_out <- "SynMutant"
  }
  return(flag_out)
}
```

Apply the "perfect check function" to the "BCperfect" dataset and add a column to indicate each BCs status.
```{r}
# Lib15
#check if this BC is a synonymous mutant
BCperfect15 <- BCperfect15 %>%
  rowwise() %>%
  mutate(synflag=cigar_syn_mutant_check(cigar))

# Lib16
#check if this BC is a synonymous mutant
BCperfect16 <- BCperfect16 %>%
  rowwise() %>%
  mutate(synflag=cigar_syn_mutant_check(cigar))
```

Count the number of "Perfects" from the dataset:
```{r class.output="goodCode"}
# Lib15
# Number of perfects
perfects.count15 <- length(which(BCperfect15$synflag=="Perfect"))

format(perfects.count15, big.mark = ",")

# Lib16
# Number of perfects
perfects.count16 <- length(which(BCperfect16$synflag=="Perfect"))

format(perfects.count16, big.mark = ",")
```

Count the number of "SynMutants" from the dataset:
```{r class.output="goodCode"}
# Lib15
# Number of synonymous mutants
synmut.count15 <- length(which(BCperfect15$synflag=="SynMutant"))

format(synmut.count15, big.mark = ",")

# Lib16
# Number of synonymous mutants
synmut.count16 <- length(which(BCperfect16$synflag=="SynMutant"))

format(synmut.count16, big.mark = ",")
```

## Perfects Stats

```{r}
#this code checks the STD among BC fitness for different ways to calc fitness

# Lib15
perfects15_std <- perfects15 %>%
  select(ID)

# Lib16
perfects16_std <- perfects16 %>%
  select(ID)
```

```{r}
#take BCperfect, for each BC with an ID, take the median and sum, calc median, join into perfects

# Lib15
perfects15 <- BCperfect15 %>%
  group_by(ID) %>%
  summarise(fitD03D01=median(D03D01fc), fitD05D03=median(D05D03fc), 
            fitD06D03=median(D06D03fc), fitD07D03=median(D07D03fc), fitD08D03=median(D08D03fc),
            fitD09D03=median(D09D03fc), fitD10D03=median(D10D03fc), fitD11D03=median(D11D03fc),
            fitE07D03=median(E07D03fc), fitE08D03=median(E08D03fc), fitE09D03=median(E09D03fc),
            fitE10D03=median(E10D03fc), fitE11D03=median(E11D03fc), fitE12D03=median(E12D03fc),
            fitF01D03=median(F01D03fc), fitF09D03=median(F09D03fc), fitF10D03=median(F10D03fc),
            fitF11D03=median(F11D03fc), fitF12D03=median(F12D03fc), fitG01D03=median(G01D03fc),
            fitG02D03=median(G02D03fc), fitG03D03=median(G03D03fc), fitG11D03=median(G11D03fc),
            fitG12D03=median(G12D03fc), fitH01D03=median(H01D03fc), fitH02D03=median(H02D03fc),
            fitH03D03=median(H03D03fc), fitH04D03=median(H04D03fc), fitH05D03=median(H05D03fc),
            ) %>%
  right_join(perfects15, by="ID")

# Lib16
perfects16 <- BCperfect16 %>%
  group_by(ID) %>%
  summarise(fitD04D02=median(D04D02fc), fitD12D04=median(D12D04fc), 
            fitE01D04=median(E01D04fc), fitE02D04=median(E02D04fc), fitE03D04=median(E03D04fc),
            fitE04D04=median(E04D04fc), fitE05D04=median(E05D04fc), fitE06D04=median(E06D04fc),
            fitF02D04=median(F02D04fc), fitF03D04=median(F03D04fc), fitF04D04=median(F04D04fc),
            fitF05D04=median(F05D04fc), fitF06D04=median(F06D04fc), fitF07D04=median(F07D04fc),
            fitF08D04=median(F08D04fc), fitG04D04=median(G04D04fc), fitG05D04=median(G05D04fc),
            fitG06D04=median(G06D04fc), fitG07D04=median(G07D04fc), fitG08D04=median(G08D04fc),
            fitG09D04=median(G09D04fc), fitG10D04=median(G10D04fc), fitH06D04=median(H06D04fc),
            fitH07D04=median(H07D04fc), fitH08D04=median(H08D04fc), fitH09D04=median(H09D04fc),
            fitH10D04=median(H10D04fc), fitH11D04=median(H11D04fc), fitH12D04=median(H12D04fc),
            ) %>%
  right_join(perfects16, by="ID")
```

Select only those perfects with a minimum number of barcode:
```{r}
### Select only those with a minimum of 2 BCs

# Lib15
perfects15_2BCs <- perfects15 %>%
  filter(numprunedBCs > 1)

# Lib16
perfects16_2BCs <- perfects16 %>%
  filter(numprunedBCs > 1)

### Select only those with a minimum of 5 BCs

# Lib15
perfects15_5BCs <- perfects15 %>%
  filter(numprunedBCs > 4)

# Lib16
perfects16_5BCs <- perfects16 %>%
  filter(numprunedBCs > 4)
```

Save an output of the data for generating trees:
```{r}
### Output 2BCs data for tree generation

# Lib15
write.table(perfects15_2BCs$mutID, file = "analysis_output/perfects15_final_ID_list_2BCs.csv", sep = ",", row.names = F,quote=F,col.names = F)

# Lib16
write.table(perfects16_2BCs$mutID, file = "analysis_output/perfects16_final_ID_list_2BCs.csv", sep = ",", row.names = F,quote=F,col.names = F)

### Output 5BCs data for tree generation

# Lib15
write.table(perfects15_5BCs$mutID, file = "analysis_output/perfects15_final_ID_list_5BCs.csv", sep = ",", row.names = F,quote=F,col.names = F)

# Lib16
write.table(perfects16_5BCs$mutID, file = "analysis_output/perfects16_final_ID_list_5BCs.csv", sep = ",", row.names = F,quote=F,col.names = F)
```

Re-order the "perfects" dataset by "numprunedBCs" in ascending order:
```{r}
# Lib15
perfects15 <- perfects15 %>%
  arrange(numprunedBCs)

# Lib16
perfects16 <- perfects16 %>%
  arrange(numprunedBCs)
```

```{r}
#make keys based on "numprunedBCs"

# Lib15
perfects15$numprunedBCs.key <- 1:length(perfects15$mutID)

# Lib16
perfects16$numprunedBCs.key <- 1:length(perfects16$mutID)
```

## Perfects Plots
The following section plots various summarizes of the Perfects BCs:

### Rank Ordered Homolog Plot
Plot the number of perfects barcodes recovered based on the rank order of homologs:

**Library 15**
```{r}
p4_15 <- ggplot(perfects15) +
  geom_point(aes(x=numprunedBCs.key,y=numprunedBCs),color = "aquamarine4") +
  ylab("Number of Barcodes") +
  xlab("Rank Ordered Homolog") +
  scale_y_log10() +
  #theme_cowplot() +
  #background_grid() +
  ggtitle("Library 15 Perfects Rank Order") +
  theme(axis.text.x = element_text(size = 12),
        axis.text.y = element_text(size = 12),
        axis.title.x = element_text(size = 14),
        axis.title.y = element_text(size = 14)) +
  theme(legend.position = "none") +
  scale_x_continuous(expand = c(0, 0), limits = c(0,1000)) +
  panel_border(color = "black")

p4_15
```

Establish rank order for plotting.
```{r}
perfects15_rank <- perfects15 %>%
  arrange(fitD05D03)

perfects15_rank$fitD05D03.key <- 1:length(perfects15_rank$ID)

BCperfect15_2 <- BCperfect15 %>%
  left_join(perfects15_rank %>%
              select(ID, fitD05D03, fitD05D03.key), by="ID")
```

Plot Perfects fitness scores for each unique Perfects based on rank ordering of the homologs:
```{r}
p0_15 <- ggplot(BCperfect15_2,aes(x=fitD05D03.key,y=D05D03fc, color=ID)) +
  geom_boxplot() +
  theme(legend.position = "none") +
  xlab("Rank Ordered Homolog") +
  ylab("Fitness (Log2 Fold Change)") +
  ggtitle("Library 15 Perfects Rank Order Fitness") +
  #theme_cowplot() +
  #background_grid() +
  theme(axis.text.x = element_text(size = 12),
        axis.text.y = element_text(size = 12),
        axis.title.x = element_text(size = 14),
        axis.title.y = element_text(size = 14)) +
  theme(legend.position = "none") +
  scale_y_continuous(expand = c(0, 0), limits = c(-12,7)) +
  guides(linetype = FALSE) + panel_border(color = "black")

p0_15
```

```{r echo=FALSE}
# Lib15
ggsave(file="plots/perfects15_rank_ordered_num_BCs.png", plot=p4_15,
       dpi=600, width = 8, height = 6, units = "in")

ggsave(file="plots/perfects15_rank_ordered_fitness.png", plot=p0_15,
       dpi=600, width = 8, height = 6, units = "in")
```


**Library 16**
```{r}
p4_16 <- ggplot(perfects16) +
  geom_point(aes(x=numprunedBCs.key,y=numprunedBCs),color = "aquamarine4") +
  ylab("Number of Barcodes") +
  xlab("Rank Ordered Homolog") +
  scale_y_log10() +
  #theme_cowplot() +
  #background_grid() +
  ggtitle("Library 16 Perfects Rank Order") +
  theme(axis.text.x = element_text(size = 12),
        axis.text.y = element_text(size = 12),
        axis.title.x = element_text(size = 14),
        axis.title.y = element_text(size = 14)) +
  theme(legend.position = "none") +
  scale_x_continuous(expand = c(0, 0), limits = c(0,1000)) +
  panel_border(color = "black")

p4_16
```

Establish rank order for plotting.
```{r}
perfects16_rank <- perfects16 %>%
  arrange(fitD12D04)

perfects16_rank$fitD12D04.key <- 1:length(perfects16_rank$ID)

BCperfect16_2 <- BCperfect16 %>%
  left_join(perfects16_rank %>%
              select(ID, fitD12D04, fitD12D04.key), by="ID")
```

Plot Perfects fitness scores for each unique Perfects based on rank ordering of the homologs:
```{r}
p0_16 <- ggplot(BCperfect16_2,aes(x=fitD12D04.key,y=D12D04fc, color=ID)) +
  geom_boxplot() +
  theme(legend.position = "none") +
  xlab("Rank Ordered Homolog") +
  ylab("Fitness (Log2 Fold Change)") +
  ggtitle("Library 16 Perfects Rank Order Fitness") +
  #theme_cowplot() +
  #background_grid() +
  theme(axis.text.x = element_text(size = 12),
        axis.text.y = element_text(size = 12),
        axis.title.x = element_text(size = 14),
        axis.title.y = element_text(size = 14)) +
  theme(legend.position = "none") +
  scale_y_continuous(expand = c(0, 0), limits = c(-12,7)) +
  guides(linetype = FALSE) + panel_border(color = "black")

p0_16
```

```{r echo=FALSE}
# Lib16
ggsave(file="plots/perfects16_rank_ordered_num_BCs.png", plot=p4_16,
       dpi=600, width = 8, height = 6, units = "in")

ggsave(file="plots/perfects16_rank_ordered_fitness.png", plot=p0_16,
       dpi=600, width = 8, height = 6, units = "in")
```

```{r echo=FALSE}
patch3 <- (p4_15 | p4_16) / (p0_15 | p0_16)
patch3
```

```{r echo=FALSE}
ggsave(file="plots/combined/lib15.16.rank.ordered.num.BCs.fitness.png", plot=patch3,
       dpi=600, width = 12, height = 10, units = "in")
```

### Perfect >2 BCs Fitness Plot

Plot the Perfects counts based on fitness values between D05 vs. D03 (Lib15) or D12 vs. D04 (Lib16) at > 2 BCs:

**Library 15**
```{r}
p5_15 <- ggplot(perfects15_2BCs,aes(x=fitD05D03)) +
  geom_histogram(binwidth=0.08, fill = "aquamarine4", color = "black") +
  xlab("Fitness (Log2 Fold Change)") +
  ylab("Counts") +
  ggtitle("Library 15 Perfects Fitness (>2 BCs)") +
  #theme_cowplot() +
  #background_grid() +
  theme(axis.text.x = element_text(size = 12),
        axis.text.y = element_text(size = 12),
        axis.title.x = element_text(size = 14),
        axis.title.y = element_text(size = 14)) +
  theme(legend.position = "none") +
  scale_x_continuous(limits = c(-6,2)) +
  scale_y_continuous(expand = c(0, 0), limits = c(0,50)) +
  guides(linetype = FALSE) +
  panel_border(color = "black")

p5_15
```

**Library 16**
```{r}
p5_16 <- ggplot(perfects16_2BCs,aes(x=fitD12D04)) +
  geom_histogram(binwidth=0.08, fill = "aquamarine4", color = "black") +
  xlab("Fitness (Log2 Fold Change)") +
  ylab("Counts") +
  ggtitle("Library 16 Perfects Fitness (>2 BCs)") +
  #theme_cowplot() +
  #background_grid() +
  theme(axis.text.x = element_text(size = 12),
        axis.text.y = element_text(size = 12),
        axis.title.x = element_text(size = 14),
        axis.title.y = element_text(size = 14)) +
  theme(legend.position = "none") +
  scale_x_continuous(limits = c(-6,2)) +
  scale_y_continuous(expand = c(0, 0), limits = c(0,50)) +
  guides(linetype = FALSE) +
  panel_border(color = "black")

p5_16
```

```{r echo=FALSE}
# Lib15
ggsave(file="plots/hist15_2BCs_fitD05D03.png", plot=p5_15,
       dpi=600, width = 8, height = 6, units = "in")

# Lib16
ggsave(file="plots/hist16_2BCs_fitD12D04.png", plot=p5_16,
       dpi=600, width = 8, height = 6, units = "in")
```

```{r echo=FALSE}
patch4 <- (p5_15 | p5_16)
patch4
```

```{r echo=FALSE}
ggsave(file="plots/combined/lib15.16.perfects.fitness.2BCs.png", plot=patch4,
       dpi=600, width = 8, height = 6, units = "in")
```

Summarize the total number of unique Perfects (>2 BCs) regardless of fitness value:
```{r class.output="goodCode"}
# Lib15
perfects15_2BCs %>%
  nrow(.)

# Lib16
perfects16_2BCs %>%
  nrow(.)
```

Summarize the number of Perfects (>2 BCs) with fitness greater than -0.5 in D05 vs. D03 (Lib15) or D12 vs. D04 (Lib16) conditions:
```{r class.output="goodCode"}
# Lib15
perfects15_2BCs %>% filter(fitD05D03>-0.5) %>%
  nrow(.)

# Lib16
perfects16_2BCs %>% filter(fitD12D04>-0.5) %>%
  nrow(.)
```

### Perfect >5 BCs Fitness Plot

Plot the Perfects counts based on fitness values between D05 vs. D03 (Lib15) or D12 vs. D04 (Lib16) at > 5 BCs:

**Library 15**
```{r}
p6_15 <- ggplot(perfects15_5BCs,aes(x=fitD05D03)) +
  geom_histogram(binwidth=0.08, fill = "aquamarine4", color = "black") +
  xlab("Fitness (Log2 Fold Change)") +
  ylab("Counts") +
  ggtitle("Library 15 Perfects Fitness (>5 BCs)") +
  #theme_cowplot() +
  #background_grid() +
  theme(axis.text.x = element_text(size = 12),
        axis.text.y = element_text(size = 12),
        axis.title.x = element_text(size = 14),
        axis.title.y = element_text(size = 14)) +
  theme(legend.position = "none") +
  scale_x_continuous(limits = c(-6,2)) +
  scale_y_continuous(expand = c(0, 0), limits = c(0,50)) +
  guides(linetype = FALSE) +
  panel_border(color = "black")

p6_15
```

**Library 16**
```{r}
p6_16 <- ggplot(perfects16_5BCs,aes(x=fitD12D04)) +
  geom_histogram(binwidth=0.08, fill = "aquamarine4", color = "black") +
  xlab("Fitness (Log2 Fold Change)") +
  ylab("Counts") +
  ggtitle("Library 16 Perfects Fitness (>5 BCs)") +
  #theme_cowplot() +
  #background_grid() +
  theme(axis.text.x = element_text(size = 12),
        axis.text.y = element_text(size = 12),
        axis.title.x = element_text(size = 14),
        axis.title.y = element_text(size = 14)) +
  theme(legend.position = "none") +
  scale_x_continuous(limits = c(-6,2)) +
  scale_y_continuous(expand = c(0, 0), limits = c(0,50)) +
  guides(linetype = FALSE) +
  panel_border(color = "black")

p6_16
```

```{r echo=FALSE}
# Lib15
ggsave(file="plots/hist15_5BCs_fitD05D03.png", plot=p6_15,
       dpi=600, width = 8, height = 6, units = "in")

# Lib16
ggsave(file="plots/hist16_5BCs_fitD12D04.png", plot=p6_16,
       dpi=600, width = 8, height = 6, units = "in")
```

```{r echo=FALSE}
patch5 <- (p6_15 | p6_16)
patch5
```

```{r echo=FALSE}
ggsave(file="plots/combined/lib15.16.perfects.fitness.5BCs.v2.png", plot=patch5,
       dpi=600, width = 8, height = 6, units = "in")
```

```{r echo=FALSE}
patch6 <- (p4_15 | p4_16) / (p5_15 | p5_16) / (p6_15 | p6_16)
patch6
```

```{r echo=FALSE}
ggsave(file="plots/combined/lib15.16.perfects.rank.order.fitness.2BCs.5BCs.v2.png", plot=patch6,
       dpi=600, width = 8, height = 10, units = "in")
```

Summarize the total number of unique Perfects (>5 BCs) regardless of fitness value:
```{r class.output="goodCode"}
# Lib15
perfects15_5BCs %>%
  nrow(.)

# Lib16
perfects16_5BCs %>%
  nrow(.)
```

Summarize the number of Perfects (>5 BCs) with fitness greater than -0.5 in D05 vs. D03 (Lib15) or D12 vs. D04 (Lib16) conditions:
```{r class.output="goodCode"}
# Lib15
perfects15_5BCs %>% filter(fitD05D03>-0.5) %>%
  nrow(.)

# Lib16
perfects16_5BCs %>% filter(fitD12D04>-0.5) %>%
  nrow(.)
```

### Perfect >5 BCs Barcode Plot
Plot the Perfects fitness scores for D05 vs. D03 (Lib15) or D12 vs. D04 (Lib16) based on the number of recovered barcodes:

**Library 15**
```{r}
p7_15 <- ggplot(perfects15_5BCs,aes(x=numprunedBCs,y=fitD05D03)) +
  geom_point(color = "aquamarine4") +
  scale_x_log10() +
  xlab("Number of Barcodes") +
  ylab("Fitness (Log2 Fold Change)") +
  ggtitle("Library 15 Perfects Fitness Scores (>5 BCs)") +
  #theme_cowplot() +
  #background_grid() +
  theme(axis.text.x = element_text(size = 12),
        axis.text.y = element_text(size = 12),
        axis.title.x = element_text(size = 14),
        axis.title.y = element_text(size = 14)) +
  theme(legend.position = "none") +
  scale_y_continuous(expand = c(0, 0), limits = c(-5.5,2.5)) +
  guides(linetype = FALSE) +
  panel_border(color = "black")

p7_15
```

**Library 16**
```{r}
p7_16 <- ggplot(perfects16_5BCs,aes(x=numprunedBCs,y=fitD12D04)) +
  geom_point(color = "aquamarine4") +
  scale_x_log10() +
  xlab("Number of Barcodes") +
  ylab("Fitness (Log2 Fold Change)") +
  ggtitle("Library 16 Perfects Fitness Scores (>5 BCs)") +
  #theme_cowplot() +
  #background_grid() +
  theme(axis.text.x = element_text(size = 12),
        axis.text.y = element_text(size = 12),
        axis.title.x = element_text(size = 14),
        axis.title.y = element_text(size = 14)) +
  theme(legend.position = "none") +
  scale_y_continuous(expand = c(0, 0), limits = c(-5.5,2.5)) +
  guides(linetype = FALSE) +
  panel_border(color = "black")

p7_16
```

```{r echo=FALSE}
# Lib15
ggsave(file="plots/perfects15_5BCs_BCs_fitD05D03.png", plot=p7_15,
       dpi=600, width = 12, height = 7.5, units = "in")

# Lib16
ggsave(file="plots/perfects16_5BCs_BCs_fitD12D04.png", plot=p7_16,
       dpi=600, width = 12, height = 7.5, units = "in")
```

```{r echo=FALSE}
patch7 <- (p7_15 | p7_16)
patch7
```

```{r echo=FALSE}
ggsave(file="plots/combined/lib15.16.perfects.fitness.scores.5BCs.png", plot=patch7,
       dpi=600, width = 10, height = 6, units = "in")
```

## DHFR Genes
Load in the organism dataset associated with DHFR genes derived from the <font color="red">NCBI database.</font>
```{r}
#load in all the organisms info for unique BCs associated with DHFR genes: 
orginfo = read.csv("DHFR_chip_3_proteins_addedseveral.csv", head=TRUE)  # read csv file
```

Subset the dataset:
```{r}
orginfo <- orginfo %>%
  select(-Lib.codon1, -Lib.codon2, -Construct., -Barcode) %>%
  dplyr::rename(IDfull=Accession) %>%
  #mutate(ID=strsplit(IDfull, split = ".")[1])
  tidyr::separate(IDfull, c("ID", "ver"), "\\.",remove = FALSE) %>%
  select(-ver)
names(orginfo)
```

Add the organism info to the perfects info as a new object called "perfects_tree":
```{r}
# Lib15
perfects15_tree <- right_join(orginfo,perfects15,by="ID") %>%
  select(-mutations,-pct_ident,-BCs)

# Lib16
perfects16_tree <- right_join(orginfo,perfects16,by="ID") %>%
  select(-mutations,-pct_ident,-BCs)
```

Create a filtered dataset that only retains perfects with 5+ BCs:
```{r}
# Lib15
perfects15_tree_5BCs <- perfects15_tree %>%
  filter(numprunedBCs > 4) %>%
  select(-IDfull)

# Lib16
perfects16_tree_5BCs <- perfects16_tree %>%
  filter(numprunedBCs > 4) %>%
  select(-IDfull)
```

### Perfects % Similar E. coli Plot

Plot the Perfects fitness scores for D05 vs. D03 (Lib15) or D12 vs. D04 (Lib16) based on percent similarity of the each variant's gene sequences to the E. coli DHFR gene sequence:

**Library 15**
```{r}
p8_15 <- ggplot(perfects15_tree_5BCs,aes(x=PctIdentEcoli,y=fitD05D03)) +
  geom_point(color = "aquamarine4") +
  xlab("Percent Similarity to E. coli DHFR") +
  ylab("Fitness (Log2 Fold Change)") +
  ggtitle("Library 15 Perfects Fitness to % Similarity E. coli (>5 BCs)") +
  #theme_cowplot() +
  #background_grid() +
  theme(axis.text.x = element_text(size = 12),
        axis.text.y = element_text(size = 12),
        axis.title.x = element_text(size = 14),
        axis.title.y = element_text(size = 14)) +
  theme(legend.position = "none") +
  scale_y_continuous(expand = c(0, 0), limits = c(-5,1)) +
  guides(linetype = FALSE) + 
  panel_border(color = "black")

p8_15
```

**Library 16**
```{r}
p8_16 <- ggplot(perfects16_tree_5BCs,aes(x=PctIdentEcoli,y=fitD12D04)) +
  geom_point(color = "aquamarine4") +
  xlab("Percent Similarity to E. coli DHFR") +
  ylab("Fitness (Log2 Fold Change)") +
  ggtitle("Library 16 Perfects Fitness to % Similarity E. coli (>5 BCs)") +
  #theme_cowplot() +
  #background_grid() +
  theme(axis.text.x = element_text(size = 12),
        axis.text.y = element_text(size = 12),
        axis.title.x = element_text(size = 14),
        axis.title.y = element_text(size = 14)) +
  theme(legend.position = "none") +
  scale_y_continuous(expand = c(0, 0), limits = c(-5,1)) +
  guides(linetype = FALSE) + 
  panel_border(color = "black")

p8_16
```

```{r echo=FALSE}
# Lib15
ggsave(file="plots/perfects15_5BCs_pctsim_vs_fitD05D03.png", plot=p8_15,
       dpi=600, width = 12, height = 7.5, units = "in")

# Lib16
ggsave(file="plots/perfects16_5BCs_pctsim_vs_fitD12D04.png", plot=p8_16,
       dpi=600, width = 12, height = 7.5, units = "in")
```

```{r echo=FALSE}
patch8 <- (p8_15 | p8_16)
patch8
```

```{r echo=FALSE}
ggsave(file="plots/combined/lib15.16.perfects.fitness.scores.5BCs.pctsim.png", plot=patch8,
       dpi=600, width = 12, height = 6, units = "in")
```

### Perfects Supplement Plot

**This section uses the `library(ggridges)` package.**

Subset the perfects_tree_5BCs object to retain only "ID" and fitness scores for select conditions. This code will transpose the individual fitness score columns into two columns with the fitness label and values, respectively, for each unique ID.
```{r}
### Lib15

# Day 1
perfects15_tree_5BCs_d1 <- perfects15_tree_5BCs %>%
  select(ID, fitD05D03, fitD06D03, fitD07D03, fitD08D03, fitD09D03, fitD10D03, fitD11D03) %>%
  pivot_longer(!ID, names_to = "fc", values_to = "val")

# Day 2
perfects15_tree_5BCs_d2 <- perfects15_tree_5BCs %>%
  select(ID, fitE07D03, fitE08D03, fitE09D03, fitE10D03, fitE11D03, fitE12D03, fitF01D03) %>%
  pivot_longer(!ID, names_to = "fc", values_to = "val")

# Day 3
perfects15_tree_5BCs_d3 <- perfects15_tree_5BCs %>%
  select(ID, fitF09D03, fitF10D03, fitF11D03, fitF12D03, fitG01D03, fitG02D03, fitG03D03) %>%
  pivot_longer(!ID, names_to = "fc", values_to = "val")

# Day 4
perfects15_tree_5BCs_d4 <- perfects15_tree_5BCs %>%
  select(ID, fitG11D03, fitG12D03, fitH01D03, fitH02D03, fitH03D03, fitH04D03, fitH05D03) %>%
  pivot_longer(!ID, names_to = "fc", values_to = "val")



### Lib16

# Day 1
perfects16_tree_5BCs_d1 <- perfects16_tree_5BCs %>%
  select(ID, fitD12D04, fitE01D04, fitE02D04, fitE03D04, fitE04D04, fitE05D04, fitE06D04) %>%
  pivot_longer(!ID, names_to = "fc", values_to = "val")

# Day 2
perfects16_tree_5BCs_d2 <- perfects16_tree_5BCs %>%
  select(ID, fitF02D04, fitF03D04, fitF04D04, fitF05D04, fitF06D04, fitF07D04, fitF08D04) %>%
  pivot_longer(!ID, names_to = "fc", values_to = "val")

# Day 3
perfects16_tree_5BCs_d3 <- perfects16_tree_5BCs %>%
  select(ID, fitG04D04, fitG05D04, fitG06D04, fitG07D04, fitG08D04, fitG09D04, fitG10D04) %>%
  pivot_longer(!ID, names_to = "fc", values_to = "val")

# Day 4
perfects16_tree_5BCs_d4 <- perfects16_tree_5BCs %>%
  select(ID, fitH06D04, fitH07D04, fitH08D04, fitH09D04, fitH10D04, fitH11D04, fitH12D04) %>%
  pivot_longer(!ID, names_to = "fc", values_to = "val")
```

Set the order of sampling conditions for plotting:
```{r}
# Lib15
lib15_level_order_d1 <- c("fitD11D03","fitD10D03","fitD09D03","fitD08D03","fitD07D03","fitD06D03","fitD05D03")
lib15_level_order_d2 <- c("fitF01D03","fitE12D03","fitE11D03","fitE10D03","fitE09D03","fitE08D03","fitE07D03")
lib15_level_order_d3 <- c("fitG03D03","fitG02D03","fitG01D03","fitF12D03","fitF11D03","fitF10D03","fitF09D03")
lib15_level_order_d4 <- c("fitH05D03","fitH04D03","fitH03D03","fitH02D03","fitH01D03","fitG12D03","fitG11D03")

# Lib16
lib16_level_order_d1 <- c("fitE06D04","fitE05D04","fitE04D04","fitE03D04","fitE02D04","fitE01D04","fitD12D04")
lib16_level_order_d2 <- c("fitF08D04","fitF07D04","fitF06D04","fitF05D04","fitF04D04","fitF03D04","fitF02D04")
lib16_level_order_d3 <- c("fitG10D04","fitG09D04","fitG08D04","fitG07D04","fitG06D04","fitG05D04","fitG04D04")
lib16_level_order_d4 <- c("fitH12D04","fitH11D04","fitH10D04","fitH09D04","fitH08D04","fitH07D04","fitH06D04")
```

Plot Perfects fitness scores based on Supplementation treatment for first sampling time point

**Library 15**
```{r}
# Day 1
p9_15_d1 <- ggplot(perfects15_tree_5BCs_d1, aes(x = val, y = factor(fc, level = lib15_level_order_d1))) + 
  geom_density_ridges(fill = "aquamarine4", color = "black") +
  scale_y_discrete(labels = c('200 ug/mL TMP',
                              '50 ug/mL TMP',
                              '10 ug/mL TMP',
                              '1 ug/mL TMP',
                              '0.5 ug/mL TMP',
                              '0.058 ug/mL TMP',
                              'M9 no supp.')) +
  xlab("Fitness (Log2 Fold Change)") +
  ylab("Selection Condition") +
  ggtitle("Library 15 Fitness - Day 1") +
  #theme_cowplot() +
  #background_grid() +
  theme(axis.text.x = element_text(size = 12),
        axis.text.y = element_text(size = 12),
        axis.title.x = element_text(size = 14),
        axis.title.y = element_text(size = 14)) +
  theme(legend.position = "none") +
  scale_x_continuous(limits = c(-15,10)) +
  guides(linetype = FALSE) +
  panel_border(color = "black")

# Day 2
p9_15_d2 <- ggplot(perfects15_tree_5BCs_d2, aes(x = val, y = factor(fc, level = lib15_level_order_d2))) + 
  geom_density_ridges(fill = "aquamarine4", color = "black") +
  scale_y_discrete(labels = c('200 ug/mL TMP',
                              '50 ug/mL TMP',
                              '10 ug/mL TMP',
                              '1 ug/mL TMP',
                              '0.5 ug/mL TMP',
                              '0.058 ug/mL TMP',
                              'M9 no supp.')) +
  xlab("Fitness (Log2 Fold Change)") +
  ylab("Selection Condition") +
  ggtitle("Library 15 Fitness - Day 2") +
  #theme_cowplot() +
  #background_grid() +
  theme(axis.text.x = element_text(size = 12),
        axis.text.y = element_blank(),
        axis.title.x = element_text(size = 14),
        axis.title.y = element_blank()) +
  theme(legend.position = "none") +
  scale_x_continuous(limits = c(-15,10)) +
  guides(linetype = FALSE) +
  panel_border(color = "black")

# Day 3
p9_15_d3 <- ggplot(perfects15_tree_5BCs_d3, aes(x = val, y = factor(fc, level = lib15_level_order_d3))) + 
  geom_density_ridges(fill = "aquamarine4", color = "black") +
  scale_y_discrete(labels = c('200 ug/mL TMP',
                              '50 ug/mL TMP',
                              '10 ug/mL TMP',
                              '1 ug/mL TMP',
                              '0.5 ug/mL TMP',
                              '0.058 ug/mL TMP',
                              'M9 no supp.')) +
  xlab("Fitness (Log2 Fold Change)") +
  ylab("Selection Condition") +
  ggtitle("Library 15 Fitness - Day 3") +
  #theme_cowplot() +
  #background_grid() +
  theme(axis.text.x = element_text(size = 12),
        axis.text.y = element_blank(),
        axis.title.x = element_text(size = 14),
        axis.title.y = element_blank()) +
  theme(legend.position = "none") +
  scale_x_continuous(limits = c(-15,10)) +
  guides(linetype = FALSE) +
  panel_border(color = "black")

# Day 4
p9_15_d4 <- ggplot(perfects15_tree_5BCs_d4, aes(x = val, y = factor(fc, level = lib15_level_order_d4))) + 
  geom_density_ridges(fill = "aquamarine4", color = "black") +
  scale_y_discrete(labels = c('200 ug/mL TMP',
                              '50 ug/mL TMP',
                              '10 ug/mL TMP',
                              '1 ug/mL TMP',
                              '0.5 ug/mL TMP',
                              '0.058 ug/mL TMP',
                              'M9 no supp.')) +
  xlab("Fitness (Log2 Fold Change)") +
  ylab("Selection Condition") +
  ggtitle("Library 15 Fitness - Day 4") +
  #theme_cowplot() +
  #background_grid() +
  theme(axis.text.x = element_text(size = 12),
        axis.text.y = element_blank(),
        axis.title.x = element_text(size = 14),
        axis.title.y = element_blank()) +
  theme(legend.position = "none") +
  scale_x_continuous(limits = c(-15,10)) +
  guides(linetype = FALSE) +
  panel_border(color = "black")

p9_15_d1
p9_15_d2
p9_15_d3
p9_15_d4
```

**Library 16**
```{r}
# Day 1
p9_16_d1 <- ggplot(perfects16_tree_5BCs_d1, aes(x = val, y = factor(fc, level = lib16_level_order_d1))) + 
  geom_density_ridges(fill = "aquamarine4", color = "black") +
  scale_y_discrete(labels = c('200 ug/mL TMP',
                              '50 ug/mL TMP',
                              '10 ug/mL TMP',
                              '1 ug/mL TMP',
                              '0.5 ug/mL TMP',
                              '0.058 ug/mL TMP',
                              'M9 no supp.')) +
  xlab("Fitness (Log2 Fold Change)") +
  ylab("Selection Condition") +
  ggtitle("Library 16 Fitness - Day 1") +
  #theme_cowplot() +
  #background_grid() +
  theme(axis.text.x = element_text(size = 12),
        axis.text.y = element_text(size = 12),
        axis.title.x = element_text(size = 14),
        axis.title.y = element_text(size = 14)) +
  theme(legend.position = "none") +
  scale_x_continuous(limits = c(-15,10)) +
  guides(linetype = FALSE) +
  panel_border(color = "black")

# Day 2
p9_16_d2 <- ggplot(perfects16_tree_5BCs_d2, aes(x = val, y = factor(fc, level = lib16_level_order_d2))) + 
  geom_density_ridges(fill = "aquamarine4", color = "black") +
  scale_y_discrete(labels = c('200 ug/mL TMP',
                              '50 ug/mL TMP',
                              '10 ug/mL TMP',
                              '1 ug/mL TMP',
                              '0.5 ug/mL TMP',
                              '0.058 ug/mL TMP',
                              'M9 no supp.')) +
  xlab("Fitness (Log2 Fold Change)") +
  ylab("Selection Condition") +
  ggtitle("Library 16 Fitness - Day 2") +
  #theme_cowplot() +
  #background_grid() +
  theme(axis.text.x = element_text(size = 12),
        axis.text.y = element_blank(),
        axis.title.x = element_text(size = 14),
        axis.title.y = element_blank()) +
  theme(legend.position = "none") +
  scale_x_continuous(limits = c(-15,10)) +
  guides(linetype = FALSE) +
  panel_border(color = "black")

# Day 3
p9_16_d3 <- ggplot(perfects16_tree_5BCs_d3, aes(x = val, y = factor(fc, level = lib16_level_order_d3))) + 
  geom_density_ridges(fill = "aquamarine4", color = "black") +
  scale_y_discrete(labels = c('200 ug/mL TMP',
                              '50 ug/mL TMP',
                              '10 ug/mL TMP',
                              '1 ug/mL TMP',
                              '0.5 ug/mL TMP',
                              '0.058 ug/mL TMP',
                              'M9 no supp.')) +
  xlab("Fitness (Log2 Fold Change)") +
  ylab("Selection Condition") +
  ggtitle("Library 16 Fitness - Day 3") +
  #theme_cowplot() +
  #background_grid() +
  theme(axis.text.x = element_text(size = 12),
        axis.text.y = element_blank(),
        axis.title.x = element_text(size = 14),
        axis.title.y = element_blank()) +
  theme(legend.position = "none") +
  scale_x_continuous(limits = c(-15,10)) +
  guides(linetype = FALSE) +
  panel_border(color = "black")

# Day 4
p9_16_d4 <- ggplot(perfects16_tree_5BCs_d4, aes(x = val, y = factor(fc, level = lib16_level_order_d4))) + 
  geom_density_ridges(fill = "aquamarine4", color = "black") +
  scale_y_discrete(labels = c('200 ug/mL TMP',
                              '50 ug/mL TMP',
                              '10 ug/mL TMP',
                              '1 ug/mL TMP',
                              '0.5 ug/mL TMP',
                              '0.058 ug/mL TMP',
                              'M9 no supp.')) +
  xlab("Fitness (Log2 Fold Change)") +
  ylab("Selection Condition") +
  ggtitle("Library 16 Fitness - Day 4") +
  #theme_cowplot() +
  #background_grid() +
  theme(axis.text.x = element_text(size = 12),
        axis.text.y = element_blank(),
        axis.title.x = element_text(size = 14),
        axis.title.y = element_blank()) +
  theme(legend.position = "none") +
  scale_x_continuous(limits = c(-15,10)) +
  guides(linetype = FALSE) +
  panel_border(color = "black")

p9_16_d1
p9_16_d2
p9_16_d3
p9_16_d4
```

```{r echo=FALSE}
patch9 <- (p9_15_d1 | p9_15_d2 | p9_15_d3 | p9_15_d4) / (p9_16_d1 | p9_16_d2 | p9_16_d3 | p9_16_d4)
patch9
```

```{r echo=FALSE}
ggsave(file="plots/combined/lib15.16.supplement.fitness.by.day.v2.png", plot=patch9,
       dpi=600, width = 14, height = 6, units = "in")
```

### Perfects Dose Response

Plot the number of perfects barcodes present under each Trimethoprim selection condition.

#### Dose Curves

**Library 15**

Filter the Perfects counts to include only those perfects with > 200 BCs, then select only the dose responses for trimethoprim concentrations on Day 1.
```{r}
perfects15_5BCs_dr1 <- perfects15_5BCs %>%
  filter(numprunedBCs>200) %>%
  select(ID,
         fitD06D03,
         fitD07D03,
         fitD08D03,
         fitD09D03,
         fitD10D03,
         fitD11D03,
         numprunedBCs) %>%
  gather("selection","fitness",-ID,-numprunedBCs)
```

Define the trimethoprim concentrations based on the fitness codes associated with sample IDs.
```{r}
perfects15_5BCs_dr1$tmp <- 0
perfects15_5BCs_dr1$tmp[perfects15_5BCs_dr1$selection== "fitD06D03"] <- 0.058
perfects15_5BCs_dr1$tmp[perfects15_5BCs_dr1$selection== "fitD07D03"] <- 0.5
perfects15_5BCs_dr1$tmp[perfects15_5BCs_dr1$selection== "fitD08D03"] <- 1
perfects15_5BCs_dr1$tmp[perfects15_5BCs_dr1$selection== "fitD09D03"] <- 10
perfects15_5BCs_dr1$tmp[perfects15_5BCs_dr1$selection== "fitD10D03"] <- 50
perfects15_5BCs_dr1$tmp[perfects15_5BCs_dr1$selection== "fitD11D03"] <- 200
```

Plot the dose response for remaining perfects (>200 BCs) over the trimethoprim gradient.
```{r}
p10_15_d1 <- ggplot(perfects15_5BCs_dr1, aes(x=tmp, y=fitness, color=ID)) +
  geom_point() +
  geom_line() +
  xlab("[Trimethoprim] (ug/mL)") +
  ylab("Fitness") +
  ggtitle("Library 15 Dose Response - Day 1") +
  scale_x_log10() +
  theme(axis.text.x = element_text(size = 12),
        axis.text.y = element_text(size = 12),
        axis.title.x = element_text(size = 14),
        axis.title.y = element_text(size = 14)) +
  theme(legend.position = "none") +
  scale_y_continuous(expand = c(0, 0), limits = c(-11,5)) +
  guides(linetype = FALSE) +
  panel_border(color = "black")

p10_15_d1
```

```{r echo=FALSE}
# Lib15
ggsave(file="plots/perfects15_5BCs_200BC_dose_response_day1.png", plot=p10_15_d1,
       dpi=600, width = 8, height = 7.5, units = "in")
```

**Library 16**

Filter the Perfects counts to include only those perfects with > 200 BCs, then select only the dose responses for trimethoprim concentrations on Day 1.
```{r}
perfects16_5BCs_dr1 <- perfects16_5BCs %>%
  filter(numprunedBCs>200) %>%
  select(ID,
         fitE01D04,
         fitE02D04,
         fitE03D04,
         fitE04D04,
         fitE05D04,
         fitE06D04,
         numprunedBCs) %>%
  gather("selection","fitness",-ID,-numprunedBCs)
```

Define the trimethoprim concentrations based on the fitness codes associated with sample IDs.
```{r}
perfects16_5BCs_dr1$tmp <- 0
perfects16_5BCs_dr1$tmp[perfects16_5BCs_dr1$selection== "fitE01D04"] <- 0.058
perfects16_5BCs_dr1$tmp[perfects16_5BCs_dr1$selection== "fitE02D04"] <- 0.5
perfects16_5BCs_dr1$tmp[perfects16_5BCs_dr1$selection== "fitE03D04"] <- 1
perfects16_5BCs_dr1$tmp[perfects16_5BCs_dr1$selection== "fitE04D04"] <- 10
perfects16_5BCs_dr1$tmp[perfects16_5BCs_dr1$selection== "fitE05D04"] <- 50
perfects16_5BCs_dr1$tmp[perfects16_5BCs_dr1$selection== "fitE06D04"] <- 200
```

Plot the dose response for remaining perfects (>200 BCs) over the trimethoprim gradient.
```{r}
p10_16_d1 <- ggplot(perfects16_5BCs_dr1, aes(x=tmp, y=fitness, color=ID)) +
  geom_point() +
  geom_line() +
  xlab("[Trimethoprim] (ug/mL)") +
  ylab("Fitness") +
  ggtitle("Library 16 Dose Response - Day 1") +
  scale_x_log10() +
  theme(axis.text.x = element_text(size = 12),
        axis.text.y = element_text(size = 12),
        axis.title.x = element_text(size = 14),
        axis.title.y = element_text(size = 14)) +
  theme(legend.position = "none") +
  scale_y_continuous(expand = c(0, 0), limits = c(-11,5)) +
  guides(linetype = FALSE) +
  panel_border(color = "black")

p10_16_d1
```

```{r echo=FALSE}
# Lib15
ggsave(file="plots/perfects16_5BCs_200BC_dose_response_day1.png", plot=p10_16_d1,
       dpi=600, width = 8, height = 7.5, units = "in")
```

```{r echo=FALSE}
patch10 <- (p10_15_d1 | p10_16_d1)
patch10
```

```{r echo=FALSE}
ggsave(file="plots/combined/lib15.16.dose.response.curves.day1.png", plot=patch10,
       dpi=600, width = 10, height = 6, units = "in")
```

#### Dose Boxplots

**Library 15 - Day 1**

Filter the Perfects counts to include only those perfects with > 200 BCs, then select only the dose responses for trimethoprim concentrations on Day 1.
```{r}
perfects15_200BCs_dr1 <- perfects15_5BCs %>%
  filter(numprunedBCs>200) %>%
  select(ID, fitD05D03, fitD06D03, fitD07D03, fitD08D03, fitD09D03, fitD10D03, fitD11D03, numprunedBCs) %>%
  gather("selection","fitness",-ID,-numprunedBCs)
```

Re-Define the trimethoprim concentrations based on the fitness codes associated with sample IDs.
```{r}
perfects15_200BCs_dr1$tmpfactor <- "0"
perfects15_200BCs_dr1$tmpfactor[perfects15_200BCs_dr1$selection== "fitD05D03"] <- "0"
perfects15_200BCs_dr1$tmpfactor[perfects15_200BCs_dr1$selection== "fitD06D03"] <- "0.058"
perfects15_200BCs_dr1$tmpfactor[perfects15_200BCs_dr1$selection== "fitD07D03"] <- "0.5"
perfects15_200BCs_dr1$tmpfactor[perfects15_200BCs_dr1$selection== "fitD08D03"] <- "1"
perfects15_200BCs_dr1$tmpfactor[perfects15_200BCs_dr1$selection== "fitD09D03"] <- "10"
perfects15_200BCs_dr1$tmpfactor[perfects15_200BCs_dr1$selection== "fitD10D03"] <- "50"
perfects15_200BCs_dr1$tmpfactor[perfects15_200BCs_dr1$selection== "fitD11D03"] <- "200"

perfects15_200BCs_dr1$tmpname <- "0"
perfects15_200BCs_dr1$tmpname[perfects15_200BCs_dr1$selection== "fitD05D03"] <- "A"
perfects15_200BCs_dr1$tmpname[perfects15_200BCs_dr1$selection== "fitD06D03"] <- "B"
perfects15_200BCs_dr1$tmpname[perfects15_200BCs_dr1$selection== "fitD07D03"] <- "C"
perfects15_200BCs_dr1$tmpname[perfects15_200BCs_dr1$selection== "fitD08D03"] <- "D"
perfects15_200BCs_dr1$tmpname[perfects15_200BCs_dr1$selection== "fitD09D03"] <- "E"
perfects15_200BCs_dr1$tmpname[perfects15_200BCs_dr1$selection== "fitD10D03"] <- "F"
perfects15_200BCs_dr1$tmpname[perfects15_200BCs_dr1$selection== "fitD11D03"] <- "G"

perfects15_200BCs_dr_labels <- c("0","0.058","0.5","1","10","50","200")
```

Plot the dose response for remaining perfects (>200 BCs) as boxplots for each trimethoprim concentration.
```{r}
p11_15_d1 <- ggplot(perfects15_200BCs_dr1, aes(x=tmpname, y=fitness)) +
  geom_boxplot(color="black", fill="aquamarine4") +
  xlab("[Trimethoprim] (ug/mL)") +
  ylab("Fitness") +
  ggtitle("Library 15 Dose Response - Day 1") +
  scale_x_discrete(labels = perfects15_200BCs_dr_labels) +
  theme(axis.text.x = element_text(size = 12),
        axis.text.y = element_text(size = 12),
        axis.title.x = element_text(size = 14),
        axis.title.y = element_text(size = 14)) +
  theme(legend.position = "none") +
  scale_y_continuous(expand = c(0, 0), limits = c(-11,5)) +
  guides(linetype = FALSE) +
  panel_border(color = "black")

p11_15_d1
```

**Library 15 - Day 2**

Filter the Perfects counts to include only those perfects with > 200 BCs, then select only the dose responses for trimethoprim concentrations on Day 2.
```{r}
perfects15_200BCs_dr2 <- perfects15_5BCs %>%
  filter(numprunedBCs>200) %>%
  select(ID,
         fitE07D03,
         fitE08D03,
         fitE09D03,
         fitE10D03,
         fitE11D03,
         fitE12D03,
         fitF01D03,
         numprunedBCs) %>%
  gather("selection","fitness",-ID,-numprunedBCs)
```

Re-Define the trimethoprim concentrations based on the fitness codes associated with sample IDs.
```{r}
perfects15_200BCs_dr2$tmpfactor <- "0"
perfects15_200BCs_dr2$tmpfactor[perfects15_200BCs_dr2$selection== "fitE07D03"] <- "0"
perfects15_200BCs_dr2$tmpfactor[perfects15_200BCs_dr2$selection== "fitE08D03"] <- "0.058"
perfects15_200BCs_dr2$tmpfactor[perfects15_200BCs_dr2$selection== "fitE09D03"] <- "0.5"
perfects15_200BCs_dr2$tmpfactor[perfects15_200BCs_dr2$selection== "fitE10D03"] <- "1"
perfects15_200BCs_dr2$tmpfactor[perfects15_200BCs_dr2$selection== "fitE11D03"] <- "10"
perfects15_200BCs_dr2$tmpfactor[perfects15_200BCs_dr2$selection== "fitE12D03"] <- "50"
perfects15_200BCs_dr2$tmpfactor[perfects15_200BCs_dr2$selection== "fitF01D03"] <- "200"

perfects15_200BCs_dr2$tmpname <- "0"
perfects15_200BCs_dr2$tmpname[perfects15_200BCs_dr2$selection== "fitE07D03"] <- "A"
perfects15_200BCs_dr2$tmpname[perfects15_200BCs_dr2$selection== "fitE08D03"] <- "B"
perfects15_200BCs_dr2$tmpname[perfects15_200BCs_dr2$selection== "fitE09D03"] <- "C"
perfects15_200BCs_dr2$tmpname[perfects15_200BCs_dr2$selection== "fitE10D03"] <- "D"
perfects15_200BCs_dr2$tmpname[perfects15_200BCs_dr2$selection== "fitE11D03"] <- "E"
perfects15_200BCs_dr2$tmpname[perfects15_200BCs_dr2$selection== "fitE12D03"] <- "F"
perfects15_200BCs_dr2$tmpname[perfects15_200BCs_dr2$selection== "fitF01D03"] <- "G"

perfects15_200BCs_dr_labels <- c("0","0.058","0.5","1","10","50","200")
```

Plot the dose response for remaining perfects (>200 BCs) as boxplots for each trimethoprim concentration.
```{r}
p11_15_d2 <- ggplot(perfects15_200BCs_dr2, aes(x=tmpname, y=fitness)) +
  geom_boxplot(color="black", fill="aquamarine4") +
  xlab("[Trimethoprim] (ug/mL)") +
  ylab("Fitness") +
  ggtitle("Library 15 Dose Response - Day 2") +
  scale_x_discrete(labels = perfects15_200BCs_dr_labels) +
  theme(axis.text.x = element_text(size = 12),
        axis.text.y = element_text(size = 12),
        axis.title.x = element_text(size = 14),
        axis.title.y = element_text(size = 14)) +
  theme(legend.position = "none") +
  scale_y_continuous(expand = c(0, 0), limits = c(-11,5)) +
  guides(linetype = FALSE) +
  panel_border(color = "black")

p11_15_d2
```

**Library 15 - Day 3**

Filter the Perfects counts to include only those perfects with > 200 BCs, then select only the dose responses for trimethoprim concentrations on Day 3.
```{r}
perfects15_200BCs_dr3 <- perfects15_5BCs %>%
  filter(numprunedBCs>200) %>%
  select(ID,
         fitF09D03,
         fitF10D03,
         fitF11D03,
         fitF12D03,
         fitG01D03,
         fitG02D03,
         fitG03D03,
         numprunedBCs) %>%
  gather("selection","fitness",-ID,-numprunedBCs)
```

Re-Define the trimethoprim concentrations based on the fitness codes associated with sample IDs.
```{r}
perfects15_200BCs_dr3$tmpfactor <- "0"
perfects15_200BCs_dr3$tmpfactor[perfects15_200BCs_dr3$selection== "fitF09D03"] <- "0"
perfects15_200BCs_dr3$tmpfactor[perfects15_200BCs_dr3$selection== "fitF10D03"] <- "0.058"
perfects15_200BCs_dr3$tmpfactor[perfects15_200BCs_dr3$selection== "fitF11D03"] <- "0.5"
perfects15_200BCs_dr3$tmpfactor[perfects15_200BCs_dr3$selection== "fitF12D03"] <- "1"
perfects15_200BCs_dr3$tmpfactor[perfects15_200BCs_dr3$selection== "fitG01D03"] <- "10"
perfects15_200BCs_dr3$tmpfactor[perfects15_200BCs_dr3$selection== "fitG02D03"] <- "50"
perfects15_200BCs_dr3$tmpfactor[perfects15_200BCs_dr3$selection== "fitG03D03"] <- "200"

perfects15_200BCs_dr3$tmpname <- "0"
perfects15_200BCs_dr3$tmpname[perfects15_200BCs_dr3$selection== "fitF09D03"] <- "A"
perfects15_200BCs_dr3$tmpname[perfects15_200BCs_dr3$selection== "fitF10D03"] <- "B"
perfects15_200BCs_dr3$tmpname[perfects15_200BCs_dr3$selection== "fitF11D03"] <- "C"
perfects15_200BCs_dr3$tmpname[perfects15_200BCs_dr3$selection== "fitF12D03"] <- "D"
perfects15_200BCs_dr3$tmpname[perfects15_200BCs_dr3$selection== "fitG01D03"] <- "E"
perfects15_200BCs_dr3$tmpname[perfects15_200BCs_dr3$selection== "fitG02D03"] <- "F"
perfects15_200BCs_dr3$tmpname[perfects15_200BCs_dr3$selection== "fitG03D03"] <- "G"

perfects15_200BCs_dr_labels <- c("0","0.058","0.5","1","10","50","200")
```

Plot the dose response for remaining perfects (>200 BCs) as boxplots for each trimethoprim concentration.
```{r}
p11_15_d3 <- ggplot(perfects15_200BCs_dr3, aes(x=tmpname, y=fitness)) +
  geom_boxplot(color="black", fill="aquamarine4") +
  xlab("[Trimethoprim] (ug/mL)") +
  ylab("Fitness") +
  ggtitle("Library 15 Dose Response - Day 3") +
  scale_x_discrete(labels = perfects15_200BCs_dr_labels) +
  theme(axis.text.x = element_text(size = 12),
        axis.text.y = element_text(size = 12),
        axis.title.x = element_text(size = 14),
        axis.title.y = element_text(size = 14)) +
  theme(legend.position = "none") +
  scale_y_continuous(expand = c(0, 0), limits = c(-11,5)) +
  guides(linetype = FALSE) +
  panel_border(color = "black")

p11_15_d3
```

**Library 15 - Day 4**

Filter the Perfects counts to include only those perfects with > 200 BCs, then select only the dose responses for trimethoprim concentrations on Day 4.
```{r}
perfects15_200BCs_dr4 <- perfects15_5BCs %>%
  filter(numprunedBCs>200) %>%
  select(ID,
         fitG11D03,
         fitG12D03,
         fitH01D03,
         fitH02D03,
         fitH03D03,
         fitH04D03,
         fitH05D03,
         numprunedBCs) %>%
  gather("selection","fitness",-ID,-numprunedBCs)
```

Re-Define the trimethoprim concentrations based on the fitness codes associated with sample IDs.
```{r}
perfects15_200BCs_dr4$tmpfactor <- "0"
perfects15_200BCs_dr4$tmpfactor[perfects15_200BCs_dr4$selection== "fitG11D03"] <- "0"
perfects15_200BCs_dr4$tmpfactor[perfects15_200BCs_dr4$selection== "fitG12D03"] <- "0.058"
perfects15_200BCs_dr4$tmpfactor[perfects15_200BCs_dr4$selection== "fitH01D03"] <- "0.5"
perfects15_200BCs_dr4$tmpfactor[perfects15_200BCs_dr4$selection== "fitH02D03"] <- "1"
perfects15_200BCs_dr4$tmpfactor[perfects15_200BCs_dr4$selection== "fitH03D03"] <- "10"
perfects15_200BCs_dr4$tmpfactor[perfects15_200BCs_dr4$selection== "fitH04D03"] <- "50"
perfects15_200BCs_dr4$tmpfactor[perfects15_200BCs_dr4$selection== "fitH05D03"] <- "200"

perfects15_200BCs_dr4$tmpname <- "0"
perfects15_200BCs_dr4$tmpname[perfects15_200BCs_dr4$selection== "fitG11D03"] <- "A"
perfects15_200BCs_dr4$tmpname[perfects15_200BCs_dr4$selection== "fitG12D03"] <- "B"
perfects15_200BCs_dr4$tmpname[perfects15_200BCs_dr4$selection== "fitH01D03"] <- "C"
perfects15_200BCs_dr4$tmpname[perfects15_200BCs_dr4$selection== "fitH02D03"] <- "D"
perfects15_200BCs_dr4$tmpname[perfects15_200BCs_dr4$selection== "fitH03D03"] <- "E"
perfects15_200BCs_dr4$tmpname[perfects15_200BCs_dr4$selection== "fitH04D03"] <- "F"
perfects15_200BCs_dr4$tmpname[perfects15_200BCs_dr4$selection== "fitH05D03"] <- "G"

perfects15_200BCs_dr_labels <- c("0","0.058","0.5","1","10","50","200")
```

Plot the dose response for remaining perfects (>200 BCs) as boxplots for each trimethoprim concentration.
```{r}
p11_15_d4 <- ggplot(perfects15_200BCs_dr4, aes(x=tmpname, y=fitness)) +
  geom_boxplot(color="black", fill="aquamarine4") +
  xlab("[Trimethoprim] (ug/mL)") +
  ylab("Fitness") +
  ggtitle("Library 15 Dose Response - Day 4") +
  scale_x_discrete(labels = perfects15_200BCs_dr_labels) +
  theme(axis.text.x = element_text(size = 12),
        axis.text.y = element_text(size = 12),
        axis.title.x = element_text(size = 14),
        axis.title.y = element_text(size = 14)) +
  theme(legend.position = "none") +
  scale_y_continuous(expand = c(0, 0), limits = c(-11,5)) +
  guides(linetype = FALSE) +
  panel_border(color = "black")

p11_15_d4
```

```{r echo=FALSE}
patch11_Lib15 <- (p11_15_d1 | p11_15_d2 | p11_15_d3 | p11_15_d4)
patch11_Lib15
```

```{r echo=FALSE}
ggsave(file="plots/combined/lib15.dose.response.by.day.png", plot=patch11_Lib15,
       dpi=600, width = 14, height = 6, units = "in")
```

**Library 16 - Day 1**

Filter the Perfects counts to include only those perfects with > 200 BCs, then select only the dose responses for trimethoprim concentrations on Day 1.
```{r}
perfects16_200BCs_dr1 <- perfects16_5BCs %>%
  filter(numprunedBCs>200) %>%
  select(ID,
         fitD12D04,
         fitE01D04,
         fitE02D04,
         fitE03D04,
         fitE04D04,
         fitE05D04,
         fitE06D04,
         numprunedBCs) %>%
  gather("selection","fitness",-ID,-numprunedBCs)
```

Re-Define the trimethoprim concentrations based on the fitness codes associated with sample IDs.
```{r}
perfects16_200BCs_dr1$tmpfactor <- "0"
perfects16_200BCs_dr1$tmpfactor[perfects16_200BCs_dr1$selection== "fitD12D04"] <- "0"
perfects16_200BCs_dr1$tmpfactor[perfects16_200BCs_dr1$selection== "fitE01D04"] <- "0.058"
perfects16_200BCs_dr1$tmpfactor[perfects16_200BCs_dr1$selection== "fitE02D04"] <- "0.5"
perfects16_200BCs_dr1$tmpfactor[perfects16_200BCs_dr1$selection== "fitE03D04"] <- "1"
perfects16_200BCs_dr1$tmpfactor[perfects16_200BCs_dr1$selection== "fitE04D04"] <- "10"
perfects16_200BCs_dr1$tmpfactor[perfects16_200BCs_dr1$selection== "fitE05D04"] <- "50"
perfects16_200BCs_dr1$tmpfactor[perfects16_200BCs_dr1$selection== "fitE06D04"] <- "200"

perfects16_200BCs_dr1$tmpname <- "0"
perfects16_200BCs_dr1$tmpname[perfects16_200BCs_dr1$selection== "fitD12D04"] <- "A"
perfects16_200BCs_dr1$tmpname[perfects16_200BCs_dr1$selection== "fitE01D04"] <- "B"
perfects16_200BCs_dr1$tmpname[perfects16_200BCs_dr1$selection== "fitE02D04"] <- "C"
perfects16_200BCs_dr1$tmpname[perfects16_200BCs_dr1$selection== "fitE03D04"] <- "D"
perfects16_200BCs_dr1$tmpname[perfects16_200BCs_dr1$selection== "fitE04D04"] <- "E"
perfects16_200BCs_dr1$tmpname[perfects16_200BCs_dr1$selection== "fitE05D04"] <- "F"
perfects16_200BCs_dr1$tmpname[perfects16_200BCs_dr1$selection== "fitE06D04"] <- "G"

perfects16_200BCs_dr_labels <- c("0","0.058","0.5","1","10","50","200")
```

Plot the dose response for remaining perfects (>200 BCs) as boxplots for each trimethoprim concentration.
```{r}
p11_16_d1 <- ggplot(perfects16_200BCs_dr1, aes(x=tmpname, y=fitness)) +
  geom_boxplot(color="black", fill="aquamarine4") +
  xlab("[Trimethoprim] (ug/mL)") +
  ylab("Fitness") +
  ggtitle("Library 16 Dose Response - Day 1") +
  scale_x_discrete(labels = perfects16_200BCs_dr_labels) +
  theme(axis.text.x = element_text(size = 12),
        axis.text.y = element_text(size = 12),
        axis.title.x = element_text(size = 14),
        axis.title.y = element_text(size = 14)) +
  theme(legend.position = "none") +
  scale_y_continuous(expand = c(0, 0), limits = c(-11,5)) +
  guides(linetype = FALSE) +
  panel_border(color = "black")

p11_16_d1
```

**Library 16 - Day 2**

Filter the Perfects counts to include only those perfects with > 200 BCs, then select only the dose responses for trimethoprim concentrations on Day 2.
```{r}
perfects16_200BCs_dr2 <- perfects16_5BCs %>%
  filter(numprunedBCs>200) %>%
  select(ID,
         fitF02D04,
         fitF03D04,
         fitF04D04,
         fitF05D04,
         fitF06D04,
         fitF07D04,
         fitF08D04,
         numprunedBCs) %>%
  gather("selection","fitness",-ID,-numprunedBCs)
```

Re-Define the trimethoprim concentrations based on the fitness codes associated with sample IDs.
```{r}
perfects16_200BCs_dr2$tmpfactor <- "0"
perfects16_200BCs_dr2$tmpfactor[perfects16_200BCs_dr2$selection== "fitF02D04"] <- "0"
perfects16_200BCs_dr2$tmpfactor[perfects16_200BCs_dr2$selection== "fitF03D04"] <- "0.058"
perfects16_200BCs_dr2$tmpfactor[perfects16_200BCs_dr2$selection== "fitF04D04"] <- "0.5"
perfects16_200BCs_dr2$tmpfactor[perfects16_200BCs_dr2$selection== "fitF05D04"] <- "1"
perfects16_200BCs_dr2$tmpfactor[perfects16_200BCs_dr2$selection== "fitF06D04"] <- "10"
perfects16_200BCs_dr2$tmpfactor[perfects16_200BCs_dr2$selection== "fitF07D04"] <- "50"
perfects16_200BCs_dr2$tmpfactor[perfects16_200BCs_dr2$selection== "fitF08D04"] <- "200"

perfects16_200BCs_dr2$tmpname <- "0"
perfects16_200BCs_dr2$tmpname[perfects16_200BCs_dr2$selection== "fitF02D04"] <- "A"
perfects16_200BCs_dr2$tmpname[perfects16_200BCs_dr2$selection== "fitF03D04"] <- "B"
perfects16_200BCs_dr2$tmpname[perfects16_200BCs_dr2$selection== "fitF04D04"] <- "C"
perfects16_200BCs_dr2$tmpname[perfects16_200BCs_dr2$selection== "fitF05D04"] <- "D"
perfects16_200BCs_dr2$tmpname[perfects16_200BCs_dr2$selection== "fitF06D04"] <- "E"
perfects16_200BCs_dr2$tmpname[perfects16_200BCs_dr2$selection== "fitF07D04"] <- "F"
perfects16_200BCs_dr2$tmpname[perfects16_200BCs_dr2$selection== "fitF08D04"] <- "G"

perfects16_200BCs_dr_labels <- c("0","0.058","0.5","1","10","50","200")
```

Plot the dose response for remaining perfects (>200 BCs) as boxplots for each trimethoprim concentration.
```{r}
p11_16_d2 <- ggplot(perfects16_200BCs_dr2, aes(x=tmpname, y=fitness)) +
  geom_boxplot(color="black", fill="aquamarine4") +
  xlab("[Trimethoprim] (ug/mL)") +
  ylab("Fitness") +
  ggtitle("Library 16 Dose Response - Day 2") +
  scale_x_discrete(labels = perfects16_200BCs_dr_labels) +
  theme(axis.text.x = element_text(size = 12),
        axis.text.y = element_text(size = 12),
        axis.title.x = element_text(size = 14),
        axis.title.y = element_text(size = 14)) +
  theme(legend.position = "none") +
  scale_y_continuous(expand = c(0, 0), limits = c(-11,5)) +
  guides(linetype = FALSE) +
  panel_border(color = "black")

p11_16_d2
```

**Library 16 - Day 3**

Filter the Perfects counts to include only those perfects with > 200 BCs, then select only the dose responses for trimethoprim concentrations on Day 3.
```{r}
perfects16_200BCs_dr3 <- perfects16_5BCs %>%
  filter(numprunedBCs>200) %>%
  select(ID,
         fitG04D04,
         fitG05D04,
         fitG06D04,
         fitG07D04,
         fitG08D04,
         fitG09D04,
         fitG10D04,
         numprunedBCs) %>%
  gather("selection","fitness",-ID,-numprunedBCs)
```

Re-Define the trimethoprim concentrations based on the fitness codes associated with sample IDs.
```{r}
perfects16_200BCs_dr3$tmpfactor <- "0"
perfects16_200BCs_dr3$tmpfactor[perfects16_200BCs_dr3$selection== "fitG04D04"] <- "0"
perfects16_200BCs_dr3$tmpfactor[perfects16_200BCs_dr3$selection== "fitG05D04"] <- "0.058"
perfects16_200BCs_dr3$tmpfactor[perfects16_200BCs_dr3$selection== "fitG06D04"] <- "0.5"
perfects16_200BCs_dr3$tmpfactor[perfects16_200BCs_dr3$selection== "fitG07D04"] <- "1"
perfects16_200BCs_dr3$tmpfactor[perfects16_200BCs_dr3$selection== "fitG08D04"] <- "10"
perfects16_200BCs_dr3$tmpfactor[perfects16_200BCs_dr3$selection== "fitG09D04"] <- "50"
perfects16_200BCs_dr3$tmpfactor[perfects16_200BCs_dr3$selection== "fitG10D04"] <- "200"

perfects16_200BCs_dr3$tmpname <- "0"
perfects16_200BCs_dr3$tmpname[perfects16_200BCs_dr3$selection== "fitG04D04"] <- "A"
perfects16_200BCs_dr3$tmpname[perfects16_200BCs_dr3$selection== "fitG05D04"] <- "B"
perfects16_200BCs_dr3$tmpname[perfects16_200BCs_dr3$selection== "fitG06D04"] <- "C"
perfects16_200BCs_dr3$tmpname[perfects16_200BCs_dr3$selection== "fitG07D04"] <- "D"
perfects16_200BCs_dr3$tmpname[perfects16_200BCs_dr3$selection== "fitG08D04"] <- "E"
perfects16_200BCs_dr3$tmpname[perfects16_200BCs_dr3$selection== "fitG09D04"] <- "F"
perfects16_200BCs_dr3$tmpname[perfects16_200BCs_dr3$selection== "fitG10D04"] <- "G"

perfects16_200BCs_dr_labels <- c("0","0.058","0.5","1","10","50","200")
```

Plot the dose response for remaining perfects (>200 BCs) as boxplots for each trimethoprim concentration.
```{r}
p11_16_d3 <- ggplot(perfects16_200BCs_dr3, aes(x=tmpname, y=fitness)) +
  geom_boxplot(color="black", fill="aquamarine4") +
  xlab("[Trimethoprim] (ug/mL)") +
  ylab("Fitness") +
  ggtitle("Library 16 Dose Response - Day 3") +
  scale_x_discrete(labels = perfects16_200BCs_dr_labels) +
  theme(axis.text.x = element_text(size = 12),
        axis.text.y = element_text(size = 12),
        axis.title.x = element_text(size = 14),
        axis.title.y = element_text(size = 14)) +
  theme(legend.position = "none") +
  scale_y_continuous(expand = c(0, 0), limits = c(-11,5)) +
  guides(linetype = FALSE) +
  panel_border(color = "black")

p11_16_d3
```

**Library 16 - Day 4**

Filter the Perfects counts to include only those perfects with > 200 BCs, then select only the dose responses for trimethoprim concentrations on Day 4.
```{r}
perfects16_200BCs_dr4 <- perfects16_5BCs %>%
  filter(numprunedBCs>200) %>%
  select(ID,
         fitH06D04,
         fitH07D04,
         fitH08D04,
         fitH09D04,
         fitH10D04,
         fitH11D04,
         fitH12D04,
         numprunedBCs) %>%
  gather("selection","fitness",-ID,-numprunedBCs)
```

Re-Define the trimethoprim concentrations based on the fitness codes associated with sample IDs.
```{r}
perfects16_200BCs_dr4$tmpfactor <- "0"
perfects16_200BCs_dr4$tmpfactor[perfects16_200BCs_dr4$selection== "fitH06D04"] <- "0"
perfects16_200BCs_dr4$tmpfactor[perfects16_200BCs_dr4$selection== "fitH07D04"] <- "0.058"
perfects16_200BCs_dr4$tmpfactor[perfects16_200BCs_dr4$selection== "fitH08D04"] <- "0.5"
perfects16_200BCs_dr4$tmpfactor[perfects16_200BCs_dr4$selection== "fitH09D04"] <- "1"
perfects16_200BCs_dr4$tmpfactor[perfects16_200BCs_dr4$selection== "fitH10D04"] <- "10"
perfects16_200BCs_dr4$tmpfactor[perfects16_200BCs_dr4$selection== "fitH11D04"] <- "50"
perfects16_200BCs_dr4$tmpfactor[perfects16_200BCs_dr4$selection== "fitH12D04"] <- "200"

perfects16_200BCs_dr4$tmpname <- "0"
perfects16_200BCs_dr4$tmpname[perfects16_200BCs_dr4$selection== "fitH06D04"] <- "A"
perfects16_200BCs_dr4$tmpname[perfects16_200BCs_dr4$selection== "fitH07D04"] <- "B"
perfects16_200BCs_dr4$tmpname[perfects16_200BCs_dr4$selection== "fitH08D04"] <- "C"
perfects16_200BCs_dr4$tmpname[perfects16_200BCs_dr4$selection== "fitH09D04"] <- "D"
perfects16_200BCs_dr4$tmpname[perfects16_200BCs_dr4$selection== "fitH10D04"] <- "E"
perfects16_200BCs_dr4$tmpname[perfects16_200BCs_dr4$selection== "fitH11D04"] <- "F"
perfects16_200BCs_dr4$tmpname[perfects16_200BCs_dr4$selection== "fitH12D04"] <- "G"

perfects16_200BCs_dr_labels <- c("0","0.058","0.5","1","10","50","200")
```

Plot the dose response for remaining perfects (>200 BCs) as boxplots for each trimethoprim concentration.
```{r}
p11_16_d4 <- ggplot(perfects16_200BCs_dr4, aes(x=tmpname, y=fitness)) +
  geom_boxplot(color="black", fill="aquamarine4") +
  xlab("[Trimethoprim] (ug/mL)") +
  ylab("Fitness") +
  ggtitle("Library 16 Dose Response - Day 4") +
  scale_x_discrete(labels = perfects16_200BCs_dr_labels) +
  theme(axis.text.x = element_text(size = 12),
        axis.text.y = element_text(size = 12),
        axis.title.x = element_text(size = 14),
        axis.title.y = element_text(size = 14)) +
  theme(legend.position = "none") +
  scale_y_continuous(expand = c(0, 0), limits = c(-11,5)) +
  guides(linetype = FALSE) +
  panel_border(color = "black")

p11_16_d4
```

```{r echo=FALSE}
patch11_Lib16 <- (p11_16_d1 | p11_16_d2 | p11_16_d3 | p11_16_d4)
patch11_Lib16
```

```{r echo=FALSE}
ggsave(file="plots/combined/lib16.dose.response.by.day.png", plot=patch11_Lib16,
       dpi=600, width = 14, height = 6, units = "in")
```

```{r echo=FALSE}
patch11_Lib15.16 <- (p11_15_d1 | p11_15_d2 | p11_15_d3 | p11_15_d4) / 
  (p11_16_d1 | p11_16_d2 | p11_16_d3 | p11_16_d4)
patch11_Lib15.16
```

```{r echo=FALSE}
ggsave(file="plots/combined/lib15.16.dose.response.by.day.png", plot=patch11_Lib15.16,
       dpi=600, width = 14, height = 12, units = "in")
```

### Perfects Fitness Counts

**Library 15**

Plot the Perfects counts based on fitness values between D11 vs. D03 conditions (>5 BCs count):
```{r}
p18_15 <- ggplot(perfects15_tree_5BCs,aes(x=fitD11D03)) +
  geom_histogram(binwidth=0.08, fill = "aquamarine4", color = "black") +
  xlab("Fitness (Log2 Fold Change)") +
  ylab("Counts") +
  ggtitle("Library 15 Perfects Fitness Counts (D11 vs. D03)") +
  #theme_cowplot() +
  #background_grid() +
  theme(axis.text.x = element_text(size = 12),
        axis.text.y = element_text(size = 12),
        axis.title.x = element_text(size = 14),
        axis.title.y = element_text(size = 14)) +
  theme(legend.position = "none") +
  scale_y_continuous(expand = c(0, 0), limits = c(0,30)) +
  guides(linetype = FALSE) + panel_border(color = "black")

p18_15
```

```{r echo=FALSE}
ggsave(file="plots/hist_Lib15_5BCs_fitD11D03.png", plot=p18_15,
       dpi=600, width = 8, height = 6, units = "in")
```


**Library 16**

Plot the Perfects counts based on fitness values between E06 vs. D04 conditions (>5 BCs count):
```{r}
p18_16 <- ggplot(perfects16_tree_5BCs,aes(x=fitE06D04)) +
  geom_histogram(binwidth=0.08, fill = "aquamarine4", color = "black") +
  xlab("Fitness (Log2 Fold Change)") +
  ylab("Counts") +
  ggtitle("Library 16 Perfects Fitness Counts (E06 vs. D04)") +
  #theme_cowplot() +
  #background_grid() +
  theme(axis.text.x = element_text(size = 12),
        axis.text.y = element_text(size = 12),
        axis.title.x = element_text(size = 14),
        axis.title.y = element_text(size = 14)) +
  theme(legend.position = "none") +
  scale_y_continuous(expand = c(0, 0), limits = c(0,30)) +
  guides(linetype = FALSE) + panel_border(color = "black")

p18_16
```

```{r echo=FALSE}
ggsave(file="plots/hist_Lib16_5BCs_fitD11D03.png", plot=p18_16,
       dpi=600, width = 8, height = 6, units = "in")
```

```{r}
patch13 <- (p18_15 | p18_16)
patch13
```

```{r echo=FALSE}
ggsave(file="plots/combined/hist_Lib15.16_5BCs.png", plot=patch13,
       dpi=600, width = 12, height = 8, units = "in")
```

## Phylogenetic Trees

**This section uses the `library(ggtree)` and `library(castor)` packages.**

### Tree File
Import the tree dataset containing all designed constructs for Library 15:
```{r}
Alltree15 <- read.tree("analysis_output/perfects_tree_Lib15_5BCs_aligned.tree")   # newick format
Alltree15
```

The following code adds the branch length distance from E. coli for each unique homolog to the "perfects_tree_5BCs" dataframe based on values in "Alltree".
```{r}
perfects15_tree_5BCs$distances = get_pairwise_distances(Alltree15, 
                                   perfects15_tree_5BCs$ID,
                                   rep(perfects15_tree_5BCs$ID[1], length(perfects15_tree_5BCs$ID)))
```

Plot fitness relative to phylogenetic distance to E. coli DHFR gene:
```{r}
p12_15 <- ggplot(perfects15_tree_5BCs,aes(x=distances, y=fitD05D03)) +
  geom_point(color = "aquamarine4") +
  xlab("Phylogenetic Distance to E. coli DHFR") +
  ylab("Fitness (Log2 Fold Change)") +
  ggtitle("Library 15 Phylogenetic Distance from E. coli") +
  #theme_cowplot() +
  #background_grid() +
  theme(axis.text.x = element_text(size = 12),
        axis.text.y = element_text(size = 12),
        axis.title.x = element_text(size = 14),
        axis.title.y = element_text(size = 14)) +
  theme(legend.position = "none") +
  scale_y_continuous(expand = c(0, 0), limits = c(-5,2)) +
  guides(linetype = FALSE) + panel_border(color = "black")

p12_15
```

```{r echo=FALSE}
ggsave(file="plots/dhfr.Lib15.phylo.fitness.png", plot=p12_15,
       dpi=600, width = 8, height = 6, units = "in")
```

### Tree Plots

#### DHFR Phylogenetic Diversity
Plot a circular phylogenetic tree with color-scale based on phylogenetic distance between each DHFR homolog relative to the E. coli DHFR gene:
```{r}
p13_15 <- ggtree(Alltree15, layout="circular", branch.length="none") %<+% 
  perfects15_tree_5BCs + 
  aes(color=PctIdentEcoli*100) +
  geom_tippoint(aes(color=PctIdentEcoli*100), size=1, alpha=0.8) +
  geom_tiplab2(aes(color=PctIdentEcoli*100, label=Source, angle=angle), size=1) +
  theme(legend.position="left") + 
  scale_colour_gradient2("Sequence\nIdentity (%)", low="blue", mid="green", high="red",
                         midpoint=(100+min(perfects15_tree_5BCs$PctIdentEcoli)*100)/2)

p13_15
```

```{r echo=FALSE}
ggsave(file="plots/trees/FastTree_Lib15_5BCs_PctIdentEcolinol.png", plot=p13_15,
       dpi=600, width = 12, height = 7.5, units = "in")
```

**Library 15**

#### Complementation Tree
Show the minimum and maximum fitness values between D05 (M9 no supp) vs. D03 (M9 full supp) dataset:
```{r class.output="goodCode"}
min(perfects15_tree_5BCs$fitD05D03)
max(perfects15_tree_5BCs$fitD05D03)
```

Plot a circular phylogenetic tree for complementation with color-scale based on fitness values between D05 vs. D03 for each DHFR homolog:

```{r}
#shade circular tree by fitness:
p14_15 <- ggtree(Alltree15, layout="circular", branch.length="none") %<+%
  perfects15_tree_5BCs +
  aes(color=fitD05D03) +
  geom_tippoint(aes(color=fitD05D03), size=1, alpha=0.8) +
  geom_tiplab2(aes(color=fitD05D03, label=Source, angle=angle), size=1) +
  theme(legend.position="left") + 
  ggtitle("Library 15 Complementation (D05 vs D03)") +
  scale_colour_gradient2("fitness",
                         low="black", 
                         mid="gold",
                         high="red",
                         na.value="gray",
                         limit = c(-5,5))

p14_15
```

```{r echo=FALSE, eval=FALSE}
ggsave(file="plots/trees/FastTree_Lib15_5BCs_fitD05D03.COMPLEMENTATION.png", plot=p14_15,
       dpi=600, width = 12, height = 7.5, units = "in")
```

Another tree version that uses the homolog IDs instead of the organism nomenclature to standardized label sizes around the tree. There's also no shading of branches to match fitness scores (no "aes()" settings):
```{r}
p15_15 <- ggtree(Alltree15, layout="circular", branch.length="none") %<+% 
  perfects15_tree_5BCs + 
  geom_tippoint(aes(color=fitD05D03), alpha=0.25) +
  geom_tiplab2(size=1, aes(color=fitD05D03,angle=angle)) +
  geom_text2(aes(subset=!isTip, label=node), size=1, color="red",hjust=-.3) +
  theme(legend.position="left") + 
  scale_colour_gradient2("fitness",
                         low="black", 
                         mid="gold",
                         high="red",
                         na.value="gray",
                         limit = c(-3,1))

p15_15
```

```{r echo=FALSE}
ggsave(file="plots/trees/FastTree_Lib15_5BCs_fitD05D03_ID_and_nodes.COMPLEMENTATION.png", plot=p15_15,
       dpi=600, width = 12, height = 7.5, units = "in")
```

Plot a circular phylogenetic tree with color-scale based on fitness values between D05 vs. D03 for each DHFR homolog. **This version uses the true branch length for plotting:**
```{r}
p16_15 <- ggtree(Alltree15, layout="circular") %<+%
  perfects15_tree_5BCs + 
  aes(color=fitD05D03) +
  geom_tippoint(aes(color=fitD05D03), size=1, alpha=0.8) +
# geom_tiplab2(aes(color=fitD05D03, label=Source, angle=angle), offset=0.1, size=1) +
  theme(legend.position="left") + 
  scale_colour_gradient2("fitness",
                         low="black", 
                         mid="gold",
                         high="red",
                         na.value="gray",
                         limit = c(-3,1))

p16_15
```

```{r echo=FALSE}
ggsave(file="plots/trees/FastTree_Lib15_5BCs_fitD05D03_truelength.COMPLEMENTATION.png", plot=p16_15,
       dpi=600, width = 12, height = 7.5, units = "in")
```

#### D11 vs. D03 Tree Plot
Show the minimum and maximum fitness values between D11 (200 mg/uL Day 1) vs. D03 (M9 no supp.) in the dataset:
```{r class.output="goodCode"}
min(perfects15_tree_5BCs$fitD11D03)
max(perfects15_tree_5BCs$fitD11D03)
```

Plot a circular phylogenetic tree with color-scale based on fitness values between D11 vs. D03 for each DHFR homolog:
```{r}
#shade circular tree by fitness:
p17_15 <- ggtree(Alltree15, layout="circular", branch.length="none") %<+% #, branch.length="none"
  perfects15_tree_5BCs +
  aes(color=fitD11D03) +
  geom_tippoint(aes(color=fitD11D03), size=1, alpha=0.8) +
  geom_tiplab2(aes(color=fitD11D03, label=Source, angle=angle), size=1) +
  theme(legend.position="left") + 
  ggtitle("Library 15 Fitness at 200 mg/uL Trimethoprim (D11 vs D03)") +
  scale_colour_gradient2("fitness",
                         low="black", 
                         mid="gold",
                         high="red",
                         na.value="gray",
                         limit = c(-10,2)) 

p17_15
```

```{r echo=FALSE}
ggsave(file="plots/trees/FastTree_Lib15_5BCs_fitD11D03.png", plot=p17_15,
       dpi=600, width = 12, height = 7.5, units = "in")
```

```{r}
patch12<- (p14_15 | p17_15)
patch12
```

```{r echo=FALSE}
ggsave(file="plots/trees/FastTree.Lib15.Complementation.Fitness.200mg.png", plot=patch12,
       dpi=600, width = 14, height = 12, units = "in")
```

# Fitness vs Distance

<font color="blue">**This section is based on the R file: "R_change_in_fitness_vs_distance.R".**</font> It determines how fitness changes with mutation distance.

## Filter Data

Grab only the BCs with up to 5 mutations (also need to ensure it has greater than 0 mutations since some BCs have negative values):
```{r}
BCs5_15 <- BCmut_15 %>%
  filter(mutations >= 0 & mutations <= 5) %>%
  left_join(mutIDinfo15 %>% select(mutID, IDalign), by="mutID") %>%
  select(BC,IDalign,mutations,D05D03fc,mutID)
```

Retain only the homologs with good data:
```{r}
fitness_distance_15 <- perfects15_tree_5BCs %>%
  select(ID,fitD05D03) %>%
  dplyr::rename(IDalign=ID)
```

Determine median and sd for 1 mutation:
```{r}
fitness_distance_15 <- BCs5_15 %>%
    filter(mutations==1) %>%
    group_by(IDalign) %>%
    summarise(mut1fit=median(D05D03fc),
              mut1sd=sd(D05D03fc),
              num1points=n()) %>%
    right_join(fitness_distance_15,by="IDalign")
```

Determine median and sd for 2 mutation:
```{r}
fitness_distance_15 <-  BCs5_15 %>%
  filter(mutations==2) %>%
  group_by(IDalign) %>%
  summarise(mut2fit=median(D05D03fc),
            mut2sd=sd(D05D03fc),
            num2points=n()) %>%
  right_join(fitness_distance_15,by="IDalign") 
```

Determine median and sd for 3 mutations:
```{r}
fitness_distance_15 <-  BCs5_15 %>%
  filter(mutations==3) %>%
  group_by(IDalign) %>%
  summarise(mut3fit=median(D05D03fc),
            mut3sd=sd(D05D03fc),
            num3points=n()) %>%
  right_join(fitness_distance_15,by="IDalign")
```

Determine median and sd for 4 mutations:
```{r}
fitness_distance_15 <-  BCs5_15 %>%
  filter(mutations==4) %>%
  group_by(IDalign) %>%
  summarise(mut4fit=median(D05D03fc),
            mut4sd=sd(D05D03fc),
            num4points=n()) %>%
  right_join(fitness_distance_15,by="IDalign")
```

Determine median and sd for 5 mutations:
```{r}
fitness_distance_15 <-  BCs5_15 %>%
  filter(mutations==5) %>%
  group_by(IDalign) %>%
  summarise(mut5fit=median(D05D03fc),
            mut5sd=sd(D05D03fc),
            num5points=n()) %>%
  right_join(fitness_distance_15,by="IDalign")
```

Determine change in fitness:
```{r}
fitness_distance_nu_15 <- fitness_distance_15 %>%
  mutate(mut1fitn=(mut1fit-fitD05D03),
         mut2fitn=(mut2fit-fitD05D03),
         mut3fitn=(mut3fit-fitD05D03),
         mut4fitn=(mut4fit-fitD05D03),
         mut5fitn=(mut5fit-fitD05D03),
         mut0fitn=0)
```

Melt data on number of mutations:
```{r}
fitness_distance_m_15 <- fitness_distance_nu_15 %>%
  select(IDalign,fitD05D03,mut0fitn,mut1fitn,mut2fitn,mut3fitn,mut4fitn,mut5fitn) %>%
  gather(mutations,fitness,mut0fitn,mut1fitn,mut2fitn,mut3fitn,mut4fitn,mut5fitn)
```

Replace names with numbers:
```{r}
fitness_distance_m_15$mutations[which(fitness_distance_m_15$mutations=="mut0fitn")] <- as.numeric(0)
fitness_distance_m_15$mutations[which(fitness_distance_m_15$mutations=="mut1fitn")] <- as.numeric(1)
fitness_distance_m_15$mutations[which(fitness_distance_m_15$mutations=="mut2fitn")] <- as.numeric(2)
fitness_distance_m_15$mutations[which(fitness_distance_m_15$mutations=="mut3fitn")] <- as.numeric(3)
fitness_distance_m_15$mutations[which(fitness_distance_m_15$mutations=="mut4fitn")] <- as.numeric(4)
fitness_distance_m_15$mutations[which(fitness_distance_m_15$mutations=="mut5fitn")] <- as.numeric(5)
```

Remove those with NA fitness:
```{r}
fitness_distance_m_15 <- fitness_distance_m_15 %>%
  filter(!is.na(fitness))
```

## Fitness vs. Distance Plots

The first plot version uses traces to display results:
```{r}
p19_15 <- ggplot(fitness_distance_m_15, aes(x=mutations, y=fitness, group=IDalign, color=IDalign)) +
  geom_point() +
  geom_line() +
  xlab("Distance from homolog (a.a.)") +
  ylab("Change in fitness relative to homolog") +
  #theme_cowplot() +
  #background_grid() +
  theme(axis.text.x = element_text(size = 12),
        axis.text.y = element_text(size = 12),
        axis.title.x = element_text(size = 14),
        axis.title.y = element_text(size = 14)) +
  theme(legend.position = "none") +
  scale_y_continuous(expand = c(0, 0), limits = c(-10,10)) +
  guides(linetype = FALSE) + panel_border(color = "black")

p19_15
```

```{r echo=FALSE}
ggsave(file="plots/fitness_vs_distance_all_traces.Lib15.png", plot=p19_15,
       dpi=600, width = 8, height = 6, units = "in")
```

The second plot version uses a boxplot to display results:
```{r}
p20_15 <- ggplot(fitness_distance_m_15, aes(x=mutations, y=fitness)) +
  geom_boxplot(color="black", fill="aquamarine4", alpha=0.8) +
  xlab("Distance from homolog (a.a.)") +
  ylab("Change in fitness relative to homolog") +
  #theme_cowplot() +
  #background_grid() +
  theme(axis.text.x = element_text(size = 12),
        axis.text.y = element_text(size = 12),
        axis.title.x = element_text(size = 14),
        axis.title.y = element_text(size = 14)) +
  theme(legend.position = "none") +
  scale_y_continuous(expand = c(0, 0), limits = c(-10,10)) +
  guides(linetype = FALSE) + panel_border(color = "black")

p20_15
```

```{r echo=FALSE}
ggsave(file="plots/fitness_vs_distance_boxplot.Lib15.png", plot=p20_15,
       dpi=600, width = 8, height = 6, units = "in")
```

Calculate Spearman correlations between fitness and distance from homolog:
```{r, warning=FALSE, class.output="goodCode"}
# Calculate Spearman coefficient:
cor(as.numeric(fitness_distance_m_15$mutations),fitness_distance_m_15$fitness,
    method=c("spearman"))

# Run correlation test:
cor.test(as.numeric(fitness_distance_m_15$mutations),fitness_distance_m_15$fitness,
         method=c("spearman"))
```































# Reproducibility

The session information is provided for full reproducibility.
```{r}
devtools::session_info()
```