---
title: "Broad Mutational Scanning for Antibiotic Resistance in the DHFR Family"
author: 'Authors: [Karl J. Romanowicz](https://kromanowicz.github.io/), Carmen Resnick, Samuel R. Hinton, Calin Plesa'
output:
  html_notebook:
    theme: spacelab
    toc: yes
    toc_depth: 5
    toc_float:
      collapsed: yes
      smooth_scroll: yes
  html_document:
    toc: yes
    toc_depth: '5'
    df_print: paged
  pdf_document:
    toc: yes
    toc_depth: '5'
---

**GitHub Repository:** [https://github.com/PlesaLab/DHFR](https://github.com/PlesaLab/DHFR)

# Experiment

This R Notebook provides complete reproducibility of the data analysis presented in ***"Broad mutational scanning of the dihydrofolate reductase protein family"*** by Romanowicz, Resnick, Hinton, and Plesa (*In Prep*).

<font color="green">This pipeline analyzes a set of 1,536 DHFR homologs and their corresponding mutants designed with two-fold redundancy (two codon libraries). It generates fitness scores from multiple time points during a multiplexed in-vivo assay that evaluates how these homologs and mutants complement the functionality of an E. coli knockout strain (dFolA|dThyA). Sequence data were generated using the Illumina NovaSeq platform using paired-end sequencing read amplicons.</font>

![Methods overview to achieve a broad-mutational scan for DHFR homologs.](images/methods.png)

```{css}
.badCode {
background-color: lightpink;
font-weight: bold;
}

.goodCode {
background-color: lightgreen;
font-weight: bold;
}

.sharedCode {
background-color: lightblue;
font-weight: bold;
}

table {
  margin: auto;
  border-top: 1px solid #666;
  border-bottom: 1px solid #666;
}
table thead th { border-bottom: 1px solid #ddd; }
th, td { padding: 5px; }
thead, tfoot, tr:nth-child(even) { background: #eee; }
```

```{r setup, include=FALSE}
# Set global options for notebook
knitr::opts_chunk$set(echo = TRUE)
knitr::opts_chunk$set(warning = TRUE, message = TRUE)
knitr::opts_chunk$set(echo = TRUE, class.source = "bg-success")

# Getting the path of your current open file and set as wd
current_path = rstudioapi::getActiveDocumentContext()$path 
setwd(dirname(current_path))
print(getwd())
```

# Packages
The following R packages must be installed prior to loading into the R session. See the **Reproducibility** tab for a complete list of packages and their versions used in this workflow.
```{r message=FALSE, results='hide', warning=FALSE}
# Make a vector of required packages
required.packages <- c("ape", "Biostrings", "castor", "cowplot", "devtools", "dplyr", "ggExtra", "ggnewscale", "ggplot2", "ggridges", "ggtree", "ggtreeExtra", "gridExtra", "knitr", "matrixStats", "patchwork", "pscl", "RColorBrewer", "reshape", "ROCR", "seqinr", "scales", "stringr", "stringi", "tidyr", "viridis")

# Load required packages
lapply(required.packages, library, character.only = TRUE)
```

```{r include=FALSE}
# set.seed is used to fix the random number generation to make the results repeatable
set.seed(123)
```

# Load Data
<font color="blue">**This section is based on the R file: "R_load_data.R".**</font> It describes how to load all of the pre-existing barcode mapping data necessary for downstream analysis.

**Pre-Existing Data Files to Import**

* **BCinfo:** Mapping Information (one for each library)
* **mutIDinfo:** Mutant Information (one for each library)
* **BCs:** Everything Sequenced (one for eacy library)
* **BCmut:** Only Mapped Barcodes (one for each library)

## Barcode Info
**Begin the analysis by loading in the barcode information for each codon-version library.** This data represents each **barcode** [`BC`] recovered from the complementation process, linked to each **homolog or mutant** [`mutID`], the **number of mutations** observed in the sequence relative to its designed homolog [`mutations`], a list of where the **mutations occurred** along the protein sequence[`cigar`], and the **number of reads** associated with each barcode [`reads`]:
```{r}
# Library 15 (Codon 1)
BCinfo15 = read.csv("for_Karl/15_HiFi_BC_mutID_all.csv", head=TRUE)  # read csv file

# List column names
names(BCinfo15)

# Library 16 (Codon 2)
BCinfo16 = read.csv("for_Karl/16_HiFi_BC_mutID_all.csv", head=TRUE)  # read csv file

# List column names
names(BCinfo16)
```

### Filter Mutations

Filter each dataset to ensure each mutant has >=0 mutations
```{r class.output="goodCode"}
## Make sure each mutant has >=0 mutations by removing rows with negative values
BCinfo15 <- BCinfo15 %>%
  filter(mutations >= 0)

# Check mutate count for Lib15 (should be "0"):
min(BCinfo15$mutations)

## Make sure each mutant has >=0 mutations by removing rows with negative values
BCinfo16 <- BCinfo16 %>%
  filter(mutations >= 0)

# Check mutate count for Lib16 (should be "0"):
min(BCinfo16$mutations)
```

<font color="green">**Preview of BCinfo**</font>
```{r echo=FALSE}
# Library 15
head(BCinfo15)

# Library 16
head(BCinfo16)
```

### Unique BC Counts

**Total BCs:** Count the number of unique barcodes for each library
```{r class.output="goodCode"}
# Count the number of unique barcodes for Library 15
BCinfo15.count <- length(unique(BCinfo15$BC))
format(BCinfo15.count, big.mark = ",")

# Count the number of unique barcodes for Library 16
BCinfo16.count <- length(unique(BCinfo16$BC))
format(BCinfo16.count, big.mark = ",")
```

**Perfects BCs:** Count the number of unique barcodes that have zero mutations (i.e., BCs associated with perfects). Note that many of these barcodes are likely associated with the same protein sequence and do not represent the total number of perfects recovered from the dataset.
```{r class.output="goodCode"}
# Count the number of rows containing zero (0) mutations for Library 15
BCinfo15.count.zeros <- sum(BCinfo15$mutations == 0)
format(BCinfo15.count.zeros, big.mark = ",")

# Count the number of rows containing zero (0) mutations for Library 16
BCinfo16.count.zeros <- sum(BCinfo16$mutations == 0)
format(BCinfo16.count.zeros, big.mark = ",")
```

**Mutant BCs:** Count the number of unique unique barcodes that have at least 1 mutation (i.e., BCs associated with mutants). Note that many of these barcodes are likely associated with the same protein sequence and do not represent the total number of mutants recovered from the dataset.
```{r class.output="goodCode"}
# Count the number of rows containing >=1 mutation for Library 15
BCinfo15.count.nonzeros <- sum(BCinfo15$mutations != 0)
format(BCinfo15.count.nonzeros, big.mark = ",")

# Count the number of rows containing >=1 mutation for Library 16
BCinfo16.count.nonzeros <- sum(BCinfo16$mutations != 0)
format(BCinfo16.count.nonzeros, big.mark = ",")
```


## Mutant Info
**Load in the mutant information for each codon-version library.** This data represents each **mutant** sequence [`mutID`] from the dataset, the **designed homolog** it mutated from [`IDalign`], the **number of barcodes** associated with each mutant [`numBCs`], the **number of mutations** in the sequence relative to the designed homolog [`mutations`], the **full sequence** of the mutant [`seq`], the **percent identity of the mutant** to the designed homolog [`pct_ident`], a list of **all barcodes** associated with the mutant [`BCs`], and the **barcode code** [`BCcode`]:
```{r}
# Library 15 (Codon 1)
mutIDinfo15 = read.csv("for_Karl/15_HiFi_mutID_info_all.csv", head=TRUE)  # read csv file

# List column names
names(mutIDinfo15)

# Library 16 (Codon 2)
mutIDinfo16 = read.csv("for_Karl/16_HiFi_mutID_info_all.csv", head=TRUE)  # read csv file

# List column names
names(mutIDinfo16)
```

### Filter Mutations

Filter each dataset to ensure each mutant has >=0 mutations
```{r class.output="goodCode"}
## Make sure each mutant has >=0 mutations by removing rows with negative values
mutIDinfo15 <- mutIDinfo15 %>%
  filter(mutations >= 0)

# Check mutate count for Lib15 (should be "0"):
min(mutIDinfo15$mutations)

## Make sure each mutant has >=0 mutations by removing rows with negative values
mutIDinfo16 <- mutIDinfo16 %>%
  filter(mutations >= 0)

# Check mutate count for Lib16 (should be "0"):
min(mutIDinfo16$mutations)
```

<font color="green">**Preview of mutIDinfo**</font>
```{r echo=FALSE}
# Library 15
head(mutIDinfo15)

# Library 16
head(mutIDinfo16)
```

### Unique Sequence Counts

**Total Sequences:** Count the number of unique protein sequences in each library
```{r class.output="goodCode"}
# Count the number of unique protein sequences for Library 15
mutIDinfo15.count <- length(unique(mutIDinfo15$seq))
format(mutIDinfo15.count, big.mark = ",")

# Count the number of unique protein sequences for Library 16
mutIDinfo16.count <- length(unique(mutIDinfo16$seq))
format(mutIDinfo16.count, big.mark = ",")
```

**Perfect Sequences:** Count the number of unique protein sequences that have zero mutations (i.e., sequences associated with perfects). These values designate the number of true homologs ("perfects") that match the designed homologs without mutations.
```{r class.output="goodCode"}
# Count the number of rows containing zero (0) mutations for Library 15
mutIDinfo15.count.zeros <- sum(mutIDinfo15$mutations == 0)
format(mutIDinfo15.count.zeros, big.mark = ",")

# Count the number of rows containing zero (0) mutations for Library 16
mutIDinfo16.count.zeros <- sum(mutIDinfo16$mutations == 0)
format(mutIDinfo16.count.zeros, big.mark = ",")
```

Count the number of shared perfect protein sequences between libraries (mutations == 0).
```{r}
# Merge the mutIDinfo datasets but keep only the mutIDs shared between libraries
mutIDinfo.15.16.zeros.shared <- merge(mutIDinfo15, mutIDinfo16, by = "mutID", all = FALSE)
```

```{r class.output="goodCode"}
# Count the number of perfects (mutations == 0) shared between both libraries
mutIDinfo.15.16.zeros.shared.count <- sum(mutIDinfo.15.16.zeros.shared$mutations.x == 0)
format(mutIDinfo.15.16.zeros.shared.count, big.mark = ",")
```

Also count the number of perfect sequences unique to one or the other library (mutations == 0).
```{r}
# Merge the mutIDinfo datasets but retain only the mutIDs unique to one library or the other
mutIDinfo.15.16.zeros.unique <- bind_rows(
  anti_join(mutIDinfo15, mutIDinfo16, by = "mutID"),
  anti_join(mutIDinfo16, mutIDinfo15, by = "mutID"))
```

```{r class.output="goodCode"}
# Count the number of perfects (mutations == 0) unique to one library or the other
mutIDinfo.15.16.zeros.unique.count <- sum(mutIDinfo.15.16.zeros.unique$mutations == 0)
format(mutIDinfo.15.16.zeros.unique.count, big.mark = ",")
```

```{r class.output="goodCode"}
mutIDinfo.15.16.zeros.all.count <- sum(mutIDinfo.15.16.zeros.shared.count + mutIDinfo.15.16.zeros.unique.count)
format(mutIDinfo.15.16.zeros.all.count, big.mark = ",")
```

**Mutant Sequences:** Count the number of unique protein sequences that have at least 1 mutation (i.e., sequences associated with mutants). These values designate the number of mutants derived from the designed homologs.
```{r class.output="goodCode"}
# Count the number of rows containing >=1 mutation for Library 15
mutIDinfo15.count.nonzeros <- sum(mutIDinfo15$mutations != 0)
format(mutIDinfo15.count.nonzeros, big.mark = ",")

# Count the number of rows containing >=1 mutation for Library 16
mutIDinfo16.count.nonzeros <- sum(mutIDinfo16$mutations != 0)
format(mutIDinfo16.count.nonzeros, big.mark = ",")
```

## Add Map Identifiers

Add the IDalign column to the "BCinfo" objects based on shared "mutID" values with the "mutIDinfo" objects:
```{r}
# Lib 15
BCinfo15$IDalign <- mutIDinfo15$IDalign[match(BCinfo15$mutID, mutIDinfo15$mutID)]

# Lib 16
BCinfo16$IDalign <- mutIDinfo16$IDalign[match(BCinfo16$mutID, mutIDinfo16$mutID)]
```

Add "Lib" column with value==1 for each BC present in the "BCinfo" object
```{r}
# Add a new column "NewColumn" with every row value set to 1

# Lib 15
BCinfo15 <- BCinfo15 %>% mutate(Lib15 = 1)

# Lib 16
BCinfo16 <- BCinfo16 %>% mutate(Lib16 = 1)
```

## Merge Mapping Files

Merge the "BCinfo" and "mutIDinfo" mapping files by shared "mutID" values for each library
```{r}
# Lib15

# Merge mapping files by "mutID"
BCs_mutID_15 <- inner_join(BCinfo15,mutIDinfo15,by="mutID")

# Select columns of interest to remove duplicates
BCs_mutID_15 <- BCs_mutID_15 %>%
  select(BC, mutID, IDalign.x, mutations.x, cigar, numBCs, seq, pct_ident, BCs, BCcode)

# Rename columns
names(BCs_mutID_15) <- c("BC","mutID","IDalign","mutations","cigar","numBCs","seq","pct_ident","BCs","BCcode")

# Lib16

# Merge mapping files by "mutID"
BCs_mutID_16 <- inner_join(BCinfo16,mutIDinfo16,by="mutID")

# Select columns of interest to remove duplicates
BCs_mutID_16 <- BCs_mutID_16 %>%
  select(BC, mutID, IDalign.x, mutations.x, cigar, numBCs, seq, pct_ident, BCs, BCcode)

# Rename columns
names(BCs_mutID_16) <- c("BC","mutID","IDalign","mutations","cigar","numBCs","seq","pct_ident","BCs","BCcode")
```

## Count Data
**Import the count data (sequences) for each sampling condition in Library 15 (D01, D03, D05:D11, E07:F01, F9:G3, G11:H5) and Library 16 (D02, D04, D12:E6, F2:F8, G4:G10, H6:H12).** Each .csv file contains two columns: **(1)** [`BC`] represents the **unique barcode** recovered from the sampling condition, and **(2)** [`D0X`] represents the **sequence counts** matching each barcode within the sampling condition.

Import all Library 15 count files:
```{r}
### Library 15

## Library 15 - T0

# D01 - Lib15 - Overnight Growth on LB - T0
D01 = read.csv("counts/D01_collapse_d1.tsv", head=FALSE, sep="\t") %>%
  select(-V3)
colnames(D01) <- c("BC","D01")

# D03 - Lib15 - Growth on M9 - Full Supplement - T0
D03 = read.csv("counts/D03_collapse_d1.tsv", head=FALSE, sep="\t") %>%
  select(-V3)
colnames(D03) <- c("BC","D03")

## Library 15 - T1

# D05 - Lib15 - Growth on M9 - 0 ug/mL TMP - T1
D05 = read.csv("counts/D05_collapse_d1.tsv", head=FALSE, sep="\t") %>%
  select(-V3)
colnames(D05) <- c("BC","D05")

# D06 - Lib15 - Growth on M9 - 0.058 ug/ml TMP - T1
D06 = read.csv("counts/D06_collapse_d1.tsv", head=FALSE, sep="\t") %>%
  select(-V3)
colnames(D06) <- c("BC","D06")

# D07 - Lib15 - Growth on M9 - 0.5 ug/ml TMP - T1
D07 = read.csv("counts/D07_collapse_d1.tsv", head=FALSE, sep="\t") %>%
  select(-V3)
colnames(D07) <- c("BC","D07")

# D08 - Lib15 - Growth on M9 - 1.0 ug/ml TMP - T1
D08 = read.csv("counts/D08_collapse_d1.tsv", head=FALSE, sep="\t") %>%
  select(-V3)
colnames(D08) <- c("BC","D08")

# D09 - Lib15 - Growth on M9 - 10.0 ug/ml TMP - T1
D09 = read.csv("counts/D09_collapse_d1.tsv", head=FALSE, sep="\t") %>%
  select(-V3)
colnames(D09) <- c("BC","D09")

# D10 - Lib15 - Growth on M9 - 50.0 ug/ml TMP - T1
D10 = read.csv("counts/D10_collapse_d1.tsv", head=FALSE, sep="\t") %>%
  select(-V3)
colnames(D10) <- c("BC","D10")

# D11 - Lib15 - Growth on M9 - 200.0 ug/ml TMP - T1
D11 = read.csv("counts/D11_collapse_d1.tsv", head=FALSE, sep="\t") %>%
  select(-V3)
colnames(D11) <- c("BC","D11")

## Library 15 - T2

# E07 - Lib15 - Growth on M9 - 0 ug/mL TMP - T2
E07 = read.csv("counts/E07_collapse_d1.tsv", head=FALSE, sep="\t") %>%
  select(-V3)
colnames(E07) <- c("BC","E07")

# E08 - Lib15 - Growth on M9 - 0.058 ug/ml TMP - T2
E08 = read.csv("counts/E08_collapse_d1.tsv", head=FALSE, sep="\t") %>%
  select(-V3)
colnames(E08) <- c("BC","E08")

# E09 - Lib15 - Growth on M9 - 0.5 ug/ml TMP - T2
E09 = read.csv("counts/E09_collapse_d1.tsv", head=FALSE, sep="\t") %>%
  select(-V3)
colnames(E09) <- c("BC","E09")

# E10 - Lib15 - Growth on M9 - 1.0 ug/ml TMP - T2
E10 = read.csv("counts/E10_collapse_d1.tsv", head=FALSE, sep="\t") %>%
  select(-V3)
colnames(E10) <- c("BC","E10")

# E11 - Lib15 - Growth on M9 - 10.0 ug/ml TMP - T2
E11 = read.csv("counts/E11_collapse_d1.tsv", head=FALSE, sep="\t") %>%
  select(-V3)
colnames(E11) <- c("BC","E11")

# E12 - Lib15 - Growth on M9 - 50.0 ug/ml TMP - T2
E12 = read.csv("counts/E12_collapse_d1.tsv", head=FALSE, sep="\t") %>%
  select(-V3)
colnames(E12) <- c("BC","E12")

# F01 - Lib15 - Growth on M9 - 200.0 ug/ml TMP - T2
F01 = read.csv("counts/F01_collapse_d1.tsv", head=FALSE, sep="\t") %>%
  select(-V3)
colnames(F01) <- c("BC","F01")

## Library 15 - T3

# F09 - Lib15 - Growth on M9 - 0 ug/mL TMP - T3
F09 = read.csv("counts/F09_collapse_d1.tsv", head=FALSE, sep="\t") %>%
  select(-V3)
colnames(F09) <- c("BC","F09")

# F10 - Lib15 - Growth on M9 - 0.058 ug/ml TMP - T3
F10 = read.csv("counts/F10_collapse_d1.tsv", head=FALSE, sep="\t") %>%
  select(-V3)
colnames(F10) <- c("BC","F10")

# F11 - Lib15 - Growth on M9 - 0.5 ug/ml TMP - T3
F11 = read.csv("counts/F11_collapse_d1.tsv", head=FALSE, sep="\t") %>%
  select(-V3)
colnames(F11) <- c("BC","F11")

# F12 - Lib15 - Growth on M9 - 1.0 ug/ml TMP - T3
F12 = read.csv("counts/F12_collapse_d1.tsv", head=FALSE, sep="\t") %>%
  select(-V3)
colnames(F12) <- c("BC","F12")

# G01 - Lib15 - Growth on M9 - 10.0 ug/ml TMP - T3
G01 = read.csv("counts/G01_collapse_d1.tsv", head=FALSE, sep="\t") %>%
  select(-V3)
colnames(G01) <- c("BC","G01")

# G02 - Lib15 - Growth on M9 - 50.0 ug/ml TMP - T3
G02 = read.csv("counts/G02_collapse_d1.tsv", head=FALSE, sep="\t") %>%
  select(-V3)
colnames(G02) <- c("BC","G02")

# G03 - Lib15 - Growth on M9 - 200.0 ug/ml TMP - T3
G03 = read.csv("counts/G03_collapse_d1.tsv", head=FALSE, sep="\t") %>%
  select(-V3)
colnames(G03) <- c("BC","G03")

## Library 15 - T4

# G11 - Lib15 - Growth on M9 - 0 ug/mL TMP - T4
G11 = read.csv("counts/G11_collapse_d1.tsv", head=FALSE, sep="\t") %>%
  select(-V3)
colnames(G11) <- c("BC","G11")

# G12 - Lib15 - Growth on M9 - 0.058 ug/ml TMP - T4
G12 = read.csv("counts/G12_collapse_d1.tsv", head=FALSE, sep="\t") %>%
  select(-V3)
colnames(G12) <- c("BC","G12")

# H01 - Lib15 - Growth on M9 - 0.5 ug/ml TMP - T4
H01 = read.csv("counts/H01_collapse_d1.tsv", head=FALSE, sep="\t") %>%
  select(-V3)
colnames(H01) <- c("BC","H01")

# H02 - Lib15 - Growth on M9 - 1.0 ug/ml TMP - T4
H02 = read.csv("counts/H02_collapse_d1.tsv", head=FALSE, sep="\t") %>%
  select(-V3)
colnames(H02) <- c("BC","H02")

# H03 - Lib15 - Growth on M9 - 10.0 ug/ml TMP - T4
H03 = read.csv("counts/H03_collapse_d1.tsv", head=FALSE, sep="\t") %>%
  select(-V3)
colnames(H03) <- c("BC","H03")

# H04 - Lib15 - Growth on M9 - 50.0 ug/ml TMP - T4
H04 = read.csv("counts/H04_collapse_d1.tsv", head=FALSE, sep="\t") %>%
  select(-V3)
colnames(H04) <- c("BC","H04")

# H05 - Lib15 - Growth on M9 - 200.0 ug/ml TMP - T4
H05 = read.csv("counts/H05_collapse_d1.tsv", head=FALSE, sep="\t") %>%
  select(-V3)
colnames(H05) <- c("BC","H05")
```

Import all Library 16 count files:
```{r}
### Library 16

## Library 16 - T0

# D02 - Lib16 - Overnight Growth on LB - T0
D02 = read.csv("counts/D02_collapse_d1.tsv", head=FALSE, sep="\t") %>%
  select(-V3)
colnames(D02) <- c("BC","D02")

# D04 - Lib16 - Growth on M9 - Full Supplement - T0
D04 = read.csv("counts/D04_collapse_d1.tsv", head=FALSE, sep="\t") %>%
  select(-V3)
colnames(D04) <- c("BC","D04")

## Library 16 - T1

# D12 - Lib16 - Growth on M9 - 0 ug/mL TMP - T1
D12 = read.csv("counts/D12_collapse_d1.tsv", head=FALSE, sep="\t") %>%
  select(-V3)
colnames(D12) <- c("BC","D12")

# E01 - Lib16 - Growth on M9 - 0.058 ug/ml TMP - T1
E01 = read.csv("counts/E01_collapse_d1.tsv", head=FALSE, sep="\t") %>%
  select(-V3)
colnames(E01) <- c("BC","E01")

# E02 - Lib16 - Growth on M9 - 0.5 ug/ml TMP - T1
E02 = read.csv("counts/E02_collapse_d1.tsv", head=FALSE, sep="\t") %>%
  select(-V3)
colnames(E02) <- c("BC","E02")

# E03 - Lib16 - Growth on M9 - 1.0 ug/ml TMP - T1
E03 = read.csv("counts/E03_collapse_d1.tsv", head=FALSE, sep="\t") %>%
  select(-V3)
colnames(E03) <- c("BC","E03")

# E04 - Lib16 - Growth on M9 - 10.0 ug/ml TMP - T1
E04 = read.csv("counts/E04_collapse_d1.tsv", head=FALSE, sep="\t") %>%
  select(-V3)
colnames(E04) <- c("BC","E04")

# E05 - Lib16 - Growth on M9 - 50.0 ug/ml TMP - T1
E05 = read.csv("counts/E05_collapse_d1.tsv", head=FALSE, sep="\t") %>%
  select(-V3)
colnames(E05) <- c("BC","E05")

# E06 - Lib16 - Growth on M9 - 200.0 ug/ml TMP - T1
E06 = read.csv("counts/E06_collapse_d1.tsv", head=FALSE, sep="\t") %>%
  select(-V3)
colnames(E06) <- c("BC","E06")

## Library 16 - T2

# F02 - Lib16 - Growth on M9 - 0 ug/mL TMP - T2
F02 = read.csv("counts/F02_collapse_d1.tsv", head=FALSE, sep="\t") %>%
  select(-V3)
colnames(F02) <- c("BC","F02")

# F03 - Lib16 - Growth on M9 - 0.058 ug/ml TMP - T2
F03 = read.csv("counts/F03_collapse_d1.tsv", head=FALSE, sep="\t") %>%
  select(-V3)
colnames(F03) <- c("BC","F03")

# F04 - Lib16 - Growth on M9 - 0.5 ug/ml TMP - T2
F04 = read.csv("counts/F04_collapse_d1.tsv", head=FALSE, sep="\t") %>%
  select(-V3)
colnames(F04) <- c("BC","F04")

# F05 - Lib16 - Growth on M9 - 1.0 ug/ml TMP - T2
F05 = read.csv("counts/F05_collapse_d1.tsv", head=FALSE, sep="\t") %>%
  select(-V3)
colnames(F05) <- c("BC","F05")

# F06 - Lib16 - Growth on M9 - 10.0 ug/ml TMP - T2
F06 = read.csv("counts/F06_collapse_d1.tsv", head=FALSE, sep="\t") %>%
  select(-V3)
colnames(F06) <- c("BC","F06")

# F07 - Lib16 - Growth on M9 - 50.0 ug/ml TMP - T2
F07 = read.csv("counts/F07_collapse_d1.tsv", head=FALSE, sep="\t") %>%
  select(-V3)
colnames(F07) <- c("BC","F07")

# F08 - Lib16 - Growth on M9 - 200.0 ug/ml TMP - T2
F08 = read.csv("counts/F08_collapse_d1.tsv", head=FALSE, sep="\t") %>%
  select(-V3)
colnames(F08) <- c("BC","F08")

## Library 16 - T3

# G04 - Lib16 - Growth on M9 - 0 ug/mL TMP - T3
G04 = read.csv("counts/G04_collapse_d1.tsv", head=FALSE, sep="\t") %>%
  select(-V3)
colnames(G04) <- c("BC","G04")

# G05 - Lib16 - Growth on M9 - 0.058 ug/ml TMP - T3
G05 = read.csv("counts/G05_collapse_d1.tsv", head=FALSE, sep="\t") %>%
  select(-V3)
colnames(G05) <- c("BC","G05")

# G06 - Lib16 - Growth on M9 - 0.5 ug/ml TMP - T3
G06 = read.csv("counts/G06_collapse_d1.tsv", head=FALSE, sep="\t") %>%
  select(-V3)
colnames(G06) <- c("BC","G06")

# G07 - Lib16 - Growth on M9 - 1.0 ug/ml TMP - T3
G07 = read.csv("counts/G07_collapse_d1.tsv", head=FALSE, sep="\t") %>%
  select(-V3)
colnames(G07) <- c("BC","G07")

# G08 - Lib16 - Growth on M9 - 10.0 ug/ml TMP - T3
G08 = read.csv("counts/G08_collapse_d1.tsv", head=FALSE, sep="\t") %>%
  select(-V3)
colnames(G08) <- c("BC","G08")

# G09 - Lib16 - Growth on M9 - 50.0 ug/ml TMP - T3
G09 = read.csv("counts/G09_collapse_d1.tsv", head=FALSE, sep="\t") %>%
  select(-V3)
colnames(G09) <- c("BC","G09")

# G10 - Lib16 - Growth on M9 - 200.0 ug/ml TMP - T3
G10 = read.csv("counts/G10_collapse_d1.tsv", head=FALSE, sep="\t") %>%
  select(-V3)
colnames(G10) <- c("BC","G10")

## Library 16 - T4

# H06 - Lib16 - Growth on M9 - 0 ug/mL TMP - T4
H06 = read.csv("counts/H06_collapse_d1.tsv", head=FALSE, sep="\t") %>%
  select(-V3)
colnames(H06) <- c("BC","H06")

# H07 - Lib16 - Growth on M9 - 0.058 ug/ml TMP - T4
H07 = read.csv("counts/H07_collapse_d1.tsv", head=FALSE, sep="\t") %>%
  select(-V3)
colnames(H07) <- c("BC","H07")

# H08 - Lib16 - Growth on M9 - 0.5 ug/ml TMP - T4
H08 = read.csv("counts/H08_collapse_d1.tsv", head=FALSE, sep="\t") %>%
  select(-V3)
colnames(H08) <- c("BC","H08")

# H09 - Lib16 - Growth on M9 - 1.0 ug/ml TMP - T4
H09 = read.csv("counts/H09_collapse_d1.tsv", head=FALSE, sep="\t") %>%
  select(-V3)
colnames(H09) <- c("BC","H09")

# H10 - Lib16 - Growth on M9 - 10.0 ug/ml TMP - T4
H10 = read.csv("counts/H10_collapse_d1.tsv", head=FALSE, sep="\t") %>%
  select(-V3)
colnames(H10) <- c("BC","H10")

# H11 - Lib16 - Growth on M9 - 50.0 ug/ml TMP - T4
H11 = read.csv("counts/H11_collapse_d1.tsv", head=FALSE, sep="\t") %>%
  select(-V3)
colnames(H11) <- c("BC","H11")

# H12 - Lib16 - Growth on M9 - 200.0 ug/ml TMP - T4
H12 = read.csv("counts/H12_collapse_d1.tsv", head=FALSE, sep="\t") %>%
  select(-V3)
colnames(H12) <- c("BC","H12")
```

# Summarized Data

**Summarize the imported count data by sampling condition for each codon-version library.** Begin by combining the count data from each sampling condition into a single dataframe for each library. Next, sum the full count values for each sampling condition (total number of sequences recovered for each sampling condition). Then, sum the number of unique barcodes recovered from each sampling condition. Define the sampling condition associated with each sample code. Finally, write the data summary to a .csv file and save in the working directory.

**Summarize Library 15**
```{r}
## Library 15

# Merge data objects together into single dataframe
norm_vals_15 <- data.frame(sample=c("D01","D03","D05","D06","D07","D08","D09","D10","D11","E07","E08","E09","E10","E11","E12","F01","F09","F10","F11","F12","G01","G02","G03","G11","G12","H01","H02","H03","H04","H05"),
                        counts=c(sum(D01$D01),sum(D03$D03),sum(D05$D05),sum(D06$D06),sum(D07$D07),sum(D08$D08),
                                 sum(D09$D09),sum(D10$D10),sum(D11$D11),sum(E07$E07),sum(E08$E08),sum(E09$E09),
                                 sum(E10$E10),sum(E11$E11),sum(E12$E12),sum(F01$F01),sum(F09$F09),sum(F10$F10),
                                 sum(F11$F11),sum(F12$F12),sum(G01$G01),sum(G02$G02),sum(G03$G03),sum(G11$G11),
                                 sum(G12$G12),sum(H01$H01),sum(H02$H02),sum(H03$H03),sum(H04$H04),sum(H05$H05)),
                        numBCs=c(length(D01$D01),length(D03$D03),length(D05$D05),length(D06$D06),length(D07$D07),
                                 length(D08$D08),length(D09$D09),length(D10$D10),length(D11$D11),length(E07$E07),
                                 length(E08$E08),length(E09$E09),length(E10$E10),length(E11$E11),length(E12$E12),
                                 length(F01$F01),length(F09$F09),length(F10$F10),length(F11$F11),length(F12$F12),
                                 length(G01$G01),length(G02$G02),length(G03$G03),length(G11$G11),length(G12$G12),
                                 length(H01$H01),length(H02$H02),length(H03$H03),length(H04$H04),length(H05$H05)),
                        sampledesc=c("D01 - Lib15 - T0 - Overnight Growth on LB",
                                     "D03 - Lib15 - T0 - Growth on M9 - Full Supplement",
                                     "D05 - Lib15 - T1 - Growth on M9 - 0.0 ug/ml TMP",
                                     "D06 - Lib15 - T1 - Growth on M9 - 0.058 ug/ml TMP",
                                     "D07 - Lib15 - T1 - Growth on M9 - 0.5 ug/ml TMP",
                                     "D08 - Lib15 - T1 - Growth on M9 - 1.0 ug/ml TMP",
                                     "D09 - Lib15 - T1 - Growth on M9 - 10.0 ug/ml TMP",
                                     "D10 - Lib15 - T1 - Growth on M9 - 50.0 ug/ml TMP",
                                     "D11 - Lib15 - T1 - Growth on M9 - 200.0 ug/ml TMP",
                                     "E07 - Lib15 - T2 - Growth on M9 - 0.0 ug/ml TMP",
                                     "E08 - Lib15 - T2 - Growth on M9 - 0.058 ug/ml TMP",
                                     "E09 - Lib15 - T2 - Growth on M9 - 0.5 ug/ml TMP",
                                     "E10 - Lib15 - T2 - Growth on M9 - 1.0 ug/ml TMP",
                                     "E11 - Lib15 - T2 - Growth on M9 - 10.0 ug/ml TMP",
                                     "E12 - Lib15 - T2 - Growth on M9 - 50.0 ug/ml TMP",
                                     "F01 - Lib15 - T2 - Growth on M9 - 200.0 ug/ml TMP",
                                     "F09 - Lib15 - T3 - Growth on M9 - 0.0 ug/ml TMP",
                                     "F10 - Lib15 - T3 - Growth on M9 - 0.058 ug/ml TMP",
                                     "F11 - Lib15 - T3 - Growth on M9 - 0.5 ug/ml TMP",
                                     "F12 - Lib15 - T3 - Growth on M9 - 1.0 ug/ml TMP",
                                     "G01 - Lib15 - T3 - Growth on M9 - 10.0 ug/ml TMP",
                                     "G02 - Lib15 - T3 - Growth on M9 - 50.0 ug/ml TMP",
                                     "G03 - Lib15 - T3 - Growth on M9 - 200.0 ug/ml TMP",
                                     "G11 - Lib15 - T4 - Growth on M9 - 0.0 ug/ml TMP",
                                     "G12 - Lib15 - T4 - Growth on M9 - 0.058 ug/ml TMP",
                                     "H01 - Lib15 - T4 - Growth on M9 - 0.5 ug/ml TMP",
                                     "H02 - Lib15 - T4 - Growth on M9 - 1.0 ug/ml TMP",
                                     "H03 - Lib15 - T4 - Growth on M9 - 10.0 ug/ml TMP",
                                     "H04 - Lib15 - T4 - Growth on M9 - 50.0 ug/ml TMP",
                                     "H05 - Lib15 - T4 - Growth on M9 - 200.0 ug/ml TMP"))

# Write dataframe to .csv table
write.csv(norm_vals_15, 'OUTPUT/novaseq_stats/novaseq_basic_stats_Lib15.csv', row.names = FALSE, quote=FALSE)
```

<font color="green">**Library 15 Summarized Count Data for Sequences and Barcodes (BC)**</font>
```{r echo=FALSE}
# Display dataframe with kable formatting
norm_vals_15_col1 <- norm_vals_15[,c(4,1,2,3)]

knitr::kable(norm_vals_15_col1,
  col.names = c('Sample Conditions', 'Sample ID', 'Seq Counts', 'BC Counts'),
  align = "lccl",
  format.args = list(big.mark = ","))
```

**Summarize Library 16**
```{r}
## Library 16

# Merge data objects together into single dataframe
norm_vals_16 <- data.frame(sample=c("D02","D04",
                                    "D12","E01","E02","E03","E04","E05","E06",
                                    "F02","F03","F04","F05","F06","F07","F08",
                                    "G04","G05","G06","G07","G08","G09","G10",
                                    "H06","H07","H08","H09","H10","H11","H12"),
                        counts=c(sum(D02$D02),sum(D04$D04),sum(D12$D12),
                                 sum(E01$E01),sum(E02$E02),sum(E03$E03),sum(E04$E04),sum(E05$E05),sum(E06$E06),
                                 sum(F02$F02),sum(F03$F03),sum(F04$F04),sum(F05$F05),sum(F06$F06),sum(F07$F07),
                                 sum(F08$F08),sum(G04$G04),sum(G05$G05),sum(G06$G06),sum(G07$G07),sum(G08$G08),
                                 sum(G09$G09),sum(G10$G10),sum(H06$H06),sum(H07$H07),sum(H08$H08),sum(H09$H09),
                                 sum(H10$H10),sum(H11$H11),sum(H12$H12)),
                        numBCs=c(length(D02$D02),length(D04$D04),length(D12$D12),length(E01$E01),length(E02$E02),
                                 length(E03$E03),length(E04$E04),length(E05$E05),length(E06$E06),length(F02$F02),
                                 length(F03$F03),length(F04$F04),length(F05$F05),length(F06$F06),length(F07$F07),
                                 length(F08$F08),length(G04$G04),length(G05$G05),length(G06$G06),length(G07$G07),
                                 length(G08$G08),length(G09$G09),length(G10$G10),length(H06$H06),length(H07$H07),
                                 length(H08$H08),length(H09$H09),length(H10$H10),length(H11$H11),length(H12$H12)),
                        sampledesc=c("D02 - Lib16 - T0 - Overnight Growth on LB",
                                     "D04 - Lib16 - T0 - Growth on M9 - Full Supplement",
                                     "D12 - Lib16 - T1 - Growth on M9 - 0.0 ug/ml TMP",
                                     "E01 - Lib16 - T1 - Growth on M9 - 0.058 ug/ml TMP",
                                     "E02 - Lib16 - T1 - Growth on M9 - 0.5 ug/ml TMP",
                                     "E03 - Lib16 - T1 - Growth on M9 - 1.0 ug/ml TMP",
                                     "E04 - Lib16 - T1 - Growth on M9 - 10.0 ug/ml TMP",
                                     "E05 - Lib16 - T1 - Growth on M9 - 50.0 ug/ml TMP",
                                     "E06 - Lib16 - T1 - Growth on M9 - 200.0 ug/ml TMP",
                                     "F02 - Lib16 - T2 - Growth on M9 - 0.0 ug/ml TMP",
                                     "F03 - Lib16 - T2 - Growth on M9 - 0.058 ug/ml TMP",
                                     "F04 - Lib16 - T2 - Growth on M9 - 0.5 ug/ml TMP",
                                     "F05 - Lib16 - T2 - Growth on M9 - 1.0 ug/ml TMP",
                                     "F06 - Lib16 - T2 - Growth on M9 - 10.0 ug/ml TMP",
                                     "F07 - Lib16 - T2 - Growth on M9 - 50.0 ug/ml TMP",
                                     "F08 - Lib16 - T2 - Growth on M9 - 200.0 ug/ml TMP",
                                     "G04 - Lib16 - T3 - Growth on M9 - 0.0 ug/ml TMP",
                                     "G05 - Lib16 - T3 - Growth on M9 - 0.058 ug/ml TMP",
                                     "G06 - Lib16 - T3 - Growth on M9 - 0.5 ug/ml TMP",
                                     "G07 - Lib16 - T3 - Growth on M9 - 1.0 ug/ml TMP",
                                     "G08 - Lib16 - T3 - Growth on M9 - 10.0 ug/ml TMP",
                                     "G09 - Lib16 - T3 - Growth on M9 - 50.0 ug/ml TMP",
                                     "G10 - Lib16 - T3 - Growth on M9 - 200.0 ug/ml TMP",
                                     "H06 - Lib16 - T4 - Growth on M9 - 0.0 ug/ml TMP",
                                     "H07 - Lib16 - T4 - Growth on M9 - 0.058 ug/ml TMP",
                                     "H08 - Lib16 - T4 - Growth on M9 - 0.5 ug/ml TMP",
                                     "H09 - Lib16 - T4 - Growth on M9 - 1.0 ug/ml TMP",
                                     "H10 - Lib16 - T4 - Growth on M9 - 10.0 ug/ml TMP",
                                     "H11 - Lib16 - T4 - Growth on M9 - 50.0 ug/ml TMP",
                                     "H12 - Lib16 - T4 - Growth on M9 - 200.0 ug/ml TMP"))

# Write dataframe to .csv table
write.csv(norm_vals_16, 'OUTPUT/novaseq_stats/novaseq_basic_stats_Lib16.csv', row.names = FALSE, quote=FALSE)
```

<font color="green">**Table of Summarized Count Data for Sequences and Barcodes (BC)**</font>
```{r echo=FALSE}
# Display dataframe with kable formatting
norm_vals_16_col1 <- norm_vals_16[,c(4,1,2,3)]

knitr::kable(norm_vals_16_col1,
  col.names = c('Sample Conditions', 'Sample ID', 'Seq Counts', 'BC Counts'),
  align = "lccl",
  format.args = list(big.mark = ","))
```

## Merged Data
Merge the count data from each library for each sampling condition by barcode for downstream analysis. 
```{r}
# Library 15
BCs15 <- D01 %>%
  full_join(D03,by="BC") %>%
  full_join(D05,by="BC") %>%
  full_join(D06,by="BC") %>%
  full_join(D07,by="BC") %>%
  full_join(D08,by="BC") %>%
  full_join(D09,by="BC") %>%
  full_join(D10,by="BC") %>%
  full_join(D11,by="BC") %>%
  full_join(E07,by="BC") %>%
  full_join(E08,by="BC") %>%
  full_join(E09,by="BC") %>%
  full_join(E10,by="BC") %>%
  full_join(E11,by="BC") %>%
  full_join(E12,by="BC") %>%
  full_join(F01,by="BC") %>%
  full_join(F09,by="BC") %>%
  full_join(F10,by="BC") %>%
  full_join(F11,by="BC") %>%
  full_join(F12,by="BC") %>%
  full_join(G01,by="BC") %>%
  full_join(G02,by="BC") %>%
  full_join(G03,by="BC") %>%
  full_join(G11,by="BC") %>%
  full_join(G12,by="BC") %>%
  full_join(H01,by="BC") %>%
  full_join(H02,by="BC") %>%
  full_join(H03,by="BC") %>%
  full_join(H04,by="BC") %>%
  full_join(H05,by="BC")

# Library 16
BCs16 <- D02 %>%
  full_join(D04,by="BC") %>%
  full_join(D12,by="BC") %>%
  full_join(E01,by="BC") %>%
  full_join(E02,by="BC") %>%
  full_join(E03,by="BC") %>%
  full_join(E04,by="BC") %>%
  full_join(E05,by="BC") %>%
  full_join(E06,by="BC") %>%
  full_join(F02,by="BC") %>%
  full_join(F03,by="BC") %>%
  full_join(F04,by="BC") %>%
  full_join(F05,by="BC") %>%
  full_join(F06,by="BC") %>%
  full_join(F07,by="BC") %>%
  full_join(F08,by="BC") %>%
  full_join(G04,by="BC") %>%
  full_join(G05,by="BC") %>%
  full_join(G06,by="BC") %>%
  full_join(G07,by="BC") %>%
  full_join(G08,by="BC") %>%
  full_join(G09,by="BC") %>%
  full_join(G10,by="BC") %>%
  full_join(H06,by="BC") %>%
  full_join(H07,by="BC") %>%
  full_join(H08,by="BC") %>%
  full_join(H09,by="BC") %>%
  full_join(H10,by="BC") %>%
  full_join(H11,by="BC") %>%
  full_join(H12,by="BC")
```

<font color="green">**Preview of BCs count data with N/A for blank values**</font>
```{r echo=FALSE}
head(BCs15)
head(BCs16)
```

Convert all "NA" values to "0" in the dataframe:
```{r}
# Library 15
BCs15$D01[is.na(BCs15$D01)] <- 0
BCs15$D03[is.na(BCs15$D03)] <- 0
BCs15$D05[is.na(BCs15$D05)] <- 0
BCs15$D06[is.na(BCs15$D06)] <- 0
BCs15$D07[is.na(BCs15$D07)] <- 0
BCs15$D08[is.na(BCs15$D08)] <- 0
BCs15$D09[is.na(BCs15$D09)] <- 0
BCs15$D10[is.na(BCs15$D10)] <- 0
BCs15$D11[is.na(BCs15$D11)] <- 0
BCs15$E07[is.na(BCs15$E07)] <- 0
BCs15$E08[is.na(BCs15$E08)] <- 0
BCs15$E09[is.na(BCs15$E09)] <- 0
BCs15$E10[is.na(BCs15$E10)] <- 0
BCs15$E11[is.na(BCs15$E11)] <- 0
BCs15$E12[is.na(BCs15$E12)] <- 0
BCs15$F01[is.na(BCs15$F01)] <- 0
BCs15$F09[is.na(BCs15$F09)] <- 0
BCs15$F10[is.na(BCs15$F10)] <- 0
BCs15$F11[is.na(BCs15$F11)] <- 0
BCs15$F12[is.na(BCs15$F12)] <- 0
BCs15$G01[is.na(BCs15$G01)] <- 0
BCs15$G02[is.na(BCs15$G02)] <- 0
BCs15$G03[is.na(BCs15$G03)] <- 0
BCs15$G11[is.na(BCs15$G11)] <- 0
BCs15$G12[is.na(BCs15$G12)] <- 0
BCs15$H01[is.na(BCs15$H01)] <- 0
BCs15$H02[is.na(BCs15$H02)] <- 0
BCs15$H03[is.na(BCs15$H03)] <- 0
BCs15$H04[is.na(BCs15$H04)] <- 0
BCs15$H05[is.na(BCs15$H05)] <- 0

# Library 16
BCs16$D02[is.na(BCs16$D02)] <- 0
BCs16$D04[is.na(BCs16$D04)] <- 0
BCs16$D12[is.na(BCs16$D12)] <- 0
BCs16$E01[is.na(BCs16$E01)] <- 0
BCs16$E02[is.na(BCs16$E02)] <- 0
BCs16$E03[is.na(BCs16$E03)] <- 0
BCs16$E04[is.na(BCs16$E04)] <- 0
BCs16$E05[is.na(BCs16$E05)] <- 0
BCs16$E06[is.na(BCs16$E06)] <- 0
BCs16$F02[is.na(BCs16$F02)] <- 0
BCs16$F03[is.na(BCs16$F03)] <- 0
BCs16$F04[is.na(BCs16$F04)] <- 0
BCs16$F05[is.na(BCs16$F05)] <- 0
BCs16$F06[is.na(BCs16$F06)] <- 0
BCs16$F07[is.na(BCs16$F07)] <- 0
BCs16$F08[is.na(BCs16$F08)] <- 0
BCs16$G04[is.na(BCs16$G04)] <- 0
BCs16$G05[is.na(BCs16$G05)] <- 0
BCs16$G06[is.na(BCs16$G06)] <- 0
BCs16$G07[is.na(BCs16$G07)] <- 0
BCs16$G08[is.na(BCs16$G08)] <- 0
BCs16$G09[is.na(BCs16$G09)] <- 0
BCs16$G10[is.na(BCs16$G10)] <- 0
BCs16$H06[is.na(BCs16$H06)] <- 0
BCs16$H07[is.na(BCs16$H07)] <- 0
BCs16$H08[is.na(BCs16$H08)] <- 0
BCs16$H09[is.na(BCs16$H09)] <- 0
BCs16$H10[is.na(BCs16$H10)] <- 0
BCs16$H11[is.na(BCs16$H11)] <- 0
BCs16$H12[is.na(BCs16$H12)] <- 0
```

<font color="green">**Preview of BCs count data with zeros ("0") for blank values for both libraries**</font>
```{r echo=FALSE}
head(BCs15)
head(BCs16)
```

**Total BCs:** Count the number of unique barcode for each library
```{r class.output="goodCode"}
# Count the number of unique barcodes for Library 15
BCs15.count <- length(unique(BCs15$BC))
format(BCs15.count, big.mark = ",")

# Count the number of unique barcodes for Library 16
BCs16.count <- length(unique(BCs16$BC))
format(BCs16.count, big.mark = ",")
```

## BC Pruning

Remove unique BCs less than 20 nucleotides in length as these do not represent true BCs from designed pool.
```{r class.output="goodCode"}
# Lib15: Sum the count of unique BC lengths in the BCs objects
BCs15.unique.table <- table(str_length(BCs15$BC))
BCs15.unique.table

# Lib16: Sum the count of unique BC lengths in the BCs objects
BCs16.unique.table <- table(str_length(BCs16$BC))
BCs16.unique.table
```

```{r}
# Filter rows where the length of BC is greater than or equal to 20

# Lib15
BCs15.prune <- BCs15 %>%
  filter(str_length(BC) >= 20)

# Lib16
BCs16.prune <- BCs16 %>%
  filter(str_length(BC) >= 20)
```

```{r class.output="goodCode"}
# Lib15: Sum the count of unique BC lengths in the BCprune objects
BCs15.prune.unique.table <- table(str_length(BCs15.prune$BC))
BCs15.prune.unique.table

# Lib16: Sum the count of unique BC lengths in the BCprune objects
BCs16.prune.unique.table <- table(str_length(BCs16.prune$BC))
BCs16.prune.unique.table
```

**Total BCs:** Count the number of unique barcode for each library after pruning by BC length
```{r class.output="goodCode"}
# Count the number of unique barcodes for Library 15
BCs15.prune.count <- length(unique(BCs15.prune$BC))
format(BCs15.prune.count, big.mark = ",")

# Count the number of unique barcodes for Library 16
BCs16.prune.count <- length(unique(BCs16.prune$BC))
format(BCs16.prune.count, big.mark = ",")
```

## Basic Checks

With the data from both libraries imported and merged, we can perform some basic checks to determine everything has been incorporated properly.

### Barcodes

**BCs matching mapping file:** How many of our barcodes (from count files) are in the mapping files for each library?
```{r class.output="goodCode"}
# Library 15
#BCs15_map <- inner_join(BCinfo15,BCs15,by="BC")
#BCs15_map <- inner_join(BCs_mutID_15,BCs15,by="BC")
BCs15_map <- inner_join(BCs_mutID_15,BCs15.prune,by="BC")

# Count number of barcodes (rows)
format(nrow(BCs15_map), big.mark = ",")

# Library 16
#BCs16_map <- inner_join(BCinfo16,BCs16,by="BC")
#BCs16_map <- inner_join(BCs_mutID_16,BCs16,by="BC")
BCs16_map <- inner_join(BCs_mutID_16,BCs16.prune,by="BC")

# Count number of barcodes (rows)
format(nrow(BCs16_map), big.mark = ",")
```

**Perfect BCs matching mapping file:** How many of our barcodes are associated with perfects (perfect sequence match with a designed homolog) for each unique library after filtering to the BC mapping file?
```{r class.output="goodCode"}
# Library 15
BCs15_perf <- BCs15_map %>%
  filter(mutations==0)

# Count number of perfects (rows)
format(nrow(BCs15_perf), big.mark = ",")

# Library 16
BCs16_perf <- BCs16_map %>%
  filter(mutations==0)

# Count number of perfects (rows)
format(nrow(BCs16_perf), big.mark = ",")
```

**Mutant BCs matching mapping file:** How many of our barcodes are associated with mutants (sequence contains at least 1 mutation from a designed homolog) for each unique library after filtering to the BC mapping file?
```{r class.output="goodCode"}
# Library 15
BCs15_mutant <- BCs15_map %>%
  filter(mutations!=0)

# Count number of perfects (rows)
format(nrow(BCs15_mutant), big.mark = ",")

# Library 16
BCs16_mutant <- BCs16_map %>%
  filter(mutations!=0)

# Count number of perfects (rows)
format(nrow(BCs16_mutant), big.mark = ",")
```

**Unique Perfects:** How many unique perfects did we recover from all sampling conditions for each library?
```{r class.output="goodCode"}
# Library 15
perf_uniq_15 <- BCs15_map %>%
  filter(mutations==0) %>%
  select(mutID) %>%
  distinct() %>%
  nrow()
format(perf_uniq_15, big.mark = ",")

# Library 16
perf_uniq_16 <- BCs16_map %>%
  filter(mutations==0) %>%
  select(mutID) %>%
  distinct() %>%
  nrow()
format(perf_uniq_16, big.mark = ",")
```

Count the number of shared perfects (mutations == 0) recovered between libraries.
```{r class.output="goodCode"}
# Merge the mutIDinfo datasets but keep only the mutIDs shared between libraries
BCs_mutID_perfects.shared <- merge(BCs15_map, BCs16_map, by = "mutID", all = FALSE)

# Retain only the rows with a specific value in the "specific_column" column
raw_data.shared <- BCs_mutID_perfects.shared %>%
  filter(mutations.x == 0)

# Count the number of perfects (mutations == 0) shared between both libraries
BCs_mutID_perfects.shared.count <- length(unique(raw_data.shared$mutID))
format(BCs_mutID_perfects.shared.count, big.mark = ",")
```

Also count the number of perfect sequences unique to one or the other library (mutations == 0).
```{r class.output="goodCode"}
# Merge the mutIDinfo datasets but retain only the mutIDs unique to one library or the other
BCs_mutID_perfects.unique <- bind_rows(
  anti_join(BCs15_map, BCs16_map, by = "mutID"),
  anti_join(BCs16_map, BCs15_map, by = "mutID"))

# Retain only the rows with a specific value in the "specific_column" column
raw_data.unique <-BCs_mutID_perfects.unique %>%
  filter(mutations == 0)

# Count the number of perfects (mutations == 0) unique to one library or the other
BCs_mutID_perfects.unique.count <- length(unique(raw_data.unique$mutID))
format(BCs_mutID_perfects.unique.count, big.mark = ",")
```

```{r class.output="goodCode"}
# Sum the number of shared and unique perfects recovered for both codon version libraries
BCs_mutID_perfects.15.16.all.count <- sum(BCs_mutID_perfects.shared.count + BCs_mutID_perfects.unique.count)
format(BCs_mutID_perfects.15.16.all.count, big.mark = ",")
```

**Unique Mutants:** How many unique mutants did we recover from all sampling conditions for unique libraries?
```{r class.output="goodCode"}
# Library 15
mutant_15 <- BCs15_map %>%
  filter(mutations!=0) %>%
  select(mutID) %>%
  distinct() %>%
  nrow()
format(mutant_15, big.mark = ",")

# Library 16
mutant_16 <- BCs16_map %>%
  filter(mutations!=0) %>%
  select(mutID) %>%
  distinct() %>%
  nrow()
format(mutant_16, big.mark = ",")
```

#-------------------------------------------------------------------------------------------

### Normalization

Count data needs to be normalized across sampling conditions prior to making any downstream comparisons. Here, we scale the sequence counts against the "T0-Overnight Growth on LB" sampling condition. This represents "D01" for Lib15 and "D02" for Lib16.

#### Scaling Values by Condition

**Lib15 Scaling:** Create a scaling value to normalize all sequence count data by the total number of sequences at "T0 - Growth on M9 - Full Supplement" sample condition (Lib15: D03; Lib16: D04).
```{r}
# Calculating the scaling factor for each sampling condition based on D01 (overnight growth on LB) as the standard
#norm_scaling_15 <- rep.int(norm_vals_15$counts[1],nrow(norm_vals_15))/norm_vals_15$counts

# Calculating the scaling factor for each sampling condition based on D03 (T0-M9-Full Supplement) as the standard
norm_scaling_15 <- rep.int(norm_vals_15$counts[2],nrow(norm_vals_15))/norm_vals_15$counts

# Add the scaling values as a column to the "norm_vals_15" dataframe
norm_vals_15$scaling <- norm_scaling_15
```

<font color="green">**Table of Summarized Count Data including the Scaling Value for each Sample Condition**</font>
```{r echo=FALSE}
# Display dataframe with kable formatting
norm_vals_15_col2 <- norm_vals_15[,c(4,1,2,3,5)]

knitr::kable(norm_vals_15_col2,
  col.names = c('Sample Conditions', 'Sample ID', 'Seq Counts', 'BC Counts', 'Scaling'),
  align = "llccc",
  format.args = list(big.mark = ","),
  digits = 2)
```

**Lib16 Scaling**
```{r}
# Calculate scaling factor for each sampling condition based on D02 (overnight growth on LB) as the standard
#norm_scaling_16 <- rep.int(norm_vals_16$counts[1],nrow(norm_vals_16))/norm_vals_16$counts

# Calculate scaling factor for each sampling condition based on D04 (T0 - M9 - Full Supplement) as the standard
norm_scaling_16 <- rep.int(norm_vals_16$counts[2],nrow(norm_vals_16))/norm_vals_16$counts

# Add the scaling values as a column to the "norm_vals_16" dataframe
norm_vals_16$scaling <- norm_scaling_16
```

<font color="green">**Table of Summarized Count Data including the Scaling Value for each Sample Condition**</font>
```{r echo=FALSE}
# Display dataframe with kable formatting
norm_vals_16_col2 <- norm_vals_16[,c(4,1,2,3,5)]

knitr::kable(norm_vals_16_col2,
  col.names = c('Sample Conditions', 'Sample ID', 'Seq Counts', 'BC Counts', 'Scaling'),
  align = "llccc",
  format.args = list(big.mark = ","),
  digits = 2)
```

#### Scaling Counts by Barcode
<font color="green">**The following table displays a preview of the original count values by unique barcode:**</font>

**Lib15 BC Scaling**
```{r echo=FALSE}
# Display dataframe with kable formatting
#BCs15_head <- head(BCs15)
BCs15_head <- head(BCs15.prune)

knitr::kable(BCs15_head,
  col.names = c('BC', 'D01', 'D03', 'D05', 'D06', 'D07', 'D08', 'D09', 'D10', 'D11', 
                'E07', 'E08', 'E09', 'E10', 'E11', 'E12', 'F01', 
                'F09', 'F10', 'F11', 'F12', 'G01', 'G02', 'G03', 
                'G11', 'G12', 'H01', 'H02', 'H03', 'H04', 'H05'),
  align = "lccccccccccc",
  format.args = list(big.mark = ","),
  digits = 0)
```

For each sample, make a normalized counts value based on sequencing depth using the scaling values derived for each sampling condition in the table above. 
```{r}
# Make new columns for the normalized counts in the BCs object:
#oldnames15 <- colnames(BCs15)
oldnames15 <- colnames(BCs15.prune)
newnamesfrac15 = c('D01n', 'D03n', 
                   'D05n', 'D06n', 'D07n', 'D08n', 'D09n', 'D10n', 'D11n', 
                   'E07n', 'E08n', 'E09n', 'E10n', 'E11n', 'E12n', 'F01n', 
                   'F09n', 'F10n', 'F11n', 'F12n', 'G01n', 'G02n', 'G03n', 
                   'G11n', 'G12n', 'H01n', 'H02n', 'H03n', 'H04n', 'H05n')

# Add the new columns to the BCs object based on the number of original columns for sampling conditions:
#numcol_15 <- ncol(BCs15)
numcol_15 <- ncol(BCs15.prune)
#D01index = which( colnames(BCs15)=="D01" )
D01index = which( colnames(BCs15.prune)=="D01" )

# For each sample, re-scale the number of reads by relative sequencing depth of the sample using the scaling values in the norm_vals object:
for (i in 1:30){
  BCs15.prune[,numcol_15+i] <- BCs15.prune[,D01index+i-1]*as.numeric(norm_vals_15$scaling[i])
}

# Add the new column names to the BCs object:
#colnames(BCs15) <- c(oldnames15,newnamesfrac15)
colnames(BCs15.prune) <- c(oldnames15,newnamesfrac15)

# Clean up objects from the global environment
#rm(oldnames15,newnamesfrac15,numcol15,D01index,i)
```

<font color="green">**The following table displays a preview of the normalized count values by unique barcode:**</font>
```{r echo=FALSE}
#BCs_norm_15 <- BCs15[,c(1,32:61)]
BCs_norm_15 <- BCs15.prune[,c(1,32:61)]

# Display dataframe with kable formatting
BCs_norm_head_15 <- head(BCs_norm_15)

knitr::kable(BCs_norm_head_15,
  col.names = c('BC', 'D01n', 'D03n', 'D05n', 'D06n', 'D07n', 'D08n', 'D09n', 'D10n', 'D11n', 
                'E07n', 'E08n', 'E09n', 'E10n', 'E11n', 'E12n', 'F01n', 
                'F09n', 'F10n', 'F11n', 'F12n', 'G01n', 'G02n', 'G03n', 
                'G11n', 'G12n', 'H01n', 'H02n', 'H03n', 'H04n', 'H05n'),
  align = "lccccccccccc",
  format.args = list(big.mark = ","),
  digits = 0)
```

**Lib16 BC Scaling**
```{r echo=FALSE}
# Display dataframe with kable formatting
#BCs16_head <- head(BCs16)
BCs16_head <- head(BCs16.prune)

knitr::kable(BCs16_head,
  col.names = c('BC', 'D02', 'D04', 'D12', 'E01', 'E02', 'E03', 'E04', 'E05', 'E06', 
                'F02', 'F03', 'F04', 'F05', 'F06', 'F07', 'F08', 
                'G04', 'G05', 'G06', 'G07', 'G08', 'G09', 'G10', 
                'H06', 'H07', 'H08', 'H09', 'H10', 'H11', 'H12'),
  align = "lccccccccccc",
  format.args = list(big.mark = ","),
  digits = 0)
```

For each sample, make a normalized counts value based on sequencing depth using the scaling values derived for each sampling condition in the table above. 
```{r}
# Make new columns for the normalized counts in the BCs object:
#oldnames16 <- colnames(BCs16)
oldnames16 <- colnames(BCs16.prune)
newnamesfrac16 = c('D02n', 'D04n', 
                   'D12n', 'E01n', 'E02n', 'E03n', 'E04n', 'E05n', 'E06n', 
                   'F02n', 'F03n', 'F04n', 'F05n', 'F06n', 'F07n', 'F08n', 
                   'G04n', 'G05n', 'G06n', 'G07n', 'G08n', 'G09n', 'G10n', 
                   'H06n', 'H07n', 'H08n', 'H09n', 'H10n', 'H11n', 'H12n')

# Add the new columns to the BCs object based on the number of original columns for sampling conditions:
#numcol_16 <- ncol(BCs16)
numcol_16 <- ncol(BCs16.prune)
#D02index = which( colnames(BCs16)=="D02" )
D02index = which( colnames(BCs16.prune)=="D02" )

# For each dilution, re-scale the number of reads by relative sequencing depth of the sample using the scaling values in the norm_vals object:
for (i in 1:30){
  BCs16.prune[,numcol_16+i] <- BCs16.prune[,D02index+i-1]*as.numeric(norm_vals_16$scaling[i])
}

# Add the new column names to the BCs object:
#colnames(BCs16) <- c(oldnames16,newnamesfrac16)
colnames(BCs16.prune) <- c(oldnames16,newnamesfrac16)

# Clean up objects from the global environment
#rm(oldnames16,newnamesfrac16,numcol16,D02index,i)
```

<font color="green">**The following table displays a preview of the normalized count values by unique barcode:**</font>
```{r echo=FALSE}
#BCs_norm_16 <- BCs16[,c(1,32:61)]
BCs_norm_16 <- BCs16.prune[,c(1,32:61)]

# Display dataframe with kable formatting
BCs_norm_head_16 <- head(BCs_norm_16)

knitr::kable(BCs_norm_head_16,
  col.names = c('BC', 'D02n', 'D04n',
                'D12n', 'E01n', 'E02n', 'E03n', 'E04n', 'E05n', 'E06n', 
                'F02n', 'F03n', 'F04n', 'F05n', 'F06n', 'F07n', 'F08n', 
                'G04n', 'G05n', 'G06n', 'G07n', 'G08n', 'G09n', 'G10n', 
                'H06n', 'H07n', 'H08n', 'H09n', 'H10n', 'H11n', 'H12n'),
  align = "lccccccccccc",
  format.args = list(big.mark = ","),
  digits = 0)
```

### Log2-Fold Change

Determine the log2-fold changes for each barcode relative to the M9-Supplementation condition (D03/D04).
```{r}
pseudocount = 1

#Library 15
BCs15.prune <- BCs15.prune %>%
  mutate(D03D01fc=log2(D03n+pseudocount)-log2(D01n+pseudocount),
         D05D03fc=log2(D05n+pseudocount)-log2(D03n+pseudocount),
         D06D03fc=log2(D06n+pseudocount)-log2(D03n+pseudocount),
         D07D03fc=log2(D07n+pseudocount)-log2(D03n+pseudocount),
         D08D03fc=log2(D08n+pseudocount)-log2(D03n+pseudocount),
         D09D03fc=log2(D09n+pseudocount)-log2(D03n+pseudocount),
         D10D03fc=log2(D10n+pseudocount)-log2(D03n+pseudocount),
         D11D03fc=log2(D11n+pseudocount)-log2(D03n+pseudocount),
         E07D03fc=log2(E07n+pseudocount)-log2(D03n+pseudocount),
         E08D03fc=log2(E08n+pseudocount)-log2(D03n+pseudocount),
         E09D03fc=log2(E09n+pseudocount)-log2(D03n+pseudocount),
         E10D03fc=log2(E10n+pseudocount)-log2(D03n+pseudocount),
         E11D03fc=log2(E11n+pseudocount)-log2(D03n+pseudocount),
         E12D03fc=log2(E12n+pseudocount)-log2(D03n+pseudocount),
         F01D03fc=log2(F01n+pseudocount)-log2(D03n+pseudocount),
         F09D03fc=log2(F09n+pseudocount)-log2(D03n+pseudocount),
         F10D03fc=log2(F10n+pseudocount)-log2(D03n+pseudocount),
         F11D03fc=log2(F11n+pseudocount)-log2(D03n+pseudocount),
         F12D03fc=log2(F12n+pseudocount)-log2(D03n+pseudocount),
         G01D03fc=log2(G01n+pseudocount)-log2(D03n+pseudocount),
         G02D03fc=log2(G02n+pseudocount)-log2(D03n+pseudocount),
         G03D03fc=log2(G03n+pseudocount)-log2(D03n+pseudocount),
         G11D03fc=log2(G11n+pseudocount)-log2(D03n+pseudocount),
         G12D03fc=log2(G12n+pseudocount)-log2(D03n+pseudocount),
         H01D03fc=log2(H01n+pseudocount)-log2(D03n+pseudocount),
         H02D03fc=log2(H02n+pseudocount)-log2(D03n+pseudocount),
         H03D03fc=log2(H03n+pseudocount)-log2(D03n+pseudocount),
         H04D03fc=log2(H04n+pseudocount)-log2(D03n+pseudocount),
         H05D03fc=log2(H05n+pseudocount)-log2(D03n+pseudocount))

#Library 16
BCs16.prune <- BCs16.prune %>%
  mutate(D04D02fc=log2(D04n+pseudocount)-log2(D02n+pseudocount),
         D12D04fc=log2(D12n+pseudocount)-log2(D04n+pseudocount),
         E01D04fc=log2(E01n+pseudocount)-log2(D04n+pseudocount),
         E02D04fc=log2(E02n+pseudocount)-log2(D04n+pseudocount),
         E03D04fc=log2(E03n+pseudocount)-log2(D04n+pseudocount),
         E04D04fc=log2(E04n+pseudocount)-log2(D04n+pseudocount),
         E05D04fc=log2(E05n+pseudocount)-log2(D04n+pseudocount),
         E06D04fc=log2(E06n+pseudocount)-log2(D04n+pseudocount),
         F02D04fc=log2(F02n+pseudocount)-log2(D04n+pseudocount),
         F03D04fc=log2(F03n+pseudocount)-log2(D04n+pseudocount),
         F04D04fc=log2(F04n+pseudocount)-log2(D04n+pseudocount),
         F05D04fc=log2(F05n+pseudocount)-log2(D04n+pseudocount),
         F06D04fc=log2(F06n+pseudocount)-log2(D04n+pseudocount),
         F07D04fc=log2(F07n+pseudocount)-log2(D04n+pseudocount),
         F08D04fc=log2(F08n+pseudocount)-log2(D04n+pseudocount),
         G04D04fc=log2(G04n+pseudocount)-log2(D04n+pseudocount),
         G05D04fc=log2(G05n+pseudocount)-log2(D04n+pseudocount),
         G06D04fc=log2(G06n+pseudocount)-log2(D04n+pseudocount),
         G07D04fc=log2(G07n+pseudocount)-log2(D04n+pseudocount),
         G08D04fc=log2(G08n+pseudocount)-log2(D04n+pseudocount),
         G09D04fc=log2(G09n+pseudocount)-log2(D04n+pseudocount),
         G10D04fc=log2(G10n+pseudocount)-log2(D04n+pseudocount),
         H06D04fc=log2(H06n+pseudocount)-log2(D04n+pseudocount),
         H07D04fc=log2(H07n+pseudocount)-log2(D04n+pseudocount),
         H08D04fc=log2(H08n+pseudocount)-log2(D04n+pseudocount),
         H09D04fc=log2(H09n+pseudocount)-log2(D04n+pseudocount),
         H10D04fc=log2(H10n+pseudocount)-log2(D04n+pseudocount),
         H11D04fc=log2(H11n+pseudocount)-log2(D04n+pseudocount),
         H12D04fc=log2(H12n+pseudocount)-log2(D04n+pseudocount))
```

**Lib15 Log2-FC**
<font color="green">**The following table displays a preview of the log2-fold change values between sampling treatments:**</font>
```{r echo=FALSE}
#BCs_norm_log_15 <- BCs15[,c(1,62:90)]
BCs_norm_log_15 <- BCs15.prune[,c(1,62:90)]

# Display dataframe with kable formatting
BCs_norm_log_head_15 <- head(BCs_norm_log_15)

knitr::kable(BCs_norm_log_head_15,
  col.names = c('BC', 'D03-D01fc', 
                'D05-D03fc', 'D06-D03fc', 'D07-D03fc', 'D08-D03fc', 'D09-D03fc', 'D10-D03fc', 'D11-D03fc', 
                'E07-D03fc', 'E08-D03fc', 'E09-D03fc', 'E10-D03fc', 'E11-D03fc', 'E12-D03fc', 'F01-D03fc', 
                'F09-D03fc', 'F10-D03fc', 'F11-D03fc', 'F12-D03fc', 'G01-D03fc', 'G02-D03fc', 'G03-D03fc', 
                'G11-D03fc', 'G12-D03fc', 'H01-D03fc', 'H02-D03fc', 'H03-D03fc', 'H04-D03fc', 'H05-D03fc'),
  align = "lcccccccccc",
  format.args = list(big.mark = ","),
  digits = 0)
```

**Lib16 Log2-FC**
<font color="green">**The following table displays a preview of the log2-fold change values between sampling treatments:**</font>
```{r echo=FALSE}
#BCs_norm_log_16 <- BCs16[,c(1,62:90)]
BCs_norm_log_16 <- BCs16.prune[,c(1,62:90)]

# Display dataframe with kable formatting
BCs_norm_log_head_16 <- head(BCs_norm_log_16)

knitr::kable(BCs_norm_log_head_16,
  col.names = c('BC', 'D04-D02fc', 
                'D12-D04fc', 'E01-D04fc', 'E02-D04fc', 'E03-D04fc', 'E04-D04fc', 'E05-D04fc', 'E06-D04fc', 
                'F02-D04fc', 'F03-D04fc', 'F04-D04fc', 'F05-D04fc', 'F06-D04fc', 'F07-D04fc', 'F08-D04fc', 
                'G04-D04fc', 'G05-D04fc', 'G06-D04fc', 'G07-D04fc', 'G08-D04fc', 'G09-D04fc', 'G10-D04fc', 
                'H06-D04fc', 'H07-D04fc', 'H08-D04fc', 'H09-D04fc', 'H10-D04fc', 'H11-D04fc', 'H12-D04fc'),
  align = "lcccccccccc",
  format.args = list(big.mark = ","),
  digits = 0)
```


### Total Reads & BCs

Add count values, normalized count values, and log2-FC values to BCmut objects
```{r class.output="goodCode"}
# Library 15

# Add the count values, normalized count values, and log2-FC values to the BCs_mutID object
#BCmut_15 <- inner_join(BCinfo15,BCs15,by="BC")
#BCmut_15 <- inner_join(BCs_mutID_15,BCs15,by="BC")
BCmut_15 <- inner_join(BCs_mutID_15,BCs15.prune,by="BC")

# Count the number of BCs prior to filtering by minimum sequence count
BCmut_15.pre.count <- BCmut_15 %>% nrow()
format(BCmut_15.pre.count, big.mark = ",")

# Library 16

# Add the count values, normalized count values, and log2-FC values to the BCs_mutID object
#BCmut_16 <- inner_join(BCinfo16,BCs16,by="BC")
#BCmut_16 <- inner_join(BCs_mutID_16,BCs16,by="BC")
BCmut_16 <- inner_join(BCs_mutID_16,BCs16.prune,by="BC")

# Count the number of BCs prior to filtering by minimum sequence count
BCmut_16.pre.count <- BCmut_16 %>% nrow()
format(BCmut_16.pre.count, big.mark = ",")
```

#### BC Filtering

Determine the number of unique barcodes based on a minimum sequence count threshold across sampling conditions. Filter the barcodes by a minimum threshold of sequence reads per unique barcode. In this case, we use a mimimum cutoff of 10 sequence reads per barcode:

**Lib15 BC Count***
```{r class.output="goodCode"}
# Filter unique barcodes by minimum sequence count. This code uses "or" such that if any of the sampling conditions contain at least 10 sequences counts/barcode, then the barcode is retained for the full dataset.
BCprune15 <- BCs15.prune %>%
  filter(D01>9 | D03>9 | D05>9 | D06>9 | D07>9 | D08>9 | D09>9 | D10>9 | D11>9 | 
           E07>9 | E08>9 | E09>9 | E10>9 | E11>9 | E12>9 | F01>9 | 
           F09>9 | F10>9 | F11>9 | F12>9 | G01>9 | G02>9 | G03>9 | 
           G11>9 | G12>9 | H01>9 | H02>9 | H03>9 | H04>9 | H05>9)

# Trim the "BCmut" dataset based on the "BCprune" barcodes retained:
#BCmut_15 <- inner_join(BCinfo15,BCprune15,by="BC")
BCmut_15 <- inner_join(BCs_mutID_15,BCprune15,by="BC")

# Count the number of BCs remaining after filtering by minimum sequence count
BCmut_15.count <- BCmut_15 %>% nrow()
format(BCmut_15.count, big.mark = ",")
```

```{r class.output="goodCode"}
# Lib15: Sum the count of unique BC lengths in the BCprune objects
BCprune15.unique.table <- table(str_length(BCprune15$BC))
BCprune15.unique.table
```

**Lib16 BC Count**
```{r class.output="goodCode"}
# Filter unique barcodes by minimum sequence count. This code uses "or" such that if any of the sampling conditions contain at least 10 sequences counts/barcode, then the barcode is retained for the full dataset.
BCprune16 <- BCs16.prune %>%
  filter(D02>9 | D04>9 | D12>9 | E01>9 | E02>9 | E03>9 | E04>9 | E05>9 | E06>9 | 
          F02>9 | F03>9 | F04>9 | F05>9 | F06>9 | F07>9 | F08>9 | 
           G04>9 | G05>9 | G06>9 | G07>9 | G08>9 | G09>9 | G10>9 | 
           H06>9 | H07>9 | H08>9 | H09>9 | H10>9 | H11>9 | H12>9)

# Trim the "BCmut" dataset based on the "BCprune" barcodes retained:
#BCmut_16 <- inner_join(BCinfo16,BCprune16,by="BC")
BCmut_16 <- inner_join(BCs_mutID_16,BCprune16,by="BC")

# Count the number of BCs remaining after filtering by minimum sequence count
BCmut_16.count <- BCmut_16 %>% nrow()
format(BCmut_16.count, big.mark = ",")
```

```{r}
# Lib16: Sum the count of unique BC lengths in the BCprune objects
BCprune16.unique.table <- table(str_length(BCprune16$BC))
BCprune16.unique.table
```

#### Summaries

**Unique BCs:** Count the number of unique BCs remaining after filtering by minimum sequence count
```{r class.output="goodCode"}
# Lib15
BCmut_15.bc.count <- length(unique(BCmut_15$BC))
format(BCmut_15.bc.count, big.mark = ",")

# Lib16
BCmut_16.bc.count <- length(unique(BCmut_16$BC))
format(BCmut_16.bc.count, big.mark = ",")
```

**Unique mutIDs:** Count the number of unique mutIDs remaining after filtering by minimum sequence count
```{r class.output="goodCode"}
# Lib15
BCmut_15.mutID.count <- length(unique(BCmut_15$mutID))
format(BCmut_15.mutID.count, big.mark = ",")

# Lib16
BCmut_16.mutID.count <- length(unique(BCmut_16$mutID))
format(BCmut_16.mutID.count, big.mark = ",")
```

**Unique Sequences:** Count the number of unique protein sequences remaining after filtering by minimum sequence count. This should be the same value as the number of unique mutIDs remaining after filtering.
```{r class.output="goodCode"}
# Lib15
BCmut_15.seq.count <- length(unique(BCmut_15$seq))
format(BCmut_15.seq.count, big.mark = ",")

# Lib16
BCmut_16.seq.count <- length(unique(BCmut_16$seq))
format(BCmut_16.seq.count, big.mark = ",")
```

**Unique Perfects:** How many unique perfects did we recover from all sampling conditions for each library after filtering by minimum sequence count?
```{r class.output="goodCode"}
# Library 15
perf_filtered_uniq_15 <- BCmut_15 %>%
  filter(mutations==0) %>%
  select(mutID) %>%
  distinct() %>%
  nrow()
format(perf_filtered_uniq_15, big.mark = ",")

# Library 16
perf_filtered_uniq_16 <- BCmut_16 %>%
  filter(mutations==0) %>%
  select(mutID) %>%
  distinct() %>%
  nrow()
format(perf_filtered_uniq_16, big.mark = ",")
```

Count the number of shared perfects (mutations == 0) recovered between libraries after filtering by minimum sequence count.
```{r class.output="badCode", eval=FALSE}
# Merge the mutIDinfo datasets but keep only the mutIDs shared between libraries
BCs_mutID_filtered_perfects.shared <- merge(BCmut_15, BCmut_16, by = "mutID", all = FALSE)

# Retain only the rows with a specific value in the "specific_column" column
filtered_data.shared <- BCs_mutID_filtered_perfects.shared %>%
  filter(mutations.x == 0)

# Count the number of perfects (mutations == 0) shared between both libraries
BCs_mutID_filtered_perfects.shared.count <- length(unique(filtered_data.shared$mutID))
format(BCs_mutID_filtered_perfects.shared.count, big.mark = ",")
```

Also count the number of perfect sequences unique to one or the other library (mutations == 0) after filtering by minimum sequence count.
```{r class.output="badCode", eval=FALSE}
# Merge the mutIDinfo datasets but retain only the mutIDs unique to one library or the other
BCs_mutID_filtered_perfects.unique <- bind_rows(
  anti_join(BCmut_15, BCmut_16, by = "mutID"),
  anti_join(BCmut_16, BCmut_15, by = "mutID"))

# Retain only the rows with a specific value in the "specific_column" column
filtered_data.unique <-BCs_mutID_filtered_perfects.unique %>%
  filter(mutations == 0)

# Count the number of perfects (mutations == 0) unique to one library or the other
BCs_mutID_filtered_perfects.unique.count <- length(unique(filtered_data.unique$mutID))
format(BCs_mutID_filtered_perfects.unique.count, big.mark = ",")
```

Sum the shared and unique counts for total perfect sequences recovered after filtering by minimum sequence count.
```{r class.output="badCode", eval=FALSE}
# Sum the number of shared and unique perfects recovered for both codon version libraries
BCs_mutID_filtered_perfects.15.16.all.count <- sum(BCs_mutID_filtered_perfects.shared.count + BCs_mutID_filtered_perfects.unique.count)
format(BCs_mutID_filtered_perfects.15.16.all.count, big.mark = ",")
```

**Unique Mutants:** How many unique mutants did we recover from all sampling conditions for unique libraries after filtering by minimum sequence count?
```{r class.output="badCode", eval=FALSE}
# Library 15
mutant_filtered_15 <- BCmut_15 %>%
  filter(mutations!=0) %>%
  select(mutID) %>%
  distinct() %>%
  nrow()
format(mutant_filtered_15, big.mark = ",")

# Library 16
mutant_filtered_16 <- BCmut_16 %>%
  filter(mutations!=0) %>%
  select(mutID) %>%
  distinct() %>%
  nrow()
format(mutant_filtered_16, big.mark = ",")
```

## Median Fitness

In this section, we calculate (1) median fitness values across time points for each unique BC, (2) median fitness values for each unique mutID based on median BC fitness values across time points, and (3) median fitness values for only those sequences represented in both libraries.

### Median BCs Fitness
Calculate median fitness values for each unique BC across time points.

**Library 15**
```{r}
# Lib15

# Calculate the median fitness values across time points for each unique BC:
BCmut_15 <- BCmut_15 %>%
  mutate(L15.0.0.TMP.fc.median = rowMedians(as.matrix(select(.,D05D03fc,E07D03fc,F09D03fc,G11D03fc)), na.rm = TRUE),
         L15.0.058.TMP.fc.median = rowMedians(as.matrix(select(.,D06D03fc,E08D03fc,F10D03fc,G12D03fc)), na.rm = TRUE),
         L15.0.5.TMP.fc.median = rowMedians(as.matrix(select(.,D07D03fc,E09D03fc,F11D03fc,H01D03fc)), na.rm = TRUE),
         L15.1.0.TMP.fc.median = rowMedians(as.matrix(select(.,D08D03fc,E10D03fc,F12D03fc,H02D03fc)), na.rm = TRUE),
         L15.10.TMP.fc.median = rowMedians(as.matrix(select(.,D09D03fc,E11D03fc,G01D03fc,H03D03fc)), na.rm = TRUE),
         L15.50.TMP.fc.median = rowMedians(as.matrix(select(.,D10D03fc,E12D03fc,G02D03fc,H04D03fc)), na.rm = TRUE),
         L15.200.TMP.fc.median = rowMedians(as.matrix(select(.,D11D03fc,F01D03fc,G03D03fc,H05D03fc)), na.rm = TRUE),
         )
```

**Library 16**
```{r}
# Lib16

# Calculate the median fitness values across time points for each unique BC:
BCmut_16 <- BCmut_16 %>%
  mutate(L16.0.0.TMP.fc.median = rowMedians(as.matrix(select(.,D12D04fc,F02D04fc,G04D04fc,H06D04fc)), na.rm = TRUE),
         L16.0.058.TMP.fc.median = rowMedians(as.matrix(select(.,E01D04fc,F03D04fc,G05D04fc,H07D04fc)), na.rm = TRUE),
         L16.0.5.TMP.fc.median = rowMedians(as.matrix(select(.,E02D04fc,F04D04fc,G06D04fc,H08D04fc)), na.rm = TRUE),
         L16.1.0.TMP.fc.median = rowMedians(as.matrix(select(.,E03D04fc,F05D04fc,G07D04fc,H09D04fc)), na.rm = TRUE),
         L16.10.TMP.fc.median = rowMedians(as.matrix(select(.,E04D04fc,F06D04fc,G08D04fc,H10D04fc)), na.rm = TRUE),
         L16.50.TMP.fc.median = rowMedians(as.matrix(select(.,E05D04fc,F07D04fc,G09D04fc,H11D04fc)), na.rm = TRUE),
         L16.200.TMP.fc.median = rowMedians(as.matrix(select(.,E06D04fc,F08D04fc,G10D04fc,H12D04fc)), na.rm = TRUE),
         )
```

### Median mutID Fitness
Calculate median counts within each sample condition based on shared mutIDs

**Library 15**
```{r}
# Lib15

# Calculate the median fitness values for each sample column based on shared mutIDs:
BCmut_15.mutID.median <- BCmut_15 %>%
  group_by(mutID) %>%
  summarize(
    L15.0.0.TMP.fc.median = median(L15.0.0.TMP.fc.median, na.rm = TRUE),
    L15.0.058.TMP.fc.median = median(L15.0.058.TMP.fc.median, na.rm = TRUE),
    L15.0.5.TMP.fc.median = median(L15.0.5.TMP.fc.median, na.rm = TRUE),
    L15.1.0.TMP.fc.median = median(L15.1.0.TMP.fc.median, na.rm = TRUE),
    L15.10.TMP.fc.median = median(L15.10.TMP.fc.median, na.rm = TRUE),
    L15.50.TMP.fc.median = median(L15.50.TMP.fc.median, na.rm = TRUE),
    L15.200.TMP.fc.median = median(L15.200.TMP.fc.median, na.rm = TRUE),
  )
```

```{r}
# Lib15

# Calculate the median fitness values for each sample column based on shared mutIDs:
BCmut_15.median <- BCmut_15 %>%
  group_by(mutID) %>%
  summarize(
    L15.D05D03fc.median = median(D05D03fc, na.rm = TRUE),
    L15.D06D03fc.median = median(D06D03fc, na.rm = TRUE),
    L15.D07D03fc.median = median(D07D03fc, na.rm = TRUE),
    L15.D08D03fc.median = median(D08D03fc, na.rm = TRUE),
    L15.D09D03fc.median = median(D09D03fc, na.rm = TRUE),
    L15.D10D03fc.median = median(D10D03fc, na.rm = TRUE),
    L15.D11D03fc.median = median(D11D03fc, na.rm = TRUE),
    L15.E07D03fc.median = median(E07D03fc, na.rm = TRUE),
    L15.E08D03fc.median = median(E08D03fc, na.rm = TRUE),
    L15.E09D03fc.median = median(E09D03fc, na.rm = TRUE),
    L15.E10D03fc.median = median(E10D03fc, na.rm = TRUE),
    L15.E11D03fc.median = median(E11D03fc, na.rm = TRUE),
    L15.E12D03fc.median = median(E12D03fc, na.rm = TRUE),
    L15.F01D03fc.median = median(F01D03fc, na.rm = TRUE),
    L15.F09D03fc.median = median(F09D03fc, na.rm = TRUE),
    L15.F10D03fc.median = median(F10D03fc, na.rm = TRUE),
    L15.F11D03fc.median = median(F11D03fc, na.rm = TRUE),
    L15.F12D03fc.median = median(F12D03fc, na.rm = TRUE),
    L15.G01D03fc.median = median(G01D03fc, na.rm = TRUE),
    L15.G02D03fc.median = median(G02D03fc, na.rm = TRUE),
    L15.G03D03fc.median = median(G03D03fc, na.rm = TRUE),
    L15.G11D03fc.median = median(G11D03fc, na.rm = TRUE),
    L15.G12D03fc.median = median(G12D03fc, na.rm = TRUE),
    L15.H01D03fc.median = median(H01D03fc, na.rm = TRUE),
    L15.H02D03fc.median = median(H02D03fc, na.rm = TRUE),
    L15.H03D03fc.median = median(H03D03fc, na.rm = TRUE),
    L15.H04D03fc.median = median(H04D03fc, na.rm = TRUE),
    L15.H05D03fc.median = median(H05D03fc, na.rm = TRUE),
  )
```

Subset the descriptive columns from BCmut_ objects and retain only the unique "mutID" values
```{r}
# Subset descriptive columns from BCmut_15 dataset
BCmut_15.subset <- BCmut_15[, 2:10]

# Use dplyr to keep only rows with unique values in the specified column
BCmut_15.subset <- BCmut_15.subset %>% 
  distinct(mutID, .keep_all = TRUE)
```

Add descriptive columns back to the median datasets
```{r}
# Merge BCmut_15.median and BCmut_15.subset based on shared values in "mutID" column
BCmut_15.median.all <- left_join(BCmut_15.median, BCmut_15.subset, by = "mutID")
BCmut_15.mutID.median.all <- left_join(BCmut_15.mutID.median, BCmut_15.subset, by = "mutID")

# Reorder columns to place columns from BCmut_15.subset at the start
BCmut_15.median.all <- BCmut_15.median.all[, c(names(BCmut_15.subset), setdiff(names(BCmut_15.median.all), names(BCmut_15.subset)))]

BCmut_15.mutID.median.all <- BCmut_15.mutID.median.all[, c(names(BCmut_15.subset), setdiff(names(BCmut_15.mutID.median.all), names(BCmut_15.subset)))]
```

Create a separate dataset with unique protein sequences
```{r}
# Add "seq" column to BCmut_15.median dataset based on shared values in "mutID" column
BCmut_15.median.seq <- merge(BCmut_15.median, BCmut_15.subset[, c("mutID", "seq")], by = "mutID", all.x = TRUE)
BCmut_15.mutID.median.seq <- merge(BCmut_15.mutID.median, BCmut_15.subset[, c("mutID", "seq")], by = "mutID", all.x = TRUE)

# Reorder columns to place "seq" column after "mutID" column in newly merged dataset
BCmut_15.median.seq <- BCmut_15.median.seq[, c("mutID", "seq", setdiff(names(BCmut_15.median.seq), c("mutID", "seq")))]
BCmut_15.mutID.median.seq <- BCmut_15.mutID.median.seq[, c("mutID", "seq", setdiff(names(BCmut_15.mutID.median.seq), c("mutID", "seq")))]

# Remove first column "mutID"
BCmut_15.median.seq <- BCmut_15.median.seq[, -1]
BCmut_15.mutID.median.seq <- BCmut_15.mutID.median.seq[, -1]
```

Create a separate dataset filtered to only include perfects (mutation==0)
```{r class.output="goodCode"}
BCmut_15.median.perfects <- BCmut_15.median.all %>% filter(mutations == 0)
BCmut_15.mutID.median.perfects <- BCmut_15.mutID.median.all %>% filter(mutations == 0)

# Count perfects
BCmut_15.median.perfects.count <- length(unique(BCmut_15.median.perfects$mutID))
format(BCmut_15.median.perfects.count, big.mark = ",")
```

Subset the median.perfects dataset to only retain specific descriptive columns
```{r}
# Subset to retain "mutID" column
BCmut_15.median.mutID.perfects <- BCmut_15.median.perfects[, c(1, 10:37)]
BCmut_15.mutID.median.mutID.perfects <- BCmut_15.mutID.median.perfects[, c(1, 10:16)]

# Subset to retain "seq" column
BCmut_15.median.seq.perfects <- BCmut_15.median.perfects[, c(6, 10:37)]
BCmut_15.mutID.median.seq.perfects <- BCmut_15.mutID.median.perfects[, c(6, 10:16)]
```

**Library 16**
```{r}
# Lib16

# Calculate the median fitness values for each sample column based on shared mutIDs:
BCmut_16.mutID.median <- BCmut_16 %>%
  group_by(mutID) %>%
  summarize(
    L16.0.0.TMP.fc.median = median(L16.0.0.TMP.fc.median, na.rm = TRUE),
    L16.0.058.TMP.fc.median = median(L16.0.058.TMP.fc.median, na.rm = TRUE),
    L16.0.5.TMP.fc.median = median(L16.0.5.TMP.fc.median, na.rm = TRUE),
    L16.1.0.TMP.fc.median = median(L16.1.0.TMP.fc.median, na.rm = TRUE),
    L16.10.TMP.fc.median = median(L16.10.TMP.fc.median, na.rm = TRUE),
    L16.50.TMP.fc.median = median(L16.50.TMP.fc.median, na.rm = TRUE),
    L16.200.TMP.fc.median = median(L16.200.TMP.fc.median, na.rm = TRUE),
  )
```

```{r}
# Lib16

# Calculate the median fitness values for each sample column based on shared mutIDs:
BCmut_16.median <- BCmut_16 %>%
  group_by(mutID) %>%
  summarize(
    L16.D12D04fc.median = median(D12D04fc, na.rm = TRUE),
    L16.E01D04fc.median = median(E01D04fc, na.rm = TRUE),
    L16.E02D04fc.median = median(E02D04fc, na.rm = TRUE),
    L16.E03D04fc.median = median(E03D04fc, na.rm = TRUE),
    L16.E04D04fc.median = median(E04D04fc, na.rm = TRUE),
    L16.E05D04fc.median = median(E05D04fc, na.rm = TRUE),
    L16.E06D04fc.median = median(E06D04fc, na.rm = TRUE),
    L16.F02D04fc.median = median(F02D04fc, na.rm = TRUE),
    L16.F03D04fc.median = median(F03D04fc, na.rm = TRUE),
    L16.F04D04fc.median = median(F04D04fc, na.rm = TRUE),
    L16.F05D04fc.median = median(F05D04fc, na.rm = TRUE),
    L16.F06D04fc.median = median(F06D04fc, na.rm = TRUE),
    L16.F07D04fc.median = median(F07D04fc, na.rm = TRUE),
    L16.F08D04fc.median = median(F08D04fc, na.rm = TRUE),
    L16.G04D04fc.median = median(G04D04fc, na.rm = TRUE),
    L16.G05D04fc.median = median(G05D04fc, na.rm = TRUE),
    L16.G06D04fc.median = median(G06D04fc, na.rm = TRUE),
    L16.G07D04fc.median = median(G07D04fc, na.rm = TRUE),
    L16.G08D04fc.median = median(G08D04fc, na.rm = TRUE),
    L16.G09D04fc.median = median(G09D04fc, na.rm = TRUE),
    L16.G10D04fc.median = median(G10D04fc, na.rm = TRUE),
    L16.H06D04fc.median = median(H06D04fc, na.rm = TRUE),
    L16.H07D04fc.median = median(H07D04fc, na.rm = TRUE),
    L16.H08D04fc.median = median(H08D04fc, na.rm = TRUE),
    L16.H09D04fc.median = median(H09D04fc, na.rm = TRUE),
    L16.H10D04fc.median = median(H10D04fc, na.rm = TRUE),
    L16.H11D04fc.median = median(H11D04fc, na.rm = TRUE),
    L16.H12D04fc.median = median(H12D04fc, na.rm = TRUE),
  )
```

Subset the descriptive columns from BCmut_ objects and retain only the unique "mutID" values
```{r}
# Subset descriptive columns from BCmut_16 dataset
BCmut_16.subset <- BCmut_16[, 2:10]

# Use dplyr to keep only rows with unique values in the specified column
BCmut_16.subset <- BCmut_16.subset %>% 
  distinct(mutID, .keep_all = TRUE)
```

Add descriptive columns back to the median datasets
```{r}
# Merge BCmut_16.median and BCmut_16.subset based on shared values in "mutID" column
BCmut_16.median.all <- left_join(BCmut_16.median, BCmut_16.subset, by = "mutID")
BCmut_16.mutID.median.all <- left_join(BCmut_16.mutID.median, BCmut_16.subset, by = "mutID")

# Reorder columns to place columns from BCmut_16.subset at the start
BCmut_16.median.all <- BCmut_16.median.all[, c(names(BCmut_16.subset), setdiff(names(BCmut_16.median.all), names(BCmut_16.subset)))]

BCmut_16.mutID.median.all <- BCmut_16.mutID.median.all[, c(names(BCmut_16.subset), setdiff(names(BCmut_16.mutID.median.all), names(BCmut_16.subset)))]
```

Create a separate dataset with unique protein sequences
```{r}
# Add "seq" column to BCmut_16.median dataset based on shared values in "mutID" column
BCmut_16.median.seq <- merge(BCmut_16.median, BCmut_16.subset[, c("mutID", "seq")], by = "mutID", all.x = TRUE)
BCmut_16.mutID.median.seq <- merge(BCmut_16.mutID.median, BCmut_16.subset[, c("mutID", "seq")], by = "mutID", all.x = TRUE)

# Reorder columns to place "seq" column after "mutID" column in newly merged dataset
BCmut_16.median.seq <- BCmut_16.median.seq[, c("mutID", "seq", setdiff(names(BCmut_16.median.seq), c("mutID", "seq")))]
BCmut_16.mutID.median.seq <- BCmut_16.mutID.median.seq[, c("mutID", "seq", setdiff(names(BCmut_16.mutID.median.seq), c("mutID", "seq")))]

# Remove first column "mutID"
BCmut_16.median.seq <- BCmut_16.median.seq[, -1]
BCmut_16.mutID.median.seq <- BCmut_16.mutID.median.seq[, -1]
```

Create a separate dataset filtered to only include perfects (mutation==0)
```{r class.output="goodCode"}
BCmut_16.median.perfects <- BCmut_16.median.all %>% filter(mutations == 0)
BCmut_16.mutID.median.perfects <- BCmut_16.mutID.median.all %>% filter(mutations == 0)

# Count perfects
BCmut_16.median.perfects.count <- length(unique(BCmut_16.median.perfects$mutID))
format(BCmut_16.median.perfects.count, big.mark = ",")
```

Subset the median.perfects dataset to only retain specific descriptive columns
```{r}
# Subset to retain "mutID" column
BCmut_16.median.mutID.perfects <- BCmut_16.median.perfects[, c(1, 10:37)]
BCmut_16.mutID.median.mutID.perfects <- BCmut_16.mutID.median.perfects[, c(1, 10:16)]

# Subset to retain "seq" column
BCmut_16.median.seq.perfects <- BCmut_16.median.perfects[, c(6, 10:37)]
BCmut_16.mutID.median.seq.perfects <- BCmut_16.mutID.median.perfects[, c(6, 10:16)]
```

### Mean numBCs

**All Shared (mutIDs):** Generate average numBCs for mutIDs shared between libraries
```{r}
# Merge BCmut_15.subset and BCmut_16.subset based on shared values in "mutID" column
numBCs_mutID_mean_shared <- merge(BCmut_15.subset, BCmut_16.subset, by = "mutID")

# Select specific columns from the merged data frame
numBCs_mutID_mean_shared <- numBCs_mutID_mean_shared[, c("mutID", "numBCs.x", "numBCs.y")]

# Calculate the average (mean) "numBCs" between the codon versions for each "mutID"
numBCs_mutID_mean_shared$numBCs_mean <- rowMeans(numBCs_mutID_mean_shared[, c("numBCs.x", "numBCs.y")])
```

**All Shared (seqs):** Generate average numBCs for seqs shared between libraries
```{r}
# Merge BCmut_15.subset and BCmut_16.subset based on shared values in "seq" column
numBCs_seq_mean_shared <- merge(BCmut_15.subset, BCmut_16.subset, by = "seq")

# Select specific columns from the merged data frame
numBCs_seq_mean_shared <- numBCs_seq_mean_shared[, c("seq", "numBCs.x", "numBCs.y")]

# Calculate the average (mean) "numBCs" between the codon versions for each "seq"
numBCs_seq_mean_shared$numBCs_mean <- rowMeans(numBCs_seq_mean_shared[, c("numBCs.x", "numBCs.y")])
```

**Perfects (mutIDs):** Generate average numBCs for perfects mutIDs shared between libraries
```{r}
# Merge BCmut_15.median.perfects and BCmut_16.median.perfects based on shared values in "mutID" column
numBCs_perfects_mutID_mean_shared <- merge(BCmut_15.median.perfects, BCmut_15.median.perfects, by = "mutID")

# Select specific columns from the merged data frame
numBCs_perfects_mutID_mean_shared <- numBCs_perfects_mutID_mean_shared[, c("mutID", "numBCs.x", "numBCs.y")]

# Calculate the average (mean) "numBCs" between the codon versions for each "mutID"
numBCs_perfects_mutID_mean_shared$numBCs_mean <- rowMeans(numBCs_perfects_mutID_mean_shared[, c("numBCs.x", "numBCs.y")])
```

**Perfects (seqs):** Generate average numBCs for perfects seqs shared between libraries
```{r}
# Merge BCmut_15.subset and BCmut_16.subset based on shared values in "seq" column
numBCs_perfects_seq_mean_shared <- merge(BCmut_15.median.perfects, BCmut_16.median.perfects, by = "seq")

# Select specific columns from the merged data frame
numBCs_perfects_seq_mean_shared <- numBCs_perfects_seq_mean_shared[, c("seq", "numBCs.x", "numBCs.y")]

# Calculate the average (mean) "numBCs" between the codon versions for each "seq"
numBCs_perfects_seq_mean_shared$numBCs_mean <- rowMeans(numBCs_perfects_seq_mean_shared[, c("numBCs.x", "numBCs.y")])
```

## Merged Libraries

### Merged All Shared

Merge the median fitness scores derived from each library into a single dataframe based on shared mutID or protein sequences between the two dataset
```{r}
# Merge by shared "mutID"
BCmut_15.16.mutID.fitness <- merge(BCmut_15.median, BCmut_16.median, by = "mutID", all = FALSE)

# Add "numBCs" column to merged dataset based on shared "mutID"
BCmut_15.16.mutID.fitness <- merge(BCmut_15.16.mutID.fitness, BCmut_15.median.all[, c("mutID", "numBCs")], by = "mutID", all.x = TRUE)

# Merged by shared "seq"
BCmut_15.16.seq.fitness <- merge(BCmut_15.median.seq, BCmut_16.median.seq, by = "seq", all = FALSE)

# Add "numBCs" column to merged dataset based on shared "seq"
BCmut_15.16.seq.fitness <- merge(BCmut_15.16.seq.fitness, BCmut_15.median.all[, c("seq", "numBCs")], by = "seq", all.x = TRUE)
```

### Merged Perfects Shared

Merge the median fitness scores for perfects derived from each library into a single dataframe based on shared mutID or protein sequences between the two dataset
```{r class.output="goodCode"}
# Merge by shared "mutID"
BCmut_15.16.perfect.mutID.fitness <- merge(BCmut_15.median.mutID.perfects, BCmut_16.median.mutID.perfects, by = "mutID", all = FALSE)

# Add "numBCs_mean" column to merged dataset based on shared "mutID"
BCmut_15.16.perfect.mutID.fitness <- merge(BCmut_15.16.perfect.mutID.fitness, numBCs_perfects_mutID_mean_shared[, c("mutID", "numBCs_mean")], by = "mutID", all.x = TRUE)

# Count the number of unique perfects shared between libraries
BCmut_15.16.perfect.mutID.fitness.count <- length(unique(BCmut_15.16.perfect.mutID.fitness$mutID))
format(BCmut_15.16.perfect.mutID.fitness.count, big.mark = ",")
```

```{r class.output="goodCode"}
# Create a new dataset retaining unique values from both datasets
unique_perfects <- bind_rows(
  anti_join(BCmut_15.median.mutID.perfects, BCmut_16.median.mutID.perfects, by = "mutID"),
  anti_join(BCmut_16.median.mutID.perfects, BCmut_15.median.mutID.perfects, by = "mutID"))

# Count the number of perfects unique to one library or the other
unique_perfects.count <- length(unique(unique_perfects$mutID))
format(unique_perfects.count, big.mark = ",")
```

**Summarize the number of perfects shared or unique between both libraries.**
```{r class.output="goodCode"}
BCmut_15.16.perfect.mutID.fitness.all.count <- sum(BCmut_15.16.perfect.mutID.fitness.count + unique_perfects.count)
format(BCmut_15.16.perfect.mutID.fitness.all.count, big.mark = ",")
```

```{r}
# Merge by shared "seq"
BCmut_15.16.perfect.seq.fitness <- merge(BCmut_15.median.seq.perfects, BCmut_16.median.seq.perfects, by = "seq", all = FALSE)

# Add "numBCs_mean" column to merged dataset based on shared "seq"
BCmut_15.16.perfect.seq.fitness <- merge(BCmut_15.16.perfect.seq.fitness, numBCs_perfects_seq_mean_shared[, c("seq", "numBCs_mean")], by = "seq", all.x = TRUE)
```

```{r}
# Merge by shared "seq"
BCmut_15.16.mutID.perfect.seq.fitness <- merge(BCmut_15.mutID.median.seq.perfects, BCmut_16.mutID.median.seq.perfects, by = "seq", all = FALSE)

# Add "numBCs_mean" column to merged dataset based on shared "seq"
BCmut_15.16.mutID.perfect.seq.fitness <- merge(BCmut_15.16.mutID.perfect.seq.fitness, numBCs_perfects_seq_mean_shared[, c("seq", "numBCs_mean")], by = "seq", all.x = TRUE)
```

#### High/Low Shared Perfects

```{r}
# Subset the dataset to retain only the fitness values for complementation and 50 ug/mL TMP at time point 1:
BCmut_15.16.perfect.mutID.fitness.subset <- BCmut_15.16.perfect.mutID.fitness[, c("mutID", "numBCs_mean", "L15.D05D03fc.median", "L16.D12D04fc.median", "L15.D10D03fc.median", "L16.E05D04fc.median")]
```

```{r}
# Calculate absolute difference
result <- BCmut_15.16.perfect.mutID.fitness.subset %>%
  mutate(Comp_Abs_Diff = abs(L15.D05D03fc.median - L16.D12D04fc.median)) %>%
  mutate(Tmp50_Abs_Diff = abs(L15.D10D03fc.median - L16.E05D04fc.median))
```

```{r}
# Retain the top 50 rows with the greatest Comp_Abs_Diff value and select certain columns
top_high_comp <- result %>% 
  top_n(50, Comp_Abs_Diff) %>%
  select(mutID, numBCs_mean, L15.D05D03fc.median, L16.D12D04fc.median, Comp_Abs_Diff)

# Retain the top 50 rows with the lowest Comp_Abs_Diff value and select certain columns
top_low_comp <- result %>% 
  top_n(-50, Comp_Abs_Diff) %>%
  select(mutID, numBCs_mean, L15.D05D03fc.median, L16.D12D04fc.median, Comp_Abs_Diff)
```

```{r}
# Retain the top 50 rows with the greatest Tmp50_Abs_Diff value and select certain columns
top_high_tmp50 <- result %>% 
  top_n(50, Tmp50_Abs_Diff) %>%
  select(mutID, numBCs_mean, L15.D10D03fc.median, L16.E05D04fc.median, Tmp50_Abs_Diff)

# Retain the top 50 rows with the lowest Tmp50_Abs_Diff value and select certain columns
top_low_tmp50 <- result %>% 
  top_n(-50, Tmp50_Abs_Diff) %>%
  select(mutID, numBCs_mean, L15.D10D03fc.median, L16.E05D04fc.median, Tmp50_Abs_Diff)
```

```{r eval=FALSE, echo=FALSE}
# Save top_high_comp dataframe as a .csv file
write.csv(top_high_comp, "OUTPUT/top_high_complementation.csv", row.names = FALSE)

# Save top_low_comp dataframe as a .csv file
write.csv(top_low_comp, "OUTPUT/top_low_complementation.csv", row.names = FALSE)

# Save top_high_Tmp50 dataframe as a .csv file
write.csv(top_high_tmp50, "OUTPUT/top_high_50tmp.csv", row.names = FALSE)

# Save top_low_Tmp50 dataframe as a .csv file
write.csv(top_low_tmp50, "OUTPUT/top_low_50tmp.csv", row.names = FALSE)
```

### Merged Lib15 v1 & v2

Import the .csv file for median fitness values of Library 15 v1 (from Carmen's thesis) and merge with the current Lib15 (v2) dataset based on shared "mutID" values.
```{r}
# Import Lib15-v1 (CR) perfects median fitness .csv file
BCmut_15.v1.perfect.mutID.fitness <- read.csv("CR_Lib15/BCmut.median.mutID.perfects.fitness.CRLib.csv", header = TRUE)

# Add "numBCs" to Lib15-v2 dataset
BCmut_15.v2.perfect.mutID.fitness <- merge(BCmut_15.median.mutID.perfects, BCmut_15.subset[, c("mutID", "numBCs")], by = "mutID", all.x = TRUE)

# Merge by shared "mutID" between Lib15 versions
BCmut_15.v1.v2.perfect.mutID.fitness <- merge(BCmut_15.v1.perfect.mutID.fitness, BCmut_15.v2.perfect.mutID.fitness, by = "mutID", all.x = FALSE)

# Calculate the average (mean) "numBCs" between the replicate library versions for each "mutID"
BCmut_15.v1.v2.perfect.mutID.fitness$numBCs_mean <- rowMeans(BCmut_15.v1.v2.perfect.mutID.fitness[, c("numBCs.v1", "numBCs")])
```

## Correlations

### Lib15 v1 vs. v2

Determine correlations:
```{r class.output="goodCode"}
# Determine median fitness correlations between Lib15 versions for T1: M9-Full Supp vs M9-No Supp conditions
cor.test(BCmut_15.v1.v2.perfect.mutID.fitness$A56fc.median,
         BCmut_15.v1.v2.perfect.mutID.fitness$L15.D05D03fc.median)$p.value
```

#### Fitness Corr

Plot median fitness correlations between Lib15 versions (perfects) for 1st time point M9-No Supp vs. M9-Full Supp
```{r}
# Plot based on shared mutID

Lib15_v1v2_M9_p1 <- ggplot(BCmut_15.v1.v2.perfect.mutID.fitness, 
                   aes(x=A56fc.median, y=L15.D05D03fc.median, color=log(numBCs_mean))) +
  labs(x = "Median fitness Lib15 v1", y ="Median fitness Lib15 v2",color="") +
  ggtitle("Perfects Median Fitness Correlation \nM9-Full Supp vs No Supp (0 mg tmp) \nLibrary 15 v1 vs v2") +
  geom_smooth(method=lm,colour="black") +
  geom_density2d(colour="black",alpha=0.2) +
  geom_point(alpha=0.7) +
  scale_colour_gradient2("",low="blue", high="red", mid="blue") + 
  theme(legend.position="left")

Lib15_v1v2_M9_p01 <- ggMarginal(Lib15_v1v2_M9_p1, type = "histogram", fill = "aquamarine4")

Lib15_v1v2_M9_p01
```

```{r echo=FALSE}
ggsave(file="plots/Lib15_v1_v2/Lib15.v1.v2.correlation.median.fitness.mutID.T1.M9.Full.vs.No.Supp.png",
       plot=Lib15_v1v2_M9_p01,
       dpi=600, width = 8, height = 6, units = "in")
```

Plot median fitness correlations between Lib15 versions (perfects) for 1st time point 0.058 TMP vs. M9-Full Supp
```{r}
# Plot based on shared mutID

Lib15_v1v2_0.058_tmp_p1 <- ggplot(BCmut_15.v1.v2.perfect.mutID.fitness, 
                   aes(x=A57fc.median, y=L15.D06D03fc.median, color=log(numBCs_mean))) +
  labs(x = "Median fitness Lib15 v1", y ="Median fitness Lib15 v2",color="") +
  ggtitle("Perfects Median Fitness Correlation \nM9-Full Supp vs 0.058 mg TMP \nLibrary 15 v1 vs v2") +
  geom_smooth(method=lm,colour="black") +
  geom_density2d(colour="black",alpha=0.2) +
  geom_point(alpha=0.7) +
  scale_colour_gradient2("",low="blue", high="red", mid="blue") + 
  theme(legend.position="left")

Lib15_v1v2_0.058_tmp_p01 <- ggMarginal(Lib15_v1v2_0.058_tmp_p1, type = "histogram", fill = "aquamarine4")

Lib15_v1v2_0.058_tmp_p01
```

```{r echo=FALSE}
ggsave(file="plots/Lib15_v1_v2/Lib15.v1.v2.correlation.median.fitness.mutID.T1.M9.Full.vs.0.058.tmp.png",
       plot=Lib15_v1v2_0.058_tmp_p01,
       dpi=600, width = 8, height = 6, units = "in")
```

Plot median fitness correlations between Lib15 versions (perfects) for 1st time point 0.5 TMP vs. M9-Full Supp
```{r}
# Plot based on shared mutID

Lib15_v1v2_0.5_tmp_p1 <- ggplot(BCmut_15.v1.v2.perfect.mutID.fitness, 
                   aes(x=A58fc.median, y=L15.D07D03fc.median, color=log(numBCs_mean))) +
  labs(x = "Median fitness Lib15 v1", y ="Median fitness Lib15 v2",color="") +
  ggtitle("Perfects Median Fitness Correlation \nM9-Full Supp vs 0.5 mg TMP \nLibrary 15 v1 vs v2") +
  geom_smooth(method=lm,colour="black") +
  geom_density2d(colour="black",alpha=0.2) +
  geom_point(alpha=0.7) +
  scale_colour_gradient2("",low="blue", high="red", mid="blue") + 
  theme(legend.position="left")

Lib15_v1v2_0.5_tmp_p01 <- ggMarginal(Lib15_v1v2_0.5_tmp_p1, type = "histogram", fill = "aquamarine4")

Lib15_v1v2_0.5_tmp_p01
```

```{r echo=FALSE}
ggsave(file="plots/Lib15_v1_v2/Lib15.v1.v2.correlation.median.fitness.mutID.T1.M9.Full.vs.0.5.tmp.png",
       plot=Lib15_v1v2_0.5_tmp_p01,
       dpi=600, width = 8, height = 6, units = "in")
```

Plot median fitness correlations between Lib15 versions (perfects) for 1st time point 1.0 TMP vs. M9-Full Supp
```{r}
# Plot based on shared mutID

Lib15_v1v2_1_tmp_p1 <- ggplot(BCmut_15.v1.v2.perfect.mutID.fitness, 
                   aes(x=A59fc.median, y=L15.D08D03fc.median, color=log(numBCs_mean))) +
  labs(x = "Median fitness Lib15 v1", y ="Median fitness Lib15 v2",color="") +
  ggtitle("Perfects Median Fitness Correlation \nM9-Full Supp vs 1.0 mg TMP \nLibrary 15 v1 vs v2") +
  geom_smooth(method=lm,colour="black") +
  geom_density2d(colour="black",alpha=0.2) +
  geom_point(alpha=0.7) +
  scale_colour_gradient2("",low="blue", high="red", mid="blue") + 
  theme(legend.position="left")

Lib15_v1v2_1_tmp_p01 <- ggMarginal(Lib15_v1v2_1_tmp_p1, type = "histogram", fill = "aquamarine4")

Lib15_v1v2_1_tmp_p01
```

```{r echo=FALSE}
ggsave(file="plots/Lib15_v1_v2/Lib15.v1.v2.correlation.median.fitness.mutID.T1.M9.Full.vs.1.tmp.png",
       plot=Lib15_v1v2_1_tmp_p01,
       dpi=600, width = 8, height = 6, units = "in")
```

Plot median fitness correlations between Lib15 versions (perfects) for 1st time point 10 TMP vs. M9-Full Supp
```{r}
# Plot based on shared mutID

Lib15_v1v2_10_tmp_p1 <- ggplot(BCmut_15.v1.v2.perfect.mutID.fitness, 
                   aes(x=A510fc.median, y=L15.D09D03fc.median, color=log(numBCs_mean))) +
  labs(x = "Median fitness Lib15 v1", y ="Median fitness Lib15 v2",color="") +
  ggtitle("Perfects Median Fitness Correlation \nM9-Full Supp vs 10 mg TMP \nLibrary 15 v1 vs v2") +
  geom_smooth(method=lm,colour="black") +
  geom_density2d(colour="black",alpha=0.2) +
  geom_point(alpha=0.7) +
  scale_colour_gradient2("",low="blue", high="red", mid="blue") + 
  theme(legend.position="left")

Lib15_v1v2_10_tmp_p01 <- ggMarginal(Lib15_v1v2_10_tmp_p1, type = "histogram", fill = "aquamarine4")

Lib15_v1v2_10_tmp_p01
```

```{r echo=FALSE}
ggsave(file="plots/Lib15_v1_v2/Lib15.v1.v2.correlation.median.fitness.mutID.T1.M9.Full.vs.10.tmp.png",
       plot=Lib15_v1v2_10_tmp_p01,
       dpi=600, width = 8, height = 6, units = "in")
```

Plot median fitness correlations between Lib15 versions (perfects) for 1st time point 50 TMP vs. M9-Full Supp
```{r}
# Plot based on shared mutID

Lib15_v1v2_50_tmp_p1 <- ggplot(BCmut_15.v1.v2.perfect.mutID.fitness, 
                   aes(x=A511fc.median, y=L15.D10D03fc.median, color=log(numBCs_mean))) +
  labs(x = "Median fitness Lib15 v1", y ="Median fitness Lib15 v2",color="") +
  ggtitle("Perfects Median Fitness Correlation \nM9-Full Supp vs 50 mg TMP \nLibrary 15 v1 vs v2") +
  geom_smooth(method=lm,colour="black") +
  geom_density2d(colour="black",alpha=0.2) +
  geom_point(alpha=0.7) +
  scale_colour_gradient2("",low="blue", high="red", mid="blue") + 
  theme(legend.position="left")

Lib15_v1v2_50_tmp_p01 <- ggMarginal(Lib15_v1v2_50_tmp_p1, type = "histogram", fill = "aquamarine4")

Lib15_v1v2_50_tmp_p01
```

```{r echo=FALSE}
ggsave(file="plots/Lib15_v1_v2/Lib15.v1.v2.correlation.median.fitness.mutID.T1.M9.Full.vs.50.tmp.png",
       plot=Lib15_v1v2_50_tmp_p01,
       dpi=600, width = 8, height = 6, units = "in")
```

Plot median fitness correlations between Lib15 versions (perfects) for 1st time point 200 TMP vs. M9-Full Supp
```{r}
# Plot based on shared mutID

Lib15_v1v2_200_tmp_p1 <- ggplot(BCmut_15.v1.v2.perfect.mutID.fitness, 
                   aes(x=A512fc.median, y=L15.D11D03fc.median, color=log(numBCs_mean))) +
  labs(x = "Median fitness Lib15 v1", y ="Median fitness Lib15 v2",color="") +
  ggtitle("Perfects Median Fitness Correlation \nM9-Full Supp vs 200 mg TMP \nLibrary 15 v1 vs v2") +
  geom_smooth(method=lm,colour="black") +
  geom_density2d(colour="black",alpha=0.2) +
  geom_point(alpha=0.7) +
  scale_colour_gradient2("",low="blue", high="red", mid="blue") + 
  theme(legend.position="left")

Lib15_v1v2_200_tmp_p01 <- ggMarginal(Lib15_v1v2_200_tmp_p1, type = "histogram", fill = "aquamarine4")

Lib15_v1v2_200_tmp_p01
```

```{r echo=FALSE}
ggsave(file="plots/Lib15_v1_v2/Lib15.v1.v2.correlation.median.fitness.mutID.T1.M9.Full.vs.200.tmp.png",
       plot=Lib15_v1v2_200_tmp_p01,
       dpi=600, width = 8, height = 6, units = "in")
```

#### Abs. Difference

Calculate the absolute difference in fitness values between Lib15 replicates for each time point.
```{r}
# Calculate absolute difference and add a new column
BCmut_15.v1.v2.perfect.mutID.fitness <- BCmut_15.v1.v2.perfect.mutID.fitness %>%
  mutate(tmp_0_abs_diff = abs(A56fc.median - L15.D05D03fc.median),
         tmp_0.058_abs_diff = abs(A57fc.median - L15.D06D03fc.median),
         tmp_0.5_abs_diff = abs(A58fc.median - L15.D07D03fc.median),
         tmp_1_abs_diff = abs(A59fc.median - L15.D08D03fc.median),
         tmp_10_abs_diff = abs(A510fc.median - L15.D09D03fc.median),
         tmp_50_abs_diff = abs(A511fc.median - L15.D10D03fc.median),
         tmp_200_abs_diff = abs(A512fc.median - L15.D11D03fc.median)
         )
```

Plot the absolute difference in median fitness between Lib15 versions (perfects) for M9-No Supp vs. M9-Full Supp
```{r}
# Create a scatter plot of absolute difference as a function of numBCs
Lib15.v1.v2.abs.diff.0.tmp.p1 <- ggplot(BCmut_15.v1.v2.perfect.mutID.fitness, 
                                  aes(x = numBCs_mean, y = tmp_0_abs_diff, color=log(numBCs_mean))) +
  geom_point(alpha=0.7) +
  scale_colour_gradient2("",low="blue", high="red", mid="blue") + 
  theme(legend.position="left") +
  labs(title = "Absolute Difference in Median Fitness \nBetween Lib15 Versions (v1 & v2) \nM9-Full Supp vs M9-No Supp",
       x = "numBCs",
       y = "Absolute Difference") +
  theme_minimal()

Lib15.v1.v2.abs.diff.0.tmp.p1
```

```{r echo=FALSE}
ggsave(file="plots/Lib15_v1_v2/Lib15.v1.v2.abs.diff.median.fitness.mutID.T1.M9.Full.vs.M9.No.png",
       plot=Lib15.v1.v2.abs.diff.0.tmp.p1,
       dpi=600, width = 8, height = 6, units = "in")
```

```{r}
Lib15.v1.v2.abs.diff.0.tmp.p3 <- ggplot(BCmut_15.v1.v2.perfect.mutID.fitness %>% 
                                          filter(numBCs_mean >= 0 & numBCs_mean <= 100 & numBCs_mean %% 2 == 0), 
                                        aes(x = factor(numBCs_mean), y = tmp_0_abs_diff, color = numBCs_mean)) +
  geom_boxplot(alpha = 0.7, notch = FALSE) +
  scale_colour_gradient2("", low = "blue", high = "red", mid = "blue") +
  scale_x_discrete(breaks = unique(factor(BCmut_15.v1.v2.perfect.mutID.fitness$numBCs_mean[BCmut_15.v1.v2.perfect.mutID.fitness$numBCs_mean %% 2 == 0]))) +
  theme(legend.position = "left") +
  labs(title = "Absolute Difference in Median Fitness \nBetween Lib15 Versions (v1 & v2) \nM9-Full Supp vs M9-No Supp",
       x = "numBCs",
       y = "Absolute Difference") +
  theme_minimal()

Lib15.v1.v2.abs.diff.0.tmp.p3
```

```{r echo=FALSE}
ggsave(file="plots/Lib15_v1_v2/Lib15.v1.v2.abs.diff.boxplot.median.fitness.mutID.T1.M9.Full.vs.M9.No.v2.png",
       plot=Lib15.v1.v2.abs.diff.0.tmp.p3,
       dpi=600, width = 12, height = 6, units = "in")
```

### Lib15 vs. Lib16
Determine correlations using median fitness values from Libraries 15 and 16:
```{r class.output="goodCode"}
# Determine median fitness correlations between Lib15 and Lib16 for Complementation (0.0-TMP)
cor.test(BCmut_15.16.mutID.perfect.seq.fitness$L15.0.0.TMP.fc.median,
         BCmut_15.16.mutID.perfect.seq.fitness$L16.0.0.TMP.fc.median)$p.value
```

#### Fit Corr Plots

Plot median fitness correlations between Lib15 & Lib16 (perfects) for Complementation:
```{r}
# Plot based on shared mutID

Lib15_16_0_TMP_p1 <- ggplot(BCmut_15.16.mutID.perfect.seq.fitness, 
                   aes(x=L15.0.0.TMP.fc.median, y=L16.0.0.TMP.fc.median, color=log(numBCs_mean))) +
  labs(x = "Median fitness Lib15", y ="Median fitness Lib16",color="") +
  ggtitle("Perfects Median Fitness Correlation \nComplementation (0 mg tmp) \nLibrary 15 vs Library 16") +
  geom_smooth(method=lm,colour="black") +
  geom_density2d(colour="black",alpha=0.2) +
  geom_point(alpha=0.7) +
  scale_colour_gradient2("",low="blue", high="red", mid="blue") + 
  theme(legend.position="left")

Lib15_16_0_TMP_p01 <- ggMarginal(Lib15_16_0_TMP_p1, type = "histogram", fill = "aquamarine4")

Lib15_16_0_TMP_p01
```

```{r echo=FALSE}
ggsave(file="plots/correlation/median_time/Lib15.16.correlation.median.fitness.seq.complementation.0.tmp.png",
       plot=Lib15_16_0_TMP_p01,
       dpi=600, width = 8, height = 6, units = "in")
```

Plot median fitness correlations between Lib15 & Lib16 (perfects) for 0.058 TMP:
```{r}
# Plot based on shared mutID

Lib15_16_0.058_TMP_p1 <- ggplot(BCmut_15.16.mutID.perfect.seq.fitness, 
                   aes(x=L15.0.058.TMP.fc.median, y=L16.0.058.TMP.fc.median, color=log(numBCs_mean))) +
  labs(x = "Median fitness Lib15", y ="Median fitness Lib16",color="") +
  ggtitle("Perfects Median Fitness Correlation \nTrimethoprim (0.058 mg tmp) \nLibrary 15 vs Library 16") +
  geom_smooth(method=lm,colour="black") +
  geom_density2d(colour="black",alpha=0.2) +
  geom_point(alpha=0.7) +
  scale_colour_gradient2("",low="blue", high="red", mid="blue") + 
  theme(legend.position="left")

Lib15_16_0.058_TMP_p01 <- ggMarginal(Lib15_16_0.058_TMP_p1, type = "histogram", fill = "aquamarine4")

Lib15_16_0.058_TMP_p01
```

```{r echo=FALSE}
ggsave(file="plots/correlation/median_time/Lib15.16.correlation.median.fitness.seq.trimethoprim.0.058.tmp.png",
       plot=Lib15_16_0.058_TMP_p01,
       dpi=600, width = 8, height = 6, units = "in")
```

Plot median fitness correlations between Lib15 & Lib16 (perfects) for 0.5 TMP:
```{r}
# Plot based on shared mutID

Lib15_16_0.5_TMP_p1 <- ggplot(BCmut_15.16.mutID.perfect.seq.fitness, 
                   aes(x=L15.0.5.TMP.fc.median, y=L16.0.5.TMP.fc.median, color=log(numBCs_mean))) +
  labs(x = "Median fitness Lib15", y ="Median fitness Lib16",color="") +
  ggtitle("Perfects Median Fitness Correlation \nTrimethoprim (0.5 mg tmp) \nLibrary 15 vs Library 16") +
  geom_smooth(method=lm,colour="black") +
  geom_density2d(colour="black",alpha=0.2) +
  geom_point(alpha=0.7) +
  scale_colour_gradient2("",low="blue", high="red", mid="blue") + 
  theme(legend.position="left")

Lib15_16_0.5_TMP_p01 <- ggMarginal(Lib15_16_0.5_TMP_p1, type = "histogram", fill = "aquamarine4")

Lib15_16_0.5_TMP_p01
```

```{r echo=FALSE}
ggsave(file="plots/correlation/median_time/Lib15.16.correlation.median.fitness.seq.trimethoprim.0.5.tmp.png",
       plot=Lib15_16_0.5_TMP_p01,
       dpi=600, width = 8, height = 6, units = "in")
```

Plot median fitness correlations between Lib15 & Lib16 (perfects) for 1.0 TMP:
```{r}
# Plot based on shared mutID

Lib15_16_1.0_TMP_p1 <- ggplot(BCmut_15.16.mutID.perfect.seq.fitness, 
                   aes(x=L15.1.0.TMP.fc.median, y=L16.1.0.TMP.fc.median, color=log(numBCs_mean))) +
  labs(x = "Median fitness Lib15", y ="Median fitness Lib16",color="") +
  ggtitle("Perfects Median Fitness Correlation \nTrimethoprim (1.0 mg tmp) \nLibrary 15 vs Library 16") +
  geom_smooth(method=lm,colour="black") +
  geom_density2d(colour="black",alpha=0.2) +
  geom_point(alpha=0.7) +
  scale_colour_gradient2("",low="blue", high="red", mid="blue") + 
  theme(legend.position="left")

Lib15_16_1.0_TMP_p01 <- ggMarginal(Lib15_16_1.0_TMP_p1, type = "histogram", fill = "aquamarine4")

Lib15_16_1.0_TMP_p01
```

```{r echo=FALSE}
ggsave(file="plots/correlation/median_time/Lib15.16.correlation.median.fitness.seq.trimethoprim.1.0.tmp.png",
       plot=Lib15_16_1.0_TMP_p01,
       dpi=600, width = 8, height = 6, units = "in")
```

Plot median fitness correlations between Lib15 & Lib16 (perfects) for 10 TMP:
```{r}
# Plot based on shared mutID

Lib15_16_10_TMP_p1 <- ggplot(BCmut_15.16.mutID.perfect.seq.fitness, 
                   aes(x=L15.10.TMP.fc.median, y=L16.10.TMP.fc.median, color=log(numBCs_mean))) +
  labs(x = "Median fitness Lib15", y ="Median fitness Lib16",color="") +
  ggtitle("Perfects Median Fitness Correlation \nTrimethoprim (10 mg tmp) \nLibrary 15 vs Library 16") +
  geom_smooth(method=lm,colour="black") +
  geom_density2d(colour="black",alpha=0.2) +
  geom_point(alpha=0.7) +
  scale_colour_gradient2("",low="blue", high="red", mid="blue") + 
  theme(legend.position="left")

Lib15_16_10_TMP_p01 <- ggMarginal(Lib15_16_10_TMP_p1, type = "histogram", fill = "aquamarine4")

Lib15_16_10_TMP_p01
```

```{r echo=FALSE}
ggsave(file="plots/correlation/median_time/Lib15.16.correlation.median.fitness.seq.trimethoprim.10.tmp.png",
       plot=Lib15_16_10_TMP_p01,
       dpi=600, width = 8, height = 6, units = "in")
```

Plot median fitness correlations between Lib15 & Lib16 (perfects) for 50 TMP:
```{r}
# Plot based on shared mutID

Lib15_16_50_TMP_p1 <- ggplot(BCmut_15.16.mutID.perfect.seq.fitness, 
                   aes(x=L15.50.TMP.fc.median, y=L16.50.TMP.fc.median, color=log(numBCs_mean))) +
  labs(x = "Median fitness Lib15", y ="Median fitness Lib16",color="") +
  ggtitle("Perfects Median Fitness Correlation \nTrimethoprim (50 mg tmp) \nLibrary 15 vs Library 16") +
  geom_smooth(method=lm,colour="black") +
  geom_density2d(colour="black",alpha=0.2) +
  geom_point(alpha=0.7) +
  scale_colour_gradient2("",low="blue", high="red", mid="blue") + 
  theme(legend.position="left")

Lib15_16_50_TMP_p01 <- ggMarginal(Lib15_16_50_TMP_p1, type = "histogram", fill = "aquamarine4")

Lib15_16_50_TMP_p01
```

```{r echo=FALSE}
ggsave(file="plots/correlation/median_time/Lib15.16.correlation.median.fitness.seq.trimethoprim.50.tmp.png",
       plot=Lib15_16_50_TMP_p01,
       dpi=600, width = 8, height = 6, units = "in")
```

Plot median fitness correlations between Lib15 & Lib16 (perfects) for 200 TMP:
```{r}
# Plot based on shared mutID

Lib15_16_200_TMP_p1 <- ggplot(BCmut_15.16.mutID.perfect.seq.fitness, 
                   aes(x=L15.200.TMP.fc.median, y=L16.200.TMP.fc.median, color=log(numBCs_mean))) +
  labs(x = "Median fitness Lib15", y ="Median fitness Lib16",color="") +
  ggtitle("Perfects Median Fitness Correlation \nTrimethoprim (200 mg tmp) \nLibrary 15 vs Library 16") +
  geom_smooth(method=lm,colour="black") +
  geom_density2d(colour="black",alpha=0.2) +
  geom_point(alpha=0.7) +
  scale_colour_gradient2("",low="blue", high="red", mid="blue") + 
  theme(legend.position="left")

Lib15_16_200_TMP_p01 <- ggMarginal(Lib15_16_200_TMP_p1, type = "histogram", fill = "aquamarine4")

Lib15_16_200_TMP_p01
```

```{r echo=FALSE}
ggsave(file="plots/correlation/median_time/Lib15.16.correlation.median.fitness.seq.trimethoprim.200.tmp.png",
       plot=Lib15_16_200_TMP_p01,
       dpi=600, width = 8, height = 6, units = "in")
```

## Mapped Reads & BCs

Calculate the total number of reads and unique barcodes recovered from each sampling condition.
```{r}
# Lib15

# Sum the total reads recovered from each sampling condition:
norm_vals_15$mappedcounts <- c(sum(BCmut_15$D01),sum(BCmut_15$D03),
                            sum(BCmut_15$D05),sum(BCmut_15$D06),
                            sum(BCmut_15$D07),sum(BCmut_15$D08),
                            sum(BCmut_15$D09),sum(BCmut_15$D10),
                            sum(BCmut_15$D11),sum(BCmut_15$E07),
                            sum(BCmut_15$E08),sum(BCmut_15$E09),
                            sum(BCmut_15$E10),sum(BCmut_15$E11),
                            sum(BCmut_15$E12),sum(BCmut_15$F01),
                            sum(BCmut_15$F09),sum(BCmut_15$F10),
                            sum(BCmut_15$F11),sum(BCmut_15$F12),
                            sum(BCmut_15$G01),sum(BCmut_15$G02),
                            sum(BCmut_15$G03),sum(BCmut_15$G11),
                            sum(BCmut_15$G12),sum(BCmut_15$H01),
                            sum(BCmut_15$H02),sum(BCmut_15$H03),
                            sum(BCmut_15$H04),sum(BCmut_15$H05))

# Determine the number of unique barcodes recovered from each sampling condition:
norm_vals_15$mappedBCs <- c(length(BCmut_15$D01[BCmut_15$D01 != 0]),
                         length(BCmut_15$D03[BCmut_15$D03 != 0]),
                         length(BCmut_15$D05[BCmut_15$D05 != 0]),
                         length(BCmut_15$D06[BCmut_15$D06 != 0]),
                         length(BCmut_15$D07[BCmut_15$D07 != 0]),
                         length(BCmut_15$D08[BCmut_15$D08 != 0]),
                         length(BCmut_15$D09[BCmut_15$D09 != 0]),
                         length(BCmut_15$D10[BCmut_15$D10 != 0]),
                         length(BCmut_15$D11[BCmut_15$D11 != 0]),
                         length(BCmut_15$E07[BCmut_15$E07 != 0]),
                         length(BCmut_15$E08[BCmut_15$E08 != 0]),
                         length(BCmut_15$E09[BCmut_15$E09 != 0]),
                         length(BCmut_15$E10[BCmut_15$E10 != 0]),
                         length(BCmut_15$E11[BCmut_15$E11 != 0]),
                         length(BCmut_15$E12[BCmut_15$E12 != 0]),
                         length(BCmut_15$F01[BCmut_15$F01 != 0]),
                         length(BCmut_15$F09[BCmut_15$F09 != 0]),
                         length(BCmut_15$F10[BCmut_15$F10 != 0]),
                         length(BCmut_15$F11[BCmut_15$F11 != 0]),
                         length(BCmut_15$F12[BCmut_15$F12 != 0]),
                         length(BCmut_15$G01[BCmut_15$G01 != 0]),
                         length(BCmut_15$G02[BCmut_15$G02 != 0]),
                         length(BCmut_15$G03[BCmut_15$G03 != 0]),
                         length(BCmut_15$G11[BCmut_15$G11 != 0]),
                         length(BCmut_15$G12[BCmut_15$G12 != 0]),
                         length(BCmut_15$H01[BCmut_15$H01 != 0]),
                         length(BCmut_15$H02[BCmut_15$H02 != 0]),
                         length(BCmut_15$H03[BCmut_15$H03 != 0]),
                         length(BCmut_15$H04[BCmut_15$H04 != 0]),
                         length(BCmut_15$H05[BCmut_15$H05 != 0]))

# Library 16

# Sum the total reads recovered from each sampling condition:
norm_vals_16$mappedcounts <- c(sum(BCmut_16$D02),sum(BCmut_16$D04),
                            sum(BCmut_16$D12),sum(BCmut_16$E01),
                            sum(BCmut_16$E02),sum(BCmut_16$E03),
                            sum(BCmut_16$E04),sum(BCmut_16$E05),
                            sum(BCmut_16$E06),sum(BCmut_16$F02),
                            sum(BCmut_16$F03),sum(BCmut_16$F04),
                            sum(BCmut_16$F05),sum(BCmut_16$F06),
                            sum(BCmut_16$F07),sum(BCmut_16$F08),
                            sum(BCmut_16$G04),sum(BCmut_16$G05),
                            sum(BCmut_16$G06),sum(BCmut_16$G07),
                            sum(BCmut_16$G08),sum(BCmut_16$G09),
                            sum(BCmut_16$G10),sum(BCmut_16$H06),
                            sum(BCmut_16$H07),sum(BCmut_16$H08),
                            sum(BCmut_16$H09),sum(BCmut_16$H10),
                            sum(BCmut_16$H11),sum(BCmut_16$H12))

# Determine the number of unique barcodes recovered from each sampling condition:
norm_vals_16$mappedBCs <- c(length(BCmut_16$D02[BCmut_16$D02 != 0]),
                         length(BCmut_16$D04[BCmut_16$D04 != 0]),
                         length(BCmut_16$D12[BCmut_16$D12 != 0]),
                         length(BCmut_16$E01[BCmut_16$E01 != 0]),
                         length(BCmut_16$E02[BCmut_16$E02 != 0]),
                         length(BCmut_16$E03[BCmut_16$E03 != 0]),
                         length(BCmut_16$E04[BCmut_16$E04 != 0]),
                         length(BCmut_16$E05[BCmut_16$E05 != 0]),
                         length(BCmut_16$E06[BCmut_16$E06 != 0]),
                         length(BCmut_16$F02[BCmut_16$F02 != 0]),
                         length(BCmut_16$F03[BCmut_16$F03 != 0]),
                         length(BCmut_16$F04[BCmut_16$F04 != 0]),
                         length(BCmut_16$F05[BCmut_16$F05 != 0]),
                         length(BCmut_16$F06[BCmut_16$F06 != 0]),
                         length(BCmut_16$F07[BCmut_16$F07 != 0]),
                         length(BCmut_16$F08[BCmut_16$F08 != 0]),
                         length(BCmut_16$G04[BCmut_16$G04 != 0]),
                         length(BCmut_16$G05[BCmut_16$G05 != 0]),
                         length(BCmut_16$G06[BCmut_16$G06 != 0]),
                         length(BCmut_16$G07[BCmut_16$G07 != 0]),
                         length(BCmut_16$G08[BCmut_16$G08 != 0]),
                         length(BCmut_16$G09[BCmut_16$G09 != 0]),
                         length(BCmut_16$G10[BCmut_16$G10 != 0]),
                         length(BCmut_16$H06[BCmut_16$H06 != 0]),
                         length(BCmut_16$H07[BCmut_16$H07 != 0]),
                         length(BCmut_16$H08[BCmut_16$H08 != 0]),
                         length(BCmut_16$H09[BCmut_16$H09 != 0]),
                         length(BCmut_16$H10[BCmut_16$H10 != 0]),
                         length(BCmut_16$H11[BCmut_16$H11 != 0]),
                         length(BCmut_16$H12[BCmut_16$H12 != 0]))
```

<font color="green">**Table of Summarized Count Data including Mapped Counts and Mapped BCs for each Sample Condition**</font>
```{r echo=FALSE}
# Display dataframe with kable formatting
norm_vals_15_col3 <- norm_vals_15[,c(4,1,2,3,5,6,7)]

knitr::kable(norm_vals_15_col3,
  col.names = c('Sample Conditions', 'Sample ID', 'Seq Counts', 'BC Counts', 'Scaling', 'Mapped Counts', 'Mapped BCs'),
  align = "llccccc",
  format.args = list(big.mark = ","),
  digits = 2)
```

```{r echo=FALSE}
# Display dataframe with kable formatting
norm_vals_16_col3 <- norm_vals_16[,c(4,1,2,3,5,6,7)]

knitr::kable(norm_vals_16_col3,
  col.names = c('Sample Conditions', 'Sample ID', 'Seq Counts', 'BC Counts', 'Scaling', 'Mapped Counts', 'Mapped BCs'),
  align = "llccccc",
  format.args = list(big.mark = ","),
  digits = 2)
```

### Frac Reads & BCs Mapped

Calculate the fraction of mapped counts and BCs relative to total counts and BCs, respectively
```{r}
# Library 15

# Calculate the fraction of mapped counts or barcodes relative to total counts or barcodes, respectively
norm_vals_15 <- norm_vals_15 %>%
  mutate(frac_mapped_counts=mappedcounts/counts,
         frac_mapped_BC=mappedBCs/numBCs)

# Library 16

# Calculate the fraction of mapped counts or barcodes relative to total counts or barcodes, respectively
norm_vals_16 <- norm_vals_16 %>%
  mutate(frac_mapped_counts=mappedcounts/counts,
         frac_mapped_BC=mappedBCs/numBCs)
```

<font color="green">**Table of Summarized Count Data including Mapped Counts and Mapped BCs for each Sample Condition**</font>
```{r echo=FALSE}
# Display dataframe with kable formatting
norm_vals_15_col4 <- norm_vals_15[,c(4,1,2,3,5,6,7,8,9)]

knitr::kable(norm_vals_15_col4,
  col.names = c('Sample Conditions', 'Sample ID', 'Seq Counts', 'BC Counts', 'Scaling', 'Mapped Counts', 'Mapped BCs', 'Frac Mapped Counts', 'Frac Mapped BCs'),
  align = "llccccccc",
  format.args = list(big.mark = ","),
  digits = 2)
```

<font color="green">**Table of Summarized Count Data including Mapped Counts and Mapped BCs for each Sample Condition**</font>
```{r echo=FALSE}
# Display dataframe with kable formatting
norm_vals_16_col4 <- norm_vals_16[,c(4,1,2,3,5,6,7,8,9)]

knitr::kable(norm_vals_16_col4,
  col.names = c('Sample Conditions', 'Sample ID', 'Seq Counts', 'BC Counts', 'Scaling', 'Mapped Counts', 'Mapped BCs', 'Frac Mapped Counts', 'Frac Mapped BCs'),
  align = "llccccccc",
  format.args = list(big.mark = ","),
  digits = 2)
```

### Plot Mapped Reads

Plot the mapped reads <font color="red">(match with a barcode)</font> as a fraction of the total reads.

**Lib15 Total Reads**
```{r}
# Plot the Fraction of Total Reads Mapped - Library 15
p1_15 <- ggplot(data=norm_vals_15, aes(x=sample, y=frac_mapped_counts)) +
  geom_bar(stat="identity",color="black", fill="aquamarine4", width = 0.8) + 
  xlab("Sample") + 
  ylab("Fraction of Total Reads Mapped") + 
  ggtitle('Library 15 Total Reads Mapped') +
  #theme_cowplot() +
  theme(axis.text.x = element_text(size = 12),
        axis.text.y = element_text(size = 12),
        axis.title.x = element_text(size = 14),
        axis.title.y = element_text(size = 14)) +
  theme(legend.position = "none") +
  theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1)) +
  scale_y_continuous(expand = c(0, 0), limits = c(0, 0.8)) +
  guides(linetype = "none") +
  panel_border(color = "black")

p1_15
```

**Lib16 Total Reads**
```{r}
# Plot the Fraction of Total Reads Mapped
p1_16 <- ggplot(data=norm_vals_16, aes(x=sample, y=frac_mapped_counts)) +
  geom_bar(stat="identity",color="black", fill="aquamarine4", width = 0.8) + 
  xlab("Sample") + 
  ylab("Fraction of Total Reads Mapped") + 
  ggtitle('Library 16 Total Reads Mapped') +
  #theme_cowplot() +
  theme(axis.text.x = element_text(size = 12),
        axis.text.y = element_text(size = 12),
        axis.title.x = element_text(size = 14),
        axis.title.y = element_text(size = 14)) +
  theme(legend.position = "none") +
  theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1)) +
  scale_y_continuous(expand = c(0, 0), limits = c(0, 0.8)) +
  guides(linetype = "none") +
  panel_border(color = "black")

p1_16
```

```{r echo=FALSE}
# Library 15
ggsave(file="plots/fraction_reads_mapped_lib15.v2.png", plot=p1_15,
       dpi=600, width = 8, height = 6, units = "in")

# Library 16
ggsave(file="plots/fraction_reads_mapped_lib16.v2.png", plot=p1_16,
       dpi=600, width = 8, height = 6, units = "in")
```

### Plot BCs Mapped

**Lib15 Total BCs**
```{r}
# Plot the Fraction of Total Barcodes Mapped
p2_15 <- ggplot(data=norm_vals_15, aes(x=sample, y=frac_mapped_BC)) +
  geom_bar(stat="identity",color="black", fill="aquamarine4", width = 0.8) + 
  xlab("Sample") + 
  ylab("Fraction of Total Barcodes Mapped") +
  ggtitle('Library 15 Total BCs Mapped') +
  #theme_cowplot() +
  theme(axis.text.x = element_text(size = 12),
        axis.text.y = element_text(size = 12),
        axis.title.x = element_text(size = 14),
        axis.title.y = element_text(size = 14)) +
  theme(legend.position = "none") +
  theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1)) +
  scale_y_continuous(expand = c(0, 0), limits = c(0, 0.8)) +
  guides(linetype = "none") +
  panel_border(color = "black")

p2_15
```

**Lib16 Total BCs**
```{r}
# Plot the Fraction of Total Barcodes Mapped
p2_16 <- ggplot(data=norm_vals_16, aes(x=sample, y=frac_mapped_BC)) +
  geom_bar(stat="identity",color="black", fill="aquamarine4", width = 0.8) + 
  xlab("Sample") + 
  ylab("Fraction of Total Barcodes Mapped") +
  ggtitle('Library 16 Total BCs Mapped') +
  #theme_cowplot() +
  theme(axis.text.x = element_text(size = 12),
        axis.text.y = element_text(size = 12),
        axis.title.x = element_text(size = 14),
        axis.title.y = element_text(size = 14)) +
  theme(legend.position = "none") +
  theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1)) +
  scale_y_continuous(expand = c(0, 0), limits = c(0, 0.8)) +
  guides(linetype = "none") +
  panel_border(color = "black")

p2_16
```

```{r echo=FALSE}
# Library 15
ggsave(file="plots/fraction_bc_mapped_lib15.v2.png", plot=p2_15,
       dpi=600, width = 8, height = 6, units = "in")

# Library 16
ggsave(file="plots/fraction_bc_mapped_lib16.v2.png", plot=p2_16,
       dpi=600, width = 8, height = 6, units = "in")
```

```{r echo=FALSE}
patch1 <- (p1_15 | p1_16)/(p2_15 | p2_16)
patch1
```

```{r echo=FALSE}
ggsave(file="plots/combined/lib15.16.total.reads.bcs.mapped.v2.png", plot=patch1,
       dpi=600, width = 12, height = 8, units = "in")
```

## Controls
The following section summarizes the log2-fold change in **CONTROL** barcodes between A05 vs. A06 sampling conditions.

### Control Barcodes
Load in the controls and their associated barcodes:
```{r}
#load controls
controls = read.csv("counts/control_BCs.csv", head=FALSE)
colnames(controls) <- c("ctrlname","colnum","BClong","BC","BClen")
```

<font color="red">What is the following code doing?</font>
```{r}
controls <- controls %>%
  mutate(BCstub=str_sub(BClong,1,-4)) %>%
  dplyr::rename(BCold=BC) %>%
  dplyr::rename(BC=BCstub)
```

Summarize the number of control barcodes from the full `BCs` dataset:
```{r, class.output="goodCode"}
# Library 15

BCcontrols_15 <- inner_join(BCs15.prune,controls,by="BC")

BCcontrols_15.count <- inner_join(BCs15.prune,controls,by="BC") %>%
  nrow()

format(BCcontrols_15.count, big.mark = ",")

# Library 16

BCcontrols_16 <- inner_join(BCs16.prune,controls,by="BC")

BCcontrols_16.count <- inner_join(BCs16.prune,controls,by="BC") %>%
  nrow()

format(BCcontrols_16.count, big.mark = ",")
```

### Plotting Controls

**Lib15 Controls**
Plot the control barcodes as the log2-fold change comparison between D05 vs. D03:
```{r}
# Plot the Fraction of Total Barcodes Mapped
p3_15 <- ggplot(data=BCcontrols_15, aes(x=ctrlname, y=D05D03fc)) +
  geom_boxplot(color="black", fill="aquamarine4", alpha=0.8) + 
  geom_point() +
  xlab("Control") + 
  ylab("log2 fold change") + 
  ggtitle("Log2-Fold Change (D05 vs. D03)") +
  #theme_cowplot() +
  theme(axis.text.x = element_text(size = 12),
        axis.text.y = element_text(size = 12),
        axis.title.x = element_text(size = 14),
        axis.title.y = element_text(size = 14)) +
  theme(legend.position = "none") +
  theme(plot.title = element_text(size=16, face="bold", hjust=0.5)) +
  panel_border(color = "black") +
  scale_y_continuous(expand = c(0, 0), limits = c(-3, 0))

p3_15
```

**Lib16 Controls**
Plot the control barcodes as the log2-fold change comparison between D12 vs. D04:
```{r}
# Plot the Fraction of Total Barcodes Mapped
p3_16 <- ggplot(data=BCcontrols_16, aes(x=ctrlname, y=D12D04fc)) +
  geom_boxplot(color="black", fill="aquamarine4", alpha=0.8) + 
  geom_point() +
  xlab("Control") + 
  ylab("log2 fold change") + 
  ggtitle("Log2-Fold Change (D12 vs. D04)") +
  #theme_cowplot() +
  theme(axis.text.x = element_text(size = 12),
        axis.text.y = element_text(size = 12),
        axis.title.x = element_text(size = 14),
        axis.title.y = element_text(size = 14)) +
  theme(legend.position = "none") +
  theme(plot.title = element_text(size=16, face="bold", hjust=0.5)) +
  panel_border(color = "black")+
  scale_y_continuous(expand = c(0, 0), limits = c(-3, 0))

p3_16
```

```{r echo=FALSE}
# Library 15
ggsave(file="plots/controls/controls_Lib15_D05D03fc.png", plot=p3_15,
       dpi=600, width = 6, height = 4, units = "in")

# Library 16
ggsave(file="plots/controls/controls_Lib16_D12D04fc.png", plot=p3_16,
       dpi=600, width = 6, height = 4, units = "in")
```

```{r echo=FALSE}
patch2 <- (p3_15 | p3_16)
patch2
```

```{r echo=FALSE}
ggsave(file="plots/controls/lib15.16.ctrls.png", plot=patch2,
       dpi=600, width = 8, height = 6, units = "in")
```

```{r echo=FALSE, results='hide'}
# Library 15
write.csv(BCcontrols_15,"OUTPUT/BC_15_controls.csv", row.names = FALSE)

# Library 16
write.csv(BCcontrols_16,"OUTPUT/BC_16_controls.csv", row.names = FALSE)
```

## Mutant Count
Begin by summarizing the mutant counts for each unique barcode and set a minimum sequence count threshold for each unique barcode.

**Lib15 Mutant Count**
```{r}
# Lib15
#how many barcodes for each mutant
mutIDinfo15 <- BCmut_15 %>%
  group_by(mutID) %>%
  summarise(numprunedBCs = n()) %>%
  right_join(mutIDinfo15, by="mutID")
```

```{r}
# Lib15
#replace NAs with 0
mutIDinfo15$numprunedBCs[is.na(mutIDinfo15$numprunedBCs)] <- 0
```

```{r}
# Lib15
#mutant must have at least 1 BC after pruning:
mutIDinfo15_1BC <- mutIDinfo15 %>%
  filter(numprunedBCs>0)
```

```{r class.output="goodCode"}
# Lib15
# Count the number of unique mutants remaining after filtering:
format(length(unique(mutIDinfo15_1BC$IDalign)), big.mark = ",")
```

**Lib16 Mutant Count**
```{r}
# Lib16
#how many barcodes for each mutant
mutIDinfo16 <- BCmut_16 %>%
  group_by(mutID) %>%
  summarise(numprunedBCs = n()) %>%
  right_join(mutIDinfo16, by="mutID")
```

```{r}
# Lib16
#replace NAs with 0
mutIDinfo16$numprunedBCs[is.na(mutIDinfo16$numprunedBCs)] <- 0
```

```{r}
# Lib16
#mutant must have at least 1 BC after pruning:
mutIDinfo16_1BC <- mutIDinfo16 %>%
  filter(numprunedBCs>0)
```

```{r class.output="goodCode"}
# Lib16
# Count the number of unique mutants remaining after filtering:
format(length(unique(mutIDinfo16_1BC$IDalign)), big.mark = ",")
```

## Wildtype E coli

The mutID for wildtype E. coli is: NP_414590

Start by pulling out the wildtype E. coli mutID from the full dataset:
```{r}
# Lib15
BCmut15_NP_414590 <- BCmut_15 %>%
  filter(mutID=="NP_414590") %>%
  select(-mutations,-cigar, -mutID)

# Lib16
BCmut16_NP_414590 <- BCmut_16 %>%
  filter(mutID=="NP_414590") %>%
  select(-mutations,-cigar, -mutID)
```

Subset the logFC data columns
```{r}
# Lib15
BCmut15_NP_414590_logfc <- BCmut15_NP_414590 %>%
  select(BC, D05D03fc, D06D03fc, D07D03fc, D08D03fc, D09D03fc, D10D03fc, D11D03fc) %>%
  pivot_longer(!BC, names_to = "fc", values_to = "val")

# Lib16
BCmut16_NP_414590_logfc <- BCmut16_NP_414590 %>%
  select(BC, D12D04fc, E01D04fc, E02D04fc, E03D04fc, E04D04fc, E05D04fc, E06D04fc) %>%
  pivot_longer(!BC, names_to = "fc", values_to = "val")
```

```{r}
# Place sample IDs in desired order for plotting

# Lib15
Lib15_WT_level_order <- c("D05D03fc", "D06D03fc", "D07D03fc", "D08D03fc", "D09D03fc", "D10D03fc", "D11D03fc")

# Lib16
Lib16_WT_level_order <- c("D12D04fc", "E01D04fc", "E02D04fc", "E03D04fc", "E04D04fc", "E05D04fc", "E06D04fc")
```

```{r}
# Lib15
p4_wt_L15 <- ggplot(BCmut15_NP_414590_logfc, aes(x= factor(fc, level = Lib15_WT_level_order), y=val)) +
  geom_violin(fill = "aquamarine4") +
  geom_boxplot(width=0.25) +
  #geom_boxplot(color="black", fill="aquamarine4", alpha=0.8)+
  geom_point()+
  ggtitle("Library 15 WT E. coli Homolog Fitness")+
  ylab("log2 fold change")+
  xlab("TMP Gradient (WT E. coli Homolog)")+
  scale_x_discrete(labels = c("0-tmp", "0.058-tmp", "0.5-tmp", "1.0-tmp", "10-tmp", "50-tmp", "200-tmp"))

p4_wt_L15
```

```{r}
# Lib16
p4_wt_L16 <- ggplot(BCmut16_NP_414590_logfc, aes(x= factor(fc, level = Lib16_WT_level_order), y=val)) +
  geom_violin(fill = "aquamarine4") +
  geom_boxplot(width=0.125) +
  #geom_boxplot(color="black", fill="aquamarine4", alpha=0.8)+
  geom_point()+
  ggtitle("Library 16 WT E. coli Homolog Fitness")+
  ylab("log2 fold change")+
  xlab("TMP Gradient (WT E. coli Homolog)")+
  scale_x_discrete(labels = c("0-tmp", "0.058-tmp", "0.5-tmp", "1.0-tmp", "10-tmp", "50-tmp", "200-tmp"))

p4_wt_L16
```

```{r echo=FALSE}
ggsave(file="plots/lib15.WT.e.coli.fitness.violin.png", plot=p4_wt_L15,
       dpi=600, width = 6, height = 4, units = "in")

ggsave(file="plots/lib16.WT.e.coli.fitness.violin.png", plot=p4_wt_L16,
       dpi=600, width = 6, height = 4, units = "in")
```



# Perfects Only
<font color="blue">**This section is based on the R file: "R_perfects_only.R".**</font> It describes how to select only for perfects (perfect match to designed homolog sequence).

## Select Perfects

Begin by selecting perfects unique mutant IDs from the "mutIDinfo" object where mutations = 0:
```{r}
# Lib15
perfects15 <- mutIDinfo15 %>%
  filter(numprunedBCs>0) %>%
  filter(mutations==0) %>%
  dplyr::rename(ID = IDalign)

# Lib16
perfects16 <- mutIDinfo16 %>%
  filter(numprunedBCs>0) %>%
  filter(mutations==0) %>%
  dplyr::rename(ID = IDalign)
```

Select perfects BCs from "BCmut" object where mutations = 0:
```{r}
# Lib15
BCperfect15 <- BCmut_15 %>%
  filter(mutations==0) %>%
  dplyr::rename(ID=mutID)

# Lib16
BCperfect16 <- BCmut_16 %>%
  filter(mutations==0) %>%
  dplyr::rename(ID=mutID)
```

Perform a quick check to verify no IDs are missing:
```{r class.output="goodCode"}
# Lib15
#check how many are missing ID
nrow(BCperfect15 %>% filter(ID == ""))

# Lib16
#check how many are missing ID
nrow(BCperfect16 %>% filter(ID == ""))
```

Create a function to check if each unique perfect sequence represents a **"true perfect"** based on the designed homolog sequence or a **"synonymous mutation"** that is functionally redundant.
```{r}
#check if this is a synonymous mutant
cigar_syn_mutant_check <- function(cigarin){
  flag_out <- "Perfect"
  if (length(grep("X",cigarin)) != 0) {
    flag_out <- "SynMutant"
  }
  return(flag_out)
}
```

Apply the "perfect check function" to the "BCperfect" dataset and add a column to indicate each BCs status.
```{r}
# Lib15
#check if this BC is a synonymous mutant
BCperfect15 <- BCperfect15 %>%
  rowwise() %>%
  mutate(synflag=cigar_syn_mutant_check(cigar))

# Lib16
#check if this BC is a synonymous mutant
BCperfect16 <- BCperfect16 %>%
  rowwise() %>%
  mutate(synflag=cigar_syn_mutant_check(cigar))
```

Count the number of "Perfects" from the dataset:
```{r class.output="goodCode"}
# Lib15
# Number of perfects
perfects.count15 <- length(which(BCperfect15$synflag=="Perfect"))

format(perfects.count15, big.mark = ",")

# Lib16
# Number of perfects
perfects.count16 <- length(which(BCperfect16$synflag=="Perfect"))

format(perfects.count16, big.mark = ",")
```

Count the number of "SynMutants" from the dataset:
```{r class.output="goodCode"}
# Lib15
# Number of synonymous mutants
synmut.count15 <- length(which(BCperfect15$synflag=="SynMutant"))

format(synmut.count15, big.mark = ",")

# Lib16
# Number of synonymous mutants
synmut.count16 <- length(which(BCperfect16$synflag=="SynMutant"))

format(synmut.count16, big.mark = ",")
```

## Perfects Stats

```{r}
#this code checks the STD among BC fitness for different ways to calc fitness

# Lib15
perfects15_std <- perfects15 %>%
  select(ID)

# Lib16
perfects16_std <- perfects16 %>%
  select(ID)
```

```{r}
#take BCperfect, for each BC with an ID, take the median and sum, calc median, join into perfects

# Lib15
perfects15 <- BCperfect15 %>%
  group_by(ID) %>%
  summarise(fitD03D01=median(D03D01fc), fitD05D03=median(D05D03fc), 
            fitD06D03=median(D06D03fc), fitD07D03=median(D07D03fc), fitD08D03=median(D08D03fc),
            fitD09D03=median(D09D03fc), fitD10D03=median(D10D03fc), fitD11D03=median(D11D03fc),
            fitE07D03=median(E07D03fc), fitE08D03=median(E08D03fc), fitE09D03=median(E09D03fc),
            fitE10D03=median(E10D03fc), fitE11D03=median(E11D03fc), fitE12D03=median(E12D03fc),
            fitF01D03=median(F01D03fc), fitF09D03=median(F09D03fc), fitF10D03=median(F10D03fc),
            fitF11D03=median(F11D03fc), fitF12D03=median(F12D03fc), fitG01D03=median(G01D03fc),
            fitG02D03=median(G02D03fc), fitG03D03=median(G03D03fc), fitG11D03=median(G11D03fc),
            fitG12D03=median(G12D03fc), fitH01D03=median(H01D03fc), fitH02D03=median(H02D03fc),
            fitH03D03=median(H03D03fc), fitH04D03=median(H04D03fc), fitH05D03=median(H05D03fc),
            ) %>%
  right_join(perfects15, by="ID")

# Lib16
perfects16 <- BCperfect16 %>%
  group_by(ID) %>%
  summarise(fitD04D02=median(D04D02fc), fitD12D04=median(D12D04fc), 
            fitE01D04=median(E01D04fc), fitE02D04=median(E02D04fc), fitE03D04=median(E03D04fc),
            fitE04D04=median(E04D04fc), fitE05D04=median(E05D04fc), fitE06D04=median(E06D04fc),
            fitF02D04=median(F02D04fc), fitF03D04=median(F03D04fc), fitF04D04=median(F04D04fc),
            fitF05D04=median(F05D04fc), fitF06D04=median(F06D04fc), fitF07D04=median(F07D04fc),
            fitF08D04=median(F08D04fc), fitG04D04=median(G04D04fc), fitG05D04=median(G05D04fc),
            fitG06D04=median(G06D04fc), fitG07D04=median(G07D04fc), fitG08D04=median(G08D04fc),
            fitG09D04=median(G09D04fc), fitG10D04=median(G10D04fc), fitH06D04=median(H06D04fc),
            fitH07D04=median(H07D04fc), fitH08D04=median(H08D04fc), fitH09D04=median(H09D04fc),
            fitH10D04=median(H10D04fc), fitH11D04=median(H11D04fc), fitH12D04=median(H12D04fc),
            ) %>%
  right_join(perfects16, by="ID")
```

Select only those perfects with a minimum number of barcode:
```{r}
### Select only those with a minimum of 2 BCs

# Lib15
perfects15_2BCs <- perfects15 %>%
  filter(numprunedBCs > 1)

# Lib16
perfects16_2BCs <- perfects16 %>%
  filter(numprunedBCs > 1)

### Select only those with a minimum of 5 BCs

# Lib15
perfects15_5BCs <- perfects15 %>%
  filter(numprunedBCs > 4)

# Lib16
perfects16_5BCs <- perfects16 %>%
  filter(numprunedBCs > 4)
```

### Perfects Counts

**Count the number of perfects after filtering by minimum BC counts (2BCs and 5BCs):**
```{r class.output="goodCode"}
# Lib15

# Count the number of unique perfects at 2BCs minimum retained in Lib15
perfects15_2BCs.count <- length(unique(perfects15_2BCs$ID))
format(perfects15_2BCs.count, big.mark = ",")

# Count the number of unique perfects at 5BCs minimum retained in Lib15
perfects15_5BCs.count <- length(unique(perfects15_5BCs$ID))
format(perfects15_5BCs.count, big.mark = ",")

# Lib16

# Count the number of unique perfects at 2BCs minimum retained in Lib16
perfects16_2BCs.count <- length(unique(perfects16_2BCs$ID))
format(perfects16_2BCs.count, big.mark = ",")

# Count the number of unique perfects at 5BCs minimum retained in Lib16
perfects16_5BCs.count <- length(unique(perfects16_5BCs$ID))
format(perfects16_5BCs.count, big.mark = ",")
```

Merge perfects derived from each library into a single dataframe based on shared ID between the two dataset
```{r class.output="goodCode"}
# Merge by shared "ID"
perfects_15_16_5BCs_shared <- merge(perfects15_5BCs, perfects16_5BCs, by = "ID", all = FALSE)
```

**Count the number of perfects with 5BC minimum shared between both libraries.**
```{r class.output="goodCode"}
# Count the number of unique perfects shared between libraries
perfects_15_16_5BCs_shared.count <- length(unique(perfects_15_16_5BCs_shared$ID))
format(perfects_15_16_5BCs_shared.count, big.mark = ",")
```

**Count the number of perfects with 5BC minimum unique to one library or the other.**
```{r class.output="goodCode"}
# Create a new dataset retaining unique values from both datasets
perfects_15_16_5BCs_unique <- bind_rows(
  anti_join(perfects15_5BCs, perfects16_5BCs, by = "ID"),
  anti_join(perfects16_5BCs, perfects15_5BCs, by = "ID"))

# Count the number of perfects unique to one library or the other
perfects_15_16_5BCs_unique.count <- length(unique(perfects_15_16_5BCs_unique$ID))
format(perfects_15_16_5BCs_unique.count, big.mark = ",")
```

**Summarize the number of perfects shared or unique between both libraries with 5BCs minimum count.**
```{r class.output="goodCode"}
perfects_15_16_5BCs_all.count <- sum(perfects_15_16_5BCs_shared.count + perfects_15_16_5BCs_unique.count)
format(perfects_15_16_5BCs_all.count, big.mark = ",")
```

Save an output of the data for generating trees:
```{r}
### Output 2BCs data for tree generation

# Lib15
write.table(perfects15_2BCs$mutID, file = "OUTPUT/perfects15_final_ID_list_2BCs.csv", sep = ",", row.names = F,quote=F,col.names = F)

# Lib16
write.table(perfects16_2BCs$mutID, file = "OUTPUT/perfects16_final_ID_list_2BCs.csv", sep = ",", row.names = F,quote=F,col.names = F)

### Output 5BCs data for tree generation

# Lib15
write.table(perfects15_5BCs$mutID, file = "OUTPUT/perfects15_final_ID_list_5BCs.csv", sep = ",", row.names = F,quote=F,col.names = F)

# Lib16
write.table(perfects16_5BCs$mutID, file = "OUTPUT/perfects16_final_ID_list_5BCs.csv", sep = ",", row.names = F,quote=F,col.names = F)
```

Re-order the "perfects" dataset by "numprunedBCs" in ascending order:
```{r}
# Lib15
perfects15 <- perfects15 %>%
  arrange(numprunedBCs)

# Lib16
perfects16 <- perfects16 %>%
  arrange(numprunedBCs)
```

```{r}
#make keys based on "numprunedBCs"

# Lib15
perfects15$numprunedBCs.key <- 1:length(perfects15$mutID)

# Lib16
perfects16$numprunedBCs.key <- 1:length(perfects16$mutID)
```

## Perfects Plots
The following section plots various summarizes of the Perfects BCs:

### Rank Ordered Homolog Plot
Plot the number of perfects barcodes recovered based on the rank order of homologs:

**Library 15**
```{r}
p4_15 <- ggplot(perfects15) +
  geom_point(aes(x=numprunedBCs.key,y=numprunedBCs),color = "aquamarine4") +
  ylab("Number of Barcodes") +
  xlab("Rank Ordered Homolog") +
  scale_y_log10() +
  #theme_cowplot() +
  #background_grid() +
  ggtitle("Library 15 Perfects Rank Order") +
  theme(axis.text.x = element_text(size = 12),
        axis.text.y = element_text(size = 12),
        axis.title.x = element_text(size = 14),
        axis.title.y = element_text(size = 14)) +
  theme(legend.position = "none") +
  scale_x_continuous(expand = c(0, 0), limits = c(0,1000)) +
  panel_border(color = "black")

p4_15
```

Establish rank order for plotting.
```{r}
perfects15_rank <- perfects15 %>%
  arrange(fitD05D03)

perfects15_rank$fitD05D03.key <- 1:length(perfects15_rank$ID)

BCperfect15_2 <- BCperfect15 %>%
  left_join(perfects15_rank %>%
              select(ID, fitD05D03, fitD05D03.key), by="ID")
```

Plot Perfects fitness scores for each unique Perfects based on rank ordering of the homologs:
```{r}
p0_15 <- ggplot(BCperfect15_2,aes(x=fitD05D03.key,y=D05D03fc, color=ID)) +
  geom_boxplot() +
  theme(legend.position = "none") +
  xlab("Rank Ordered Homolog") +
  ylab("Fitness (Log2 Fold Change)") +
  ggtitle("Library 15 Perfects Rank Order Fitness") +
  #theme_cowplot() +
  #background_grid() +
  theme(axis.text.x = element_text(size = 12),
        axis.text.y = element_text(size = 12),
        axis.title.x = element_text(size = 14),
        axis.title.y = element_text(size = 14)) +
  theme(legend.position = "none") +
  scale_y_continuous(expand = c(0, 0), limits = c(-12,7)) +
  guides(linetype = FALSE) + panel_border(color = "black")

p0_15
```

```{r echo=FALSE}
# Lib15
ggsave(file="plots/perfects15_rank_ordered_num_BCs.png", plot=p4_15,
       dpi=600, width = 8, height = 6, units = "in")

ggsave(file="plots/perfects15_rank_ordered_fitness.png", plot=p0_15,
       dpi=600, width = 8, height = 6, units = "in")
```


**Library 16**
```{r}
p4_16 <- ggplot(perfects16) +
  geom_point(aes(x=numprunedBCs.key,y=numprunedBCs),color = "aquamarine4") +
  ylab("Number of Barcodes") +
  xlab("Rank Ordered Homolog") +
  scale_y_log10() +
  #theme_cowplot() +
  #background_grid() +
  ggtitle("Library 16 Perfects Rank Order") +
  theme(axis.text.x = element_text(size = 12),
        axis.text.y = element_text(size = 12),
        axis.title.x = element_text(size = 14),
        axis.title.y = element_text(size = 14)) +
  theme(legend.position = "none") +
  scale_x_continuous(expand = c(0, 0), limits = c(0,1000)) +
  panel_border(color = "black")

p4_16
```

Establish rank order for plotting.
```{r}
perfects16_rank <- perfects16 %>%
  arrange(fitD12D04)

perfects16_rank$fitD12D04.key <- 1:length(perfects16_rank$ID)

BCperfect16_2 <- BCperfect16 %>%
  left_join(perfects16_rank %>%
              select(ID, fitD12D04, fitD12D04.key), by="ID")
```

Plot Perfects fitness scores for each unique Perfects based on rank ordering of the homologs:
```{r}
p0_16 <- ggplot(BCperfect16_2,aes(x=fitD12D04.key,y=D12D04fc, color=ID)) +
  geom_boxplot() +
  theme(legend.position = "none") +
  xlab("Rank Ordered Homolog") +
  ylab("Fitness (Log2 Fold Change)") +
  ggtitle("Library 16 Perfects Rank Order Fitness") +
  #theme_cowplot() +
  #background_grid() +
  theme(axis.text.x = element_text(size = 12),
        axis.text.y = element_text(size = 12),
        axis.title.x = element_text(size = 14),
        axis.title.y = element_text(size = 14)) +
  theme(legend.position = "none") +
  scale_y_continuous(expand = c(0, 0), limits = c(-12,7)) +
  guides(linetype = FALSE) + panel_border(color = "black")

p0_16
```

```{r echo=FALSE}
# Lib16
ggsave(file="plots/perfects16_rank_ordered_num_BCs.png", plot=p4_16,
       dpi=600, width = 8, height = 6, units = "in")

ggsave(file="plots/perfects16_rank_ordered_fitness.png", plot=p0_16,
       dpi=600, width = 8, height = 6, units = "in")
```

```{r echo=FALSE}
patch3 <- (p4_15 | p4_16) / (p0_15 | p0_16)
patch3
```

```{r echo=FALSE}
ggsave(file="plots/combined/lib15.16.rank.ordered.num.BCs.fitness.png", plot=patch3,
       dpi=600, width = 12, height = 10, units = "in")
```

### Perfect >2 BCs Fitness Plot

Plot the Perfects counts based on fitness values between D05 vs. D03 (Lib15) or D12 vs. D04 (Lib16) at > 2 BCs:

**Library 15**
```{r}
p5_15 <- ggplot(perfects15_2BCs,aes(x=fitD05D03)) +
  geom_histogram(binwidth=0.08, fill = "aquamarine4", color = "black") +
  xlab("Fitness (Log2 Fold Change)") +
  ylab("Counts") +
  ggtitle("Library 15 Perfects Fitness (>2 BCs)") +
  #theme_cowplot() +
  #background_grid() +
  theme(axis.text.x = element_text(size = 12),
        axis.text.y = element_text(size = 12),
        axis.title.x = element_text(size = 14),
        axis.title.y = element_text(size = 14)) +
  theme(legend.position = "none") +
  scale_x_continuous(limits = c(-6,2)) +
  scale_y_continuous(expand = c(0, 0), limits = c(0,50)) +
  guides(linetype = FALSE) +
  panel_border(color = "black")

p5_15
```

**Library 16**
```{r}
p5_16 <- ggplot(perfects16_2BCs,aes(x=fitD12D04)) +
  geom_histogram(binwidth=0.08, fill = "aquamarine4", color = "black") +
  xlab("Fitness (Log2 Fold Change)") +
  ylab("Counts") +
  ggtitle("Library 16 Perfects Fitness (>2 BCs)") +
  #theme_cowplot() +
  #background_grid() +
  theme(axis.text.x = element_text(size = 12),
        axis.text.y = element_text(size = 12),
        axis.title.x = element_text(size = 14),
        axis.title.y = element_text(size = 14)) +
  theme(legend.position = "none") +
  scale_x_continuous(limits = c(-6,2)) +
  scale_y_continuous(expand = c(0, 0), limits = c(0,50)) +
  guides(linetype = FALSE) +
  panel_border(color = "black")

p5_16
```

```{r echo=FALSE}
# Lib15
ggsave(file="plots/hist15_2BCs_fitD05D03.png", plot=p5_15,
       dpi=600, width = 8, height = 6, units = "in")

# Lib16
ggsave(file="plots/hist16_2BCs_fitD12D04.png", plot=p5_16,
       dpi=600, width = 8, height = 6, units = "in")
```

```{r echo=FALSE}
patch4 <- (p5_15 | p5_16)
patch4
```

```{r echo=FALSE}
ggsave(file="plots/combined/lib15.16.perfects.fitness.2BCs.png", plot=patch4,
       dpi=600, width = 8, height = 6, units = "in")
```

Summarize the total number of unique Perfects (>2 BCs) regardless of fitness value:
```{r class.output="goodCode"}
# Lib15
perfects15_2BCs %>%
  nrow(.)

# Lib16
perfects16_2BCs %>%
  nrow(.)
```

Summarize the number of Perfects (>2 BCs) with fitness greater than -0.5 in D05 vs. D03 (Lib15) or D12 vs. D04 (Lib16) conditions:
```{r class.output="goodCode"}
# Lib15
perfects15_2BCs %>% filter(fitD05D03>-0.5) %>%
  nrow(.)

# Lib16
perfects16_2BCs %>% filter(fitD12D04>-0.5) %>%
  nrow(.)
```

### Perfect >5 BCs Fitness Plot

Plot the Perfects counts based on fitness values between D05 vs. D03 (Lib15) or D12 vs. D04 (Lib16) at > 5 BCs:

**Library 15**
```{r}
p6_15 <- ggplot(perfects15_5BCs,aes(x=fitD05D03)) +
  geom_histogram(binwidth=0.08, fill = "aquamarine4", color = "black") +
  xlab("Fitness (Log2 Fold Change)") +
  ylab("Counts") +
  ggtitle("Library 15 Perfects Fitness (>5 BCs)") +
  #theme_cowplot() +
  #background_grid() +
  theme(axis.text.x = element_text(size = 12),
        axis.text.y = element_text(size = 12),
        axis.title.x = element_text(size = 14),
        axis.title.y = element_text(size = 14)) +
  theme(legend.position = "none") +
  scale_x_continuous(limits = c(-6,2)) +
  scale_y_continuous(expand = c(0, 0), limits = c(0,50)) +
  guides(linetype = FALSE) +
  panel_border(color = "black")

p6_15
```

**Library 16**
```{r}
p6_16 <- ggplot(perfects16_5BCs,aes(x=fitD12D04)) +
  geom_histogram(binwidth=0.08, fill = "aquamarine4", color = "black") +
  xlab("Fitness (Log2 Fold Change)") +
  ylab("Counts") +
  ggtitle("Library 16 Perfects Fitness (>5 BCs)") +
  #theme_cowplot() +
  #background_grid() +
  theme(axis.text.x = element_text(size = 12),
        axis.text.y = element_text(size = 12),
        axis.title.x = element_text(size = 14),
        axis.title.y = element_text(size = 14)) +
  theme(legend.position = "none") +
  scale_x_continuous(limits = c(-6,2)) +
  scale_y_continuous(expand = c(0, 0), limits = c(0,50)) +
  guides(linetype = FALSE) +
  panel_border(color = "black")

p6_16
```

```{r echo=FALSE}
# Lib15
ggsave(file="plots/hist15_5BCs_fitD05D03.png", plot=p6_15,
       dpi=600, width = 8, height = 6, units = "in")

# Lib16
ggsave(file="plots/hist16_5BCs_fitD12D04.png", plot=p6_16,
       dpi=600, width = 8, height = 6, units = "in")
```

```{r echo=FALSE}
patch5 <- (p6_15 | p6_16)
patch5
```

```{r echo=FALSE}
ggsave(file="plots/combined/lib15.16.perfects.fitness.5BCs.v2.png", plot=patch5,
       dpi=600, width = 8, height = 6, units = "in")
```

```{r echo=FALSE}
patch6 <- (p4_15 | p4_16) / (p5_15 | p5_16) / (p6_15 | p6_16)
patch6
```

```{r echo=FALSE}
ggsave(file="plots/combined/lib15.16.perfects.rank.order.fitness.2BCs.5BCs.v2.png", plot=patch6,
       dpi=600, width = 8, height = 10, units = "in")
```

Summarize the total number of unique Perfects (>5 BCs) regardless of fitness value:
```{r class.output="goodCode"}
# Lib15
perfects15_5BCs %>%
  nrow(.)

# Lib16
perfects16_5BCs %>%
  nrow(.)
```

Summarize the number of Perfects (>5 BCs) with fitness greater than -0.5 in D05 vs. D03 (Lib15) or D12 vs. D04 (Lib16) conditions:
```{r class.output="goodCode"}
# Lib15
perfects15_5BCs %>% filter(fitD05D03>-0.5) %>%
  nrow(.)

# Lib16
perfects16_5BCs %>% filter(fitD12D04>-0.5) %>%
  nrow(.)
```

### Perfect >5 BCs Barcode Plot
Plot the Perfects fitness scores for D05 vs. D03 (Lib15) or D12 vs. D04 (Lib16) based on the number of recovered barcodes:

**Library 15**
```{r}
p7_15 <- ggplot(perfects15_5BCs,aes(x=numprunedBCs,y=fitD05D03)) +
  geom_point(color = "aquamarine4") +
  scale_x_log10() +
  xlab("Number of Barcodes") +
  ylab("Fitness (Log2 Fold Change)") +
  ggtitle("Library 15 Perfects Fitness Scores (>5 BCs)") +
  #theme_cowplot() +
  #background_grid() +
  theme(axis.text.x = element_text(size = 12),
        axis.text.y = element_text(size = 12),
        axis.title.x = element_text(size = 14),
        axis.title.y = element_text(size = 14)) +
  theme(legend.position = "none") +
  scale_y_continuous(expand = c(0, 0), limits = c(-5.5,2.5)) +
  guides(linetype = FALSE) +
  panel_border(color = "black")

p7_15
```

**Library 16**
```{r}
p7_16 <- ggplot(perfects16_5BCs,aes(x=numprunedBCs,y=fitD12D04)) +
  geom_point(color = "aquamarine4") +
  scale_x_log10() +
  xlab("Number of Barcodes") +
  ylab("Fitness (Log2 Fold Change)") +
  ggtitle("Library 16 Perfects Fitness Scores (>5 BCs)") +
  #theme_cowplot() +
  #background_grid() +
  theme(axis.text.x = element_text(size = 12),
        axis.text.y = element_text(size = 12),
        axis.title.x = element_text(size = 14),
        axis.title.y = element_text(size = 14)) +
  theme(legend.position = "none") +
  scale_y_continuous(expand = c(0, 0), limits = c(-5.5,2.5)) +
  guides(linetype = FALSE) +
  panel_border(color = "black")

p7_16
```

```{r echo=FALSE}
# Lib15
ggsave(file="plots/perfects15_5BCs_BCs_fitD05D03.png", plot=p7_15,
       dpi=600, width = 12, height = 7.5, units = "in")

# Lib16
ggsave(file="plots/perfects16_5BCs_BCs_fitD12D04.png", plot=p7_16,
       dpi=600, width = 12, height = 7.5, units = "in")
```

```{r echo=FALSE}
patch7 <- (p7_15 | p7_16)
patch7
```

```{r echo=FALSE}
ggsave(file="plots/combined/lib15.16.perfects.fitness.scores.5BCs.png", plot=patch7,
       dpi=600, width = 10, height = 6, units = "in")
```

Gather the average fitness of Perfects at each Trimethoprim concentration based on filtered 5BC dataset.

**Library 15**
```{r}
# Filter perfects15_5BCs dataset by minimum of numprunedBCs > 200
perfects15_5BCsdr <- perfects15_5BCs %>%
  filter(numprunedBCs>10) %>%
  select(ID, fitD05D03, fitD06D03, fitD07D03, fitD08D03, fitD09D03, fitD10D03, fitD11D03, numprunedBCs) %>%#
  gather("selection","fitness",-ID,-numprunedBCs)

# Create naming factors to convert column names into trimethoprim concentrations for plotting
perfects15_5BCsdr$tmpfactor <- "0"
perfects15_5BCsdr$tmpfactor[perfects15_5BCsdr$selection== "fitD05D03"] <- "0"
perfects15_5BCsdr$tmpfactor[perfects15_5BCsdr$selection== "fitD06D03"] <- "0.0058"
perfects15_5BCsdr$tmpfactor[perfects15_5BCsdr$selection== "fitD07D03"] <- "0.5"
perfects15_5BCsdr$tmpfactor[perfects15_5BCsdr$selection== "fitD08D03"] <- "1"
perfects15_5BCsdr$tmpfactor[perfects15_5BCsdr$selection== "fitD09D03"] <- "10"
perfects15_5BCsdr$tmpfactor[perfects15_5BCsdr$selection== "fitD10D03"] <- "50"
perfects15_5BCsdr$tmpfactor[perfects15_5BCsdr$selection== "fitD11D03"] <- "200"

perfects15_5BCsdr$tmpname[perfects15_5BCsdr$selection== "fitD05D03"] <- "A"
perfects15_5BCsdr$tmpname[perfects15_5BCsdr$selection== "fitD06D03"] <- "B"
perfects15_5BCsdr$tmpname[perfects15_5BCsdr$selection== "fitD07D03"] <- "C"
perfects15_5BCsdr$tmpname[perfects15_5BCsdr$selection== "fitD08D03"] <- "D"
perfects15_5BCsdr$tmpname[perfects15_5BCsdr$selection== "fitD09D03"] <- "E"
perfects15_5BCsdr$tmpname[perfects15_5BCsdr$selection== "fitD10D03"] <- "F"
perfects15_5BCsdr$tmpname[perfects15_5BCsdr$selection== "fitD11D03"] <- "G"

perfects15_5BCsdrlabs <- c("0","0.058","0.5","1","10","50","200")
```

Plot the average fitness of Perfects at each Trimethoprim concentration based on filtered 5BC dataset:
```{r}
p32_15 <- ggplot(perfects15_5BCsdr,aes(x=tmpname,y=fitness)) + 
  geom_violin(fill = "aquamarine4") +
  geom_boxplot(width=0.15) +
  xlab("[Trimethoprim] (ug/mL)") +
  ylab("Fitness") +
  scale_x_discrete(labels= perfects15_5BCsdrlabs) +
  #theme_cowplot(14) +
  ggtitle("Library 15 Dose Response") +
  theme(axis.text.x = element_text(size = 12),
        axis.text.y = element_text(size = 12),
        axis.title.x = element_text(size = 14),
        axis.title.y = element_text(size = 14)) +
  theme(legend.position = "none") +
  scale_y_continuous(expand = c(0, 0), limits = c(-11,8)) +
  guides(linetype = FALSE) +
  panel_border(color = "black")

p32_15
```

```{r echo=FALSE}
# Lib15
ggsave(file="plots/perfects15_5BCs_dose_response.10numprunedBCs.png", plot=p32_15,
       dpi=600, width = 12, height = 7.5, units = "in")
```

**Library 16**
```{r}
# Filter perfects16_5BCs dataset by minimum of numprunedBCs > 200
perfects16_5BCsdr <- perfects16_5BCs %>%
  filter(numprunedBCs>10) %>%
  select(ID, fitD12D04, fitE01D04, fitE02D04, fitE03D04, fitE04D04, fitE05D04, fitE06D04, numprunedBCs) %>%#
  gather("selection","fitness",-ID,-numprunedBCs)

# Create naming factors to convert column names into trimethoprim concentrations for plotting
perfects16_5BCsdr$tmpfactor <- "0"
perfects16_5BCsdr$tmpfactor[perfects16_5BCsdr$selection== "fitD12D04"] <- "0"
perfects16_5BCsdr$tmpfactor[perfects16_5BCsdr$selection== "fitE01D04"] <- "0.0058"
perfects16_5BCsdr$tmpfactor[perfects16_5BCsdr$selection== "fitE02D04"] <- "0.5"
perfects16_5BCsdr$tmpfactor[perfects16_5BCsdr$selection== "fitE03D04"] <- "1"
perfects16_5BCsdr$tmpfactor[perfects16_5BCsdr$selection== "fitE04D04"] <- "10"
perfects16_5BCsdr$tmpfactor[perfects16_5BCsdr$selection== "fitE05D04"] <- "50"
perfects16_5BCsdr$tmpfactor[perfects16_5BCsdr$selection== "fitE06D04"] <- "200"

perfects16_5BCsdr$tmpname[perfects16_5BCsdr$selection== "fitD12D04"] <- "A"
perfects16_5BCsdr$tmpname[perfects16_5BCsdr$selection== "fitE01D04"] <- "B"
perfects16_5BCsdr$tmpname[perfects16_5BCsdr$selection== "fitE02D04"] <- "C"
perfects16_5BCsdr$tmpname[perfects16_5BCsdr$selection== "fitE03D04"] <- "D"
perfects16_5BCsdr$tmpname[perfects16_5BCsdr$selection== "fitE04D04"] <- "E"
perfects16_5BCsdr$tmpname[perfects16_5BCsdr$selection== "fitE05D04"] <- "F"
perfects16_5BCsdr$tmpname[perfects16_5BCsdr$selection== "fitE06D04"] <- "G"

perfects16_5BCsdrlabs <- c("0","0.058","0.5","1","10","50","200")
```

Plot the average fitness of Perfects at each Trimethoprim concentration based on filtered 5BC dataset:
```{r}
p32_16 <- ggplot(perfects16_5BCsdr,aes(x=tmpname,y=fitness)) + 
  geom_violin(fill = "aquamarine4") +
  geom_boxplot(width=0.15) +
  xlab("[Trimethoprim] (ug/mL)") +
  ylab("Fitness") +
  scale_x_discrete(labels= perfects16_5BCsdrlabs) +
  #theme_cowplot(14) +
  ggtitle("Library 16 Dose Response") +
  theme(axis.text.x = element_text(size = 12),
        axis.text.y = element_text(size = 12),
        axis.title.x = element_text(size = 14),
        axis.title.y = element_text(size = 14)) +
  theme(legend.position = "none") +
  scale_y_continuous(expand = c(0, 0), limits = c(-11,8)) +
  guides(linetype = FALSE) +
  panel_border(color = "black")

p32_16
```

```{r echo=FALSE}
# Lib16
ggsave(file="plots/perfects16_5BCs_dose_response.10numprunedBCs.png", plot=p32_16,
       dpi=600, width = 12, height = 7.5, units = "in")
```

```{r echo=FALSE}
patch20 <- (p32_15 | p32_16)
patch20
```

```{r echo=FALSE}
ggsave(file="plots/combined/lib15.16.perfects.5BCs.violin.dose.response.10numprunedBCs.png", plot=patch20,
       dpi=600, width = 10, height = 6, units = "in")
```

## DHFR Genes
Load in the organism dataset associated with DHFR genes derived from the <font color="red">NCBI database.</font>
```{r}
#load in all the organisms info for unique BCs associated with DHFR genes: 
orginfo = read.csv("OUTPUT/DHFR_chip_3_proteins_addedseveral.csv", head=TRUE)  # read csv file
```

Subset the dataset:
```{r}
orginfo <- orginfo %>%
  select(-Lib.codon1, -Lib.codon2, -Construct., -Barcode) %>%
  dplyr::rename(IDfull=Accession) %>%
  #mutate(ID=strsplit(IDfull, split = ".")[1])
  tidyr::separate(IDfull, c("ID", "ver"), "\\.",remove = FALSE) %>%
  select(-ver)
names(orginfo)
```

Add the organism info to the perfects info as a new object called "perfects_tree":
```{r}
# Lib15
perfects15_tree <- right_join(orginfo,perfects15,by="ID") %>%
  select(-mutations,-pct_ident,-BCs)

# Lib16
perfects16_tree <- right_join(orginfo,perfects16,by="ID") %>%
  select(-mutations,-pct_ident,-BCs)
```

Create a filtered dataset that only retains perfects with 5+ BCs:
```{r}
# Lib15
perfects15_tree_5BCs <- perfects15_tree %>%
  filter(numprunedBCs > 4) %>%
  select(-IDfull)

# Lib16
perfects16_tree_5BCs <- perfects16_tree %>%
  filter(numprunedBCs > 4) %>%
  select(-IDfull)
```

### Perfects % Similar E. coli Plot

Plot the Perfects fitness scores for D05 vs. D03 (Lib15) or D12 vs. D04 (Lib16) based on percent similarity of the each variant's gene sequences to the E. coli DHFR gene sequence:

**Library 15**
```{r}
p8_15 <- ggplot(perfects15_tree_5BCs,aes(x=PctIdentEcoli,y=fitD05D03)) +
  geom_point(color = "aquamarine4") +
  xlab("Percent Similarity to E. coli DHFR") +
  ylab("Fitness (Log2 Fold Change)") +
  ggtitle("Library 15 Perfects Fitness to % Similarity E. coli (>5 BCs)") +
  #theme_cowplot() +
  #background_grid() +
  theme(axis.text.x = element_text(size = 12),
        axis.text.y = element_text(size = 12),
        axis.title.x = element_text(size = 14),
        axis.title.y = element_text(size = 14)) +
  theme(legend.position = "none") +
  scale_y_continuous(expand = c(0, 0), limits = c(-6,1)) +
  guides(linetype = FALSE) + 
  panel_border(color = "black")

p8_15
```

Bin the % Similarity values in 10% increments and replot as boxplots:
```{r}
# Define bin widths
bin_width <- 0.1

p8_15_box <- ggplot(perfects15_tree_5BCs, aes(x = cut(PctIdentEcoli, breaks = seq(0, 1.0, by = bin_width)), y = fitD05D03)) +
  geom_boxplot(color = "aquamarine4", fill = "aquamarine4") +
  xlab("Percent Similarity to E. coli DHFR (Binned)") +
  ylab("Fitness (Log2 Fold Change)") +
  ggtitle("Library 15 Perfects Fitness to % Similarity E. coli (>5 BCs)") +
  theme(axis.text.x = element_text(size = 12),
        axis.text.y = element_text(size = 12),
        axis.title.x = element_text(size = 14),
        axis.title.y = element_text(size = 14)) +
  scale_x_discrete(labels = c("0.3-0.4", "0.4-0.5", "0.5-0.6", "0.6-0.7", "0.7-0.8", "0.8-0.9", "0.9-1.0")) +
  scale_y_continuous(expand = c(0, 0), limits = c(-6, 1)) +
  guides(linetype = FALSE) + 
  panel_border(color = "black")

p8_15_box
```

**Library 16**
```{r}
p8_16 <- ggplot(perfects16_tree_5BCs,aes(x=PctIdentEcoli,y=fitD12D04)) +
  geom_point(color = "aquamarine4") +
  xlab("Percent Similarity to E. coli DHFR") +
  ylab("Fitness (Log2 Fold Change)") +
  ggtitle("Library 16 Perfects Fitness to % Similarity E. coli (>5 BCs)") +
  #theme_cowplot() +
  #background_grid() +
  theme(axis.text.x = element_text(size = 12),
        axis.text.y = element_text(size = 12),
        axis.title.x = element_text(size = 14),
        axis.title.y = element_text(size = 14)) +
  theme(legend.position = "none") +
  scale_y_continuous(expand = c(0, 0), limits = c(-6,1)) +
  guides(linetype = FALSE) + 
  panel_border(color = "black")

p8_16
```

Bin the % Similarity values in 10% increments and replot as boxplots:
```{r}
# Define bin widths
bin_width <- 0.1

p8_16_box <- ggplot(perfects16_tree_5BCs, aes(x = cut(PctIdentEcoli, breaks = seq(0, 1.0, by = bin_width)), y = fitD12D04)) +
  geom_boxplot(color = "aquamarine4", fill = "aquamarine4") +
  xlab("Percent Similarity to E. coli DHFR (Binned)") +
  ylab("Fitness (Log2 Fold Change)") +
  ggtitle("Library 16 Perfects Fitness to % Similarity E. coli (>5 BCs)") +
  theme(axis.text.x = element_text(size = 12),
        axis.text.y = element_text(size = 12),
        axis.title.x = element_text(size = 14),
        axis.title.y = element_text(size = 14)) +
  scale_x_discrete(labels = c("0.3-0.4", "0.4-0.5", "0.5-0.6", "0.6-0.7", "0.7-0.8", "0.8-0.9", "0.9-1.0")) +
  scale_y_continuous(expand = c(0, 0), limits = c(-6, 1)) +
  guides(linetype = FALSE) + 
  panel_border(color = "black")

p8_16_box
```

```{r echo=FALSE}
# Lib15
ggsave(file="plots/perfects15_5BCs_pctsim_vs_fitD05D03.png", plot=p8_15,
       dpi=600, width = 12, height = 7.5, units = "in")

ggsave(file="plots/perfects15_5BCs_pctsim_vs_fitD05D03.boxplot.png", plot=p8_15_box,
       dpi=600, width = 12, height = 7.5, units = "in")

# Lib16
ggsave(file="plots/perfects16_5BCs_pctsim_vs_fitD12D04.png", plot=p8_16,
       dpi=600, width = 12, height = 7.5, units = "in")

ggsave(file="plots/perfects16_5BCs_pctsim_vs_fitD12D04.boxplot.png", plot=p8_16_box,
       dpi=600, width = 12, height = 7.5, units = "in")
```

```{r echo=FALSE}
patch8 <- (p8_15 | p8_16)
patch8
```

```{r echo=FALSE}
patch8_box <- (p8_15_box | p8_16_box)
patch8_box
```

```{r}
patch8_all <- (p8_15 | p8_16) / (p8_15_box | p8_16_box)
patch8_all
```


```{r echo=FALSE}
ggsave(file="plots/combined/lib15.16.perfects.fitness.scores.5BCs.pctsim.boxplot.png", plot=patch8_box,
       dpi=600, width = 12, height = 6, units = "in")

ggsave(file="plots/combined/lib15.16.perfects.fitness.scores.5BCs.pctsim.scatter.boxplot.png", plot=patch8_all,
       dpi=600, width = 12, height = 12, units = "in")
```

### Perfects Supplement Plot

**This section uses the `library(ggridges)` package.**

Subset the perfects_tree_5BCs object to retain only "ID" and fitness scores for select conditions. This code will transpose the individual fitness score columns into two columns with the fitness label and values, respectively, for each unique ID.
```{r}
### Lib15

# Day 1
perfects15_tree_5BCs_d1 <- perfects15_tree_5BCs %>%
  select(ID, fitD05D03, fitD06D03, fitD07D03, fitD08D03, fitD09D03, fitD10D03, fitD11D03) %>%
  pivot_longer(!ID, names_to = "fc", values_to = "val")

# Day 2
perfects15_tree_5BCs_d2 <- perfects15_tree_5BCs %>%
  select(ID, fitE07D03, fitE08D03, fitE09D03, fitE10D03, fitE11D03, fitE12D03, fitF01D03) %>%
  pivot_longer(!ID, names_to = "fc", values_to = "val")

# Day 3
perfects15_tree_5BCs_d3 <- perfects15_tree_5BCs %>%
  select(ID, fitF09D03, fitF10D03, fitF11D03, fitF12D03, fitG01D03, fitG02D03, fitG03D03) %>%
  pivot_longer(!ID, names_to = "fc", values_to = "val")

# Day 4
perfects15_tree_5BCs_d4 <- perfects15_tree_5BCs %>%
  select(ID, fitG11D03, fitG12D03, fitH01D03, fitH02D03, fitH03D03, fitH04D03, fitH05D03) %>%
  pivot_longer(!ID, names_to = "fc", values_to = "val")



### Lib16

# Day 1
perfects16_tree_5BCs_d1 <- perfects16_tree_5BCs %>%
  select(ID, fitD12D04, fitE01D04, fitE02D04, fitE03D04, fitE04D04, fitE05D04, fitE06D04) %>%
  pivot_longer(!ID, names_to = "fc", values_to = "val")

# Day 2
perfects16_tree_5BCs_d2 <- perfects16_tree_5BCs %>%
  select(ID, fitF02D04, fitF03D04, fitF04D04, fitF05D04, fitF06D04, fitF07D04, fitF08D04) %>%
  pivot_longer(!ID, names_to = "fc", values_to = "val")

# Day 3
perfects16_tree_5BCs_d3 <- perfects16_tree_5BCs %>%
  select(ID, fitG04D04, fitG05D04, fitG06D04, fitG07D04, fitG08D04, fitG09D04, fitG10D04) %>%
  pivot_longer(!ID, names_to = "fc", values_to = "val")

# Day 4
perfects16_tree_5BCs_d4 <- perfects16_tree_5BCs %>%
  select(ID, fitH06D04, fitH07D04, fitH08D04, fitH09D04, fitH10D04, fitH11D04, fitH12D04) %>%
  pivot_longer(!ID, names_to = "fc", values_to = "val")
```

Set the order of sampling conditions for plotting:
```{r}
# Lib15
lib15_level_order_d1 <- c("fitD11D03","fitD10D03","fitD09D03","fitD08D03","fitD07D03","fitD06D03","fitD05D03")
lib15_level_order_d2 <- c("fitF01D03","fitE12D03","fitE11D03","fitE10D03","fitE09D03","fitE08D03","fitE07D03")
lib15_level_order_d3 <- c("fitG03D03","fitG02D03","fitG01D03","fitF12D03","fitF11D03","fitF10D03","fitF09D03")
lib15_level_order_d4 <- c("fitH05D03","fitH04D03","fitH03D03","fitH02D03","fitH01D03","fitG12D03","fitG11D03")

# Lib16
lib16_level_order_d1 <- c("fitE06D04","fitE05D04","fitE04D04","fitE03D04","fitE02D04","fitE01D04","fitD12D04")
lib16_level_order_d2 <- c("fitF08D04","fitF07D04","fitF06D04","fitF05D04","fitF04D04","fitF03D04","fitF02D04")
lib16_level_order_d3 <- c("fitG10D04","fitG09D04","fitG08D04","fitG07D04","fitG06D04","fitG05D04","fitG04D04")
lib16_level_order_d4 <- c("fitH12D04","fitH11D04","fitH10D04","fitH09D04","fitH08D04","fitH07D04","fitH06D04")
```

Plot Perfects fitness scores based on Supplementation treatment for first sampling time point

**Library 15**
```{r}
# Day 1
p9_15_d1 <- ggplot(perfects15_tree_5BCs_d1, aes(x = val, y = factor(fc, level = lib15_level_order_d1))) + 
  geom_density_ridges(fill = "aquamarine4", color = "black") +
  scale_y_discrete(labels = c('200 ug/mL TMP',
                              '50 ug/mL TMP',
                              '10 ug/mL TMP',
                              '1 ug/mL TMP',
                              '0.5 ug/mL TMP',
                              '0.058 ug/mL TMP',
                              'M9 no supp.')) +
  xlab("Fitness (Log2 Fold Change)") +
  ylab("Selection Condition") +
  ggtitle("Library 15 Fitness - Day 1") +
  #theme_cowplot() +
  #background_grid() +
  theme(axis.text.x = element_text(size = 12),
        axis.text.y = element_text(size = 12),
        axis.title.x = element_text(size = 14),
        axis.title.y = element_text(size = 14)) +
  theme(legend.position = "none") +
  scale_x_continuous(limits = c(-15,10)) +
  guides(linetype = FALSE) +
  panel_border(color = "black")

# Day 2
p9_15_d2 <- ggplot(perfects15_tree_5BCs_d2, aes(x = val, y = factor(fc, level = lib15_level_order_d2))) + 
  geom_density_ridges(fill = "aquamarine4", color = "black") +
  scale_y_discrete(labels = c('200 ug/mL TMP',
                              '50 ug/mL TMP',
                              '10 ug/mL TMP',
                              '1 ug/mL TMP',
                              '0.5 ug/mL TMP',
                              '0.058 ug/mL TMP',
                              'M9 no supp.')) +
  xlab("Fitness (Log2 Fold Change)") +
  ylab("Selection Condition") +
  ggtitle("Library 15 Fitness - Day 2") +
  #theme_cowplot() +
  #background_grid() +
  theme(axis.text.x = element_text(size = 12),
        axis.text.y = element_blank(),
        axis.title.x = element_text(size = 14),
        axis.title.y = element_blank()) +
  theme(legend.position = "none") +
  scale_x_continuous(limits = c(-15,10)) +
  guides(linetype = FALSE) +
  panel_border(color = "black")

# Day 3
p9_15_d3 <- ggplot(perfects15_tree_5BCs_d3, aes(x = val, y = factor(fc, level = lib15_level_order_d3))) + 
  geom_density_ridges(fill = "aquamarine4", color = "black") +
  scale_y_discrete(labels = c('200 ug/mL TMP',
                              '50 ug/mL TMP',
                              '10 ug/mL TMP',
                              '1 ug/mL TMP',
                              '0.5 ug/mL TMP',
                              '0.058 ug/mL TMP',
                              'M9 no supp.')) +
  xlab("Fitness (Log2 Fold Change)") +
  ylab("Selection Condition") +
  ggtitle("Library 15 Fitness - Day 3") +
  #theme_cowplot() +
  #background_grid() +
  theme(axis.text.x = element_text(size = 12),
        axis.text.y = element_blank(),
        axis.title.x = element_text(size = 14),
        axis.title.y = element_blank()) +
  theme(legend.position = "none") +
  scale_x_continuous(limits = c(-15,10)) +
  guides(linetype = FALSE) +
  panel_border(color = "black")

# Day 4
p9_15_d4 <- ggplot(perfects15_tree_5BCs_d4, aes(x = val, y = factor(fc, level = lib15_level_order_d4))) + 
  geom_density_ridges(fill = "aquamarine4", color = "black") +
  scale_y_discrete(labels = c('200 ug/mL TMP',
                              '50 ug/mL TMP',
                              '10 ug/mL TMP',
                              '1 ug/mL TMP',
                              '0.5 ug/mL TMP',
                              '0.058 ug/mL TMP',
                              'M9 no supp.')) +
  xlab("Fitness (Log2 Fold Change)") +
  ylab("Selection Condition") +
  ggtitle("Library 15 Fitness - Day 4") +
  #theme_cowplot() +
  #background_grid() +
  theme(axis.text.x = element_text(size = 12),
        axis.text.y = element_blank(),
        axis.title.x = element_text(size = 14),
        axis.title.y = element_blank()) +
  theme(legend.position = "none") +
  scale_x_continuous(limits = c(-15,10)) +
  guides(linetype = FALSE) +
  panel_border(color = "black")

p9_15_d1
p9_15_d2
p9_15_d3
p9_15_d4
```

**Library 16**
```{r}
# Day 1
p9_16_d1 <- ggplot(perfects16_tree_5BCs_d1, aes(x = val, y = factor(fc, level = lib16_level_order_d1))) + 
  geom_density_ridges(fill = "aquamarine4", color = "black") +
  scale_y_discrete(labels = c('200 ug/mL TMP',
                              '50 ug/mL TMP',
                              '10 ug/mL TMP',
                              '1 ug/mL TMP',
                              '0.5 ug/mL TMP',
                              '0.058 ug/mL TMP',
                              'M9 no supp.')) +
  xlab("Fitness (Log2 Fold Change)") +
  ylab("Selection Condition") +
  ggtitle("Library 16 Fitness - Day 1") +
  #theme_cowplot() +
  #background_grid() +
  theme(axis.text.x = element_text(size = 12),
        axis.text.y = element_text(size = 12),
        axis.title.x = element_text(size = 14),
        axis.title.y = element_text(size = 14)) +
  theme(legend.position = "none") +
  scale_x_continuous(limits = c(-15,10)) +
  guides(linetype = FALSE) +
  panel_border(color = "black")

# Day 2
p9_16_d2 <- ggplot(perfects16_tree_5BCs_d2, aes(x = val, y = factor(fc, level = lib16_level_order_d2))) + 
  geom_density_ridges(fill = "aquamarine4", color = "black") +
  scale_y_discrete(labels = c('200 ug/mL TMP',
                              '50 ug/mL TMP',
                              '10 ug/mL TMP',
                              '1 ug/mL TMP',
                              '0.5 ug/mL TMP',
                              '0.058 ug/mL TMP',
                              'M9 no supp.')) +
  xlab("Fitness (Log2 Fold Change)") +
  ylab("Selection Condition") +
  ggtitle("Library 16 Fitness - Day 2") +
  #theme_cowplot() +
  #background_grid() +
  theme(axis.text.x = element_text(size = 12),
        axis.text.y = element_blank(),
        axis.title.x = element_text(size = 14),
        axis.title.y = element_blank()) +
  theme(legend.position = "none") +
  scale_x_continuous(limits = c(-15,10)) +
  guides(linetype = FALSE) +
  panel_border(color = "black")

# Day 3
p9_16_d3 <- ggplot(perfects16_tree_5BCs_d3, aes(x = val, y = factor(fc, level = lib16_level_order_d3))) + 
  geom_density_ridges(fill = "aquamarine4", color = "black") +
  scale_y_discrete(labels = c('200 ug/mL TMP',
                              '50 ug/mL TMP',
                              '10 ug/mL TMP',
                              '1 ug/mL TMP',
                              '0.5 ug/mL TMP',
                              '0.058 ug/mL TMP',
                              'M9 no supp.')) +
  xlab("Fitness (Log2 Fold Change)") +
  ylab("Selection Condition") +
  ggtitle("Library 16 Fitness - Day 3") +
  #theme_cowplot() +
  #background_grid() +
  theme(axis.text.x = element_text(size = 12),
        axis.text.y = element_blank(),
        axis.title.x = element_text(size = 14),
        axis.title.y = element_blank()) +
  theme(legend.position = "none") +
  scale_x_continuous(limits = c(-15,10)) +
  guides(linetype = FALSE) +
  panel_border(color = "black")

# Day 4
p9_16_d4 <- ggplot(perfects16_tree_5BCs_d4, aes(x = val, y = factor(fc, level = lib16_level_order_d4))) + 
  geom_density_ridges(fill = "aquamarine4", color = "black") +
  scale_y_discrete(labels = c('200 ug/mL TMP',
                              '50 ug/mL TMP',
                              '10 ug/mL TMP',
                              '1 ug/mL TMP',
                              '0.5 ug/mL TMP',
                              '0.058 ug/mL TMP',
                              'M9 no supp.')) +
  xlab("Fitness (Log2 Fold Change)") +
  ylab("Selection Condition") +
  ggtitle("Library 16 Fitness - Day 4") +
  #theme_cowplot() +
  #background_grid() +
  theme(axis.text.x = element_text(size = 12),
        axis.text.y = element_blank(),
        axis.title.x = element_text(size = 14),
        axis.title.y = element_blank()) +
  theme(legend.position = "none") +
  scale_x_continuous(limits = c(-15,10)) +
  guides(linetype = FALSE) +
  panel_border(color = "black")

p9_16_d1
p9_16_d2
p9_16_d3
p9_16_d4
```

```{r echo=FALSE}
patch9 <- (p9_15_d1 | p9_15_d2 | p9_15_d3 | p9_15_d4) / (p9_16_d1 | p9_16_d2 | p9_16_d3 | p9_16_d4)
patch9
```

```{r echo=FALSE}
ggsave(file="plots/combined/lib15.16.supplement.fitness.by.day.v2.png", plot=patch9,
       dpi=600, width = 14, height = 6, units = "in")
```

### Perfects Dose Response

Plot the number of perfects barcodes present under each Trimethoprim selection condition.

#### Dose Curves

**Library 15**

Filter the Perfects counts to include only those perfects with > 200 BCs, then select only the dose responses for trimethoprim concentrations on Day 1.
```{r}
perfects15_5BCs_dr1 <- perfects15_5BCs %>%
  filter(numprunedBCs>200) %>%
  select(ID,
         fitD06D03,
         fitD07D03,
         fitD08D03,
         fitD09D03,
         fitD10D03,
         fitD11D03,
         numprunedBCs) %>%
  gather("selection","fitness",-ID,-numprunedBCs)
```

Define the trimethoprim concentrations based on the fitness codes associated with sample IDs.
```{r}
perfects15_5BCs_dr1$tmp <- 0
perfects15_5BCs_dr1$tmp[perfects15_5BCs_dr1$selection== "fitD06D03"] <- 0.058
perfects15_5BCs_dr1$tmp[perfects15_5BCs_dr1$selection== "fitD07D03"] <- 0.5
perfects15_5BCs_dr1$tmp[perfects15_5BCs_dr1$selection== "fitD08D03"] <- 1
perfects15_5BCs_dr1$tmp[perfects15_5BCs_dr1$selection== "fitD09D03"] <- 10
perfects15_5BCs_dr1$tmp[perfects15_5BCs_dr1$selection== "fitD10D03"] <- 50
perfects15_5BCs_dr1$tmp[perfects15_5BCs_dr1$selection== "fitD11D03"] <- 200
```

Plot the dose response for remaining perfects (>200 BCs) over the trimethoprim gradient.
```{r}
p10_15_d1 <- ggplot(perfects15_5BCs_dr1, aes(x=tmp, y=fitness, color=ID)) +
  geom_point() +
  geom_line() +
  xlab("[Trimethoprim] (ug/mL)") +
  ylab("Fitness") +
  ggtitle("Library 15 Dose Response - Day 1") +
  scale_x_log10() +
  theme(axis.text.x = element_text(size = 12),
        axis.text.y = element_text(size = 12),
        axis.title.x = element_text(size = 14),
        axis.title.y = element_text(size = 14)) +
  theme(legend.position = "none") +
  scale_y_continuous(expand = c(0, 0), limits = c(-11,5)) +
  guides(linetype = FALSE) +
  panel_border(color = "black")

p10_15_d1
```

```{r echo=FALSE}
# Lib15
ggsave(file="plots/perfects15_5BCs_200BC_dose_response_day1.png", plot=p10_15_d1,
       dpi=600, width = 8, height = 7.5, units = "in")
```

**Library 16**

Filter the Perfects counts to include only those perfects with > 200 BCs, then select only the dose responses for trimethoprim concentrations on Day 1.
```{r}
perfects16_5BCs_dr1 <- perfects16_5BCs %>%
  filter(numprunedBCs>200) %>%
  select(ID,
         fitE01D04,
         fitE02D04,
         fitE03D04,
         fitE04D04,
         fitE05D04,
         fitE06D04,
         numprunedBCs) %>%
  gather("selection","fitness",-ID,-numprunedBCs)
```

Define the trimethoprim concentrations based on the fitness codes associated with sample IDs.
```{r}
perfects16_5BCs_dr1$tmp <- 0
perfects16_5BCs_dr1$tmp[perfects16_5BCs_dr1$selection== "fitE01D04"] <- 0.058
perfects16_5BCs_dr1$tmp[perfects16_5BCs_dr1$selection== "fitE02D04"] <- 0.5
perfects16_5BCs_dr1$tmp[perfects16_5BCs_dr1$selection== "fitE03D04"] <- 1
perfects16_5BCs_dr1$tmp[perfects16_5BCs_dr1$selection== "fitE04D04"] <- 10
perfects16_5BCs_dr1$tmp[perfects16_5BCs_dr1$selection== "fitE05D04"] <- 50
perfects16_5BCs_dr1$tmp[perfects16_5BCs_dr1$selection== "fitE06D04"] <- 200
```

Plot the dose response for remaining perfects (>200 BCs) over the trimethoprim gradient.
```{r}
p10_16_d1 <- ggplot(perfects16_5BCs_dr1, aes(x=tmp, y=fitness, color=ID)) +
  geom_point() +
  geom_line() +
  xlab("[Trimethoprim] (ug/mL)") +
  ylab("Fitness") +
  ggtitle("Library 16 Dose Response - Day 1") +
  scale_x_log10() +
  theme(axis.text.x = element_text(size = 12),
        axis.text.y = element_text(size = 12),
        axis.title.x = element_text(size = 14),
        axis.title.y = element_text(size = 14)) +
  theme(legend.position = "none") +
  scale_y_continuous(expand = c(0, 0), limits = c(-11,5)) +
  guides(linetype = FALSE) +
  panel_border(color = "black")

p10_16_d1
```

```{r echo=FALSE}
# Lib15
ggsave(file="plots/perfects16_5BCs_200BC_dose_response_day1.png", plot=p10_16_d1,
       dpi=600, width = 8, height = 7.5, units = "in")
```

```{r echo=FALSE}
patch10 <- (p10_15_d1 | p10_16_d1)
patch10
```

```{r echo=FALSE}
ggsave(file="plots/combined/lib15.16.dose.response.curves.day1.png", plot=patch10,
       dpi=600, width = 10, height = 6, units = "in")
```

#### Dose Boxplots

**Library 15 - Day 1**

Filter the Perfects counts to include only those perfects with > 200 BCs, then select only the dose responses for trimethoprim concentrations on Day 1.
```{r}
perfects15_200BCs_dr1 <- perfects15_5BCs %>%
  filter(numprunedBCs>200) %>%
  select(ID, fitD05D03, fitD06D03, fitD07D03, fitD08D03, fitD09D03, fitD10D03, fitD11D03, numprunedBCs) %>%
  gather("selection","fitness",-ID,-numprunedBCs)
```

Re-Define the trimethoprim concentrations based on the fitness codes associated with sample IDs.
```{r}
perfects15_200BCs_dr1$tmpfactor <- "0"
perfects15_200BCs_dr1$tmpfactor[perfects15_200BCs_dr1$selection== "fitD05D03"] <- "0"
perfects15_200BCs_dr1$tmpfactor[perfects15_200BCs_dr1$selection== "fitD06D03"] <- "0.058"
perfects15_200BCs_dr1$tmpfactor[perfects15_200BCs_dr1$selection== "fitD07D03"] <- "0.5"
perfects15_200BCs_dr1$tmpfactor[perfects15_200BCs_dr1$selection== "fitD08D03"] <- "1"
perfects15_200BCs_dr1$tmpfactor[perfects15_200BCs_dr1$selection== "fitD09D03"] <- "10"
perfects15_200BCs_dr1$tmpfactor[perfects15_200BCs_dr1$selection== "fitD10D03"] <- "50"
perfects15_200BCs_dr1$tmpfactor[perfects15_200BCs_dr1$selection== "fitD11D03"] <- "200"

perfects15_200BCs_dr1$tmpname <- "0"
perfects15_200BCs_dr1$tmpname[perfects15_200BCs_dr1$selection== "fitD05D03"] <- "A"
perfects15_200BCs_dr1$tmpname[perfects15_200BCs_dr1$selection== "fitD06D03"] <- "B"
perfects15_200BCs_dr1$tmpname[perfects15_200BCs_dr1$selection== "fitD07D03"] <- "C"
perfects15_200BCs_dr1$tmpname[perfects15_200BCs_dr1$selection== "fitD08D03"] <- "D"
perfects15_200BCs_dr1$tmpname[perfects15_200BCs_dr1$selection== "fitD09D03"] <- "E"
perfects15_200BCs_dr1$tmpname[perfects15_200BCs_dr1$selection== "fitD10D03"] <- "F"
perfects15_200BCs_dr1$tmpname[perfects15_200BCs_dr1$selection== "fitD11D03"] <- "G"

perfects15_200BCs_dr_labels <- c("0","0.058","0.5","1","10","50","200")
```

Plot the dose response for remaining perfects (>200 BCs) as boxplots for each trimethoprim concentration.
```{r}
p11_15_d1 <- ggplot(perfects15_200BCs_dr1, aes(x=tmpname, y=fitness)) +
  geom_boxplot(color="black", fill="aquamarine4") +
  xlab("[Trimethoprim] (ug/mL)") +
  ylab("Fitness") +
  ggtitle("Library 15 Dose Response - Day 1") +
  scale_x_discrete(labels = perfects15_200BCs_dr_labels) +
  theme(axis.text.x = element_text(size = 12),
        axis.text.y = element_text(size = 12),
        axis.title.x = element_text(size = 14),
        axis.title.y = element_text(size = 14)) +
  theme(legend.position = "none") +
  scale_y_continuous(expand = c(0, 0), limits = c(-11,5)) +
  guides(linetype = FALSE) +
  panel_border(color = "black")

p11_15_d1
```

**Library 15 - Day 2**

Filter the Perfects counts to include only those perfects with > 200 BCs, then select only the dose responses for trimethoprim concentrations on Day 2.
```{r}
perfects15_200BCs_dr2 <- perfects15_5BCs %>%
  filter(numprunedBCs>200) %>%
  select(ID,
         fitE07D03,
         fitE08D03,
         fitE09D03,
         fitE10D03,
         fitE11D03,
         fitE12D03,
         fitF01D03,
         numprunedBCs) %>%
  gather("selection","fitness",-ID,-numprunedBCs)
```

Re-Define the trimethoprim concentrations based on the fitness codes associated with sample IDs.
```{r}
perfects15_200BCs_dr2$tmpfactor <- "0"
perfects15_200BCs_dr2$tmpfactor[perfects15_200BCs_dr2$selection== "fitE07D03"] <- "0"
perfects15_200BCs_dr2$tmpfactor[perfects15_200BCs_dr2$selection== "fitE08D03"] <- "0.058"
perfects15_200BCs_dr2$tmpfactor[perfects15_200BCs_dr2$selection== "fitE09D03"] <- "0.5"
perfects15_200BCs_dr2$tmpfactor[perfects15_200BCs_dr2$selection== "fitE10D03"] <- "1"
perfects15_200BCs_dr2$tmpfactor[perfects15_200BCs_dr2$selection== "fitE11D03"] <- "10"
perfects15_200BCs_dr2$tmpfactor[perfects15_200BCs_dr2$selection== "fitE12D03"] <- "50"
perfects15_200BCs_dr2$tmpfactor[perfects15_200BCs_dr2$selection== "fitF01D03"] <- "200"

perfects15_200BCs_dr2$tmpname <- "0"
perfects15_200BCs_dr2$tmpname[perfects15_200BCs_dr2$selection== "fitE07D03"] <- "A"
perfects15_200BCs_dr2$tmpname[perfects15_200BCs_dr2$selection== "fitE08D03"] <- "B"
perfects15_200BCs_dr2$tmpname[perfects15_200BCs_dr2$selection== "fitE09D03"] <- "C"
perfects15_200BCs_dr2$tmpname[perfects15_200BCs_dr2$selection== "fitE10D03"] <- "D"
perfects15_200BCs_dr2$tmpname[perfects15_200BCs_dr2$selection== "fitE11D03"] <- "E"
perfects15_200BCs_dr2$tmpname[perfects15_200BCs_dr2$selection== "fitE12D03"] <- "F"
perfects15_200BCs_dr2$tmpname[perfects15_200BCs_dr2$selection== "fitF01D03"] <- "G"

perfects15_200BCs_dr_labels <- c("0","0.058","0.5","1","10","50","200")
```

Plot the dose response for remaining perfects (>200 BCs) as boxplots for each trimethoprim concentration.
```{r}
p11_15_d2 <- ggplot(perfects15_200BCs_dr2, aes(x=tmpname, y=fitness)) +
  geom_boxplot(color="black", fill="aquamarine4") +
  xlab("[Trimethoprim] (ug/mL)") +
  ylab("Fitness") +
  ggtitle("Library 15 Dose Response - Day 2") +
  scale_x_discrete(labels = perfects15_200BCs_dr_labels) +
  theme(axis.text.x = element_text(size = 12),
        axis.text.y = element_text(size = 12),
        axis.title.x = element_text(size = 14),
        axis.title.y = element_text(size = 14)) +
  theme(legend.position = "none") +
  scale_y_continuous(expand = c(0, 0), limits = c(-11,5)) +
  guides(linetype = FALSE) +
  panel_border(color = "black")

p11_15_d2
```

**Library 15 - Day 3**

Filter the Perfects counts to include only those perfects with > 200 BCs, then select only the dose responses for trimethoprim concentrations on Day 3.
```{r}
perfects15_200BCs_dr3 <- perfects15_5BCs %>%
  filter(numprunedBCs>200) %>%
  select(ID,
         fitF09D03,
         fitF10D03,
         fitF11D03,
         fitF12D03,
         fitG01D03,
         fitG02D03,
         fitG03D03,
         numprunedBCs) %>%
  gather("selection","fitness",-ID,-numprunedBCs)
```

Re-Define the trimethoprim concentrations based on the fitness codes associated with sample IDs.
```{r}
perfects15_200BCs_dr3$tmpfactor <- "0"
perfects15_200BCs_dr3$tmpfactor[perfects15_200BCs_dr3$selection== "fitF09D03"] <- "0"
perfects15_200BCs_dr3$tmpfactor[perfects15_200BCs_dr3$selection== "fitF10D03"] <- "0.058"
perfects15_200BCs_dr3$tmpfactor[perfects15_200BCs_dr3$selection== "fitF11D03"] <- "0.5"
perfects15_200BCs_dr3$tmpfactor[perfects15_200BCs_dr3$selection== "fitF12D03"] <- "1"
perfects15_200BCs_dr3$tmpfactor[perfects15_200BCs_dr3$selection== "fitG01D03"] <- "10"
perfects15_200BCs_dr3$tmpfactor[perfects15_200BCs_dr3$selection== "fitG02D03"] <- "50"
perfects15_200BCs_dr3$tmpfactor[perfects15_200BCs_dr3$selection== "fitG03D03"] <- "200"

perfects15_200BCs_dr3$tmpname <- "0"
perfects15_200BCs_dr3$tmpname[perfects15_200BCs_dr3$selection== "fitF09D03"] <- "A"
perfects15_200BCs_dr3$tmpname[perfects15_200BCs_dr3$selection== "fitF10D03"] <- "B"
perfects15_200BCs_dr3$tmpname[perfects15_200BCs_dr3$selection== "fitF11D03"] <- "C"
perfects15_200BCs_dr3$tmpname[perfects15_200BCs_dr3$selection== "fitF12D03"] <- "D"
perfects15_200BCs_dr3$tmpname[perfects15_200BCs_dr3$selection== "fitG01D03"] <- "E"
perfects15_200BCs_dr3$tmpname[perfects15_200BCs_dr3$selection== "fitG02D03"] <- "F"
perfects15_200BCs_dr3$tmpname[perfects15_200BCs_dr3$selection== "fitG03D03"] <- "G"

perfects15_200BCs_dr_labels <- c("0","0.058","0.5","1","10","50","200")
```

Plot the dose response for remaining perfects (>200 BCs) as boxplots for each trimethoprim concentration.
```{r}
p11_15_d3 <- ggplot(perfects15_200BCs_dr3, aes(x=tmpname, y=fitness)) +
  geom_boxplot(color="black", fill="aquamarine4") +
  xlab("[Trimethoprim] (ug/mL)") +
  ylab("Fitness") +
  ggtitle("Library 15 Dose Response - Day 3") +
  scale_x_discrete(labels = perfects15_200BCs_dr_labels) +
  theme(axis.text.x = element_text(size = 12),
        axis.text.y = element_text(size = 12),
        axis.title.x = element_text(size = 14),
        axis.title.y = element_text(size = 14)) +
  theme(legend.position = "none") +
  scale_y_continuous(expand = c(0, 0), limits = c(-11,5)) +
  guides(linetype = FALSE) +
  panel_border(color = "black")

p11_15_d3
```

**Library 15 - Day 4**

Filter the Perfects counts to include only those perfects with > 200 BCs, then select only the dose responses for trimethoprim concentrations on Day 4.
```{r}
perfects15_200BCs_dr4 <- perfects15_5BCs %>%
  filter(numprunedBCs>200) %>%
  select(ID,
         fitG11D03,
         fitG12D03,
         fitH01D03,
         fitH02D03,
         fitH03D03,
         fitH04D03,
         fitH05D03,
         numprunedBCs) %>%
  gather("selection","fitness",-ID,-numprunedBCs)
```

Re-Define the trimethoprim concentrations based on the fitness codes associated with sample IDs.
```{r}
perfects15_200BCs_dr4$tmpfactor <- "0"
perfects15_200BCs_dr4$tmpfactor[perfects15_200BCs_dr4$selection== "fitG11D03"] <- "0"
perfects15_200BCs_dr4$tmpfactor[perfects15_200BCs_dr4$selection== "fitG12D03"] <- "0.058"
perfects15_200BCs_dr4$tmpfactor[perfects15_200BCs_dr4$selection== "fitH01D03"] <- "0.5"
perfects15_200BCs_dr4$tmpfactor[perfects15_200BCs_dr4$selection== "fitH02D03"] <- "1"
perfects15_200BCs_dr4$tmpfactor[perfects15_200BCs_dr4$selection== "fitH03D03"] <- "10"
perfects15_200BCs_dr4$tmpfactor[perfects15_200BCs_dr4$selection== "fitH04D03"] <- "50"
perfects15_200BCs_dr4$tmpfactor[perfects15_200BCs_dr4$selection== "fitH05D03"] <- "200"

perfects15_200BCs_dr4$tmpname <- "0"
perfects15_200BCs_dr4$tmpname[perfects15_200BCs_dr4$selection== "fitG11D03"] <- "A"
perfects15_200BCs_dr4$tmpname[perfects15_200BCs_dr4$selection== "fitG12D03"] <- "B"
perfects15_200BCs_dr4$tmpname[perfects15_200BCs_dr4$selection== "fitH01D03"] <- "C"
perfects15_200BCs_dr4$tmpname[perfects15_200BCs_dr4$selection== "fitH02D03"] <- "D"
perfects15_200BCs_dr4$tmpname[perfects15_200BCs_dr4$selection== "fitH03D03"] <- "E"
perfects15_200BCs_dr4$tmpname[perfects15_200BCs_dr4$selection== "fitH04D03"] <- "F"
perfects15_200BCs_dr4$tmpname[perfects15_200BCs_dr4$selection== "fitH05D03"] <- "G"

perfects15_200BCs_dr_labels <- c("0","0.058","0.5","1","10","50","200")
```

Plot the dose response for remaining perfects (>200 BCs) as boxplots for each trimethoprim concentration.
```{r}
p11_15_d4 <- ggplot(perfects15_200BCs_dr4, aes(x=tmpname, y=fitness)) +
  geom_boxplot(color="black", fill="aquamarine4") +
  xlab("[Trimethoprim] (ug/mL)") +
  ylab("Fitness") +
  ggtitle("Library 15 Dose Response - Day 4") +
  scale_x_discrete(labels = perfects15_200BCs_dr_labels) +
  theme(axis.text.x = element_text(size = 12),
        axis.text.y = element_text(size = 12),
        axis.title.x = element_text(size = 14),
        axis.title.y = element_text(size = 14)) +
  theme(legend.position = "none") +
  scale_y_continuous(expand = c(0, 0), limits = c(-11,5)) +
  guides(linetype = FALSE) +
  panel_border(color = "black")

p11_15_d4
```

```{r echo=FALSE}
patch11_Lib15 <- (p11_15_d1 | p11_15_d2 | p11_15_d3 | p11_15_d4)
patch11_Lib15
```

```{r echo=FALSE}
ggsave(file="plots/combined/lib15.dose.response.by.day.png", plot=patch11_Lib15,
       dpi=600, width = 14, height = 6, units = "in")
```

**Library 16 - Day 1**

Filter the Perfects counts to include only those perfects with > 200 BCs, then select only the dose responses for trimethoprim concentrations on Day 1.
```{r}
perfects16_200BCs_dr1 <- perfects16_5BCs %>%
  filter(numprunedBCs>200) %>%
  select(ID,
         fitD12D04,
         fitE01D04,
         fitE02D04,
         fitE03D04,
         fitE04D04,
         fitE05D04,
         fitE06D04,
         numprunedBCs) %>%
  gather("selection","fitness",-ID,-numprunedBCs)
```

Re-Define the trimethoprim concentrations based on the fitness codes associated with sample IDs.
```{r}
perfects16_200BCs_dr1$tmpfactor <- "0"
perfects16_200BCs_dr1$tmpfactor[perfects16_200BCs_dr1$selection== "fitD12D04"] <- "0"
perfects16_200BCs_dr1$tmpfactor[perfects16_200BCs_dr1$selection== "fitE01D04"] <- "0.058"
perfects16_200BCs_dr1$tmpfactor[perfects16_200BCs_dr1$selection== "fitE02D04"] <- "0.5"
perfects16_200BCs_dr1$tmpfactor[perfects16_200BCs_dr1$selection== "fitE03D04"] <- "1"
perfects16_200BCs_dr1$tmpfactor[perfects16_200BCs_dr1$selection== "fitE04D04"] <- "10"
perfects16_200BCs_dr1$tmpfactor[perfects16_200BCs_dr1$selection== "fitE05D04"] <- "50"
perfects16_200BCs_dr1$tmpfactor[perfects16_200BCs_dr1$selection== "fitE06D04"] <- "200"

perfects16_200BCs_dr1$tmpname <- "0"
perfects16_200BCs_dr1$tmpname[perfects16_200BCs_dr1$selection== "fitD12D04"] <- "A"
perfects16_200BCs_dr1$tmpname[perfects16_200BCs_dr1$selection== "fitE01D04"] <- "B"
perfects16_200BCs_dr1$tmpname[perfects16_200BCs_dr1$selection== "fitE02D04"] <- "C"
perfects16_200BCs_dr1$tmpname[perfects16_200BCs_dr1$selection== "fitE03D04"] <- "D"
perfects16_200BCs_dr1$tmpname[perfects16_200BCs_dr1$selection== "fitE04D04"] <- "E"
perfects16_200BCs_dr1$tmpname[perfects16_200BCs_dr1$selection== "fitE05D04"] <- "F"
perfects16_200BCs_dr1$tmpname[perfects16_200BCs_dr1$selection== "fitE06D04"] <- "G"

perfects16_200BCs_dr_labels <- c("0","0.058","0.5","1","10","50","200")
```

Plot the dose response for remaining perfects (>200 BCs) as boxplots for each trimethoprim concentration.
```{r}
p11_16_d1 <- ggplot(perfects16_200BCs_dr1, aes(x=tmpname, y=fitness)) +
  geom_boxplot(color="black", fill="aquamarine4") +
  xlab("[Trimethoprim] (ug/mL)") +
  ylab("Fitness") +
  ggtitle("Library 16 Dose Response - Day 1") +
  scale_x_discrete(labels = perfects16_200BCs_dr_labels) +
  theme(axis.text.x = element_text(size = 12),
        axis.text.y = element_text(size = 12),
        axis.title.x = element_text(size = 14),
        axis.title.y = element_text(size = 14)) +
  theme(legend.position = "none") +
  scale_y_continuous(expand = c(0, 0), limits = c(-11,5)) +
  guides(linetype = FALSE) +
  panel_border(color = "black")

p11_16_d1
```

**Library 16 - Day 2**

Filter the Perfects counts to include only those perfects with > 200 BCs, then select only the dose responses for trimethoprim concentrations on Day 2.
```{r}
perfects16_200BCs_dr2 <- perfects16_5BCs %>%
  filter(numprunedBCs>200) %>%
  select(ID,
         fitF02D04,
         fitF03D04,
         fitF04D04,
         fitF05D04,
         fitF06D04,
         fitF07D04,
         fitF08D04,
         numprunedBCs) %>%
  gather("selection","fitness",-ID,-numprunedBCs)
```

Re-Define the trimethoprim concentrations based on the fitness codes associated with sample IDs.
```{r}
perfects16_200BCs_dr2$tmpfactor <- "0"
perfects16_200BCs_dr2$tmpfactor[perfects16_200BCs_dr2$selection== "fitF02D04"] <- "0"
perfects16_200BCs_dr2$tmpfactor[perfects16_200BCs_dr2$selection== "fitF03D04"] <- "0.058"
perfects16_200BCs_dr2$tmpfactor[perfects16_200BCs_dr2$selection== "fitF04D04"] <- "0.5"
perfects16_200BCs_dr2$tmpfactor[perfects16_200BCs_dr2$selection== "fitF05D04"] <- "1"
perfects16_200BCs_dr2$tmpfactor[perfects16_200BCs_dr2$selection== "fitF06D04"] <- "10"
perfects16_200BCs_dr2$tmpfactor[perfects16_200BCs_dr2$selection== "fitF07D04"] <- "50"
perfects16_200BCs_dr2$tmpfactor[perfects16_200BCs_dr2$selection== "fitF08D04"] <- "200"

perfects16_200BCs_dr2$tmpname <- "0"
perfects16_200BCs_dr2$tmpname[perfects16_200BCs_dr2$selection== "fitF02D04"] <- "A"
perfects16_200BCs_dr2$tmpname[perfects16_200BCs_dr2$selection== "fitF03D04"] <- "B"
perfects16_200BCs_dr2$tmpname[perfects16_200BCs_dr2$selection== "fitF04D04"] <- "C"
perfects16_200BCs_dr2$tmpname[perfects16_200BCs_dr2$selection== "fitF05D04"] <- "D"
perfects16_200BCs_dr2$tmpname[perfects16_200BCs_dr2$selection== "fitF06D04"] <- "E"
perfects16_200BCs_dr2$tmpname[perfects16_200BCs_dr2$selection== "fitF07D04"] <- "F"
perfects16_200BCs_dr2$tmpname[perfects16_200BCs_dr2$selection== "fitF08D04"] <- "G"

perfects16_200BCs_dr_labels <- c("0","0.058","0.5","1","10","50","200")
```

Plot the dose response for remaining perfects (>200 BCs) as boxplots for each trimethoprim concentration.
```{r}
p11_16_d2 <- ggplot(perfects16_200BCs_dr2, aes(x=tmpname, y=fitness)) +
  geom_boxplot(color="black", fill="aquamarine4") +
  xlab("[Trimethoprim] (ug/mL)") +
  ylab("Fitness") +
  ggtitle("Library 16 Dose Response - Day 2") +
  scale_x_discrete(labels = perfects16_200BCs_dr_labels) +
  theme(axis.text.x = element_text(size = 12),
        axis.text.y = element_text(size = 12),
        axis.title.x = element_text(size = 14),
        axis.title.y = element_text(size = 14)) +
  theme(legend.position = "none") +
  scale_y_continuous(expand = c(0, 0), limits = c(-11,5)) +
  guides(linetype = FALSE) +
  panel_border(color = "black")

p11_16_d2
```

**Library 16 - Day 3**

Filter the Perfects counts to include only those perfects with > 200 BCs, then select only the dose responses for trimethoprim concentrations on Day 3.
```{r}
perfects16_200BCs_dr3 <- perfects16_5BCs %>%
  filter(numprunedBCs>200) %>%
  select(ID,
         fitG04D04,
         fitG05D04,
         fitG06D04,
         fitG07D04,
         fitG08D04,
         fitG09D04,
         fitG10D04,
         numprunedBCs) %>%
  gather("selection","fitness",-ID,-numprunedBCs)
```

Re-Define the trimethoprim concentrations based on the fitness codes associated with sample IDs.
```{r}
perfects16_200BCs_dr3$tmpfactor <- "0"
perfects16_200BCs_dr3$tmpfactor[perfects16_200BCs_dr3$selection== "fitG04D04"] <- "0"
perfects16_200BCs_dr3$tmpfactor[perfects16_200BCs_dr3$selection== "fitG05D04"] <- "0.058"
perfects16_200BCs_dr3$tmpfactor[perfects16_200BCs_dr3$selection== "fitG06D04"] <- "0.5"
perfects16_200BCs_dr3$tmpfactor[perfects16_200BCs_dr3$selection== "fitG07D04"] <- "1"
perfects16_200BCs_dr3$tmpfactor[perfects16_200BCs_dr3$selection== "fitG08D04"] <- "10"
perfects16_200BCs_dr3$tmpfactor[perfects16_200BCs_dr3$selection== "fitG09D04"] <- "50"
perfects16_200BCs_dr3$tmpfactor[perfects16_200BCs_dr3$selection== "fitG10D04"] <- "200"

perfects16_200BCs_dr3$tmpname <- "0"
perfects16_200BCs_dr3$tmpname[perfects16_200BCs_dr3$selection== "fitG04D04"] <- "A"
perfects16_200BCs_dr3$tmpname[perfects16_200BCs_dr3$selection== "fitG05D04"] <- "B"
perfects16_200BCs_dr3$tmpname[perfects16_200BCs_dr3$selection== "fitG06D04"] <- "C"
perfects16_200BCs_dr3$tmpname[perfects16_200BCs_dr3$selection== "fitG07D04"] <- "D"
perfects16_200BCs_dr3$tmpname[perfects16_200BCs_dr3$selection== "fitG08D04"] <- "E"
perfects16_200BCs_dr3$tmpname[perfects16_200BCs_dr3$selection== "fitG09D04"] <- "F"
perfects16_200BCs_dr3$tmpname[perfects16_200BCs_dr3$selection== "fitG10D04"] <- "G"

perfects16_200BCs_dr_labels <- c("0","0.058","0.5","1","10","50","200")
```

Plot the dose response for remaining perfects (>200 BCs) as boxplots for each trimethoprim concentration.
```{r}
p11_16_d3 <- ggplot(perfects16_200BCs_dr3, aes(x=tmpname, y=fitness)) +
  geom_boxplot(color="black", fill="aquamarine4") +
  xlab("[Trimethoprim] (ug/mL)") +
  ylab("Fitness") +
  ggtitle("Library 16 Dose Response - Day 3") +
  scale_x_discrete(labels = perfects16_200BCs_dr_labels) +
  theme(axis.text.x = element_text(size = 12),
        axis.text.y = element_text(size = 12),
        axis.title.x = element_text(size = 14),
        axis.title.y = element_text(size = 14)) +
  theme(legend.position = "none") +
  scale_y_continuous(expand = c(0, 0), limits = c(-11,5)) +
  guides(linetype = FALSE) +
  panel_border(color = "black")

p11_16_d3
```

**Library 16 - Day 4**

Filter the Perfects counts to include only those perfects with > 200 BCs, then select only the dose responses for trimethoprim concentrations on Day 4.
```{r}
perfects16_200BCs_dr4 <- perfects16_5BCs %>%
  filter(numprunedBCs>200) %>%
  select(ID,
         fitH06D04,
         fitH07D04,
         fitH08D04,
         fitH09D04,
         fitH10D04,
         fitH11D04,
         fitH12D04,
         numprunedBCs) %>%
  gather("selection","fitness",-ID,-numprunedBCs)
```

Re-Define the trimethoprim concentrations based on the fitness codes associated with sample IDs.
```{r}
perfects16_200BCs_dr4$tmpfactor <- "0"
perfects16_200BCs_dr4$tmpfactor[perfects16_200BCs_dr4$selection== "fitH06D04"] <- "0"
perfects16_200BCs_dr4$tmpfactor[perfects16_200BCs_dr4$selection== "fitH07D04"] <- "0.058"
perfects16_200BCs_dr4$tmpfactor[perfects16_200BCs_dr4$selection== "fitH08D04"] <- "0.5"
perfects16_200BCs_dr4$tmpfactor[perfects16_200BCs_dr4$selection== "fitH09D04"] <- "1"
perfects16_200BCs_dr4$tmpfactor[perfects16_200BCs_dr4$selection== "fitH10D04"] <- "10"
perfects16_200BCs_dr4$tmpfactor[perfects16_200BCs_dr4$selection== "fitH11D04"] <- "50"
perfects16_200BCs_dr4$tmpfactor[perfects16_200BCs_dr4$selection== "fitH12D04"] <- "200"

perfects16_200BCs_dr4$tmpname <- "0"
perfects16_200BCs_dr4$tmpname[perfects16_200BCs_dr4$selection== "fitH06D04"] <- "A"
perfects16_200BCs_dr4$tmpname[perfects16_200BCs_dr4$selection== "fitH07D04"] <- "B"
perfects16_200BCs_dr4$tmpname[perfects16_200BCs_dr4$selection== "fitH08D04"] <- "C"
perfects16_200BCs_dr4$tmpname[perfects16_200BCs_dr4$selection== "fitH09D04"] <- "D"
perfects16_200BCs_dr4$tmpname[perfects16_200BCs_dr4$selection== "fitH10D04"] <- "E"
perfects16_200BCs_dr4$tmpname[perfects16_200BCs_dr4$selection== "fitH11D04"] <- "F"
perfects16_200BCs_dr4$tmpname[perfects16_200BCs_dr4$selection== "fitH12D04"] <- "G"

perfects16_200BCs_dr_labels <- c("0","0.058","0.5","1","10","50","200")
```

Plot the dose response for remaining perfects (>200 BCs) as boxplots for each trimethoprim concentration.
```{r}
p11_16_d4 <- ggplot(perfects16_200BCs_dr4, aes(x=tmpname, y=fitness)) +
  geom_boxplot(color="black", fill="aquamarine4") +
  xlab("[Trimethoprim] (ug/mL)") +
  ylab("Fitness") +
  ggtitle("Library 16 Dose Response - Day 4") +
  scale_x_discrete(labels = perfects16_200BCs_dr_labels) +
  theme(axis.text.x = element_text(size = 12),
        axis.text.y = element_text(size = 12),
        axis.title.x = element_text(size = 14),
        axis.title.y = element_text(size = 14)) +
  theme(legend.position = "none") +
  scale_y_continuous(expand = c(0, 0), limits = c(-11,5)) +
  guides(linetype = FALSE) +
  panel_border(color = "black")

p11_16_d4
```

```{r echo=FALSE}
patch11_Lib16 <- (p11_16_d1 | p11_16_d2 | p11_16_d3 | p11_16_d4)
patch11_Lib16
```

```{r echo=FALSE}
ggsave(file="plots/combined/lib16.dose.response.by.day.png", plot=patch11_Lib16,
       dpi=600, width = 14, height = 6, units = "in")
```

```{r echo=FALSE}
patch11_Lib15.16 <- (p11_15_d1 | p11_15_d2 | p11_15_d3 | p11_15_d4) / 
  (p11_16_d1 | p11_16_d2 | p11_16_d3 | p11_16_d4)
patch11_Lib15.16
```

```{r echo=FALSE}
ggsave(file="plots/combined/lib15.16.dose.response.by.day.png", plot=patch11_Lib15.16,
       dpi=600, width = 14, height = 12, units = "in")
```

### Perfects Fitness Counts

**Library 15**

Plot the Perfects counts based on fitness values between D11 vs. D03 conditions (>5 BCs count):
```{r}
p18_15 <- ggplot(perfects15_tree_5BCs,aes(x=fitD11D03)) +
  geom_histogram(binwidth=0.08, fill = "aquamarine4", color = "black") +
  xlab("Fitness (Log2 Fold Change)") +
  ylab("Counts") +
  ggtitle("Library 15 Perfects Fitness Counts (D11 vs. D03)") +
  #theme_cowplot() +
  #background_grid() +
  theme(axis.text.x = element_text(size = 12),
        axis.text.y = element_text(size = 12),
        axis.title.x = element_text(size = 14),
        axis.title.y = element_text(size = 14)) +
  theme(legend.position = "none") +
  scale_y_continuous(expand = c(0, 0), limits = c(0,30)) +
  guides(linetype = FALSE) + panel_border(color = "black")

p18_15
```

```{r echo=FALSE}
ggsave(file="plots/hist_Lib15_5BCs_fitD11D03.png", plot=p18_15,
       dpi=600, width = 8, height = 6, units = "in")
```


**Library 16**

Plot the Perfects counts based on fitness values between E06 vs. D04 conditions (>5 BCs count):
```{r}
p18_16 <- ggplot(perfects16_tree_5BCs,aes(x=fitE06D04)) +
  geom_histogram(binwidth=0.08, fill = "aquamarine4", color = "black") +
  xlab("Fitness (Log2 Fold Change)") +
  ylab("Counts") +
  ggtitle("Library 16 Perfects Fitness Counts (E06 vs. D04)") +
  #theme_cowplot() +
  #background_grid() +
  theme(axis.text.x = element_text(size = 12),
        axis.text.y = element_text(size = 12),
        axis.title.x = element_text(size = 14),
        axis.title.y = element_text(size = 14)) +
  theme(legend.position = "none") +
  scale_y_continuous(expand = c(0, 0), limits = c(0,30)) +
  guides(linetype = FALSE) + panel_border(color = "black")

p18_16
```

```{r echo=FALSE}
ggsave(file="plots/hist_Lib16_5BCs_fitD11D03.png", plot=p18_16,
       dpi=600, width = 8, height = 6, units = "in")
```

```{r}
patch13 <- (p18_15 | p18_16)
patch13
```

```{r echo=FALSE}
ggsave(file="plots/combined/hist_Lib15.16_5BCs.png", plot=patch13,
       dpi=600, width = 12, height = 8, units = "in")
```

## Phylogenetic Trees

**This section uses the `library(ggtree)` and `library(castor)` packages.**

### Tree File
Import the tree dataset containing all designed constructs for Library 15:
```{r}
Alltree15 <- read.tree("OUTPUT/perfects_tree_Lib15_5BCs_aligned.tree")   # newick format
Alltree15
```

The following code adds the branch length distance from E. coli for each unique homolog to the "perfects_tree_5BCs" dataframe based on values in "Alltree".
```{r}
perfects15_tree_5BCs$distances = get_pairwise_distances(Alltree15, 
                                   perfects15_tree_5BCs$ID,
                                   rep(perfects15_tree_5BCs$ID[1], length(perfects15_tree_5BCs$ID)))
```

Plot fitness relative to phylogenetic distance to E. coli DHFR gene:
```{r}
p12_15 <- ggplot(perfects15_tree_5BCs,aes(x=distances, y=fitD05D03)) +
  geom_point(color = "aquamarine4") +
  xlab("Phylogenetic Distance to E. coli DHFR") +
  ylab("Fitness (Log2 Fold Change)") +
  ggtitle("Library 15 Phylogenetic Distance from E. coli") +
  #theme_cowplot() +
  #background_grid() +
  theme(axis.text.x = element_text(size = 12),
        axis.text.y = element_text(size = 12),
        axis.title.x = element_text(size = 14),
        axis.title.y = element_text(size = 14)) +
  theme(legend.position = "none") +
  scale_y_continuous(expand = c(0, 0), limits = c(-5,2)) +
  guides(linetype = FALSE) + panel_border(color = "black")

p12_15
```

```{r echo=FALSE}
ggsave(file="plots/dhfr.Lib15.phylo.fitness.png", plot=p12_15,
       dpi=600, width = 8, height = 6, units = "in")
```

### Tree Plots

#### DHFR Phylogenetic Diversity
Plot a circular phylogenetic tree with color-scale based on phylogenetic distance between each DHFR homolog relative to the E. coli DHFR gene:
```{r}
p13_15 <- ggtree(Alltree15, layout="circular", branch.length="none") %<+% 
  perfects15_tree_5BCs + 
  aes(color=PctIdentEcoli*100) +
  geom_tippoint(aes(color=PctIdentEcoli*100), size=1, alpha=0.8) +
  geom_tiplab2(aes(color=PctIdentEcoli*100, label=Source, angle=angle), size=1) +
  theme(legend.position="left") + 
  scale_colour_gradient2("Sequence\nIdentity (%)", low="blue", mid="green", high="red",
                         midpoint=(100+min(perfects15_tree_5BCs$PctIdentEcoli)*100)/2)

p13_15
```

```{r echo=FALSE}
ggsave(file="plots/trees/FastTree_Lib15_5BCs_PctIdentEcolinol.png", plot=p13_15,
       dpi=600, width = 12, height = 7.5, units = "in")
```

**Library 15**

#### Complementation Tree
Show the minimum and maximum fitness values between D05 (M9 no supp) vs. D03 (M9 full supp) dataset:
```{r class.output="goodCode"}
min(perfects15_tree_5BCs$fitD05D03)
max(perfects15_tree_5BCs$fitD05D03)
```

Plot a circular phylogenetic tree for complementation with color-scale based on fitness values between D05 vs. D03 for each DHFR homolog:

```{r}
#shade circular tree by fitness:
p14_15 <- ggtree(Alltree15, layout="circular", branch.length="none") %<+%
  perfects15_tree_5BCs +
  aes(color=fitD05D03) +
  geom_tippoint(aes(color=fitD05D03), size=1, alpha=0.8) +
  geom_tiplab2(aes(color=fitD05D03, label=Source, angle=angle), size=1) +
  theme(legend.position="left") + 
  ggtitle("Library 15 Complementation (D05 vs D03)") +
  scale_colour_gradient2("fitness",
                         low="black", 
                         mid="gold",
                         high="red",
                         na.value="gray",
                         limit = c(-5,5))

p14_15
```

```{r echo=FALSE, eval=FALSE}
ggsave(file="plots/trees/FastTree_Lib15_5BCs_fitD05D03.COMPLEMENTATION.png", plot=p14_15,
       dpi=600, width = 12, height = 7.5, units = "in")
```

Another tree version that uses the homolog IDs instead of the organism nomenclature to standardized label sizes around the tree. There's also no shading of branches to match fitness scores (no "aes()" settings):
```{r}
p15_15 <- ggtree(Alltree15, layout="circular", branch.length="none") %<+% 
  perfects15_tree_5BCs + 
  geom_tippoint(aes(color=fitD05D03), alpha=0.25) +
  geom_tiplab2(size=1, aes(color=fitD05D03,angle=angle)) +
  geom_text2(aes(subset=!isTip, label=node), size=1, color="red",hjust=-.3) +
  theme(legend.position="left") + 
  scale_colour_gradient2("fitness",
                         low="black", 
                         mid="gold",
                         high="red",
                         na.value="gray",
                         limit = c(-3,1))

p15_15
```

```{r echo=FALSE}
ggsave(file="plots/trees/FastTree_Lib15_5BCs_fitD05D03_ID_and_nodes.COMPLEMENTATION.png", plot=p15_15,
       dpi=600, width = 12, height = 7.5, units = "in")
```

Plot a circular phylogenetic tree with color-scale based on fitness values between D05 vs. D03 for each DHFR homolog. **This version uses the true branch length for plotting:**
```{r}
p16_15 <- ggtree(Alltree15, layout="circular") %<+%
  perfects15_tree_5BCs + 
  aes(color=fitD05D03) +
  geom_tippoint(aes(color=fitD05D03), size=1, alpha=0.8) +
# geom_tiplab2(aes(color=fitD05D03, label=Source, angle=angle), offset=0.1, size=1) +
  theme(legend.position="left") + 
  scale_colour_gradient2("fitness",
                         low="black", 
                         mid="gold",
                         high="red",
                         na.value="gray",
                         limit = c(-3,1))

p16_15
```

```{r echo=FALSE}
ggsave(file="plots/trees/FastTree_Lib15_5BCs_fitD05D03_truelength.COMPLEMENTATION.png", plot=p16_15,
       dpi=600, width = 12, height = 7.5, units = "in")
```

#### D11 vs. D03 Tree Plot
Show the minimum and maximum fitness values between D11 (200 mg/uL Day 1) vs. D03 (M9 no supp.) in the dataset:
```{r class.output="goodCode"}
min(perfects15_tree_5BCs$fitD11D03)
max(perfects15_tree_5BCs$fitD11D03)
```

Plot a circular phylogenetic tree with color-scale based on fitness values between D11 vs. D03 for each DHFR homolog:
```{r}
#shade circular tree by fitness:
p17_15 <- ggtree(Alltree15, layout="circular", branch.length="none") %<+% #, branch.length="none"
  perfects15_tree_5BCs +
  aes(color=fitD11D03) +
  geom_tippoint(aes(color=fitD11D03), size=1, alpha=0.8) +
  geom_tiplab2(aes(color=fitD11D03, label=Source, angle=angle), size=1) +
  theme(legend.position="left") + 
  ggtitle("Library 15 Fitness at 200 mg/uL Trimethoprim (D11 vs D03)") +
  scale_colour_gradient2("fitness",
                         low="black", 
                         mid="gold",
                         high="red",
                         na.value="gray",
                         limit = c(-10,2)) 

p17_15
```

```{r echo=FALSE}
ggsave(file="plots/trees/FastTree_Lib15_5BCs_fitD11D03.png", plot=p17_15,
       dpi=600, width = 12, height = 7.5, units = "in")
```

```{r}
patch12<- (p14_15 | p17_15)
patch12
```

```{r echo=FALSE}
ggsave(file="plots/trees/FastTree.Lib15.Complementation.Fitness.200mg.png", plot=patch12,
       dpi=600, width = 14, height = 12, units = "in")
```

# Fitness vs Distance

<font color="blue">**This section is based on the R file: "R_change_in_fitness_vs_distance.R".**</font> It determines how fitness changes with mutation distance.

## Filter Data

Grab only the BCs with up to 5 mutations (also need to ensure it has greater than 0 mutations since some BCs have negative values):
```{r}
BCs5_15 <- BCmut_15 %>%
  filter(mutations >= 0 & mutations <= 5) %>%
  left_join(mutIDinfo15 %>% select(mutID), by="mutID") %>%
  select(BC,IDalign,mutID,mutations,D05D03fc)
```

Retain only the homologs with good data:
```{r}
fitness_distance_15 <- perfects15_tree_5BCs %>%
  select(ID,fitD05D03) %>%
  dplyr::rename(IDalign=ID)
```

Determine median and sd for 1 mutation:
```{r}
fitness_distance_15 <- BCs5_15 %>%
    filter(mutations==1) %>%
    group_by(IDalign) %>%
    summarise(mut1fit=median(D05D03fc),
              mut1sd=sd(D05D03fc),
              num1points=n()) %>%
    right_join(fitness_distance_15,by="IDalign")
```

Determine median and sd for 2 mutation:
```{r}
fitness_distance_15 <-  BCs5_15 %>%
  filter(mutations==2) %>%
  group_by(IDalign) %>%
  summarise(mut2fit=median(D05D03fc),
            mut2sd=sd(D05D03fc),
            num2points=n()) %>%
  right_join(fitness_distance_15,by="IDalign") 
```

Determine median and sd for 3 mutations:
```{r}
fitness_distance_15 <-  BCs5_15 %>%
  filter(mutations==3) %>%
  group_by(IDalign) %>%
  summarise(mut3fit=median(D05D03fc),
            mut3sd=sd(D05D03fc),
            num3points=n()) %>%
  right_join(fitness_distance_15,by="IDalign")
```

Determine median and sd for 4 mutations:
```{r}
fitness_distance_15 <-  BCs5_15 %>%
  filter(mutations==4) %>%
  group_by(IDalign) %>%
  summarise(mut4fit=median(D05D03fc),
            mut4sd=sd(D05D03fc),
            num4points=n()) %>%
  right_join(fitness_distance_15,by="IDalign")
```

Determine median and sd for 5 mutations:
```{r}
fitness_distance_15 <-  BCs5_15 %>%
  filter(mutations==5) %>%
  group_by(IDalign) %>%
  summarise(mut5fit=median(D05D03fc),
            mut5sd=sd(D05D03fc),
            num5points=n()) %>%
  right_join(fitness_distance_15,by="IDalign")
```

Determine change in fitness:
```{r}
fitness_distance_nu_15 <- fitness_distance_15 %>%
  mutate(mut1fitn=(mut1fit-fitD05D03),
         mut2fitn=(mut2fit-fitD05D03),
         mut3fitn=(mut3fit-fitD05D03),
         mut4fitn=(mut4fit-fitD05D03),
         mut5fitn=(mut5fit-fitD05D03),
         mut0fitn=0)
```

Melt data on number of mutations:
```{r}
fitness_distance_m_15 <- fitness_distance_nu_15 %>%
  select(IDalign,fitD05D03,mut0fitn,mut1fitn,mut2fitn,mut3fitn,mut4fitn,mut5fitn) %>%
  gather(mutations,fitness,mut0fitn,mut1fitn,mut2fitn,mut3fitn,mut4fitn,mut5fitn)
```

Replace names with numbers:
```{r}
fitness_distance_m_15$mutations[which(fitness_distance_m_15$mutations=="mut0fitn")] <- as.numeric(0)
fitness_distance_m_15$mutations[which(fitness_distance_m_15$mutations=="mut1fitn")] <- as.numeric(1)
fitness_distance_m_15$mutations[which(fitness_distance_m_15$mutations=="mut2fitn")] <- as.numeric(2)
fitness_distance_m_15$mutations[which(fitness_distance_m_15$mutations=="mut3fitn")] <- as.numeric(3)
fitness_distance_m_15$mutations[which(fitness_distance_m_15$mutations=="mut4fitn")] <- as.numeric(4)
fitness_distance_m_15$mutations[which(fitness_distance_m_15$mutations=="mut5fitn")] <- as.numeric(5)
```

Remove those with NA fitness:
```{r}
fitness_distance_m_15 <- fitness_distance_m_15 %>%
  filter(!is.na(fitness))
```

## Fitness vs. Distance Plots

The first plot version uses traces to display results:
```{r}
p19_15 <- ggplot(fitness_distance_m_15, aes(x=mutations, y=fitness, group=IDalign, color=IDalign)) +
  geom_point() +
  geom_line() +
  xlab("Distance from homolog (a.a.)") +
  ylab("Change in fitness relative to homolog") +
  #theme_cowplot() +
  #background_grid() +
  theme(axis.text.x = element_text(size = 12),
        axis.text.y = element_text(size = 12),
        axis.title.x = element_text(size = 14),
        axis.title.y = element_text(size = 14)) +
  theme(legend.position = "none") +
  scale_y_continuous(expand = c(0, 0), limits = c(-10,10)) +
  guides(linetype = FALSE) + panel_border(color = "black")

p19_15
```

```{r echo=FALSE, eval=FALSE}
ggsave(file="plots/fitness_vs_distance_all_traces.Lib15.png", plot=p19_15,
       dpi=600, width = 8, height = 6, units = "in")
```

The second plot version uses a boxplot to display results:
```{r}
p20_15 <- ggplot(fitness_distance_m_15, aes(x=mutations, y=fitness)) +
  geom_boxplot(color="black", fill="aquamarine4", alpha=0.8) +
  xlab("Distance from homolog (a.a.)") +
  ylab("Change in fitness relative to homolog") +
  #theme_cowplot() +
  #background_grid() +
  theme(axis.text.x = element_text(size = 12),
        axis.text.y = element_text(size = 12),
        axis.title.x = element_text(size = 14),
        axis.title.y = element_text(size = 14)) +
  theme(legend.position = "none") +
  scale_y_continuous(expand = c(0, 0), limits = c(-10,10)) +
  guides(linetype = FALSE) + panel_border(color = "black")

p20_15
```

```{r echo=FALSE}
ggsave(file="plots/fitness_vs_distance_boxplot.Lib15.png", plot=p20_15,
       dpi=600, width = 8, height = 6, units = "in")
```

Calculate Spearman correlations between fitness and distance from homolog:
```{r, warning=FALSE, class.output="goodCode"}
# Calculate Spearman coefficient:
cor(as.numeric(fitness_distance_m_15$mutations),fitness_distance_m_15$fitness,
    method=c("spearman"))

# Run correlation test:
cor.test(as.numeric(fitness_distance_m_15$mutations),fitness_distance_m_15$fitness,
         method=c("spearman"))
```

# Plot Mutants
<font color="blue">**This section is based on the R file: "R_plot_all_mutants.R".**</font> It describes how to plot all mutants per homolog independently.

Determine the number of mutants per homolog:
```{r}
# Lib15
mutantsperhomolog15 <- mutIDinfo15 %>%
  filter(mutations != 0) %>%
  select(mutID,IDalign) %>%
  distinct() %>%
  group_by(IDalign) %>%
  summarise(count=n())

# Lib16
mutantsperhomolog16 <- mutIDinfo16 %>%
  filter(mutations != 0) %>%
  select(mutID,IDalign) %>%
  distinct() %>%
  group_by(IDalign) %>%
  summarise(count=n())
```

Add a column and label each ID for the library it comes from to keep track of the data source:
```{r}
# Lib15
mutantsperhomolog15$lib <- "Lib15"

# Lib16
mutantsperhomolog16$lib <- "Lib16"
```

Plot the mutant counts for Lib. 15:
```{r}
p21_15 <- ggplot(mutantsperhomolog15, aes(x=lib, y=count)) +
  geom_violin(color="black", fill="aquamarine4") +
  #geom_boxplot(color="black", fill="aquamarine4", alpha=0.8) +
  xlab("") +
  ylab("Mutants per homolog") +
  #theme_cowplot() +
  #background_grid() +
  theme(axis.text.x = element_text(size = 12),
        axis.text.y = element_text(size = 12),
        axis.title.x = element_text(size = 14),
        axis.title.y = element_text(size = 14)) +
  theme(legend.position = "none") +
  scale_y_continuous(expand = c(0, 0), limits = c(0,4000)) +
  guides(linetype = FALSE) +
  panel_border(color = "black")
  
p21_15
```

Plot the mutant counts for Lib. 16:
```{r}
p21_16 <- ggplot(mutantsperhomolog16, aes(x=lib, y=count)) +
  geom_violin(color="black", fill="aquamarine4") +
  #geom_boxplot(color="black", fill="aquamarine4", alpha=0.8) +
  xlab("") +
  ylab("Mutants per homolog") +
  #theme_cowplot() +
  #background_grid() +
  theme(axis.text.x = element_text(size = 12),
        axis.text.y = element_text(size = 12),
        axis.title.x = element_text(size = 14),
        axis.title.y = element_text(size = 14)) +
  theme(legend.position = "none") +
  scale_y_continuous(expand = c(0, 0), limits = c(0,4000)) +
  guides(linetype = FALSE) +
  panel_border(color = "black")
  
p21_16
```

```{r}
patch14<- (p21_15 | p21_16)
patch14
```

```{r echo=FALSE}
ggsave(file="plots/combined/lib15.16.mutants.per.homolog.violin.png", plot=patch14,
       dpi=600, width = 14, height = 12, units = "in")
```

Calculate the mean and median mutant counts for each distinct homolog:
```{r class.output="goodCode"}
#Lib15

#Mean mutant count per homolog
mean(mutantsperhomolog15$count)

#Median mutant count per homolog
median(mutantsperhomolog15$count)

#Lib16

#Mean mutant count per homolog
mean(mutantsperhomolog16$count)

#Median mutant count per homolog
median(mutantsperhomolog16$count)
```

<font color="green">If both measures (mean and median) are considerably different, this indicates that the data are skewed (i.e. they are far from being normally distributed) and the **MEDIAN** generally gives a more appropriate idea of the data distribution.</font>

## Top 10 Mutants

Arrange by biotin enrich score
```{r}
#Lib15
mutantsperhomolog15 <- mutantsperhomolog15 %>%
  arrange(-count)

#Lib16
mutantsperhomolog16 <- mutantsperhomolog16 %>%
  arrange(-count)
```

<font color="green">**Select the top 10 mutants based on greatest count values**</font>
```{r echo=FALSE}
#Lib15

# Display dataframe with kable formatting
mutantsperhomolog15_10.count <- head(mutantsperhomolog15, 10)

knitr::kable(mutantsperhomolog15_10.count,
  col.names = c('ID Align', 'Count', 'Library'),
  align = "lll",
  format.args = list(big.mark = ","))

#Lib16

# Display dataframe with kable formatting
mutantsperhomolog16_10.count <- head(mutantsperhomolog16, 10)

knitr::kable(mutantsperhomolog16_10.count,
  col.names = c('ID Align', 'Count', 'Library'),
  align = "lll",
  format.args = list(big.mark = ","))
```

Make keys for plotting:
```{r}
#Lib15

mutantsperhomolog15$key <- 1:length(mutantsperhomolog15$count)

mutantsperhomolog15_10 <- mutantsperhomolog15 %>%
  filter(key<11)

mutantsdist15 <- mutIDinfo15 %>%
  filter(IDalign %in% mutantsperhomolog15_10$IDalign) %>%
  filter(mutations>-1)

#Lib16

mutantsperhomolog16$key <- 1:length(mutantsperhomolog16$count)

mutantsperhomolog16_10 <- mutantsperhomolog16 %>%
  filter(key<11)

mutantsdist16 <- mutIDinfo16 %>%
  filter(IDalign %in% mutantsperhomolog16_10$IDalign) %>%
  filter(mutations>-1)
```

Plot the top 10 mutants with mean and SD based on total counts:
```{r}
p22_15 <- ggplot(mutantsdist15, aes(x=IDalign, y=mutations))+
  geom_boxplot(color="black", fill="aquamarine4") +
  ggtitle("Library 15") +
  xlab("") +
  ylab("Distribution of Mutants at Distance (a.a.)") +
  coord_flip() +
  #theme_cowplot() +
  #background_grid() +
  theme(axis.text.x = element_text(size = 12),
        axis.text.y = element_text(size = 12),
        axis.title.x = element_text(size = 14),
        axis.title.y = element_text(size = 14)) +
  theme(legend.position = "none") +
  #scale_y_continuous(expand = c(0, 0), limits = c(0,180)) +
  guides(linetype = FALSE) +
  panel_border(color = "black")

p22_15
```

```{r}
p22_16 <- ggplot(mutantsdist16, aes(x=IDalign, y=mutations))+
  geom_boxplot(color="black", fill="aquamarine4") +
  ggtitle("Library 16") +
  xlab("") +
  ylab("Distribution of Mutants at Distance (a.a.)") +
  coord_flip() +
  #theme_cowplot() +
  #background_grid() +
  theme(axis.text.x = element_text(size = 12),
        axis.text.y = element_text(size = 12),
        axis.title.x = element_text(size = 14),
        axis.title.y = element_text(size = 14)) +
  theme(legend.position = "none") +
  #scale_y_continuous(expand = c(0, 0), limits = c(0,180)) +
  guides(linetype = FALSE) +
  panel_border(color = "black")

p22_16
```

```{r}
patch15<- (p22_15 | p22_16)
patch15
```

```{r echo=FALSE}
ggsave(file="plots/combined/lib15.16.mutants.at.distance.top10.png", plot=patch15,
       dpi=600, width = 14, height = 12, units = "in")
```

Summarize the total number of reads associated with all BCs:

<font color="red">**Does this value include the original designed homologs AND all mutants, or just the total number of mutants?**</font>
```{r class.output="goodCode"}
#Lib15
BCinfo.read.count15 <- sum(BCinfo15$reads)

format(BCinfo.read.count15, big.mark = ",")

#Lib16
BCinfo.read.count16 <- sum(BCinfo16$reads)

format(BCinfo.read.count16, big.mark = ",")
```

## Mutant Fitness

Calculate fitness values for all mutants independently:
```{r}
#Lib15

BCprune15 <- BCprune15 %>%
  left_join(BCinfo15, by="BC")

mutants15 <- mutIDinfo15_1BC

mutants15 <- BCprune15 %>%
  group_by(mutID) %>%
  summarise(fitD05D03=median(D05D03fc),
            fitD06D03=median(D06D03fc),
            fitD07D03=median(D07D03fc),
            fitD08D03=median(D08D03fc),
            fitD09D03=median(D09D03fc),
            fitD10D03=median(D10D03fc),
            fitD11D03=median(D11D03fc)) %>%
  right_join(mutants15, by="mutID")
```

```{r}
#Lib16

BCprune16 <- BCprune16 %>%
  left_join(BCinfo16, by="BC")

mutants16 <- mutIDinfo16_1BC

mutants16 <- BCprune16 %>%
  group_by(mutID) %>%
  summarise(fitD12D04=median(D12D04fc),
            fitE01D04=median(E01D04fc),
            fitE02D04=median(E02D04fc),
            fitE03D04=median(E03D04fc),
            fitE04D04=median(E04D04fc),
            fitE05D04=median(E05D04fc),
            fitE06D04=median(E06D04fc)) %>%
  right_join(mutants16, by="mutID")
```

Bin mutants by their percent similarity to designed homologs:
```{r}
#Lib15
mutants15$identbins <- cut(mutants15$pct_ident,
                         breaks = seq(0,1.005,1/100),
                         labels=as.character(seq(0.005,0.995,1/100)))

#Lib16
mutants16$identbins <- cut(mutants16$pct_ident,
                         breaks = seq(0,1.005,1/100),
                         labels=as.character(seq(0.005,0.995,1/100)))
```

Determine the **minimum** percent similarity mutant in the dataset (most different from designed homolog):
```{r class.output="goodCode"}
#Lib15
min(mutants15$pct_ident)

#Lib16
min(mutants16$pct_ident)
```

Determine the **maximum** percent similarity mutant in the dataset (most similar to designed homolog):
```{r class.output="goodCode"}
#Lib15
max(mutants15$pct_ident)

#Lib16
max(mutants16$pct_ident)
```

Calculate the total number of mutants with at least 2 barcodes recovered:
```{r class.output="goodCode"}
#Lib15
mut15.2BCs.count <- nrow(mutants15 %>% filter(numprunedBCs >= 2))
format(mut15.2BCs.count, big.mark = ",")

#Lib16
mut16.2BCs.count <- nrow(mutants16 %>% filter(numprunedBCs >= 2))
format(mut16.2BCs.count, big.mark = ",")
```

### Histogram Plots

Plot these mutants (>2 BCs recovered) with histograms:
```{r warning=FALSE}
#Lib15
p23_15 <- mutants15 %>%
  filter(numprunedBCs >= 2) %>%
  ggplot(aes(x=pct_ident, y=fitD05D03)) +
  labs(x = "Fractional Sequence Identity", y ="Fitness",color="") +
  geom_point(alpha=0.3,color='aquamarine4') +
  #theme_cowplot() +
  #background_grid() +
  ggtitle("Library 15") +
  theme(axis.text.x = element_text(size = 12),
        axis.text.y = element_text(size = 12),
        axis.title.x = element_text(size = 14),
        axis.title.y = element_text(size = 14)) +
  theme(legend.position = "none") +
  #scale_y_continuous(expand = c(0, 0), limits = c(0,180)) +
  guides(linetype = FALSE) +
  panel_border(color = "black")

p24_15 <- ggMarginal(p23_15, type = "histogram", fill = "aquamarine4", bins=40)
p24_15

#Lib16
p23_16 <- mutants16 %>%
  filter(numprunedBCs >= 2) %>%
  ggplot(aes(x=pct_ident, y=fitD12D04)) +
  labs(x = "Fractional Sequence Identity", y ="Fitness",color="") +
  geom_point(alpha=0.3,color='aquamarine4') +
  #theme_cowplot() +
  #background_grid() +
  ggtitle("Library 16") +
  theme(axis.text.x = element_text(size = 12),
        axis.text.y = element_text(size = 12),
        axis.title.x = element_text(size = 14),
        axis.title.y = element_text(size = 14)) +
  theme(legend.position = "none") +
  #scale_y_continuous(expand = c(0, 0), limits = c(0,180)) +
  guides(linetype = FALSE) +
  panel_border(color = "black")

p24_16 <- ggMarginal(p23_16, type = "histogram", fill = "aquamarine4", bins=40)
p24_16
```

```{r echo=FALSE}
#Lib15
ggsave(file="plots/mutants/lib15.mutants.fitD05D03.min1BC.png", plot=p24_15,
       dpi=600, width = 8, height = 6, units = "in")

#Lib16
ggsave(file="plots/mutants/lib16.mutants.fitD12D04.min1BC.png", plot=p24_16,
       dpi=600, width = 8, height = 6, units = "in")
```

### Boxplots

Plot boxplots based on mutant sequence identity to E. coli for fitness between M9 (No Supp) vs. M9 (Full Supp)
```{r warning=FALSE}
#Lib15
p25_15 <- mutants15 %>%
  filter(numprunedBCs >= 2) %>%
  ggplot(aes(x=as.numeric(as.character(identbins))*100,y=fitD05D03,color=identbins)) +
  labs(x = "Sequence Identity (%)", y ="Fitness",color="") +
  ggtitle("Lib15: M9 No vs M9 Full") +
  geom_boxplot() +
  theme(axis.text.x = element_text(size = 12),
        axis.text.y = element_text(size = 12),
        axis.title.x = element_text(size = 14),
        axis.title.y = element_text(size = 14)) +
  theme(legend.position = "none") +
  panel_border(color = "black") +
  scale_y_continuous(expand = c(0, 0), limits = c(-14,7))

p25_15

#Lib16
p25_16 <- mutants16 %>%
  filter(numprunedBCs >= 2) %>%
  ggplot(aes(x=as.numeric(as.character(identbins))*100,y=fitD12D04,color=identbins)) +
  labs(x = "Sequence Identity (%)", y ="Fitness",color="") +
  ggtitle("Lib16: M9 No vs M9 Full") +
  geom_boxplot() +
  theme(axis.text.x = element_text(size = 12),
        axis.text.y = element_text(size = 12),
        axis.title.x = element_text(size = 14),
        axis.title.y = element_text(size = 14)) +
  theme(legend.position = "none") +
  panel_border(color = "black") +
  scale_y_continuous(expand = c(0, 0), limits = c(-14,7))

p25_16
```

```{r}
patch16<- (p25_15 | p25_16)
patch16
```

```{r echo=FALSE, eval=FALSE}
ggsave(file="plots/mutants/lib15.16.mutants.by.seq.ident.min1BC.M9-No.vs.M9-Full.png", plot=patch16,
       dpi=600, width = 18, height = 12, units = "in")
```

Plot boxplots based on mutant sequence identity to E. coli for fitness between 50 ug/ml TMP vs. M9 (Full Supp)
```{r}
#Lib15
p26_15 <- mutants15 %>%
  filter(numprunedBCs >= 2) %>%
  ggplot(aes(x=as.numeric(as.character(identbins))*100,y=fitD10D03,color=identbins)) +
  labs(x = "Sequence Identity (%)", y ="Fitness",color="") +
  ggtitle("Lib15: 50 TMP vs M9 Full") +
  geom_boxplot() +
  theme(axis.text.x = element_text(size = 12),
        axis.text.y = element_text(size = 12),
        axis.title.x = element_text(size = 14),
        axis.title.y = element_text(size = 14)) +
  theme(legend.position = "none") +
  panel_border(color = "black") +
  scale_y_continuous(expand = c(0, 0), limits = c(-14,7))

p26_15

#Lib16
p26_16 <- mutants16 %>%
  filter(numprunedBCs >= 2) %>%
  ggplot(aes(x=as.numeric(as.character(identbins))*100,y=fitE05D04,color=identbins)) +
  labs(x = "Sequence Identity (%)", y ="Fitness",color="") +
  ggtitle("Lib16: 50 TMP vs M9 Full") +
  geom_boxplot() +
  theme(axis.text.x = element_text(size = 12),
        axis.text.y = element_text(size = 12),
        axis.title.x = element_text(size = 14),
        axis.title.y = element_text(size = 14)) +
  theme(legend.position = "none") +
  panel_border(color = "black") +
  scale_y_continuous(expand = c(0, 0), limits = c(-14,7))

p26_16
```

```{r}
patch17<- (p26_15 | p26_16)
patch17
```

```{r echo=FALSE, eval=FALSE}
ggsave(file="plots/mutants/lib15.16.mutants.by.seq.ident.min1BC.50-TMP.vs.M9-Full.png", plot=patch17,
       dpi=600, width = 18, height = 12, units = "in")
```

Plot boxplots based on mutant sequence identity to E. coli for fitness between 200 ug/ml TMP vs. M9 (Full Supp)
```{r}
#Lib15
p27_15 <- mutants15 %>%
  filter(numprunedBCs >= 2) %>%
  ggplot(aes(x=as.numeric(as.character(identbins))*100,y=fitD11D03,color=identbins)) +
  labs(x = "Sequence Identity (%)", y ="Fitness",color="") +
  ggtitle("Lib15: 200 TMP vs M9 Full") +
  geom_boxplot() +
  theme(axis.text.x = element_text(size = 12),
        axis.text.y = element_text(size = 12),
        axis.title.x = element_text(size = 14),
        axis.title.y = element_text(size = 14)) +
  theme(legend.position = "none") +
  panel_border(color = "black") +
  scale_y_continuous(expand = c(0, 0), limits = c(-14,7))

p27_15

#Lib16
p27_16 <- mutants16 %>%
  filter(numprunedBCs >= 2) %>%
  ggplot(aes(x=as.numeric(as.character(identbins))*100,y=fitE06D04,color=identbins)) +
  labs(x = "Sequence Identity (%)", y ="Fitness",color="") +
  ggtitle("Lib16: 200 TMP vs M9 Full") +
  geom_boxplot() +
  theme(axis.text.x = element_text(size = 12),
        axis.text.y = element_text(size = 12),
        axis.title.x = element_text(size = 14),
        axis.title.y = element_text(size = 14)) +
  theme(legend.position = "none") +
  panel_border(color = "black") +
  scale_y_continuous(expand = c(0, 0), limits = c(-14,7))

p27_16
```

```{r}
patch18<- (p27_15 | p27_16)
patch18
```

```{r echo=FALSE, eval=FALSE}
ggsave(file="plots/mutants/lib15.16.mutants.by.seq.ident.min1BC.200-TMP.vs.M9-Full.png", plot=patch18,
       dpi=600, width = 18, height = 12, units = "in")
```

Combine them all:
```{r}
patch19<- (p25_15 | p26_15 | p27_15) / (p25_16 | p26_16 | p27_16)
patch19
```

```{r echo=FALSE, eval=FALSE}
ggsave(file="plots/mutants/lib15.16.mutants.by.seq.ident.min1BC.All.Combined.png", plot=patch19,
       dpi=600, width = 18, height = 12, units = "in")
```

# Collapse Mutants
<font color="blue">**This section is based on the R file: "R_collapse_mutants.R".**</font>

## 2 AA Distance

The following code describes how to collapse mutant BCs on their designed homologs within a distance of 2 amino acids.

### Lib15

Begin by selecting all mutants within a distance of 2 AA from their designed homologs, but only if the "ID" contains a designed variant (mutations == 0):
```{r}
mut_collapse_15 <- mutants15 %>%
  filter(mutations >= 0 & mutations < 3) %>%
  dplyr::rename(ID = IDalign) %>%
  group_by(ID) %>%
  filter(0 %in% mutations) %>%
  ungroup()
```

Summarize the number of unique "ID" with mutations==0 (perfects) after filtering:
```{r}
result15 <- mut_collapse_15 %>%
  filter(mutations == 0) %>%
  summarise(unique_rows = n_distinct(ID))

print(result15)
```

Summarize the number of unique "ID" at each mutation level:
```{r class.output="goodCode"}
summary_table15 <- mut_collapse_15 %>%
  group_by(mutations) %>%
  summarise(unique_IDs = n_distinct(ID))

# View the summary table
print(summary_table15)
```

Summarize the number of collapsed homologs after filtering
```{r class.output="goodCode"}
# Count the number of unique designed homologs retained in the filtered dataset:
format(length(unique(mut_collapse_15$ID)), big.mark = ",")

# Count the number of unique "mutID" after excluding rows with mutations==0 (retains mutants only)
format(length(unique(mut_collapse_15$mutID[mut_collapse_15$mutations != 0])), big.mark = ",")

# Now, count the number of unique "mutID" (this includes designed homologs and their mutants)
format(length(unique(mut_collapse_15$mutID)), big.mark = ",")
```

#### Mutation Similarity

Next, run correlation analyses between designed homologs and their corresponding mutant versions (1, 2, or 3 mutations) to determine if mutations share similar fitness values with designed variants

```{r}
# Lib15

# Filter the dataframe for mutations = 0 and mutations = 1
mutations15_0 <- subset(mut_collapse_15, mutations == 0)
mutations15_1 <- subset(mut_collapse_15, mutations == 1)
mutations15_2 <- subset(mut_collapse_15, mutations == 2)

# Merge the dataframes based on shared "ID"
Mut0vsMut1_15 <- merge(mutations15_0, mutations15_1, by = "ID", suffixes = c("_mutations_0", "_mutations_1"))
Mut0vsMut2_15 <- merge(mutations15_0, mutations15_2, by = "ID", suffixes = c("_mutations_0", "_mutations_2"))
Mut1vsMut2_15 <- merge(mutations15_1, mutations15_2, by = "ID", suffixes = c("_mutations_1", "_mutations_2"))

# Perform correlation analysis
cor_result15_Mut0vsMut1 <- cor(Mut0vsMut1_15$fitD05D03_mutations_0, Mut0vsMut1_15$fitD05D03_mutations_1)
cor_result15_Mut0vsMut2 <- cor(Mut0vsMut2_15$fitD05D03_mutations_0, Mut0vsMut2_15$fitD05D03_mutations_2)
cor_result15_Mut1vsMut2 <- cor(Mut1vsMut2_15$fitD05D03_mutations_1, Mut1vsMut2_15$fitD05D03_mutations_2)

cor_result15_Mut0vsMut1
cor_result15_Mut0vsMut2
cor_result15_Mut1vsMut2


# Calculate correlation and p-value
cor_test15_Mut0vsMut1 <- cor.test(Mut0vsMut1_15$fitD05D03_mutations_0, Mut0vsMut1_15$fitD05D03_mutations_1)
cor_test15_Mut0vsMut2 <- cor.test(Mut0vsMut2_15$fitD05D03_mutations_0, Mut0vsMut2_15$fitD05D03_mutations_2)
cor_test15_Mut1vsMut2 <- cor.test(Mut1vsMut2_15$fitD05D03_mutations_1, Mut1vsMut2_15$fitD05D03_mutations_2)

cor_test15_Mut0vsMut1
cor_test15_Mut0vsMut2
cor_test15_Mut1vsMut2
```

#### Plot Correlations
```{r}
# Extract correlation value from cor_result15_Mut0vsMut1 object
cor_value_Mut0vsMut1 <- cor_test15_Mut0vsMut1$estimate
cor_value_Mut0vsMut2 <- cor_test15_Mut0vsMut2$estimate


# Format p-value in scientific notation
p_value_scientific15_v1 <- format(cor_test15_Mut0vsMut1$p.value, scientific = TRUE, digits = 4)
p_value_scientific15_v2 <- format(cor_test15_Mut0vsMut2$p.value, scientific = TRUE, digits = 4)

# Plot the correlation (Mut0vsMut1)
p28.1_15 <- ggplot(Mut0vsMut1_15, 
             aes(x = fitD05D03_mutations_0, y = fitD05D03_mutations_1, color=numBCs_mutations_0)) +
  labs(x = "fitD05D03 (mutations = 0)", y = "fitD05D03 (mutations = 1)", color="", title = "Lib15 Complementation (Mut=0 vs Mut=1)") +
  geom_smooth(method=lm,colour="black") +
  geom_density2d(colour="black",alpha=0.2) +
  geom_point(alpha=0.7) +
  scale_colour_gradient2("BCs",low="blue", high="red", mid="blue") +
  theme(legend.position="left") +
  annotate("text", x = max(Mut0vsMut1_15$fitD05D03_mutations_0), y = min(Mut0vsMut1_15$fitD05D03_mutations_1), 
           label = paste("p-value =", p_value_scientific15_v1), hjust = 1, vjust = 0) +
  annotate("text", x = max(Mut0vsMut1_15$fitD05D03_mutations_0), y = min(Mut0vsMut1_15$fitD05D03_mutations_1),
            label = paste("Correlation =", round(cor_value_Mut0vsMut1, 2)), hjust = 1, vjust = -1.5)

p28.2_15 <- ggMarginal(p28.1_15, type = "histogram", fill = "aquamarine4") #add side histograms
p28.2_15

# Plot the correlation (Mut0vsMut2)
p29.1_15 <- ggplot(Mut0vsMut2_15, 
             aes(x = fitD05D03_mutations_0, y = fitD05D03_mutations_2, color=numBCs_mutations_0)) +
  labs(x = "fitD05D03 (mutations = 0)", y = "fitD05D03 (mutations = 2)", color="", title = "Lib15 Complementation (Mut=0 vs Mut=2)") +
  geom_smooth(method=lm,colour="black") +
  geom_density2d(colour="black",alpha=0.2) +
  geom_point(alpha=0.7) +
  scale_colour_gradient2("BCs",low="blue", high="red", mid="blue") +
  theme(legend.position="left") +
  annotate("text", x = max(Mut0vsMut2_15$fitD05D03_mutations_0), y = min(Mut0vsMut2_15$fitD05D03_mutations_2), 
           label = paste("p-value =", p_value_scientific15_v2), hjust = 1, vjust = 0)+
  annotate("text", x = max(Mut0vsMut2_15$fitD05D03_mutations_0), y = min(Mut0vsMut2_15$fitD05D03_mutations_2),
            label = paste("Correlation =", round(cor_value_Mut0vsMut2, 2)), hjust = 1, vjust = -1.5)

p29.2_15 <- ggMarginal(p29.1_15, type = "histogram", fill = "aquamarine4") #add side histograms
p29.2_15
```

```{r echo=FALSE, eval=FALSE}
ggsave(file="plots/correlation/L15.complementation.mut0.vs.mut1.png",
       plot=p28.2_15,
       dpi=600, width = 8, height = 6, units = "in")

ggsave(file="plots/correlation/L15.complementation.mut0.vs.mut2.png",
       plot=p29.2_15,
       dpi=600, width = 8, height = 6, units = "in")
```

### Lib16

Begin by selecting all mutants within a distance of 2 AA from their designed homologs, but only if the "ID" contains a designed variant (mutations == 0):
```{r}
mut_collapse_16 <- mutants16 %>%
  filter(mutations >= 0 & mutations < 3) %>%
  dplyr::rename(ID = IDalign) %>%
  group_by(ID) %>%
  filter(0 %in% mutations) %>%
  ungroup()
```

Summarize the number of unique "ID" with mutations==0 (perfects) after filtering:
```{r}
result16 <- mut_collapse_16 %>%
  filter(mutations == 0) %>%
  summarise(unique_rows = n_distinct(ID))

print(result16)
```

Summarize the number of unique "ID" at each mutation level:
```{r class.output="goodCode"}
summary_table16 <- mut_collapse_16 %>%
  group_by(mutations) %>%
  summarise(unique_IDs = n_distinct(ID))

# View the summary table
print(summary_table16)
```

Summarize the number of collapsed homologs after filtering
```{r class.output="goodCode"}
# Count the number of unique designed homologs retained in the filtered dataset:
format(length(unique(mut_collapse_16$ID)), big.mark = ",")

# Count the number of unique "mutID" after excluding rows with mutations==0 (retains mutants only)
format(length(unique(mut_collapse_16$mutID[mut_collapse_16$mutations != 0])), big.mark = ",")

# Now, count the number of unique "mutID" (this includes designed homologs and their mutants)
format(length(unique(mut_collapse_16$mutID)), big.mark = ",")
```

#### Mutation Similarity

Next, run correlation analyses between designed homologs and their corresponding mutant versions (1, 2, or 3 mutations) to determine if mutations share similar fitness values with designed variants

```{r}
# Filter the dataframe for mutations = 0 and mutations = 1
mutations16_0 <- subset(mut_collapse_16, mutations == 0)
mutations16_1 <- subset(mut_collapse_16, mutations == 1)
mutations16_2 <- subset(mut_collapse_16, mutations == 2)

# Merge the dataframes based on shared "ID"
Mut0vsMut1_16 <- merge(mutations16_0, mutations16_1, by = "ID", suffixes = c("_mutations_0", "_mutations_1"))
Mut0vsMut2_16 <- merge(mutations16_0, mutations16_2, by = "ID", suffixes = c("_mutations_0", "_mutations_2"))
Mut1vsMut2_16 <- merge(mutations16_1, mutations16_2, by = "ID", suffixes = c("_mutations_1", "_mutations_2"))

# Perform correlation analysis
cor_result16_Mut0vsMut1 <- cor(Mut0vsMut1_16$fitD12D04_mutations_0, Mut0vsMut1_16$fitD12D04_mutations_1)
cor_result16_Mut0vsMut2 <- cor(Mut0vsMut2_16$fitD12D04_mutations_0, Mut0vsMut2_16$fitD12D04_mutations_2)
cor_result16_Mut1vsMut2 <- cor(Mut1vsMut2_16$fitD12D04_mutations_1, Mut1vsMut2_16$fitD12D04_mutations_2)

cor_result16_Mut0vsMut1
cor_result16_Mut0vsMut2
cor_result16_Mut1vsMut2


# Calculate correlation and p-value
cor_test16_Mut0vsMut1 <- cor.test(Mut0vsMut1_16$fitD12D04_mutations_0, Mut0vsMut1_16$fitD12D04_mutations_1)
cor_test16_Mut0vsMut2 <- cor.test(Mut0vsMut2_16$fitD12D04_mutations_0, Mut0vsMut2_16$fitD12D04_mutations_2)
cor_test16_Mut1vsMut2 <- cor.test(Mut1vsMut2_16$fitD12D04_mutations_1, Mut1vsMut2_16$fitD12D04_mutations_2)

cor_test16_Mut0vsMut1
cor_test16_Mut0vsMut2
cor_test16_Mut1vsMut2
```

#### Plot Correlations
```{r}
# Extract correlation value from cor_result15_Mut0vsMut1 object
cor_value_16_Mut0vsMut1 <- cor_test16_Mut0vsMut1$estimate
cor_value_16_Mut0vsMut2 <- cor_test16_Mut0vsMut2$estimate


# Format p-value in scientific notation
p_value_scientific16_v1 <- format(cor_test16_Mut0vsMut1$p.value, scientific = TRUE, digits = 4)
p_value_scientific16_v2 <- format(cor_test16_Mut0vsMut2$p.value, scientific = TRUE, digits = 4)

# Plot the correlation (Mut0vsMut1)
p30.1_16 <- ggplot(Mut0vsMut1_16, 
             aes(x = fitD12D04_mutations_0, y = fitD12D04_mutations_1, color=numBCs_mutations_0)) +
  labs(x = "fitD12D04 (mutations = 0)", y = "fitD12D04 (mutations = 1)", color="", title = "Lib16 Complementation (Mut=0 vs Mut=1)") +
  geom_smooth(method=lm,colour="black") +
  geom_density2d(colour="black",alpha=0.2) +
  geom_point(alpha=0.7) +
  scale_colour_gradient2("BCs",low="blue", high="red", mid="blue") +
  theme(legend.position="left") +
  annotate("text", x = max(Mut0vsMut1_16$fitD12D04_mutations_0), y = min(Mut0vsMut1_16$fitD12D04_mutations_1), 
           label = paste("p-value =", p_value_scientific15_v1), hjust = 1, vjust = 0) +
  annotate("text", x = max(Mut0vsMut1_16$fitD12D04_mutations_0), y = min(Mut0vsMut1_16$fitD12D04_mutations_1),
            label = paste("Correlation =", round(cor_value_16_Mut0vsMut1, 2)), hjust = 1, vjust = -1.5)

p30.2_16 <- ggMarginal(p30.1_16, type = "histogram", fill = "aquamarine4") #add side histograms
p30.2_16

# Plot the correlation (Mut0vsMut2)
p31.1_16 <- ggplot(Mut0vsMut2_16, 
             aes(x = fitD12D04_mutations_0, y = fitD12D04_mutations_2, color=numBCs_mutations_0)) +
  labs(x = "fitD12D04 (mutations = 0)", y = "fitD12D04 (mutations = 2)", color="", title = "Lib16 Complementation (Mut=0 vs Mut=2)") +
  geom_smooth(method=lm,colour="black") +
  geom_density2d(colour="black",alpha=0.2) +
  geom_point(alpha=0.7) +
  scale_colour_gradient2("BCs",low="blue", high="red", mid="blue") +
  theme(legend.position="left") +
  annotate("text", x = max(Mut0vsMut2_16$fitD12D04_mutations_0), y = min(Mut0vsMut2_16$fitD12D04_mutations_2), 
           label = paste("p-value =", p_value_scientific16_v2), hjust = 1, vjust = 0)+
  annotate("text", x = max(Mut0vsMut2_16$fitD12D04_mutations_0), y = min(Mut0vsMut2_16$fitD12D04_mutations_2),
            label = paste("Correlation =", round(cor_value_16_Mut0vsMut2, 2)), hjust = 1, vjust = -1.5)

p31.2_16 <- ggMarginal(p31.1_16, type = "histogram", fill = "aquamarine4") #add side histograms
p31.2_16
```

```{r echo=FALSE, eval=FALSE}
ggsave(file="plots/correlation/L16.complementation.mut0.vs.mut1.png",
       plot=p30.2_16,
       dpi=600, width = 8, height = 6, units = "in")

ggsave(file="plots/correlation/L16.complementation.mut0.vs.mut2.png",
       plot=p31.2_16,
       dpi=600, width = 8, height = 6, units = "in")
```


## Generate FASTA

Generate a FASTA file for each library containing each unique mutID (designed homologs and mutants up to 2 AA difference) and their corresponding protein sequence for use in gain-of-function (GOF) analysis and broad mutational scanning (BMS).
```{r}
# Lib15

# Create the sequences in FASTA format
fasta_content_15 <- paste(">", mut_collapse_15$mutID, "\n", mut_collapse_15$seq, "\n", sep = "", collapse = "")

# Define the file path in the working directory
fasta_file_path_15 <- file.path(getwd(), "OUTPUT/mutant.collapse.2BC.Lib15.fasta")

# Write the FASTA content to the file
writeLines(fasta_content_15, con = fasta_file_path_15)

# Lib15

# Create the sequences in FASTA format
fasta_content_16 <- paste(">", mut_collapse_16$mutID, "\n", mut_collapse_16$seq, "\n", sep = "", collapse = "")

# Define the file path in the working directory
fasta_file_path_16 <- file.path(getwd(), "OUTPUT/mutant.collapse.2BC.Lib16.fasta")

# Write the FASTA content to the file
writeLines(fasta_content_16, con = fasta_file_path_16)
```

## Median Fitness
Generate median fitness values for each homolog and associated mutants (up to 2 A.A. distance) by grouping by "ID"

**Lib15**
```{r}
#Calculate median fitness for each homolog and associated mutants and sum the total number of BCs (numBCs and numprunedBCs)
mut_collapse_15info <- mut_collapse_15 %>%
  group_by(ID) %>%
  summarise(medD05D03=median(fitD05D03),
            medD06D03=median(fitD06D03),
            medD07D03=median(fitD07D03),
            medD08D03=median(fitD08D03),
            medD09D03=median(fitD09D03),
            medD10D03=median(fitD10D03),
            medD11D03=median(fitD11D03),
            totalnumBCs.L15=sum(numBCs),
            totalnumprunedBCs.L15=sum(numprunedBCs))
```

Count the number of unique IDs after collapsing mutants up to 2 A.A. distance:
```{r class.output="goodCode"}
format(length(unique(mut_collapse_15info$ID)), big.mark = ",")
```

**Lib16**
```{r}
#Calculate median fitness for each homolog and associated mutants and sum the total number of BCs (numBCs and numprunedBCs)
mut_collapse_16info <- mut_collapse_16 %>%
  group_by(ID) %>%
  summarise(medD12D04=median(fitD12D04),
            medE01D04=median(fitE01D04),
            medE02D04=median(fitE02D04),
            medE03D04=median(fitE03D04),
            medE04D04=median(fitE04D04),
            medE05D04=median(fitE05D04),
            medE06D04=median(fitE06D04),
            totalnumBCs.L16=sum(numBCs),
            totalnumprunedBCs.L16=sum(numprunedBCs))
```

Count the number of unique IDs after collapsing mutants up to 2 A.A. distance:
```{r class.output="goodCode"}
format(length(unique(mut_collapse_16info$ID)), big.mark = ",")
```

Generate a new dataframe retaining only the unique IDs shared between libraries:
```{r}
shared_mut_collapse_15.16info <- merge(mut_collapse_15info, mut_collapse_16info, by = "ID")
```

Count the number of unique IDs shared between libraries:
```{r class.output="goodCode"}
format(length(unique(shared_mut_collapse_15.16info$ID)), big.mark = ",")
```

Calculate correlation between median fitness values for unique IDs shared between libraries:
```{r}
# Calculate correlation and p-value between Lib15 and Lib16 Complementation
cor_test_shared_mut_collapse_15.16info <- cor.test(shared_mut_collapse_15.16info$medD05D03, shared_mut_collapse_15.16info$medD12D04)

cor_test_shared_mut_collapse_15.16info
```

Plot the median fitness correlation between unique IDs shared between Lib15 and Lib16 (median with collapsed mutants):
```{r}
# Extract correlation value from cor_test_shared_mut_collapse_15.16info object
cor_value_shared <- cor_test_shared_mut_collapse_15.16info$estimate

# Format p-value in scientific notation
p_value_scientific_shared <- format(cor_test_shared_mut_collapse_15.16info$p.value, 
                                                    scientific = TRUE, digits = 4)

p33.1 <- ggplot(shared_mut_collapse_15.16info, 
             aes(x = medD05D03, y = medD12D04, color=totalnumprunedBCs.L15)) +
  labs(x = "Median Lib15 Fitness (Complementation)", y = "Median Lib16 Fitness (Complementation)", color="", title = "Median Fitness Correlation between Shared IDs") +
  geom_smooth(method=lm,colour="black") +
  geom_density2d(colour="black",alpha=0.2) +
  geom_point(alpha=0.7) +
  scale_colour_gradient2("BCs",low="blue", high="red", mid="blue") +
  theme(legend.position="left") +
  annotate("text", x = max(shared_mut_collapse_15.16info$medD05D03), y = min(shared_mut_collapse_15.16info$medD12D04), 
           label = paste("p-value =", p_value_scientific_shared), hjust = 1, vjust = 0) +
  annotate("text", x = max(shared_mut_collapse_15.16info$medD05D03), y = min(shared_mut_collapse_15.16info$medD12D04),
            label = paste("Correlation =", round(cor_value_shared, 2)), hjust = 1, vjust = -1.5)

p33.2 <- ggMarginal(p33.1, type = "histogram", fill = "aquamarine4") #add side histograms
p33.2
```

```{r echo=FALSE, eval=FALSE}
ggsave(file="plots/correlation/L15.L16.shared.collapsed.mutants.median.complementation.png",
       plot=p33.2,
       dpi=600, width = 8, height = 6, units = "in")
```

# BMS Analysis

<font color="blue">**This section is based on the R file: "R_BMS_all.Lib15.R".**</font> It describes how to generate MSA mapping files for each homolog and map their fitness values by sequential amino acid positions aligned to E. coli as the reference sequence. The end result is a heatmap displaying median fitness values for each amino acid at each position along the protein sequence.

## BMS Python

**Broad Mutational Scanning (BMS) Analysis Script:**
This set of scripts facilitates broad mutational scanning analysis where data from many related protein homologs and their mutants is combined and collapsed for further analysis and visualization. This script is based on the approach used in: Plesa C, Sidore AM, Lubock N, Zhang D, Kosuri S. Multiplexed Gene Synthesis in Emulsions for Exploring Protein Functional Landscapes. Science 359, 343–347 (2018), [DOI: 10.1126/science.aao5167](https://www.science.org/doi/10.1126/science.aao5167).

**Procedure**
Download the following python scripts via [GitHub](link_to_files) and place in working directory:

- dssp.py
- clustalo
- FastTree
  
The analysis can be carried out by running the following scripts, which include <ins>python</ins> and <ins>bash</ins>:

- `clustalo` Alignment python script based on fasta files generated for each library
- `map.aligned.residues.py` Residue mapping from aligned fasta file generated for each library
- `parse_dssp.py` Generate RSA and SS information files from the WT E. coli DHFR dssp file

**Alignment:** Use the `clustalo` executable to align the protein sequences associated with designed homologs and associated mutants up to two A.A. distance as generated in the **mutant.collapse.2BC.Lib_.fasta** files.
```{bash eval=FALSE}
# May need to enable permissions to run the executable:
# chmod +x ./clustalo
./clustalo -i OUTPUT/mutant.collapse.2BC.Lib15.fasta -o OUTPUT/mutant.collapse.2BC.Lib15_aligned.fa --outfmt=a2m --force
```

**Mapping Residues:** Use the following `map.aligned.residues.py` python script to generate csv files for each designed homolog and associated mutant that maps residue positions of each A.A. from alignment fasta:

- `map_aligned_residues.py`
This will parse the alignments and generate tables of which homolog's residue corresponds to which position in the alignment table so that co-aligned residues can be determined.
```{python eval=FALSE}
import time
import csv
from Bio.Seq import Seq
from Bio.SeqRecord import SeqRecord
from Bio import SeqIO

#
# Note; modify the file to remove blank lines
#

##################################
#FUNCTIONS:
def getFastaSeqs(filename):
    fastaseqs = []
    handle = open(filename)
    for seqrec in SeqIO.parse(handle,"fasta"):
        fastaseqs.append(seqrec)
    handle.close()
    return fastaseqs

##################################
#INPUTS:

base_path = ""
trees_path_prefix = base_path+""

#clustal format alignment file
align_file_in = [trees_path_prefix+"perfects_tree_5BCs_aligned_mod.aln"]

#number of seqs in each alignment file
num_samples_in_file = [852]

##################################
#OUTPUTS:

msa_map_out_path = [trees_path_prefix+"MSA_BMS_1/"]


for alni in range(1):#len(align_file_in)):
    print(alni)
    
    ##################################
    #VARIABLES:
    
    #ID as key, align as value
    align_dict = dict()
    
    #num_samples = 454
    num_samples = num_samples_in_file[alni]
    
    #pos key, consensus pos val
    IDaadictlist = [dict() for x in range(num_samples)]
    
    IDtoindexdict = dict()
    indexdtoIDict = dict()
    
    ##################################
    #CODE:
    
    line_count = 0
    #loop over all alignments:
    print(align_file_in[alni])
    for line in open(align_file_in[alni]):
        #skip header
        if line_count > 1:
            listWords = line.split('    ')
            ID = listWords[0]
            align = line[16:].rstrip()
            if ID.strip() != "":
                align_dict[ID] = align_dict.get(ID, "") + align.replace(" ", "")
        line_count += 1
    
    print("NP_414590")
    print(align_dict["NP_414590"])

    counter = 0
    for ID in align_dict:
        #print(ID)
        #print(align_dict[ID])
        IDtoindexdict[ID] = counter
        indexdtoIDict[counter]=ID
        align = align_dict[ID]
        
        aacounter = 1
        
        
        for i in range(len(align)):
            if align[i] != "-":
                
                #print(str(counter)+" "+str(aacounter))
                IDaadictlist[counter][aacounter]=i+1
                aacounter += 1
        counter += 1
        
    print(len(IDaadictlist))
    for i in range(len(IDaadictlist)-1):
        #print(indexdtoIDict[i])
        #print(i)
        #print(alni)
        #print(indexdtoIDict[i])
        csvfile = open(str(msa_map_out_path[alni]+indexdtoIDict[i]+".csv"), 'w')
        fieldnames = ['orth_aanum','msa_aanum']
        writer = csv.DictWriter(csvfile, fieldnames=fieldnames)
        writer.writeheader()
        for j in IDaadictlist[i]:
            #print(str(j)+" "+str(IDaadictlist[i][j]))
            #save all data:
            writer.writerow({'orth_aanum':str(j),'msa_aanum':str(IDaadictlist[i][j])})
        csvfile.close()
```

- `parse_dssp.py`
This is used to generate Relative-Solvent-Accessibility and Secondary-Structure information files from the WT E. coli DHFR dssp file. This uses Jesse Bloom's dssp module from mapmuts.

  - Script requires a properly formatted dssp file derived from the WT E. coli PDB file available online
  - **PDB File for WT E. coli (4KJK):** Download PDB file from  [https://www.rcsb.org/structure/4KJK](https://www.rcsb.org/structure/4KJK) 
  - **DSSP File Generator (4KJK):** Navigate to [http://bioinformatica.isa.cnr.it/SUSAN/DSSP-web/](http://bioinformatica.isa.cnr.it/SUSAN/DSSP-web/)
    - Open the PDB file in text editor, select all, copy and paste in text box on Bioinformatica website
    - Select "DSSP file" from option list and hit "SEND"
    - Select entire dssp output from browser window and copy and paste in text editor
    - Save the text file as: <font color="green">4kjk.dssp.txt</font>
    - Remove the last two rows and save as: <font color="green">4kjk.dsspmod1.txt</font> (necessary for python script to work)
  - Run the following `parse_dssp.py` script calling in the <font color="green">4kjk.dsspmod1.txt</font> file
    - Python script will generate the following two text files:
      - <font color="green">4kjk.SSs.txt</font> (Secondary Structures)
      - <font color="green">4kjk.RSAs.txt</font> (Relative-Solvent-Accessibility)
  

**parse_dssp.py** python script
```{python eval=FALSE}
import dssp

dsspdat = dssp.ReadDSSP('OUTPUT/structures/4kjk.dsspmod1.txt', 'Tien2013', 'A')

frsa = open('OUTPUT/structures/4kjk.RSAs.txt', 'w')
frsa.write('#Relative solvent accessibilities from a DSSP analysis of the 4KJK, taking results for chain A. Absolute solvent accessibilities are normalized to relative ones using the maximum solvent accessibilities of Tien et al, PLoS One, 8:e80635.\n#SITE RSA\n')
fss = open('OUTPUT/structures/4kjk.SSs.txt', 'w')
fss.write('#Secondary structures from a DSSP analysis of the 4KJK, taking results for chain A.\n#SITE SS\n')
sites = sorted(dsspdat.keys())
for site in sites:
    fss.write('%d %s\n' % (site, dsspdat[site]['SS_CLASS']))
    frsa.write('%d %s\n' % (site, dsspdat[site]['RSA']))
fss.close()
frsa.close()
```

## MSA Mapping
Define the directory containing the parse MSA mapping files (.csv) for each collapsed homolog (<2 A.A. distance):
```{r}
BMS_MSA_directory = "MSA_BMS_1"

# AA map + X
BMS_aa_list<-data.frame(aa=c('G','P','A','V','L','I','M','C','F','Y','W','H','K','R','Q','N','E','D','S','T','X'),
                        aanum=c(1:21))

BMS_aa_dim <- length(BMS_aa_list$aa)

# Wildtype E. coli is:
#NP_414590

# AA length of homolog which collapsing on
BMS_ref_len <- 159

# Minimum number of BCs to use a mutant
BMS_min_BCs = 1

# Largest number of mutations to use
#currently limited to 5 due to mutation naming scheme
BMS_max_mutations = 5

# Minimum fitness to include the mutation
BMS_min_fitness <- -1
```

```{r}
# Grab perfects (perfects15_2BCs) %>%
BMS_homolog15_ID_list <- perfects15_2BCs %>%
  filter(fitD05D03 > BMS_min_fitness) %>%
  dplyr::select(ID)

write.table(BMS_homolog15_ID_list, file = paste("OUTPUT/BMS_Fitness/BMS_homolog15_ID_list_min_fitness_",as.character(BMS_min_fitness),".csv",sep=""), sep = ",", row.names = F,quote=F,col.names = F)
```

```{r}
#make sure all of the perfects are in the MSA
BMS_index_of_mut_to_drop = numeric()
for (i in 1:length(BMS_homolog15_ID_list$ID)){
  mutant_current_temp <- BMS_homolog15_ID_list$ID[i]
  if(!file.exists(paste(BMS_MSA_directory,"/",mutant_current_temp,".csv",sep=""))){
    print(mutant_current_temp)
    BMS_index_of_mut_to_drop[length(BMS_index_of_mut_to_drop)+1] <- which(BMS_homolog15_ID_list$ID==mutant_current_temp)
  }
}
if (length(BMS_index_of_mut_to_drop) > 0){
  BMS_homolog15_ID_list <- BMS_homolog15_ID_list[-BMS_index_of_mut_to_drop,]
}

#rm(BMS_index_of_mut_to_drop,
#   mutant_current_temp)
```

```{r}
#load ecoli MSA mapping
BMS_ecoli_map <- read.csv(file=paste(BMS_MSA_directory,"/NP_414590.csv",sep=""),
                          head=TRUE,
                          sep=",")

#E.coli PPAT seq:
BMS_ecoli_seq <- "MISLIAALAVDRVIGMENAMPWNLPADLAWFKRNTLNKPVIMGRHTWESIGRPLPGRKNIILSSQPGTDDRVTWVKSVDEAIAACGDVPEIMVIGGGRVYEQFLPKAQKLYLTHIDAEVEGDTHFPDYEPDDWESVFSEFHDADAQNSHSYCFEILERR"
```

Determine the total number of mutants included in BMS analysis
```{r class.output="goodCode"}
mutants15 %>%
  filter(mutations > 0 & mutations < 6 & numprunedBCs >= BMS_min_BCs) %>%
  dplyr::rename(ID=IDalign) %>%
  semi_join(BMS_homolog15_ID_list,by="ID") %>%
  ungroup() %>%
  dplyr::rename(IDalign=ID) %>%
  nrow(.)
```

## BMS Perfects

Grab the perfects with this number of mutations:
```{r}
BMS_mutants15_temp <- mutants15 %>%
  filter(mutations == 0 &
           numprunedBCs >= BMS_min_BCs) %>%
  dplyr::rename(ID=IDalign) %>%
  semi_join(BMS_homolog15_ID_list,by="ID") %>%#filtering join
  ungroup() %>%
  dplyr::rename(IDalign=ID)
```

Make a new data frame which will keep info:
```{r}
BMS_fitness15_map <- data.frame(position=numeric(),
                              aa=character(),
                              mutations=numeric(),
                              fitness=numeric(),
                              posortho=numeric(),
                              ingap=character(),
                              mutID=character(),
                              ID=character())
```

Loop over all perfects:
```{r}
for (i in 1:nrow(BMS_mutants15_temp)) {
  #current ortholog:
  mutant_current_temp <- BMS_mutants15_temp$IDalign[i]
  
  #get the MSA mapping:
  mutant_map_temp <- read.csv(file=paste(BMS_MSA_directory,"/",mutant_current_temp,".csv",sep=""), head=TRUE, sep=",")
  
  #load the full seq and add an M at start
  BMS_seq_temp <- paste("M", as.character(BMS_mutants15_temp$seq[i]), sep="")
  
  #get fitness for this ortholog
  BMS_fit_temp <- BMS_mutants15_temp$fitD05D03[i]
  
  #loop over all residues
  for (j in 1:nchar(BMS_seq_temp)) {
    #find the corresponding residue in E.coli using MSA
    BMS_cons_aanum <- mutant_map_temp$msa_aanum[which(mutant_map_temp$orth_aanum == j)]
    
    #does this map to a non-gap residue in E.coli?
    if (BMS_cons_aanum %in% BMS_ecoli_map$msa_aanum) {
      #get the E.coli residue
      e_coli_residue <- BMS_ecoli_map$orth_aanum[which(BMS_ecoli_map$msa_aanum == BMS_cons_aanum)]
      
      #aa at this residue
      BMS_aa_temp <- substr(BMS_seq_temp, j, j)
      
      #add info for this residue to df
      BMS_fitness15_map <- rbind(BMS_fitness15_map,
                               data.frame(position=e_coli_residue,
                                          aa=BMS_aa_temp,
                                          mutations=0,
                                          fitness=BMS_fit_temp,
                                          posortho=j,
                                          ingap="No",
                                          mutID=mutant_current_temp,
                                          ID=mutant_current_temp))
    } else {
      #if it's here it maps to a gap
      BMS_aa_temp <- substr(BMS_seq_temp, j, j)
      BMS_fitness15_map <- rbind(BMS_fitness15_map,
                               data.frame(position=-1,
                                          aa=BMS_aa_temp,
                                          mutations=0,
                                          fitness=BMS_fit_temp,
                                          posortho=j,
                                          ingap="Yes",
                                          mutID=mutant_current_temp,
                                          ID=mutant_current_temp))
    }
  }
}
```

```{r eval=FALSE, echo=FALSE}
write.table(BMS_fitness15_map, file = paste("OUTPUT/BMS_Fitness/BMS_fitness15_map.csv",sep=""), sep = ",", row.names = F,quote=F,col.names = T)
```

Collapse fitness values for each position onto the BMS Fitness Map:
```{r}
BMS_fitness15_collapsed <- BMS_fitness15_map %>%
  filter(position > 0) %>%
  group_by(position, aa) %>%
  summarise(fitval=median(fitness),
            numpoints=n(),
            stdfit=sd(fitness))
```

These matrices have the fitness for each aa at each position:
```{r}
BMS_matrix15_perfects = matrix(rep(NA, BMS_ref_len*BMS_aa_dim),nrow=BMS_aa_dim,ncol=BMS_ref_len)
BMS_matrix15_perfects_num = matrix(rep(NA, BMS_ref_len*BMS_aa_dim),nrow=BMS_aa_dim,ncol=BMS_ref_len)
BMS_matrix15_perfects_sd = matrix(rep(NA, BMS_ref_len*BMS_aa_dim),nrow=BMS_aa_dim,ncol=BMS_ref_len)
```

Populate matrix:
```{r}
for (i in 1:nrow(BMS_fitness15_collapsed)){
  BMS_matrix15_perfects[which(BMS_aa_list$aa==as.character(BMS_fitness15_collapsed$aa[i])),
                      BMS_fitness15_collapsed$position[i]] <- as.numeric(BMS_fitness15_collapsed$fitval[i])
  BMS_matrix15_perfects_num[which(BMS_aa_list$aa==as.character(BMS_fitness15_collapsed$aa[i])),
                          BMS_fitness15_collapsed$position[i]] <- as.numeric(BMS_fitness15_collapsed$numpoints[i])
  BMS_matrix15_perfects_sd[which(BMS_aa_list$aa==as.character(BMS_fitness15_collapsed$aa[i])),
                         BMS_fitness15_collapsed$position[i]] <- as.numeric(BMS_fitness15_collapsed$stdfit[i])
}
```

Assign AA names and position numbers to matrices:
```{r warning=FALSE}
rownames(BMS_matrix15_perfects)<-BMS_aa_list$aa
colnames(BMS_matrix15_perfects)<-c(1:BMS_ref_len)
rownames(BMS_matrix15_perfects_num)<-BMS_aa_list$aa
colnames(BMS_matrix15_perfects_num)<-c(1:BMS_ref_len)
rownames(BMS_matrix15_perfects_sd)<-BMS_aa_list$aa
colnames(BMS_matrix15_perfects_sd)<-c(1:BMS_ref_len)

BMS_matrix15_perfects_melt <- melt(BMS_matrix15_perfects)
BMS_matrix15_perfects_num_melt <- melt(BMS_matrix15_perfects_num)
BMS_matrix15_perfects_sd_melt <- melt(BMS_matrix15_perfects_sd)
```

### Perfects Plotting

Plot the fitness data:
```{r}
ggplot(data = BMS_matrix15_perfects_melt, aes(x=X2, y=X1, fill=value)) +
  geom_tile()+ labs(x = "Position (aa)", y ="Amino acid",color="") +
  scale_fill_gradient2(low = "blue", high = "red", mid="gold",name="Fitness",na.value="grey", limit = c(-1,1)) +
  theme_minimal() + 
  scale_x_continuous(breaks=seq(0,155,5))
```

```{r echo=FALSE, eval=FALSE}
ggsave(file=paste("plots/BMS/BMS_allpos15_min",toString(BMS_min_BCs),"_0_fit.pdf",sep=""))
```

Plot the coverage data:
```{r}
ggplot(data = BMS_matrix15_perfects_num_melt, aes(x=X2, y=X1, fill=log(value))) +
  geom_tile()+ labs(x = "Position (aa)", y ="Amino acid",color="") +
  scale_fill_gradient(low = "blue", high = "red",name="log(#points)",na.value="grey", limit = c(0,max(BMS_matrix15_perfects_num_melt$value))) +
  theme_minimal() + 
  scale_x_continuous(breaks=seq(0,155,5))
```

```{r echo=FALSE, eval=FALSE}
ggsave(file=paste("plots/BMS/BMS_allpos15_min",toString(BMS_min_BCs),"_0_num.pdf",sep=""))
```

Plot the STD Data:
```{r}
ggplot(data = BMS_matrix15_perfects_sd_melt, aes(x=X2, y=X1, fill=value)) +
  geom_tile()+ labs(x = "Position (aa)", y ="Amino acid",color="") +
  scale_fill_gradient2(low = "blue", high = "red", mid="gold",name="std(Fitness)",na.value="grey", limit = c(0,1.1*max(BMS_matrix15_perfects_sd_melt$value))) +
  theme_minimal() + 
  scale_x_continuous(breaks=seq(0,155,5))
```

```{r echo=FALSE, eval=FALSE}
ggsave(file=paste("plots/BMS/BMS_allpos15_min",toString(BMS_min_BCs),"_0_sd.pdf",sep=""))
```

## BMS Mutants

Grab the mutants associated with the perfects:
```{r}
# Make a copy of the dataframe for mutants
BMS_fitness15_map_all <- BMS_fitness15_map
```

Loop over mutants at some distance:
```{r warning=FALSE}
for (BMS_cur_mut_num in 1:BMS_max_mutations){
  
  #grab the mutants
  BMS_mutants15_temp <- mutants15 %>%
    filter(mutations == BMS_cur_mut_num &
             numprunedBCs >= BMS_min_BCs) %>%
    dplyr::rename(ID=IDalign) %>%
    semi_join(BMS_homolog15_ID_list,by="ID") %>%#filtering join
    ungroup() %>%
    dplyr::rename(IDalign=ID)
  
  #initialize new df
  BMS_fitness15_map1 <- data.frame(position=numeric(),
                                 aa=character(),
                                 mutations=numeric(),
                                 fitness=numeric(),
                                 posortho=numeric(),
                                 ingap=character(),
                                 mutID=character(),
                                 ID=character())
  
  #loop over mutants
  for (i in 1:nrow(BMS_mutants15_temp)){
    
    #mutant base name
    mutant_current_temp <- BMS_mutants15_temp$IDalign[i]
    
    #length of the name
    name_size = nchar(paste(mutant_current_temp,"_",sep=""))
    
    #residue mappings
    mutant_map_temp <- read.csv(file=paste(BMS_MSA_directory,"/",mutant_current_temp,".csv",sep=""),head=TRUE,sep=",")
    
    #this mutants fitness
    BMS_fit_temp <- BMS_mutants15_temp$fitD05D03[i]
    
    #grab the mut name
    mutations15_names <- as.character(BMS_mutants15_temp$mutID[i])
    
    #grab only the relevant portion of the name
    mutations15_names <- substr(mutations15_names, name_size+1, nchar(mutations15_names))
    
    ## split mutation string at non-digits
    s <- strsplit(mutations15_names, "_")
    
    for (mutnum in 1:BMS_cur_mut_num){
      
      #grab the corresponding mutation string
      mutcurr<-s[[1]][mutnum]
      
      #get the position
      mutpos <- as.numeric(str_extract(mutcurr, "[0-9]+"))
      
      #get ending aa
      to_aa <- substr(mutcurr, nchar(mutpos)+2, nchar(mutcurr))
      
      #find the number in the consensus seq
      BMS_cons_aanum <- mutant_map_temp$msa_aanum[which(mutant_map_temp$orth_aanum == mutpos)]
      
      #does this map to a non-gap
      if (BMS_cons_aanum %in% BMS_ecoli_map$msa_aanum){
        
        #the corresponding e.coli residue
        e_coli_residue <- BMS_ecoli_map$orth_aanum[which(BMS_ecoli_map$msa_aanum == BMS_cons_aanum)]
        
        #add this point to the data
        BMS_fitness15_map1 <- rbind(BMS_fitness15_map1,
                                  data.frame(position=e_coli_residue,
                                             aa=to_aa,
                                             mutations=BMS_cur_mut_num,
                                             fitness=BMS_fit_temp,
                                             posortho=mutpos,
                                             ingap="No",
                                             mutID=as.character(BMS_mutants15_temp$mutID[i]),
                                             ID=mutant_current_temp))
        
      } else {
        #if it's here it maps to a gap
        
        #add this point to the data
        BMS_fitness15_map1 <- rbind(BMS_fitness15_map1,
                                  data.frame(position=-1,
                                             aa=to_aa,
                                             mutations=BMS_cur_mut_num,
                                             fitness=BMS_fit_temp,
                                             posortho=mutpos,
                                             ingap="Yes",
                                             mutID=as.character(BMS_mutants15_temp$mutID[i]),
                                             ID=mutant_current_temp))
        
      }
      
    }
    
  }

  #Add these mutants onto the existing data:
  BMS_fitness15_map_all <- rbind(BMS_fitness15_map_all,BMS_fitness15_map1)

  write.table(BMS_fitness15_map1, file = paste("OUTPUT/BMS_Fitness/BMS_fitness15_map_",as.character(BMS_cur_mut_num),".csv",sep=""), 
            sep = ",", row.names = F,quote=F,col.names = T)
  
  #calc all data and stats
  BMS_fitness15_collapsed_all <- BMS_fitness15_map_all %>%
    filter(position > 0) %>%
    group_by(position, aa) %>%
    summarise(fitval=median(fitness),
              numpoints=n(),
              stdfit=sd(fitness))
  
  #these matrices have the fitness/num/sd for each aa at each position:
  BMS_matrix15_perfects_and_1 = matrix(rep(NA, BMS_ref_len*BMS_aa_dim),nrow=BMS_aa_dim,ncol=BMS_ref_len)
  BMS_matrix15_perfects_and_1_num = matrix(rep(NA, BMS_ref_len*BMS_aa_dim),nrow=BMS_aa_dim,ncol=BMS_ref_len)
  BMS_matrix15_perfects_and_1_sd = matrix(rep(NA, BMS_ref_len*BMS_aa_dim),nrow=BMS_aa_dim,ncol=BMS_ref_len)
  
  #populate matrix
  for (i in 1:nrow(BMS_fitness15_collapsed_all)){
    
    BMS_matrix15_perfects_and_1[which(BMS_aa_list$aa==as.character(BMS_fitness15_collapsed_all$aa[i])),BMS_fitness15_collapsed_all$position[i]] <- as.numeric(BMS_fitness15_collapsed_all$fitval[i])
    BMS_matrix15_perfects_and_1_num[which(BMS_aa_list$aa==as.character(BMS_fitness15_collapsed_all$aa[i])),BMS_fitness15_collapsed_all$position[i]] <- as.numeric(BMS_fitness15_collapsed_all$numpoints[i])
    BMS_matrix15_perfects_and_1_sd[which(BMS_aa_list$aa==as.character(BMS_fitness15_collapsed_all$aa[i])),BMS_fitness15_collapsed_all$position[i]] <- as.numeric(BMS_fitness15_collapsed_all$stdfit[i])
  }
  
  rownames(BMS_matrix15_perfects_and_1)<-BMS_aa_list$aa
  colnames(BMS_matrix15_perfects_and_1)<-c(1:BMS_ref_len)
  rownames(BMS_matrix15_perfects_and_1_num)<-BMS_aa_list$aa
  colnames(BMS_matrix15_perfects_and_1_num)<-c(1:BMS_ref_len)
  rownames(BMS_matrix15_perfects_and_1_sd)<-BMS_aa_list$aa
  colnames(BMS_matrix15_perfects_and_1_sd)<-c(1:BMS_ref_len)
  
  BMS_matrix15_perfects_and_1_melt <- melt(BMS_matrix15_perfects_and_1)
  BMS_matrix15_perfects_and_1_num_melt <- melt(BMS_matrix15_perfects_and_1_num)
  BMS_matrix15_perfects_and_1_sd_melt <- melt(BMS_matrix15_perfects_and_1_sd)
  
  #plot the data from these mutants:
  ggplot(data = BMS_matrix15_perfects_and_1_melt, aes(x=X2, y=X1, fill=value)) +
    geom_tile()+ labs(x = "Position (aa)", y ="Amino acid",color="") +
    scale_fill_gradient2(low = "blue", high = "red", mid="gold",name="Fitness",na.value="grey", limit = c(-5,5)) +
    theme_minimal() + 
    scale_x_continuous(breaks=seq(0,155,5))
  ggsave(file=paste("plots/BMS/BMS_allpos15_min",toString(BMS_min_BCs),"_max",toString(BMS_cur_mut_num),"_fit.pdf",sep=""))
  
  ggplot(data = BMS_matrix15_perfects_and_1_num_melt, aes(x=X2, y=X1, fill=log(value))) +
    geom_tile()+ labs(x = "Position (aa)", y ="Amino acid",color="") +
    scale_fill_gradient(low = "blue", high = "red",name="log(#points)",na.value="grey", limit = c(0,max(BMS_matrix15_perfects_and_1_num_melt$value))) +
    theme_minimal() + 
    scale_x_continuous(breaks=seq(0,155,5))
  ggsave(file=paste("plots/BMS/BMS_allpos15_min",toString(BMS_min_BCs),"_max",toString(BMS_cur_mut_num),"_num.pdf",sep=""))
  
  ggplot(data = BMS_matrix15_perfects_and_1_sd_melt, aes(x=X2, y=X1, fill=value)) +
    geom_tile()+ labs(x = "Position (aa)", y ="Amino acid",color="") +
    scale_fill_gradient2(low = "blue", high = "red", mid="gold",name="std(Fitness)",na.value="grey", limit = c(0,1.1*max(BMS_matrix15_perfects_and_1_sd_melt$value))) +
    theme_minimal() + 
    scale_x_continuous(breaks=seq(0,155,5))
  ggsave(file=paste("plots/BMS/BMS_allpos15_min",toString(BMS_min_BCs),"_max",toString(BMS_cur_mut_num),"_sd.pdf",sep=""))
}
```

```{r eval=FALSE, echo=FALSE}
write.table(BMS_matrix15_perfects_and_1_melt, 
            file = paste("OUTPUT/BMS_Fitness/BMS_matrix_perfects_and_1_melt.csv",sep=""),
            sep = ",", row.names = F,quote=F,col.names = T)
write.table(BMS_matrix15_perfects_and_1_num_melt, 
            file = paste("OUTPUT/BMS_Fitness/BMS_matrix_perfects_and_1_num_melt.csv",sep=""),
            sep = ",", row.names = F,quote=F,col.names = T)
write.table(BMS_matrix15_perfects_and_1_sd_melt, 
            file = paste("OUTPUT/BMS_Fitness/BMS_matrix_perfects_and_1_sd_melt.csv",sep=""),
            sep = ",", row.names = F,quote=F,col.names = T)
```

## BMS Custom Build

Begin by adding a new column for the WT residues:
```{r}
BMS_matrix15_perfects_and_1_melt$WTcolor <- NA
#X1 is aa, X2 is position

BMS_matrix15_perfects_and_1_melt$aanum <- 0

for (i in 1:nrow(BMS_matrix15_perfects_and_1_melt)){
  
  #find the corresponding residue in E.coli using MSA
  BMS_cons_aanum <- BMS_matrix15_perfects_and_1_melt$X2[i]
  
  #is this the WT E.coli residue?
  if (BMS_matrix15_perfects_and_1_melt$X1[i] == substr(BMS_ecoli_seq, BMS_cons_aanum, BMS_cons_aanum)){
    
    #assign WT color
    BMS_matrix15_perfects_and_1_melt$WTcolor[i] <- "red"
    
  }
  
  BMS_matrix15_perfects_and_1_melt$aanum[i] <- which(BMS_aa_list$aa == BMS_matrix15_perfects_and_1_melt$X1[i])
  
}
```

Read in the RSA and SS files generated from the `parse_dssp.py` script above:
```{r}
RSA_1H1T <- read.table( "OUTPUT/structures/4KJK.RSAs.txt" , skip = 2 , header = FALSE )
SS_1H1T <- read.table( "OUTPUT/structures/4KJK.SSs.txt" , skip = 2 , header = FALSE )
cons_1H1T <- read.table( "OUTPUT/structures/4KJK.JSD_out.txt" , skip = 5 , header = FALSE )
```

<font color="red">**I have no idea what this code is doing:**</font>
```{r}
colnames(cons_1H1T) <- c("pos","cons","col")
cons_1H1T$pos <- cons_1H1T$pos+1

protein_info_1H1T <- RSA_1H1T %>%
  right_join(SS_1H1T, by="V1")

colnames(protein_info_1H1T) <- c("pos","RSA","SS")

protein_info_1H1T$cons <- 0

for (i in 1:nrow(cons_1H1T)){
  #get MSA position: cons_1H1T$pos[i]
  #find if this position exists in the e.coli MSA
  if (cons_1H1T$pos[i] %in% BMS_ecoli_map$msa_aanum){
    #get corresponding residue
    e_coli_residue <- BMS_ecoli_map$orth_aanum[which(BMS_ecoli_map$msa_aanum==cons_1H1T$pos[i])]
    protein_info_1H1T$cons[which(protein_info_1H1T$pos==e_coli_residue)] <- cons_1H1T$cons[i]
  }
}

protein_info_1H1T <- BMS_matrix15_perfects_and_1_melt %>%
  filter(X1 != "X") %>%
  group_by(X2) %>%
  dplyr::rename(pos=X2) %>%
  summarise(avgfit=mean(value,na.rm=TRUE),
            numcov=20-sum(is.na(value))) %>%
  right_join(protein_info_1H1T, by="pos")

protein_info_1H1T_melt <- melt(as.data.frame(protein_info_1H1T),
                               id=c("pos"))

protein_info_1H1T_melt$yval <- 0
protein_info_1H1T_melt$yval[which(protein_info_1H1T_melt$variable=="avgfit")] <- 23
protein_info_1H1T_melt$yval[which(protein_info_1H1T_melt$variable=="RSA")] <- 24
protein_info_1H1T_melt$yval[which(protein_info_1H1T_melt$variable=="SS")] <- 25

BMS_matrix15_perfects_and_1_melt$aanum[which(BMS_matrix15_perfects_and_1_melt$X1=="X")] <- 21.3
```

Determine coverage:
```{r warning=FALSE, class.output="goodCode"}
(nrow(BMS_matrix15_perfects_and_1_melt %>%
            filter(X1 != "X") %>%
            dplyr::select(value)) -
  sum(is.na(BMS_matrix15_perfects_and_1_melt %>%
              filter(X1 != "X") %>%
              dplyr::select(value))))/nrow(BMS_matrix15_perfects_and_1_melt %>%
                                             filter(X1 != "X") %>%
                                             dplyr::select(value))



BMS_Fig_yminlim <- -2
BMS_Fig_ymaxlim <- 30
BMS_Fig_xminlim <- -1
BMS_Fig_xmaxlim <- 160

BMS_1mut15_del <- read.table( "OUTPUT/BMS_Fitness/BMS_fitness15_map_1.csv" , skip = 0 , sep=",",header = TRUE )
#calc all data and stats
BMS_1mut15_del_collapsed_all <- BMS_1mut15_del %>%
  filter(position > 0) %>%
  group_by(position, aa) %>%
  summarise(fitval=median(fitness),
            numpoints=n(),
            stdfit=sd(fitness))

#these matrices have the fitness/num/sd for each aa at each position:
BMS_1mut15_del_matrix = matrix(rep(NA, BMS_ref_len*BMS_aa_dim),nrow=BMS_aa_dim,ncol=BMS_ref_len)

#populate matrix
for (i in 1:nrow(BMS_1mut15_del_collapsed_all)){
  
  BMS_1mut15_del_matrix[which(BMS_aa_list$aa==as.character(BMS_1mut15_del_collapsed_all$aa[i])),BMS_1mut15_del_collapsed_all$position[i]] <- as.numeric(BMS_1mut15_del_collapsed_all$fitval[i])
}

rownames(BMS_1mut15_del_matrix)<-BMS_aa_list$aa
colnames(BMS_1mut15_del_matrix)<-c(1:BMS_ref_len)

BMS_1mut15_del_matrix_melt <- melt(BMS_1mut15_del_matrix)
BMS_1mut15_del_matrix_melt <- BMS_1mut15_del_matrix_melt %>%
  filter(X1=="X")
BMS_1mut15_del_matrix_melt$WTcolor <- NA
BMS_1mut15_del_matrix_melt$aanum <- 21.3
  
BMS_matrix15_perfects_and_1_melt <- BMS_matrix15_perfects_and_1_melt %>%
  filter(X1!="X")

BMS_matrix15_perfects_and_1_melt <- rbind(BMS_matrix15_perfects_and_1_melt,BMS_1mut15_del_matrix_melt)

BMS_matrix15_perfects_and_1_melt$aanum[which(BMS_matrix15_perfects_and_1_melt$X1=="A")] <- 12
BMS_matrix15_perfects_and_1_melt$aanum[which(BMS_matrix15_perfects_and_1_melt$X1=="C")] <- 10
BMS_matrix15_perfects_and_1_melt$aanum[which(BMS_matrix15_perfects_and_1_melt$X1=="D")] <- 5
BMS_matrix15_perfects_and_1_melt$aanum[which(BMS_matrix15_perfects_and_1_melt$X1=="E")] <- 4
BMS_matrix15_perfects_and_1_melt$aanum[which(BMS_matrix15_perfects_and_1_melt$X1=="F")] <- 19
BMS_matrix15_perfects_and_1_melt$aanum[which(BMS_matrix15_perfects_and_1_melt$X1=="G")] <- 11
BMS_matrix15_perfects_and_1_melt$aanum[which(BMS_matrix15_perfects_and_1_melt$X1=="H")] <- 3
BMS_matrix15_perfects_and_1_melt$aanum[which(BMS_matrix15_perfects_and_1_melt$X1=="I")] <- 15
BMS_matrix15_perfects_and_1_melt$aanum[which(BMS_matrix15_perfects_and_1_melt$X1=="K")] <- 1
BMS_matrix15_perfects_and_1_melt$aanum[which(BMS_matrix15_perfects_and_1_melt$X1=="L")] <- 14
BMS_matrix15_perfects_and_1_melt$aanum[which(BMS_matrix15_perfects_and_1_melt$X1=="M")] <- 16
BMS_matrix15_perfects_and_1_melt$aanum[which(BMS_matrix15_perfects_and_1_melt$X1=="N")] <- 6
BMS_matrix15_perfects_and_1_melt$aanum[which(BMS_matrix15_perfects_and_1_melt$X1=="P")] <- 17
BMS_matrix15_perfects_and_1_melt$aanum[which(BMS_matrix15_perfects_and_1_melt$X1=="Q")] <- 7
BMS_matrix15_perfects_and_1_melt$aanum[which(BMS_matrix15_perfects_and_1_melt$X1=="R")] <- 2
BMS_matrix15_perfects_and_1_melt$aanum[which(BMS_matrix15_perfects_and_1_melt$X1=="S")] <- 9
BMS_matrix15_perfects_and_1_melt$aanum[which(BMS_matrix15_perfects_and_1_melt$X1=="T")] <- 8
BMS_matrix15_perfects_and_1_melt$aanum[which(BMS_matrix15_perfects_and_1_melt$X1=="V")] <- 13
BMS_matrix15_perfects_and_1_melt$aanum[which(BMS_matrix15_perfects_and_1_melt$X1=="W")] <- 20
BMS_matrix15_perfects_and_1_melt$aanum[which(BMS_matrix15_perfects_and_1_melt$X1=="Y")] <- 18
BMS_matrix15_perfects_and_1_melt$aanum[which(BMS_matrix15_perfects_and_1_melt$X1=="X")] <- 22.6

BMS_matrix15_perfects_and_1_melt_WT <- BMS_matrix15_perfects_and_1_melt %>%
  filter(WTcolor=="red")
```

## BMS Custom Plot

Bring it all together to build a custom BMS plot for COMPLEMENTATION (fitD05D03) for LIBRARY 15:
```{r}
BMS_plot15 <- ggplot() +
  #BMS matrix
  geom_rect(data=BMS_matrix15_perfects_and_1_melt,aes(xmin=X2, xmax=X2+1, ymin=aanum, ymax=aanum+1, fill=value))+
  #WT seq
  geom_rect(data=BMS_matrix15_perfects_and_1_melt_WT,aes(xmin=X2, xmax=X2+1, ymin=aanum, ymax=aanum+1), color="green",alpha=0)+
  #avg fit
  geom_rect(data=protein_info_1H1T,aes(xmin=pos, xmax=pos+1, ymin=21.3, ymax=22.3,fill=avgfit))+
  labs(x = "Position (aa)", y ="Amino acid",color="") +
  scale_fill_gradient2(low = "blue", high = "red", mid="gold",name="Fitness",na.value="grey", limit = c(-2,2)) +     #1.1*max(BMS_matrix_perfects_and_1_melt$value)
  geom_text(data=BMS_matrix15_perfects_and_1_melt[1:21,], aes(x=-1, y=aanum+0.5, label=X1), size=3.5)+
  geom_text(data=data.frame(pos=seq(0,150,30)),aes(x=pos+0.5,y=0,label=pos))+
  geom_segment(aes(x = 14.5, y = 1, xend = 14.5, yend = 0), colour = "blue")+
  geom_segment(aes(x = 15.5, y = 1, xend = 15.5, yend = 0), colour = "blue")+
  geom_segment(aes(x = 16.5, y = 1, xend = 16.5, yend = 0), colour = "blue")+
  geom_segment(aes(x = 17.5, y = 1, xend = 17.5, yend = 0), colour = "blue")+
  geom_segment(aes(x = 20.5, y = 1, xend = 20.5, yend = 0), colour = "blue")+
  geom_segment(aes(x = 27.5, y = 1, xend = 27.5, yend = 0), colour = "blue")+
  geom_segment(aes(x = 31.5, y = 1, xend = 31.5, yend = 0), colour = "blue")+
  geom_segment(aes(x = 32.5, y = 1, xend = 32.5, yend = 0), colour = "blue")+
  geom_segment(aes(x = 46.5, y = 1, xend = 46.5, yend = 0), colour = "blue")+
  geom_segment(aes(x = 52.5, y = 1, xend = 52.5, yend = 0), colour = "blue")+
  geom_segment(aes(x = 57.5, y = 1, xend = 57.5, yend = 0), colour = "blue")+
  geom_segment(aes(x = 94.5, y = 1, xend = 94.5, yend = 0), colour = "blue")+
  geom_segment(aes(x = 100.5, y = 1, xend = 100.5, yend = 0), colour = "blue")+
  theme(panel.grid.major = element_blank(), 
        panel.grid.minor = element_blank(),
        axis.ticks.x=element_blank(),
        axis.ticks.y=element_blank(),
        axis.text.x=element_blank(),
        axis.text.y=element_blank(),
        axis.line=element_blank())+
  xlim(BMS_Fig_xminlim,BMS_Fig_xmaxlim)+
  ylim(BMS_Fig_yminlim,BMS_Fig_ymaxlim)

BMS_plot15
```

```{r eval=FALSE, echo=FALSE}
#Saving 13.3 x 6.51 in images
ggsave(file="plots/BMS/BMS_Lib15_Complementation_Legend.pdf", plot=BMS_plot15,
       width=13.3, height=6.5, units="in")

#Plot without legend and save copy
BMS_plot15_2 <- BMS_plot15 + theme(legend.position="none")

#Saving 13.3 x 6.51 in images
ggsave(file="plots/BMS/BMS_Lib15_Complementation_no_Legend.pdf", plot=BMS_plot15_2,
       width=13.3, height=6.5, units="in")
```

Determine the minimum and maximum average fitness values:
```{r class.output="goodCode"}
min(protein_info_1H1T$avgfit)
max(protein_info_1H1T$avgfit)
```

Plot the SS string to go above the custom BMS plot:
```{r}
protein_info_1H1T_no_loop <- protein_info_1H1T %>%
  filter(SS != "loop")

# Plot SS

SS_plot15 <- ggplot()+
  geom_segment(aes(x = 1, y = 24.5, xend = 160, yend = 24.5), colour = "black")+
  geom_rect(data=protein_info_1H1T_no_loop,aes(xmin=pos, xmax=pos+1, ymin=24, ymax=25, fill=SS))+
  xlim(BMS_Fig_xminlim,BMS_Fig_xmaxlim)+
  ylim(BMS_Fig_yminlim,BMS_Fig_ymaxlim)+
  labs(x = "", y ="",color="") +
  theme(panel.grid.major = element_blank(), 
        panel.grid.minor = element_blank(),
        axis.ticks.x=element_blank(),
        axis.ticks.y=element_blank(),
        axis.text.x=element_blank(),
        axis.text.y=element_blank(),
        axis.line=element_blank())

SS_plot15
```

```{r eval=FALSE, echo=FALSE}
#Saving 13.3 x 6.51 in images
ggsave(file="plots/BMS/SS_Lib15_Complementation_Legend.pdf", plot=SS_plot15,
       width=13.3, height=6.5, units="in")

#Plot without legend and save copy
SS_plot15_2 <- SS_plot15 + theme(legend.position="none")

#Saving 13.3 x 6.51 in images
ggsave(file="plots/BMS/SS_Lib15_Complementation_no_Legend.pdf", plot=SS_plot15_2,
       width=13.3, height=6.5, units="in")
```

Plot the RSA (protein) string:
```{r}
RSA_plot15 <- ggplot(protein_info_1H1T)+
  geom_rect(aes(xmin=pos, xmax=pos+1, ymin=25.3, ymax=26.3, fill=RSA))+
  xlim(BMS_Fig_xminlim,BMS_Fig_xmaxlim)+
  ylim(BMS_Fig_yminlim,BMS_Fig_ymaxlim)+
  labs(x = "", y ="",color="") +
  scale_fill_gradient(low = "white", high = "red",name="RSA",na.value="grey") +
  theme(panel.grid.major = element_blank(), 
        panel.grid.minor = element_blank(),
        axis.ticks.x=element_blank(),
        axis.ticks.y=element_blank(),
        axis.text.x=element_blank(),
        axis.text.y=element_blank(),
        axis.line=element_blank())

RSA_plot15
```

```{r eval=FALSE, echo=FALSE}
#Saving 13.3 x 6.51 in images
ggsave(file="plots/BMS/RSA_Lib15_Complementation_Legend.pdf", plot=RSA_plot15,
       width=13.3, height=6.5, units="in")

#Plot without legend and save copy
RSA_plot15_2 <- RSA_plot15 + theme(legend.position="none")

#Saving 13.3 x 6.51 in images
ggsave(file="plots/BMS/RSA_Lib15_Complementation_no_Legend.pdf", plot=RSA_plot15_2,
       width=13.3, height=6.5, units="in")
```

Plot the RSA again with different color parameters:
```{r}
RSA2_plot15 <- ggplot(protein_info_1H1T)+
  geom_rect(aes(xmin=pos, xmax=pos+1, ymin=26.6, ymax=27.6, fill=cons))+
  xlim(BMS_Fig_xminlim,BMS_Fig_xmaxlim)+
  ylim(BMS_Fig_yminlim,BMS_Fig_ymaxlim)+
  labs(x = "", y ="",color="") +
  scale_fill_gradient(low = "white", high = "red",name="Cons",na.value="grey") +
  theme(panel.grid.major = element_blank(), 
        panel.grid.minor = element_blank(),
        axis.ticks.x=element_blank(),
        axis.ticks.y=element_blank(),
        axis.text.x=element_blank(),
        axis.text.y=element_blank(),
        axis.line=element_blank())

RSA2_plot15
```

```{r eval=FALSE, echo=FALSE}
#Saving 13.3 x 6.51 in images
ggsave(file="plots/BMS/RSA2_Lib15_Complementation_Legend.pdf", plot=RSA2_plot15,
       width=13.3, height=6.5, units="in")

#Plot without legend and save copy
RSA2_plot15_2 <- RSA2_plot15 + theme(legend.position="none")

#Saving 13.3 x 6.51 in images
ggsave(file="plots/BMS/RSA2_Lib15_Complementation_no_Legend.pdf", plot=RSA2_plot15_2,
       width=13.3, height=6.5, units="in")
```

## Summary Plots

Plot site coverage versus average fitness and conservation:
```{r}
RSA_cons_plot15 <- ggplot(protein_info_1H1T,aes(x=avgfit,y=numcov/20*100))+
  geom_smooth(fill="aquamarine4")+
  geom_point()+
  labs(x = "Average fitness at position", y ="Position mutational coverage (%)",color="")

RSA_cons_plot15
```

```{r eval=FALSE, echo=FALSE}
#Saving 6 x 5 in images
ggsave(file="plots/BMS/Lib15_Complementation_Coverage_vs_Avg_Fit.pdf", plot=RSA_cons_plot15,
       width=6, height=5, units="in")
```

Correlation Tests:
```{r class.output="goodCode"}
cor.test(protein_info_1H1T$avgfit,protein_info_1H1T$numcov/20*100)
cor(protein_info_1H1T$avgfit,protein_info_1H1T$numcov)
```

Plot Average Fit versus SS:
```{r}
SS_fit_plot15 <- ggplot(protein_info_1H1T,aes(x=SS,y=avgfit))+
  geom_boxplot(color="black", fill="aquamarine4")+
  geom_jitter()+
  labs(x = "Secondary structure", y ="Average fitness at position",color="")

SS_fit_plot15
```

```{r eval=FALSE, echo=FALSE}
#Saving 6 x 5 in images
ggsave(file="plots/BMS/Lib15_Complementation_Avg_Fit_vs_SS.pdf", plot=SS_fit_plot15,
       width=6, height=5, units="in")
```

Plot Average Fit versus RSA:
```{r}
RSA_fit_plot15 <- ggplot(protein_info_1H1T,aes(x=RSA,y=avgfit))+
  geom_smooth()+
  geom_point(alpha=0.7)+
  labs(x = "Relative solvent accesibility", y ="Average fitness at position",color="")+
  ylim(-2.7,2.7)

RSA_fit_plot15 <- ggExtra::ggMarginal(RSA_fit_plot15,type = "histogram",
                         xparams = list(bins=30),
                         yparams = list(bins=20),
                         col = 'black',
                         fill = 'aquamarine4')

RSA_fit_plot15
```

```{r eval=FALSE, echo=FALSE}
#Saving 6 x 5 in images
ggsave(file="plots/BMS/Lib15_Complementation_Avg_Fit_vs_RSA.pdf", plot=RSA_fit_plot15,
       width=6, height=5, units="in")
```

Correlation Tests:
```{r class.output="goodCode"}
cor.test(protein_info_1H1T$RSA,protein_info_1H1T$avgfit)
cor(protein_info_1H1T$RSA,protein_info_1H1T$avgfit)
```

Put the plots together:
```{r}
plot_BMS_2SUP <- plot_grid(SS_fit_plot15, RSA_fit_plot15, labels = c("A", "B"))
plot_BMS_2SUP
```

```{r}
save_plot("plots/BMS/avg_fit_vs_cons_vs_RSA.pdf", plot_BMS_2SUP,
          ncol = 2, # we're saving a grid plot of 2 columns
          # each individual subplot should have an aspect ratio of 1.3
          base_aspect_ratio = 1.3)
```

```{r}
patch21<- (RSA_cons_plot15 | SS_fit_plot15) / (RSA_fit_plot15)
patch21
```

```{r echo=FALSE, eval=FALSE}
ggsave(file="plots/BMS/Lib15.Avg_Fit_vs_Cons_vs_RSA.png", plot=patch21,
       dpi=600, width = 18, height = 14, units = "in")
```

```{r}
RSA_WT_plot15 <- ggplot(BMS_matrix15_perfects_and_1_melt,aes(x=WTcolor,y=value))+
  geom_boxplot(color="black", fill="aquamarine4")+
  labs(x = "Residue", y ="Fitness",color="")+ 
  scale_x_discrete(name ="Residue type",
                   labels=c("Wildtype E. coli DHFR","Other"))

RSA_WT_plot15
```

```{r eval=FALSE, echo=FALSE}
#Saving 6 x 5 in images
ggsave(file="plots/BMS/Lib15_Complementation_WT_vs_other_fitness.pdf", plot=RSA_WT_plot15,
       width=6, height=5, units="in")
```

Summary Stats:
```{r}
# WT Summary Stats
mean(as.numeric(unlist(BMS_matrix15_perfects_and_1_melt %>% filter(WTcolor=="red") %>% dplyr::select(value))))
median(as.numeric(unlist(BMS_matrix15_perfects_and_1_melt %>% filter(WTcolor=="red") %>% dplyr::select(value))))
sd(as.numeric(unlist(BMS_matrix15_perfects_and_1_melt %>% filter(WTcolor=="red") %>% dplyr::select(value))))

# Non-WT Summary Stats
mean(as.numeric(unlist(BMS_matrix15_perfects_and_1_melt %>% filter(is.na(WTcolor) & !is.na(value)) %>% dplyr::select(value))))
median(as.numeric(unlist(BMS_matrix15_perfects_and_1_melt %>% filter(is.na(WTcolor) & !is.na(value)) %>% dplyr::select(value))))
sd(as.numeric(unlist(BMS_matrix15_perfects_and_1_melt %>% filter(is.na(WTcolor) & !is.na(value)) %>% dplyr::select(value))))
```

## BMS STD Plots

Establish plotting parameters:
```{r}
BMS_matrix15_perfects_and_1_sd_melt <- BMS_matrix15_perfects_and_1_sd_melt %>%
  filter(X1!="X")

BMS_matrix15_perfects_and_1_sd_melt$aanum <- 0

BMS_matrix15_perfects_and_1_sd_melt$aanum[which(BMS_matrix15_perfects_and_1_sd_melt$X1=="A")] <- 12
BMS_matrix15_perfects_and_1_sd_melt$aanum[which(BMS_matrix15_perfects_and_1_sd_melt$X1=="C")] <- 10
BMS_matrix15_perfects_and_1_sd_melt$aanum[which(BMS_matrix15_perfects_and_1_sd_melt$X1=="D")] <- 5
BMS_matrix15_perfects_and_1_sd_melt$aanum[which(BMS_matrix15_perfects_and_1_sd_melt$X1=="E")] <- 4
BMS_matrix15_perfects_and_1_sd_melt$aanum[which(BMS_matrix15_perfects_and_1_sd_melt$X1=="F")] <- 19
BMS_matrix15_perfects_and_1_sd_melt$aanum[which(BMS_matrix15_perfects_and_1_sd_melt$X1=="G")] <- 11
BMS_matrix15_perfects_and_1_sd_melt$aanum[which(BMS_matrix15_perfects_and_1_sd_melt$X1=="H")] <- 3
BMS_matrix15_perfects_and_1_sd_melt$aanum[which(BMS_matrix15_perfects_and_1_sd_melt$X1=="I")] <- 15
BMS_matrix15_perfects_and_1_sd_melt$aanum[which(BMS_matrix15_perfects_and_1_sd_melt$X1=="K")] <- 1
BMS_matrix15_perfects_and_1_sd_melt$aanum[which(BMS_matrix15_perfects_and_1_sd_melt$X1=="L")] <- 14
BMS_matrix15_perfects_and_1_sd_melt$aanum[which(BMS_matrix15_perfects_and_1_sd_melt$X1=="M")] <- 16
BMS_matrix15_perfects_and_1_sd_melt$aanum[which(BMS_matrix15_perfects_and_1_sd_melt$X1=="N")] <- 6
BMS_matrix15_perfects_and_1_sd_melt$aanum[which(BMS_matrix15_perfects_and_1_sd_melt$X1=="P")] <- 17
BMS_matrix15_perfects_and_1_sd_melt$aanum[which(BMS_matrix15_perfects_and_1_sd_melt$X1=="Q")] <- 7
BMS_matrix15_perfects_and_1_sd_melt$aanum[which(BMS_matrix15_perfects_and_1_sd_melt$X1=="R")] <- 2
BMS_matrix15_perfects_and_1_sd_melt$aanum[which(BMS_matrix15_perfects_and_1_sd_melt$X1=="S")] <- 9
BMS_matrix15_perfects_and_1_sd_melt$aanum[which(BMS_matrix15_perfects_and_1_sd_melt$X1=="T")] <- 8
BMS_matrix15_perfects_and_1_sd_melt$aanum[which(BMS_matrix15_perfects_and_1_sd_melt$X1=="V")] <- 13
BMS_matrix15_perfects_and_1_sd_melt$aanum[which(BMS_matrix15_perfects_and_1_sd_melt$X1=="W")] <- 20
BMS_matrix15_perfects_and_1_sd_melt$aanum[which(BMS_matrix15_perfects_and_1_sd_melt$X1=="Y")] <- 18
BMS_matrix15_perfects_and_1_sd_melt$aanum[which(BMS_matrix15_perfects_and_1_sd_melt$X1=="X")] <- 22.6
```

Plot the STD values of the BMS analysis:
```{r}
BMS_std_plot15 <- ggplot() +
  #BMS matrix
  geom_rect(data=BMS_matrix15_perfects_and_1_sd_melt,aes(xmin=X2, xmax=X2+1, ymin=aanum, ymax=aanum+1, fill=value))+
  #WT seq
  geom_rect(data=BMS_matrix15_perfects_and_1_melt_WT,aes(xmin=X2, xmax=X2+1, ymin=aanum, ymax=aanum+1), color="green",alpha=0)+
  labs(x = "Position (aa)", y ="Amino acid",color="") +
  scale_fill_gradient2(low = "blue", high = "red", mid="gold",name="Std",na.value="grey", midpoint = 3, limit = c(0,1.1*max(BMS_matrix15_perfects_and_1_sd_melt$value))) +#
  theme(panel.grid.major = element_blank(), 
        panel.grid.minor = element_blank(),
        axis.ticks.x=element_blank(),
        axis.ticks.y=element_blank(),
        axis.text.x=element_blank(),
        axis.text.y=element_blank(),
        axis.line=element_blank())

BMS_std_plot15
```

Plot epistatis:
```{r}
BMS_epistasis15 <- BMS_matrix15_perfects_and_1_sd_melt %>%
  dplyr::rename(sd=value) %>%
  inner_join(BMS_matrix15_perfects_and_1_num_melt, by=c("X1","X2")) %>%
  dplyr::rename(num=value) %>%
  filter(num>4)

BMS_epistasis_std_plot15 <- ggplot() +
  #BMS matrix
  geom_rect(data=BMS_epistasis15,aes(xmin=X2, xmax=X2+1, ymin=aanum, ymax=aanum+1, fill=sd))+
  #WT seq
  geom_rect(data=BMS_matrix15_perfects_and_1_melt_WT,aes(xmin=X2, xmax=X2+1, ymin=aanum, ymax=aanum+1), color="green",alpha=0)+
  geom_text(data=BMS_matrix15_perfects_and_1_melt[1:21,], aes(x=-1, y=aanum+0.5, label=X1), size=3.5)+
  geom_text(data=data.frame(pos=seq(0,150,30)),aes(x=pos+0.5,y=0,label=pos))+
  labs(x = "Position (aa)", y ="Amino acid",color="") +
  scale_fill_gradient(low = "gold", high = "red", name="Std",na.value="grey", limit = c(3,1.1*max(BMS_epistasis15$sd))) +#
  theme(panel.grid.major = element_blank(), 
        panel.grid.minor = element_blank(),
        axis.ticks.x=element_blank(),
        axis.ticks.y=element_blank(),
        axis.text.x=element_blank(),
        axis.text.y=element_blank(),
        axis.line=element_blank())

BMS_epistasis_std_plot15
```

```{r}
BMS_epistasis15_pos <- BMS_epistasis15 %>%
  group_by(X2) %>%
  summarise(numaa=n(),meansd=mean(sd),stdsd=sd(sd))
  
ggplot(BMS_epistasis15_pos,aes(x=X2,y=numaa)) +
  geom_point()
ggplot(BMS_epistasis15_pos,aes(x=X2,y=meansd)) +
  geom_point()
ggplot(BMS_epistasis15_pos,aes(x=X2,y=stdsd)) +
  geom_point()
```

Save BMS Info as dataframe (and write to csv file):
```{r}
BMS_info15 <- right_join(BMS_matrix15_perfects_and_1_melt %>%
                         select(X1,X2,value) %>%
                         dplyr::rename(AA=X1,Pos=X2,avgfitness=value),
                       BMS_matrix15_perfects_and_1_num_melt %>%
                         select(X1,X2,value) %>%
                         dplyr::rename(AA=X1,Pos=X2,numpoints=value),
                       by=c("AA", "Pos"))

BMS_info15 <- right_join(BMS_info15,
                       BMS_matrix15_perfects_and_1_sd_melt%>%
                         select(X1,X2,value) %>%
                         dplyr::rename(AA=X1,Pos=X2,sd=value),
                       by=c("AA", "Pos"))
```

```{r eval=FALSE, echo=FALSE}
#save CSV
write.table(BMS_info15, file = paste("OUTPUT/BMS_Fitness/BMS_info_L15.csv",sep=""), 
            sep = ",", row.names = F,quote=F,col.names = T)
```

# GOF Mutants
<font color="blue">**This section is based on the R file: "R_dropout_GOF.R".**</font> It describes how to determine if certain mutant versions of a designed homolog increase in fitness under trimethoprim selection (i.e., gain of function after mutation).

## Dropout Mutants

Start by retrieving all dropout mutants with a log-fold change value less than -2.5.
```{r}
# Lib15

# Get dropouts
dropout_mutants15_GOF <- mut_collapse_15 %>%
  filter(fitD05D03 < -2.5) %>%
  dplyr::select(ID, mutID, mutations, fitD05D03)
```

Make sure none of the collapsed info has corresponding perfects with high fitness:
```{r}
dropout_fake_index = numeric()
for (i in 1:nrow(dropout_mutants15_GOF)){
  if (dropout_mutants15_GOF$ID[i] %in% perfects15$ID){
    if (perfects15$fitD05D03[which(perfects15$ID==dropout_mutants15_GOF$ID[i])] > -2.5){
      dropout_fake_index[length(dropout_fake_index)+1] <- i
    }
  }
}
if (length(dropout_fake_index) > 0){
  dropout_mutants15_GOF <- dropout_mutants15_GOF[-dropout_fake_index,]
}
```

Create a temporary dataframe compiling the total number of mutants for dropouts:
```{r}
deltemp <- mutants15 %>%
  dplyr::rename(ID=IDalign) %>%
  filter(mutations > 0 &
           mutations < 6 &
           numprunedBCs > 0 &
           (ID %in% dropout_mutants15_GOF$ID))
```

Gather all dropout mutants and select those with positive fitness
```{r}
# Mutants15 df is generated in R_plot_all_mutants.R above
dropout_mutants15_GOF <- mutants15 %>%
  dplyr::rename(ID=IDalign) %>%
  filter(fitD05D03 > 0 &
           mutations > 0 &
           mutations < 6 &
           numprunedBCs > 0) %>%
  inner_join(dropout_mutants15_GOF, by="ID") %>%
  ungroup()
```

Sum the number of unique dropouts with GOF mutants
```{r class.output="goodCode"}
length(unique(dropout_mutants15_GOF$ID))
```

## GOF Plotting

Find GOF mutants for dropouts:
```{r eval=FALSE}
mutants15_to_plot <- data.frame(ID=unique(dropout_mutants15_GOF$ID))
```

Add median fitness values for each unique ID

Make sure all of the chosen dropouts are in the MSA:
```{r eval=FALSE}
mutants15_to_plot_index_of_mut_to_drop = numeric()
for (i in 1:nrow(mutants15_to_plot)){
  mutant15_current_temp <- mutants15_to_plot$ID[i]
  if(!file.exists(paste("MSAdropouts/",mutant15_current_temp,".csv",sep=""))){
    print(mutant15_current_temp)
    mutants15_to_plot_index_of_mut_to_drop[length(mutants15_to_plot_index_of_mut_to_drop)+1] <- which(mutants15_to_plot$ID==mutant15_current_temp)
  }
}
```
Determine if any are not in MSA:
```{r eval=FALSE}
if (length(mutants15_to_plot_index_of_mut_to_drop) > 0){
  mutants15_to_plot <- mutants15_to_plot[-mutants15_to_plot_index_of_mut_to_drop,]
}
```

Read in the E. coli map:
```{r eval=FALSE}
ecoli_map <- read.csv(file=paste("PPATtrees/MSAdropouts/NP_418091.csv",sep=""),head=TRUE,sep=",")
ecoli_map_all <- read.csv(file=paste("PPATtrees/MSAmap_all/NP_418091.csv",sep=""),head=TRUE,sep=",")
```

# Reproducibility

The session information is provided for full reproducibility.
```{r}
devtools::session_info()
```