## Broad Mutational Scanning of the Dihydrofolate Reductase Protein Family

### [Karl J. Romanowicz](https://kromanowicz.github.io/), Carmen Resnick, Samuel R. Hinton, Calin Plesa
_____________________________________

This GitHub repository hosts R Markdown (RMD) files and links to the rendered code for reproducing the data analysis in the following preprint:

Romanowicz KJ, Resnick C, Hinton SR, and Plesa C. ***Exploring Antibiotic Resistance in Diverse Homologs of the Dihydrofolate Reductase Protein Family through Broad Mutational Scanning.*** [biorXiv]().

**Links to Rendered Code:**

[Mapping Files Analysis]()

[Sequencing Counts Analysis]()

[Perfect Homologs Analysis]()

[Mutant Homologs Analysis]()

[Broad Mutational Scanning Analysis]()

[Gain-of-Function Mutants Analysis]()

[Resistant Taxa Fitness Analysis]()

[Dialout Variants Fitness Analysis]()

**NCBI BioProject:** [PRJNA1189478](https://www.ncbi.nlm.nih.gov/bioproject/1189478) for raw .fastq files
_____________________________________

This pipeline processes a library of 1,536 DHFR homologs and their associated mutants in two-fold redundancy (two codon versions). It derives fitness scores from multiple time points during a multiplexed in-vivo assay, characterizing the ability of these homologs and their mutants to complement functionality of an *E. coli* knockout strain and to tolerate treatment against the antibiotic trimethoprim. This analysis helps us to understand how antibiotic resistance emerges given many evolutionarily divergent starting points within the protein family. Sequence data were generated using the Illumina NovaSeq platform using 100 bp paired-end sequencing read amplicons.

![](Images/DHFR.Diagram.png)

**Direct any questions regarding this repository to the corresponding author:** [Calin Plesa](mailto:calin@uoregon.edu).
